<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-08-31">

<title>Thomas Sandmann’s blog - Adventures with parquet: Storing &amp; querying gene expression data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Adventures with parquet: Storing &amp; quering gene expression data</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">R</div>
                <div class="quarto-category">parquet</div>
                <div class="quarto-category">TIL</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">August 31, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#loading-gene-expression-datasets" id="toc-loading-gene-expression-datasets" class="nav-link" data-scroll-target="#loading-gene-expression-datasets">Loading gene expression datasets</a></li>
  <li><a href="#write-parquet-files" id="toc-write-parquet-files" class="nav-link" data-scroll-target="#write-parquet-files">Write parquet files</a></li>
  <li><a href="#querying" id="toc-querying" class="nav-link" data-scroll-target="#querying">Querying</a>
  <ul>
  <li><a href="#arrow" id="toc-arrow" class="nav-link" data-scroll-target="#arrow">Arrow</a></li>
  <li><a href="#duckdb" id="toc-duckdb" class="nav-link" data-scroll-target="#duckdb">Duckdb</a>
  <ul class="collapse">
  <li><a href="#using-dbplyr" id="toc-using-dbplyr" class="nav-link" data-scroll-target="#using-dbplyr">Using dbplyr</a></li>
  <li><a href="#using-sql" id="toc-using-sql" class="nav-link" data-scroll-target="#using-sql">Using SQL</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#spark" id="toc-spark" class="nav-link" data-scroll-target="#spark">Spark</a>
  <ul>
  <li><a href="#dbplyr" id="toc-dbplyr" class="nav-link" data-scroll-target="#dbplyr">dbplyr</a></li>
  <li><a href="#sql" id="toc-sql" class="nav-link" data-scroll-target="#sql">SQL</a></li>
  </ul></li>
  <li><a href="#reproducibility" id="toc-reproducibility" class="nav-link" data-scroll-target="#reproducibility">Reproducibility</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<p>Today I explored storing gene expression data in <a href="https://parquet.apache.org/">parquet files</a>, and querying them with the <code>arrow</code>, <code>duckdb</code> or <code>sparklyr</code> R packages.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (lib <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"arrow"</span>, <span class="st">"dplyr"</span>, <span class="st">"duckdb"</span>, <span class="st">"edgeR"</span>, <span class="st">"fs"</span>, <span class="st">"glue"</span>, <span class="st">"memoise"</span>,</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>              <span class="st">"rnaseqExamples"</span>, <span class="st">"sparklyr"</span>, <span class="st">"tictoc"</span>, <span class="st">"tidyr"</span>, <span class="st">"DESeq2"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>              <span class="st">"Homo.sapiens"</span>, <span class="st">"Mus.musculus"</span>)) {</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">suppressPackageStartupMessages</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(lib, <span class="at">character.only =</span> <span class="cn">TRUE</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>, <span class="at">warn.conflicts =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A while back, I created the <a href="https://tomsing1.github.io/rnaseq-examples/index.html">rnaseq-examples</a> R package with serialized <code>SummarizedExperiment</code> objects for three published RNA-seq experiments <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. Here, I will</p>
<ol type="1">
<li>Load the data from all three experiments into my R session,</li>
<li>Extract the raw counts and TMM-normalized gene expression measurements (counts per million) into (tidy) R data.frames,</li>
<li>Store the data in separate parquet files, one for each experiment and</li>
<li>Query the parquet files with the</li>
</ol>
<ul>
<li><a href="https://arrow.apache.org/docs/r/">arrow</a>,</li>
<li><a href="https://duckdb.org/docs/archive/0.8.1/api/r">duckdb</a> and</li>
<li><a href="https://spark.rstudio.com/">sparklyr</a> R packages.</li>
</ul>
<p>I was amazed by the speed of <code>duckdb</code> on my local system, and am looking forward to distributing large datasets across nodes of a Spark cluster in the future. Either way, parquet files are a highly portable and language-agnostic file format that I am happy to add to my toolkit.</p>
</section>
<section id="loading-gene-expression-datasets" class="level2">
<h2 class="anchored" data-anchor-id="loading-gene-expression-datasets">Loading gene expression datasets</h2>
<p>The <code>rnaseqExamplesR</code> package is <a href="https://github.com/tomsing1/rnaseq-examples">available from github</a> and can be installed via</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"tomsing1/rnaseq-examples"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once installed, we can load the three <code>SummarizedExperiment</code> objects it contains into our R session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>datasets <span class="ot">&lt;-</span> <span class="fu">data</span>(<span class="at">package =</span> <span class="st">"rnaseqExamples"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">data.frame</span>(datasets<span class="sc">$</span>results[, <span class="fu">c</span>(<span class="st">"Item"</span>, <span class="st">"Title"</span>)]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 8%">
<col style="width: 91%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Item</th>
<th style="text-align: left;">Title</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">rnai</td>
<td style="text-align: left;">Naguib et al: Global effects of SUPT4H1 RNAi on gene expression</td>
</tr>
<tr class="even">
<td style="text-align: left;">sarm1</td>
<td style="text-align: left;">Zhu et al: Effects of SARM1 deficiency on gene expression</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tau</td>
<td style="text-align: left;">Wang et al: time course of microglia from the rTg4510 mouse model</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"tau"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, I define the <code>tidy()</code> helper function to calculate <code>sizeFactors</code> for normalization with <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">Robinson’s and Oshlack’s TMM method</a> and export raw counts alongside (normalized) counts per million (CPM) in a data.frame.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Note that I am not exporting sample annotations in this example. This information could be added to the data.frame (with a join operation) or stored in a different format, e.g.&nbsp;for efficient row-wise queries.</li>
<li>The helper function is wrapped in a to <code>memoise::memoise()</code>, which caches the outputs in memory. That way, repeated calling of the same function avoids additional computations / joins. This functionality is not really required here, but I will try to remember it for future reference!</li>
</ul>
</div>
</div>
</div>
<div class="cell">
<details>
<summary>A helper function to coerce a SummarizedExperiment into a tibbbe</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#' Coerce a DGEList or a SummarizedExperiment into a tibble</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param x Either a `DGEList` or a `SummarizedExperiment`</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#' @return A tibble</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>.tidy <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove non-ensembl feature identifiers (e.g. spike-ins)</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> x[<span class="fu">grepl</span>(<span class="st">"^ENS"</span>, <span class="fu">row.names</span>(x)), ]</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  annotation <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all</span>(<span class="fu">grepl</span>(<span class="st">"^ENSG"</span>, <span class="fu">row.names</span>(x))) <span class="sc">~</span> <span class="st">"Homo.sapiens"</span>,</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all</span>(<span class="fu">grepl</span>(<span class="st">"^ENSMUS"</span>, <span class="fu">row.names</span>(x))) <span class="sc">~</span> <span class="st">"Mus.musculus"</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">is.na</span>(annotation))</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(annotation, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract raw counts</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(x)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> y<span class="sc">$</span>counts <span class="sc">%&gt;%</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(y), </span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"count"</span>)</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract cpms</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  cpms <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(y, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>, <span class="at">log =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(y), </span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"cpm"</span>)</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add gene annotations &amp; alternative identifiers</span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">inner_join</span>(</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>    counts, cpms, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"feature_id"</span>, <span class="st">"sample_id"</span>)</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(y<span class="sc">$</span>genes, <span class="st">"feature_id"</span>),</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">ensembl =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>      <span class="at">entrez =</span> <span class="fu">suppressMessages</span>({</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mapIds</span>(<span class="fu">get</span>(annotation), <span class="at">keys =</span> .data<span class="sc">$</span>ensembl, </span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>               <span class="at">column =</span> <span class="st">"ENTREZID"</span>, <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>               <span class="at">multiVals =</span> <span class="st">"first"</span>)</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">any_of</span>(<span class="st">"spikein"</span>))</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>tidy <span class="ot">&lt;-</span> <span class="fu">memoise</span>(.tidy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="write-parquet-files" class="level2">
<h2 class="anchored" data-anchor-id="write-parquet-files">Write parquet files</h2>
<p>Now we extract the gene expression measurements from each of our <code>SummarizedExperiment</code> objects and then write it to a parquet file in a temporary directory on the local file system. Alternatively, I could store the files on a cloud system, e.g.&nbsp;AWS S3, and access them remotely.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a temporary directory</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>out_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"parquet"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(out_dir)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (dataset <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"rnai"</span>, <span class="st">"sarm1"</span>)) {</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">tidy</span>(<span class="fu">get</span>(dataset))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>study <span class="ot">&lt;-</span> dataset  <span class="co"># add a columsn with the name of the experiment</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  arrow<span class="sc">::</span><span class="fu">write_parquet</span>(</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> df, </span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">sink =</span> <span class="fu">file.path</span>(out_dir, <span class="fu">paste0</span>(dataset, <span class="st">".parquet"</span>))</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># list the contents of the temporary directory</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_info</span>(out_dir) <span class="sc">%&gt;%</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(path, size) <span class="sc">%&gt;%</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">path =</span> <span class="fu">basename</span>(path))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  path                 size
  &lt;chr&gt;         &lt;fs::bytes&gt;
1 rnai.parquet        5.22M
2 sarm1.parquet       5.29M
3 tau.parquet         8.22M</code></pre>
</div>
</div>
</section>
<section id="querying" class="level2">
<h2 class="anchored" data-anchor-id="querying">Querying</h2>
<section id="arrow" class="level3">
<h3 class="anchored" data-anchor-id="arrow">Arrow</h3>
<p>Even though we created three separate files, the <a href="https://arrow.apache.org/docs/r/">arrow R package</a> can abstract them into a single <code>Dataset</code>. We simply point the <code>open_dataset()</code> function at the directory containing the <code>.parquet</code> files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> arrow<span class="sc">::</span><span class="fu">open_dataset</span>(out_dir)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FileSystemDataset with 3 Parquet files
feature_id: string
sample_id: string
count: int32
cpm: double
ensembl: string
symbol: string
gene_type: string
entrez: string
study: string

See $metadata for additional Schema metadata</code></pre>
</div>
</div>
<p>We can use <code>dbplr</code> verbs to query this <code>FileSystemDataset</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">"arrow"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ds <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
# Groups:   symbol [2]
  symbol study     n
  &lt;chr&gt;  &lt;chr&gt; &lt;int&gt;
1 GAPDH  rnai     12
2 Gapdh  sarm1    16
3 Gapdh  tau      32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>arrow: 1.717 sec elapsed</code></pre>
</div>
</div>
<p>On my system, it takes ~ 1 second to retrieve the results, and about the same amount of time is needed to retrieve the full <code>rnai</code> dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>ds <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 694,236 × 9
   feature_id      sample_id count     cpm ensembl symbol gene_type entrez study
 * &lt;chr&gt;           &lt;chr&gt;     &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
 1 ENSG00000163106 S320          0  0      ENSG00… HPGDS  protein_… 27306  rnai 
 2 ENSG00000163106 S168          2  0.0318 ENSG00… HPGDS  protein_… 27306  rnai 
 3 ENSG00000163106 S255          1  0.0111 ENSG00… HPGDS  protein_… 27306  rnai 
 4 ENSG00000163106 S190          0  0      ENSG00… HPGDS  protein_… 27306  rnai 
 5 ENSG00000163110 S912       1607 26.0    ENSG00… PDLIM5 protein_… 10611  rnai 
 6 ENSG00000163110 S832       2320 29.2    ENSG00… PDLIM5 protein_… 10611  rnai 
 7 ENSG00000163110 S783       1022 22.7    ENSG00… PDLIM5 protein_… 10611  rnai 
 8 ENSG00000163110 S623       1159 28.9    ENSG00… PDLIM5 protein_… 10611  rnai 
 9 ENSG00000163110 S622       1384 20.7    ENSG00… PDLIM5 protein_… 10611  rnai 
10 ENSG00000163110 S487       1023 32.6    ENSG00… PDLIM5 protein_… 10611  rnai 
# ℹ 694,226 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1.09 sec elapsed</code></pre>
</div>
</div>
</section>
<section id="duckdb" class="level3">
<h3 class="anchored" data-anchor-id="duckdb">Duckdb</h3>
<p>Alternatively, we can use <a href="https://duckdb.org/">duckdb</a> to query our parquet files. The <a href="https://duckdb.org/docs/api/r">duckdb R API</a> supports both dbplyr verbs and raw SQL queries.</p>
<p>First, we establish a connection to the <code>duckdb</code> backend:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="using-dbplyr" class="level4">
<h4 class="anchored" data-anchor-id="using-dbplyr">Using dbplyr</h4>
<p>First, we execute the same dbplyr query we used above, which translates it into SQL for us and passes it on to duckdb.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">"duckdb"</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tbl</span>(con, <span class="fu">sprintf</span>(<span class="st">"read_parquet('%s/*.parquet')"</span>, out_dir)) <span class="sc">%&gt;%</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  symbol study     n
  &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt;
1 GAPDH  rnai     12
2 Gapdh  tau      32
3 Gapdh  sarm1    16</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>duckdb: 0.065 sec elapsed</code></pre>
</div>
</div>
<p>On my system, <code>duckdb</code> returns results more than 10x faster than arrow’s implementation (see above).</p>
<p>Retrieval of the full dataset for the <code>rnai</code> study is also completed in less than half a second:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tbl</span>(con, <span class="fu">glue</span>(<span class="st">"read_parquet('{out_dir}/*.parquet')"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 694,236 × 9
   feature_id      sample_id count   cpm ensembl   symbol gene_type entrez study
   &lt;chr&gt;           &lt;chr&gt;     &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
 1 ENSG00000000003 S912       1942  31.4 ENSG0000… TSPAN6 protein_… 7105   rnai 
 2 ENSG00000000003 S832       2942  37.1 ENSG0000… TSPAN6 protein_… 7105   rnai 
 3 ENSG00000000003 S783        905  20.1 ENSG0000… TSPAN6 protein_… 7105   rnai 
 4 ENSG00000000003 S623        960  24.0 ENSG0000… TSPAN6 protein_… 7105   rnai 
 5 ENSG00000000003 S622       2002  29.9 ENSG0000… TSPAN6 protein_… 7105   rnai 
 6 ENSG00000000003 S487        792  25.3 ENSG0000… TSPAN6 protein_… 7105   rnai 
 7 ENSG00000000003 S458       1549  20.9 ENSG0000… TSPAN6 protein_… 7105   rnai 
 8 ENSG00000000003 S420       1021  25.3 ENSG0000… TSPAN6 protein_… 7105   rnai 
 9 ENSG00000000003 S320       1432  21.5 ENSG0000… TSPAN6 protein_… 7105   rnai 
10 ENSG00000000003 S168       1773  28.2 ENSG0000… TSPAN6 protein_… 7105   rnai 
# ℹ 694,226 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.244 sec elapsed</code></pre>
</div>
</div>
</section>
<section id="using-sql" class="level4">
<h4 class="anchored" data-anchor-id="using-sql">Using SQL</h4>
<p>Because <code>duckdb</code> is a SQL database, we can also query the parquet files directly with raw SQL, realizing another gain in execution speed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> con,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue_sql</span>(</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT symbol, study, COUNT(*) AS n </span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM read_parquet({paste0(out_dir, '/*.parquet')}) </span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE UPPER(symbol) = 'GAPDH' </span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="st">     GROUP BY symbol, study"</span>, </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">.con =</span> con)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  symbol study  n
1  GAPDH  rnai 12
2  Gapdh sarm1 16
3  Gapdh   tau 32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.024 sec elapsed</code></pre>
</div>
</div>
<p>Similarly, reading all data for the <code>rnai</code> dataset into memory is faster than with arrow’s implementation as well:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> con,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue_sql</span>(</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT * </span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM read_parquet({paste0(out_dir, '/*.parquet')}) </span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE study = 'rnai'"</span>, </span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">.con =</span> con)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       feature_id sample_id count      cpm         ensembl symbol
1 ENSG00000000003      S912  1942 31.43893 ENSG00000000003 TSPAN6
2 ENSG00000000003      S832  2942 37.08823 ENSG00000000003 TSPAN6
3 ENSG00000000003      S783   905 20.10531 ENSG00000000003 TSPAN6
4 ENSG00000000003      S623   960 23.97706 ENSG00000000003 TSPAN6
5 ENSG00000000003      S622  2002 29.94743 ENSG00000000003 TSPAN6
6 ENSG00000000003      S487   792 25.26281 ENSG00000000003 TSPAN6
       gene_type entrez study
1 protein_coding   7105  rnai
2 protein_coding   7105  rnai
3 protein_coding   7105  rnai
4 protein_coding   7105  rnai
5 protein_coding   7105  rnai
6 protein_coding   7105  rnai</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.211 sec elapsed</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="spark" class="level2">
<h2 class="anchored" data-anchor-id="spark">Spark</h2>
<p>Finally, and mainly just for future reference, the <a href="https://spark.rstudio.com/">sparklyr R package</a> provides an R interface to leverage a Spark cluster and its distributed analysis libraries.</p>
<p>First, we establish a connection to the Spark cluster. Here am running a local Spark cluster (e.g.&nbsp;a single local node) on my laptop:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">spark_connect</span>(<span class="at">master =</span> <span class="st">"local"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we import the datasets into the cluster by creating a new <code>Spark DataFrame</code> with the name <code>gene_expression</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sdf <span class="ot">&lt;-</span> <span class="fu">spark_read_parquet</span>(<span class="at">sc =</span> sc, <span class="at">name =</span> <span class="st">"gene_expression"</span>,</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">path =</span> <span class="fu">paste0</span>(out_dir, <span class="st">'/*.parquet'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="dbplyr" class="level3">
<h3 class="anchored" data-anchor-id="dbplyr">dbplyr</h3>
<p>The <code>tbl_spark</code> object returned by <code>sparklyr::spark_read_parquet()</code> can be queried with <code>dbplyr</code> verbs, e.g.&nbsp;to translate our now familiar queries into <code>Spark SQL</code> statements on the fly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>sdf <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  symbol study     n
  &lt;chr&gt;  &lt;chr&gt; &lt;int&gt;
1 Gapdh  sarm1    16
2 GAPDH  rnai     12
3 Gapdh  tau      32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1.472 sec elapsed</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>sdf <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 694,236 × 9
   feature_id      sample_id count   cpm ensembl   symbol gene_type entrez study
   &lt;chr&gt;           &lt;chr&gt;     &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
 1 ENSG00000000003 S912       1942  31.4 ENSG0000… TSPAN6 protein_… 7105   rnai 
 2 ENSG00000000003 S832       2942  37.1 ENSG0000… TSPAN6 protein_… 7105   rnai 
 3 ENSG00000000003 S783        905  20.1 ENSG0000… TSPAN6 protein_… 7105   rnai 
 4 ENSG00000000003 S623        960  24.0 ENSG0000… TSPAN6 protein_… 7105   rnai 
 5 ENSG00000000003 S622       2002  29.9 ENSG0000… TSPAN6 protein_… 7105   rnai 
 6 ENSG00000000003 S487        792  25.3 ENSG0000… TSPAN6 protein_… 7105   rnai 
 7 ENSG00000000003 S458       1549  20.9 ENSG0000… TSPAN6 protein_… 7105   rnai 
 8 ENSG00000000003 S420       1021  25.3 ENSG0000… TSPAN6 protein_… 7105   rnai 
 9 ENSG00000000003 S320       1432  21.5 ENSG0000… TSPAN6 protein_… 7105   rnai 
10 ENSG00000000003 S168       1773  28.2 ENSG0000… TSPAN6 protein_… 7105   rnai 
# ℹ 694,226 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3.329 sec elapsed</code></pre>
</div>
</div>
</section>
<section id="sql" class="level3">
<h3 class="anchored" data-anchor-id="sql">SQL</h3>
<p>Alternatively, we can use SQL queries to query the cluster’s <code>gene_expression</code> table directly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> sc,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue</span>(</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT symbol, study, COUNT(*) AS n </span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM gene_expression </span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE UPPER(symbol) = 'GAPDH' </span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="st">     GROUP BY symbol, study"</span>)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  symbol study  n
1  Gapdh sarm1 16
2  GAPDH  rnai 12
3  Gapdh   tau 32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1.493 sec elapsed</code></pre>
</div>
</div>
<p>My local Spark instance performs these queries more slowly than e.g.&nbsp;duckdb. But Spark’s <em>real power</em> is in deploying <a href="https://spark.rstudio.com/guides/mlib.html">Machine learning models</a> across a (potentially large) cluster, enabling parallel processing of very large datasets by distributing both data and computation across nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">spark_disconnect</span>(sc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="reproducibility" class="level2">
<h2 class="anchored" data-anchor-id="reproducibility">Reproducibility</h2>
<details>
<summary>
Session Information
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="st">"attached"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       macOS Ventura 13.5.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Los_Angeles
 date     2023-09-01
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 ! package                            * version    date (UTC) lib source
 P AnnotationDbi                      * 1.62.2     2023-07-02 [?] Bioconductor
 P arrow                              * 13.0.0     2023-08-30 [?] CRAN (R 4.3.0)
 P Biobase                            * 2.60.0     2023-05-08 [?] Bioconductor
 P BiocGenerics                       * 0.46.0     2023-06-04 [?] Bioconductor
 P DBI                                * 1.1.3      2022-06-18 [?] CRAN (R 4.3.0)
 P DESeq2                             * 1.40.2     2023-07-02 [?] Bioconductor
 P dplyr                              * 1.1.2      2023-04-20 [?] CRAN (R 4.3.0)
 P duckdb                             * 0.8.1-3    2023-09-01 [?] CRAN (R 4.3.1)
 P edgeR                              * 3.42.4     2023-06-04 [?] Bioconductor
 P fs                                 * 1.6.3      2023-07-20 [?] CRAN (R 4.3.0)
 P GenomeInfoDb                       * 1.36.1     2023-07-02 [?] Bioconductor
 P GenomicFeatures                    * 1.52.2     2023-08-27 [?] Bioconductor
 P GenomicRanges                      * 1.52.0     2023-05-08 [?] Bioconductor
 P glue                               * 1.6.2      2022-02-24 [?] CRAN (R 4.3.0)
 P GO.db                              * 3.17.0     2023-08-25 [?] Bioconductor
 P Homo.sapiens                       * 1.3.1      2023-09-01 [?] Bioconductor
 P IRanges                            * 2.34.1     2023-07-02 [?] Bioconductor
 P limma                              * 3.56.2     2023-06-04 [?] Bioconductor
 P MatrixGenerics                     * 1.12.3     2023-07-30 [?] Bioconductor
 P matrixStats                        * 1.0.0      2023-06-02 [?] CRAN (R 4.3.0)
 P memoise                            * 2.0.1      2021-11-26 [?] CRAN (R 4.3.0)
 P Mus.musculus                       * 1.3.1      2023-09-01 [?] Bioconductor
 P org.Hs.eg.db                       * 3.17.0     2023-08-25 [?] Bioconductor
 P org.Mm.eg.db                       * 3.17.0     2023-08-23 [?] Bioconductor
 P OrganismDbi                        * 1.42.0     2023-05-08 [?] Bioconductor
 P rnaseqExamples                     * 0.0.0.9000 2023-09-01 [?] Github (tomsing1/rnaseq-examples@ac35304)
 P S4Vectors                          * 0.38.1     2023-05-08 [?] Bioconductor
 P sparklyr                           * 1.8.2      2023-07-01 [?] CRAN (R 4.3.0)
 P SummarizedExperiment               * 1.30.2     2023-06-11 [?] Bioconductor
 P tictoc                             * 1.2        2023-04-23 [?] CRAN (R 4.3.0)
 P tidyr                              * 1.3.0      2023-01-24 [?] CRAN (R 4.3.0)
 P TxDb.Hsapiens.UCSC.hg19.knownGene  * 3.2.2      2023-09-01 [?] Bioconductor
 P TxDb.Mmusculus.UCSC.mm10.knownGene * 3.10.0     2023-09-01 [?] Bioconductor

 [1] /Users/sandmann/repositories/blog/renv/library/R-4.3/aarch64-apple-darwin20
 [2] /Users/sandmann/Library/Caches/org.R-project.R/R/renv/sandbox/R-4.3/aarch64-apple-darwin20/ac5c2659

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>


<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>If you are curious, please consult the <a href="https://tomsing1.github.io/rnaseq-examples/articles/index.html">vignettes</a> to see examples of differential expression analyses for each dataset.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb52" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Adventures with parquet: Storing &amp; querying gene expression data"</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-08-31"</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [R, parquet, TIL]</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a>Today I explored storing gene expression data in</span>
<span id="cb52-25"><a href="#cb52-25" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">parquet files</span><span class="co">](https://parquet.apache.org/)</span>,</span>
<span id="cb52-26"><a href="#cb52-26" aria-hidden="true" tabindex="-1"></a>and querying them with the <span class="in">`arrow`</span>, <span class="in">`duckdb`</span> or <span class="in">`sparklyr`</span> R packages.</span>
<span id="cb52-27"><a href="#cb52-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-28"><a href="#cb52-28" aria-hidden="true" tabindex="-1"></a><span class="fu">## Introduction</span></span>
<span id="cb52-29"><a href="#cb52-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-32"><a href="#cb52-32" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-33"><a href="#cb52-33" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (lib <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"arrow"</span>, <span class="st">"dplyr"</span>, <span class="st">"duckdb"</span>, <span class="st">"edgeR"</span>, <span class="st">"fs"</span>, <span class="st">"glue"</span>, <span class="st">"memoise"</span>,</span>
<span id="cb52-34"><a href="#cb52-34" aria-hidden="true" tabindex="-1"></a>              <span class="st">"rnaseqExamples"</span>, <span class="st">"sparklyr"</span>, <span class="st">"tictoc"</span>, <span class="st">"tidyr"</span>, <span class="st">"DESeq2"</span>, </span>
<span id="cb52-35"><a href="#cb52-35" aria-hidden="true" tabindex="-1"></a>              <span class="st">"Homo.sapiens"</span>, <span class="st">"Mus.musculus"</span>)) {</span>
<span id="cb52-36"><a href="#cb52-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">suppressPackageStartupMessages</span>(</span>
<span id="cb52-37"><a href="#cb52-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(lib, <span class="at">character.only =</span> <span class="cn">TRUE</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>, <span class="at">warn.conflicts =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-38"><a href="#cb52-38" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb52-39"><a href="#cb52-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-40"><a href="#cb52-40" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-41"><a href="#cb52-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-42"><a href="#cb52-42" aria-hidden="true" tabindex="-1"></a>A while back, I created the</span>
<span id="cb52-43"><a href="#cb52-43" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">rnaseq-examples</span><span class="co">](https://tomsing1.github.io/rnaseq-examples/index.html)</span></span>
<span id="cb52-44"><a href="#cb52-44" aria-hidden="true" tabindex="-1"></a>R package with serialized <span class="in">`SummarizedExperiment`</span> objects for three</span>
<span id="cb52-45"><a href="#cb52-45" aria-hidden="true" tabindex="-1"></a>published RNA-seq experiments <span class="ot">[^1]</span>. Here, I will </span>
<span id="cb52-46"><a href="#cb52-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-47"><a href="#cb52-47" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Load the data from all three experiments into my R session,</span>
<span id="cb52-48"><a href="#cb52-48" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Extract the raw counts and TMM-normalized gene expression measurements</span>
<span id="cb52-49"><a href="#cb52-49" aria-hidden="true" tabindex="-1"></a>   (counts per million) into (tidy) R data.frames,</span>
<span id="cb52-50"><a href="#cb52-50" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Store the data in separate parquet files, one for each experiment and</span>
<span id="cb52-51"><a href="#cb52-51" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Query the parquet files with the</span>
<span id="cb52-52"><a href="#cb52-52" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">arrow</span><span class="co">](https://arrow.apache.org/docs/r/)</span>,</span>
<span id="cb52-53"><a href="#cb52-53" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">duckdb</span><span class="co">](https://duckdb.org/docs/archive/0.8.1/api/r)</span> and</span>
<span id="cb52-54"><a href="#cb52-54" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">sparklyr</span><span class="co">](https://spark.rstudio.com/)</span></span>
<span id="cb52-55"><a href="#cb52-55" aria-hidden="true" tabindex="-1"></a>  R packages.</span>
<span id="cb52-56"><a href="#cb52-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-57"><a href="#cb52-57" aria-hidden="true" tabindex="-1"></a>I was amazed by the speed of <span class="in">`duckdb`</span> on my local system, and am looking forward</span>
<span id="cb52-58"><a href="#cb52-58" aria-hidden="true" tabindex="-1"></a>to distributing large datasets across nodes of a Spark cluster in the future. </span>
<span id="cb52-59"><a href="#cb52-59" aria-hidden="true" tabindex="-1"></a>Either way, parquet files are a highly portable and language-agnostic file </span>
<span id="cb52-60"><a href="#cb52-60" aria-hidden="true" tabindex="-1"></a>format that I am happy to add to my toolkit.</span>
<span id="cb52-61"><a href="#cb52-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-62"><a href="#cb52-62" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>If you are curious, please consult the </span>
<span id="cb52-63"><a href="#cb52-63" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">vignettes</span><span class="co">](https://tomsing1.github.io/rnaseq-examples/articles/index.html)</span></span>
<span id="cb52-64"><a href="#cb52-64" aria-hidden="true" tabindex="-1"></a>to see examples of differential expression analyses for each dataset.</span>
<span id="cb52-65"><a href="#cb52-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-66"><a href="#cb52-66" aria-hidden="true" tabindex="-1"></a><span class="fu">## Loading gene expression datasets</span></span>
<span id="cb52-67"><a href="#cb52-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-68"><a href="#cb52-68" aria-hidden="true" tabindex="-1"></a>The <span class="in">`rnaseqExamplesR`</span> package is </span>
<span id="cb52-69"><a href="#cb52-69" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">available from github</span><span class="co">](https://github.com/tomsing1/rnaseq-examples)</span> and </span>
<span id="cb52-70"><a href="#cb52-70" aria-hidden="true" tabindex="-1"></a>can be installed via </span>
<span id="cb52-71"><a href="#cb52-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-74"><a href="#cb52-74" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-75"><a href="#cb52-75" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-76"><a href="#cb52-76" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"tomsing1/rnaseq-examples"</span>)</span>
<span id="cb52-77"><a href="#cb52-77" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-78"><a href="#cb52-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-79"><a href="#cb52-79" aria-hidden="true" tabindex="-1"></a>Once installed, we can load the three <span class="in">`SummarizedExperiment`</span> objects it contains</span>
<span id="cb52-80"><a href="#cb52-80" aria-hidden="true" tabindex="-1"></a>into our R session.</span>
<span id="cb52-81"><a href="#cb52-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-84"><a href="#cb52-84" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-85"><a href="#cb52-85" aria-hidden="true" tabindex="-1"></a>datasets <span class="ot">&lt;-</span> <span class="fu">data</span>(<span class="at">package =</span> <span class="st">"rnaseqExamples"</span>)</span>
<span id="cb52-86"><a href="#cb52-86" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">data.frame</span>(datasets<span class="sc">$</span>results[, <span class="fu">c</span>(<span class="st">"Item"</span>, <span class="st">"Title"</span>)]))</span>
<span id="cb52-87"><a href="#cb52-87" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-88"><a href="#cb52-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-91"><a href="#cb52-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-92"><a href="#cb52-92" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"tau"</span>)</span>
<span id="cb52-93"><a href="#cb52-93" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-94"><a href="#cb52-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-95"><a href="#cb52-95" aria-hidden="true" tabindex="-1"></a>Next, I define the <span class="in">`tidy()`</span> helper function to calculate <span class="in">`sizeFactors`</span></span>
<span id="cb52-96"><a href="#cb52-96" aria-hidden="true" tabindex="-1"></a>for normalization with </span>
<span id="cb52-97"><a href="#cb52-97" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Robinson's and Oshlack's TMM method</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25)</span></span>
<span id="cb52-98"><a href="#cb52-98" aria-hidden="true" tabindex="-1"></a>and export raw counts alongside (normalized) counts per million (CPM) in a</span>
<span id="cb52-99"><a href="#cb52-99" aria-hidden="true" tabindex="-1"></a>data.frame.</span>
<span id="cb52-100"><a href="#cb52-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-101"><a href="#cb52-101" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true"}</span>
<span id="cb52-102"><a href="#cb52-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-103"><a href="#cb52-103" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Note that I am not exporting sample annotations in this example. This</span>
<span id="cb52-104"><a href="#cb52-104" aria-hidden="true" tabindex="-1"></a>  information could be added to the data.frame (with a join operation) or </span>
<span id="cb52-105"><a href="#cb52-105" aria-hidden="true" tabindex="-1"></a>  stored in a different format, e.g. for efficient row-wise queries.</span>
<span id="cb52-106"><a href="#cb52-106" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The helper function is wrapped in a to <span class="in">`memoise::memoise()`</span>, which caches</span>
<span id="cb52-107"><a href="#cb52-107" aria-hidden="true" tabindex="-1"></a>  the outputs in memory. That way, repeated calling of the same function avoids</span>
<span id="cb52-108"><a href="#cb52-108" aria-hidden="true" tabindex="-1"></a>  additional computations / joins. This functionality is not really required</span>
<span id="cb52-109"><a href="#cb52-109" aria-hidden="true" tabindex="-1"></a>  here, but I will try to remember it for future reference!</span>
<span id="cb52-110"><a href="#cb52-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-111"><a href="#cb52-111" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb52-112"><a href="#cb52-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-115"><a href="#cb52-115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-116"><a href="#cb52-116" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb52-117"><a href="#cb52-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "A helper function to coerce a SummarizedExperiment into a tibble"</span></span>
<span id="cb52-118"><a href="#cb52-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-119"><a href="#cb52-119" aria-hidden="true" tabindex="-1"></a><span class="co">#' Coerce a DGEList or a SummarizedExperiment into a tibble</span></span>
<span id="cb52-120"><a href="#cb52-120" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb52-121"><a href="#cb52-121" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param x Either a `DGEList` or a `SummarizedExperiment`</span></span>
<span id="cb52-122"><a href="#cb52-122" aria-hidden="true" tabindex="-1"></a><span class="co">#' @return A tibble</span></span>
<span id="cb52-123"><a href="#cb52-123" aria-hidden="true" tabindex="-1"></a>.tidy <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb52-124"><a href="#cb52-124" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-125"><a href="#cb52-125" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove non-ensembl feature identifiers (e.g. spike-ins)</span></span>
<span id="cb52-126"><a href="#cb52-126" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> x[<span class="fu">grepl</span>(<span class="st">"^ENS"</span>, <span class="fu">row.names</span>(x)), ]</span>
<span id="cb52-127"><a href="#cb52-127" aria-hidden="true" tabindex="-1"></a>  annotation <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb52-128"><a href="#cb52-128" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all</span>(<span class="fu">grepl</span>(<span class="st">"^ENSG"</span>, <span class="fu">row.names</span>(x))) <span class="sc">~</span> <span class="st">"Homo.sapiens"</span>,</span>
<span id="cb52-129"><a href="#cb52-129" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all</span>(<span class="fu">grepl</span>(<span class="st">"^ENSMUS"</span>, <span class="fu">row.names</span>(x))) <span class="sc">~</span> <span class="st">"Mus.musculus"</span></span>
<span id="cb52-130"><a href="#cb52-130" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb52-131"><a href="#cb52-131" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">is.na</span>(annotation))</span>
<span id="cb52-132"><a href="#cb52-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(annotation, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-133"><a href="#cb52-133" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-134"><a href="#cb52-134" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract raw counts</span></span>
<span id="cb52-135"><a href="#cb52-135" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(x)</span>
<span id="cb52-136"><a href="#cb52-136" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> y<span class="sc">$</span>counts <span class="sc">%&gt;%</span></span>
<span id="cb52-137"><a href="#cb52-137" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb52-138"><a href="#cb52-138" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-139"><a href="#cb52-139" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(y), </span>
<span id="cb52-140"><a href="#cb52-140" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"count"</span>)</span>
<span id="cb52-141"><a href="#cb52-141" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-142"><a href="#cb52-142" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract cpms</span></span>
<span id="cb52-143"><a href="#cb52-143" aria-hidden="true" tabindex="-1"></a>  cpms <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(y, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>, <span class="at">log =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-144"><a href="#cb52-144" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb52-145"><a href="#cb52-145" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-146"><a href="#cb52-146" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(y), </span>
<span id="cb52-147"><a href="#cb52-147" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"cpm"</span>)</span>
<span id="cb52-148"><a href="#cb52-148" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-149"><a href="#cb52-149" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add gene annotations &amp; alternative identifiers</span></span>
<span id="cb52-150"><a href="#cb52-150" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">inner_join</span>(</span>
<span id="cb52-151"><a href="#cb52-151" aria-hidden="true" tabindex="-1"></a>    counts, cpms, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"feature_id"</span>, <span class="st">"sample_id"</span>)</span>
<span id="cb52-152"><a href="#cb52-152" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb52-153"><a href="#cb52-153" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb52-154"><a href="#cb52-154" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(y<span class="sc">$</span>genes, <span class="st">"feature_id"</span>),</span>
<span id="cb52-155"><a href="#cb52-155" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-156"><a href="#cb52-156" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">ensembl =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-157"><a href="#cb52-157" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb52-158"><a href="#cb52-158" aria-hidden="true" tabindex="-1"></a>      <span class="at">entrez =</span> <span class="fu">suppressMessages</span>({</span>
<span id="cb52-159"><a href="#cb52-159" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mapIds</span>(<span class="fu">get</span>(annotation), <span class="at">keys =</span> .data<span class="sc">$</span>ensembl, </span>
<span id="cb52-160"><a href="#cb52-160" aria-hidden="true" tabindex="-1"></a>               <span class="at">column =</span> <span class="st">"ENTREZID"</span>, <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb52-161"><a href="#cb52-161" aria-hidden="true" tabindex="-1"></a>               <span class="at">multiVals =</span> <span class="st">"first"</span>)</span>
<span id="cb52-162"><a href="#cb52-162" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb52-163"><a href="#cb52-163" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb52-164"><a href="#cb52-164" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">any_of</span>(<span class="st">"spikein"</span>))</span>
<span id="cb52-165"><a href="#cb52-165" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-166"><a href="#cb52-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-167"><a href="#cb52-167" aria-hidden="true" tabindex="-1"></a>tidy <span class="ot">&lt;-</span> <span class="fu">memoise</span>(.tidy)</span>
<span id="cb52-168"><a href="#cb52-168" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-169"><a href="#cb52-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-170"><a href="#cb52-170" aria-hidden="true" tabindex="-1"></a><span class="fu">## Write parquet files</span></span>
<span id="cb52-171"><a href="#cb52-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-172"><a href="#cb52-172" aria-hidden="true" tabindex="-1"></a>Now we extract the gene expression measurements from each of our </span>
<span id="cb52-173"><a href="#cb52-173" aria-hidden="true" tabindex="-1"></a><span class="in">`SummarizedExperiment`</span> objects and then write it to a parquet file in a </span>
<span id="cb52-174"><a href="#cb52-174" aria-hidden="true" tabindex="-1"></a>temporary directory on the local file system. Alternatively, I could store</span>
<span id="cb52-175"><a href="#cb52-175" aria-hidden="true" tabindex="-1"></a>the files on a cloud system, e.g. AWS S3, and access them remotely.</span>
<span id="cb52-176"><a href="#cb52-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-179"><a href="#cb52-179" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-180"><a href="#cb52-180" aria-hidden="true" tabindex="-1"></a><span class="co"># create a temporary directory</span></span>
<span id="cb52-181"><a href="#cb52-181" aria-hidden="true" tabindex="-1"></a>out_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"parquet"</span>)</span>
<span id="cb52-182"><a href="#cb52-182" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(out_dir)</span>
<span id="cb52-183"><a href="#cb52-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-184"><a href="#cb52-184" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (dataset <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"rnai"</span>, <span class="st">"sarm1"</span>)) {</span>
<span id="cb52-185"><a href="#cb52-185" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">tidy</span>(<span class="fu">get</span>(dataset))</span>
<span id="cb52-186"><a href="#cb52-186" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>study <span class="ot">&lt;-</span> dataset  <span class="co"># add a columns with the name of the experiment</span></span>
<span id="cb52-187"><a href="#cb52-187" aria-hidden="true" tabindex="-1"></a>  arrow<span class="sc">::</span><span class="fu">write_parquet</span>(</span>
<span id="cb52-188"><a href="#cb52-188" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> df, </span>
<span id="cb52-189"><a href="#cb52-189" aria-hidden="true" tabindex="-1"></a>    <span class="at">sink =</span> <span class="fu">file.path</span>(out_dir, <span class="fu">paste0</span>(dataset, <span class="st">".parquet"</span>))</span>
<span id="cb52-190"><a href="#cb52-190" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb52-191"><a href="#cb52-191" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-192"><a href="#cb52-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-193"><a href="#cb52-193" aria-hidden="true" tabindex="-1"></a><span class="co"># list the contents of the temporary directory</span></span>
<span id="cb52-194"><a href="#cb52-194" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_info</span>(out_dir) <span class="sc">%&gt;%</span></span>
<span id="cb52-195"><a href="#cb52-195" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(path, size) <span class="sc">%&gt;%</span></span>
<span id="cb52-196"><a href="#cb52-196" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">path =</span> <span class="fu">basename</span>(path))</span>
<span id="cb52-197"><a href="#cb52-197" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-198"><a href="#cb52-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-199"><a href="#cb52-199" aria-hidden="true" tabindex="-1"></a><span class="fu">## Querying</span></span>
<span id="cb52-200"><a href="#cb52-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-201"><a href="#cb52-201" aria-hidden="true" tabindex="-1"></a><span class="fu">### Arrow</span></span>
<span id="cb52-202"><a href="#cb52-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-203"><a href="#cb52-203" aria-hidden="true" tabindex="-1"></a>Even though we created three separate files, the</span>
<span id="cb52-204"><a href="#cb52-204" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">arrow R package</span><span class="co">](https://arrow.apache.org/docs/r/)</span></span>
<span id="cb52-205"><a href="#cb52-205" aria-hidden="true" tabindex="-1"></a>can abstract them into a single <span class="in">`Dataset`</span>. We simply point the <span class="in">`open_dataset()`</span></span>
<span id="cb52-206"><a href="#cb52-206" aria-hidden="true" tabindex="-1"></a>function at the directory containing the <span class="in">`.parquet`</span> files.</span>
<span id="cb52-207"><a href="#cb52-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-210"><a href="#cb52-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-211"><a href="#cb52-211" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> arrow<span class="sc">::</span><span class="fu">open_dataset</span>(out_dir)</span>
<span id="cb52-212"><a href="#cb52-212" aria-hidden="true" tabindex="-1"></a>ds</span>
<span id="cb52-213"><a href="#cb52-213" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-214"><a href="#cb52-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-215"><a href="#cb52-215" aria-hidden="true" tabindex="-1"></a>We can use <span class="in">`dbplr`</span> verbs to query this <span class="in">`FileSystemDataset`</span>:</span>
<span id="cb52-216"><a href="#cb52-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-219"><a href="#cb52-219" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-220"><a href="#cb52-220" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb52-221"><a href="#cb52-221" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">"arrow"</span>)</span>
<span id="cb52-222"><a href="#cb52-222" aria-hidden="true" tabindex="-1"></a>ds <span class="sc">%&gt;%</span></span>
<span id="cb52-223"><a href="#cb52-223" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb52-224"><a href="#cb52-224" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb52-225"><a href="#cb52-225" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb52-226"><a href="#cb52-226" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-227"><a href="#cb52-227" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-228"><a href="#cb52-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-229"><a href="#cb52-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-230"><a href="#cb52-230" aria-hidden="true" tabindex="-1"></a>On my system, it takes ~ 1 second to retrieve the results, and about the</span>
<span id="cb52-231"><a href="#cb52-231" aria-hidden="true" tabindex="-1"></a>same amount of time is needed to retrieve the full <span class="in">`rnai`</span> dataset:</span>
<span id="cb52-232"><a href="#cb52-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-235"><a href="#cb52-235" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-236"><a href="#cb52-236" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb52-237"><a href="#cb52-237" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-238"><a href="#cb52-238" aria-hidden="true" tabindex="-1"></a>ds <span class="sc">%&gt;%</span></span>
<span id="cb52-239"><a href="#cb52-239" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-240"><a href="#cb52-240" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-241"><a href="#cb52-241" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-242"><a href="#cb52-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-243"><a href="#cb52-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-244"><a href="#cb52-244" aria-hidden="true" tabindex="-1"></a><span class="fu">### Duckdb</span></span>
<span id="cb52-245"><a href="#cb52-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-246"><a href="#cb52-246" aria-hidden="true" tabindex="-1"></a>Alternatively, we can use </span>
<span id="cb52-247"><a href="#cb52-247" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">duckdb</span><span class="co">](https://duckdb.org/)</span></span>
<span id="cb52-248"><a href="#cb52-248" aria-hidden="true" tabindex="-1"></a>to query our parquet files. The</span>
<span id="cb52-249"><a href="#cb52-249" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">duckdb R API</span><span class="co">](https://duckdb.org/docs/api/r)</span></span>
<span id="cb52-250"><a href="#cb52-250" aria-hidden="true" tabindex="-1"></a>supports both dbplyr verbs and raw SQL queries.</span>
<span id="cb52-251"><a href="#cb52-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-252"><a href="#cb52-252" aria-hidden="true" tabindex="-1"></a>First, we establish a connection to the <span class="in">`duckdb`</span> backend:</span>
<span id="cb52-253"><a href="#cb52-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-256"><a href="#cb52-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-257"><a href="#cb52-257" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span>
<span id="cb52-258"><a href="#cb52-258" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-259"><a href="#cb52-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-260"><a href="#cb52-260" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Using dbplyr</span></span>
<span id="cb52-261"><a href="#cb52-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-262"><a href="#cb52-262" aria-hidden="true" tabindex="-1"></a>First, we execute the same dbplyr query we used above, which translates it</span>
<span id="cb52-263"><a href="#cb52-263" aria-hidden="true" tabindex="-1"></a>into SQL for us and passes it on to duckdb.</span>
<span id="cb52-264"><a href="#cb52-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-267"><a href="#cb52-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-268"><a href="#cb52-268" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">"duckdb"</span>)</span>
<span id="cb52-269"><a href="#cb52-269" aria-hidden="true" tabindex="-1"></a><span class="fu">tbl</span>(con, <span class="fu">sprintf</span>(<span class="st">"read_parquet('%s/*.parquet')"</span>, out_dir)) <span class="sc">%&gt;%</span></span>
<span id="cb52-270"><a href="#cb52-270" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb52-271"><a href="#cb52-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb52-272"><a href="#cb52-272" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb52-273"><a href="#cb52-273" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-274"><a href="#cb52-274" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-275"><a href="#cb52-275" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-276"><a href="#cb52-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-277"><a href="#cb52-277" aria-hidden="true" tabindex="-1"></a>On my system, <span class="in">`duckdb`</span> returns results more than 10x faster than arrow's</span>
<span id="cb52-278"><a href="#cb52-278" aria-hidden="true" tabindex="-1"></a>implementation (see above).</span>
<span id="cb52-279"><a href="#cb52-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-280"><a href="#cb52-280" aria-hidden="true" tabindex="-1"></a>Retrieval of the full dataset for the <span class="in">`rnai`</span> study is also completed in less</span>
<span id="cb52-281"><a href="#cb52-281" aria-hidden="true" tabindex="-1"></a>than half a second:</span>
<span id="cb52-282"><a href="#cb52-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-285"><a href="#cb52-285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-286"><a href="#cb52-286" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-287"><a href="#cb52-287" aria-hidden="true" tabindex="-1"></a><span class="fu">tbl</span>(con, <span class="fu">glue</span>(<span class="st">"read_parquet('{out_dir}/*.parquet')"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb52-288"><a href="#cb52-288" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-289"><a href="#cb52-289" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-290"><a href="#cb52-290" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-291"><a href="#cb52-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-292"><a href="#cb52-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-293"><a href="#cb52-293" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Using SQL</span></span>
<span id="cb52-294"><a href="#cb52-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-295"><a href="#cb52-295" aria-hidden="true" tabindex="-1"></a>Because <span class="in">`duckdb`</span> is a SQL database, we can also query the parquet files directly</span>
<span id="cb52-296"><a href="#cb52-296" aria-hidden="true" tabindex="-1"></a>with raw SQL, realizing another gain in execution speed:</span>
<span id="cb52-297"><a href="#cb52-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-300"><a href="#cb52-300" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-301"><a href="#cb52-301" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-302"><a href="#cb52-302" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb52-303"><a href="#cb52-303" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> con,</span>
<span id="cb52-304"><a href="#cb52-304" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue_sql</span>(</span>
<span id="cb52-305"><a href="#cb52-305" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT symbol, study, COUNT(*) AS n </span></span>
<span id="cb52-306"><a href="#cb52-306" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM read_parquet({paste0(out_dir, '/*.parquet')}) </span></span>
<span id="cb52-307"><a href="#cb52-307" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE UPPER(symbol) = 'GAPDH' </span></span>
<span id="cb52-308"><a href="#cb52-308" aria-hidden="true" tabindex="-1"></a><span class="st">     GROUP BY symbol, study"</span>, </span>
<span id="cb52-309"><a href="#cb52-309" aria-hidden="true" tabindex="-1"></a>    <span class="at">.con =</span> con)</span>
<span id="cb52-310"><a href="#cb52-310" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb52-311"><a href="#cb52-311" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-312"><a href="#cb52-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-313"><a href="#cb52-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-314"><a href="#cb52-314" aria-hidden="true" tabindex="-1"></a>Similarly, reading all data for the <span class="in">`rnai`</span> dataset into memory is faster than</span>
<span id="cb52-315"><a href="#cb52-315" aria-hidden="true" tabindex="-1"></a>with arrow's implementation as well:</span>
<span id="cb52-316"><a href="#cb52-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-319"><a href="#cb52-319" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-320"><a href="#cb52-320" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-321"><a href="#cb52-321" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb52-322"><a href="#cb52-322" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> con,</span>
<span id="cb52-323"><a href="#cb52-323" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue_sql</span>(</span>
<span id="cb52-324"><a href="#cb52-324" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT * </span></span>
<span id="cb52-325"><a href="#cb52-325" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM read_parquet({paste0(out_dir, '/*.parquet')}) </span></span>
<span id="cb52-326"><a href="#cb52-326" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE study = 'rnai'"</span>, </span>
<span id="cb52-327"><a href="#cb52-327" aria-hidden="true" tabindex="-1"></a>    <span class="at">.con =</span> con)</span>
<span id="cb52-328"><a href="#cb52-328" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb52-329"><a href="#cb52-329" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span>
<span id="cb52-330"><a href="#cb52-330" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-331"><a href="#cb52-331" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-332"><a href="#cb52-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-333"><a href="#cb52-333" aria-hidden="true" tabindex="-1"></a><span class="fu">## Spark</span></span>
<span id="cb52-334"><a href="#cb52-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-335"><a href="#cb52-335" aria-hidden="true" tabindex="-1"></a>Finally, and mainly just for future reference, the</span>
<span id="cb52-336"><a href="#cb52-336" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">sparklyr R package</span><span class="co">](https://spark.rstudio.com/)</span></span>
<span id="cb52-337"><a href="#cb52-337" aria-hidden="true" tabindex="-1"></a>provides an R interface to leverage a Spark cluster and its</span>
<span id="cb52-338"><a href="#cb52-338" aria-hidden="true" tabindex="-1"></a>distributed analysis libraries.</span>
<span id="cb52-339"><a href="#cb52-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-340"><a href="#cb52-340" aria-hidden="true" tabindex="-1"></a>First, we establish a connection to the Spark cluster. Here am running a local</span>
<span id="cb52-341"><a href="#cb52-341" aria-hidden="true" tabindex="-1"></a>Spark cluster (e.g. a single local node) on my laptop:</span>
<span id="cb52-342"><a href="#cb52-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-345"><a href="#cb52-345" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-346"><a href="#cb52-346" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">spark_connect</span>(<span class="at">master =</span> <span class="st">"local"</span>)</span>
<span id="cb52-347"><a href="#cb52-347" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-348"><a href="#cb52-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-349"><a href="#cb52-349" aria-hidden="true" tabindex="-1"></a>Next, we import the datasets into the cluster by creating a new </span>
<span id="cb52-350"><a href="#cb52-350" aria-hidden="true" tabindex="-1"></a><span class="in">`Spark DataFrame`</span> with the name <span class="in">`gene_expression`</span>. </span>
<span id="cb52-351"><a href="#cb52-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-354"><a href="#cb52-354" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-355"><a href="#cb52-355" aria-hidden="true" tabindex="-1"></a>sdf <span class="ot">&lt;-</span> <span class="fu">spark_read_parquet</span>(<span class="at">sc =</span> sc, <span class="at">name =</span> <span class="st">"gene_expression"</span>,</span>
<span id="cb52-356"><a href="#cb52-356" aria-hidden="true" tabindex="-1"></a>                          <span class="at">path =</span> <span class="fu">paste0</span>(out_dir, <span class="st">'/*.parquet'</span>))</span>
<span id="cb52-357"><a href="#cb52-357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-358"><a href="#cb52-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-359"><a href="#cb52-359" aria-hidden="true" tabindex="-1"></a><span class="fu">### dbplyr</span></span>
<span id="cb52-360"><a href="#cb52-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-361"><a href="#cb52-361" aria-hidden="true" tabindex="-1"></a>The <span class="in">`tbl_spark`</span> object returned by <span class="in">`sparklyr::spark_read_parquet()`</span></span>
<span id="cb52-362"><a href="#cb52-362" aria-hidden="true" tabindex="-1"></a>can be queried with <span class="in">`dbplyr`</span> verbs, e.g. to translate our now familiar queries</span>
<span id="cb52-363"><a href="#cb52-363" aria-hidden="true" tabindex="-1"></a>into <span class="in">`Spark SQL`</span> statements on the fly:</span>
<span id="cb52-364"><a href="#cb52-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-367"><a href="#cb52-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-368"><a href="#cb52-368" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-369"><a href="#cb52-369" aria-hidden="true" tabindex="-1"></a>sdf <span class="sc">%&gt;%</span></span>
<span id="cb52-370"><a href="#cb52-370" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>, <span class="st">"Gapdh"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb52-371"><a href="#cb52-371" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(symbol, study) <span class="sc">%&gt;%</span></span>
<span id="cb52-372"><a href="#cb52-372" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb52-373"><a href="#cb52-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-374"><a href="#cb52-374" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-375"><a href="#cb52-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-376"><a href="#cb52-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-379"><a href="#cb52-379" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-380"><a href="#cb52-380" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-381"><a href="#cb52-381" aria-hidden="true" tabindex="-1"></a>sdf <span class="sc">%&gt;%</span></span>
<span id="cb52-382"><a href="#cb52-382" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">"rnai"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-383"><a href="#cb52-383" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb52-384"><a href="#cb52-384" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-385"><a href="#cb52-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-386"><a href="#cb52-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-387"><a href="#cb52-387" aria-hidden="true" tabindex="-1"></a><span class="fu">### SQL</span></span>
<span id="cb52-388"><a href="#cb52-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-389"><a href="#cb52-389" aria-hidden="true" tabindex="-1"></a>Alternatively, we can use SQL queries to query the cluster's <span class="in">`gene_expression`</span></span>
<span id="cb52-390"><a href="#cb52-390" aria-hidden="true" tabindex="-1"></a>table directly:</span>
<span id="cb52-391"><a href="#cb52-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-394"><a href="#cb52-394" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-395"><a href="#cb52-395" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>()</span>
<span id="cb52-396"><a href="#cb52-396" aria-hidden="true" tabindex="-1"></a><span class="fu">dbGetQuery</span>(</span>
<span id="cb52-397"><a href="#cb52-397" aria-hidden="true" tabindex="-1"></a>  <span class="at">con =</span> sc,</span>
<span id="cb52-398"><a href="#cb52-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glue</span>(</span>
<span id="cb52-399"><a href="#cb52-399" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SELECT symbol, study, COUNT(*) AS n </span></span>
<span id="cb52-400"><a href="#cb52-400" aria-hidden="true" tabindex="-1"></a><span class="st">     FROM gene_expression </span></span>
<span id="cb52-401"><a href="#cb52-401" aria-hidden="true" tabindex="-1"></a><span class="st">     WHERE UPPER(symbol) = 'GAPDH' </span></span>
<span id="cb52-402"><a href="#cb52-402" aria-hidden="true" tabindex="-1"></a><span class="st">     GROUP BY symbol, study"</span>)</span>
<span id="cb52-403"><a href="#cb52-403" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb52-404"><a href="#cb52-404" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb52-405"><a href="#cb52-405" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-406"><a href="#cb52-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-407"><a href="#cb52-407" aria-hidden="true" tabindex="-1"></a>My local Spark instance performs these queries more slowly than e.g. duckdb.</span>
<span id="cb52-408"><a href="#cb52-408" aria-hidden="true" tabindex="-1"></a>But Spark's _real power_ is in deploying </span>
<span id="cb52-409"><a href="#cb52-409" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Machine learning models</span><span class="co">](https://spark.rstudio.com/guides/mlib.html)</span></span>
<span id="cb52-410"><a href="#cb52-410" aria-hidden="true" tabindex="-1"></a>across a (potentially large) cluster, enabling parallel processing of very</span>
<span id="cb52-411"><a href="#cb52-411" aria-hidden="true" tabindex="-1"></a>large datasets by distributing both data and computation across nodes.</span>
<span id="cb52-412"><a href="#cb52-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-415"><a href="#cb52-415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-416"><a href="#cb52-416" aria-hidden="true" tabindex="-1"></a><span class="fu">spark_disconnect</span>(sc)</span>
<span id="cb52-417"><a href="#cb52-417" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-418"><a href="#cb52-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-419"><a href="#cb52-419" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reproducibility</span></span>
<span id="cb52-420"><a href="#cb52-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-421"><a href="#cb52-421" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb52-422"><a href="#cb52-422" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;</span></span>
<span id="cb52-423"><a href="#cb52-423" aria-hidden="true" tabindex="-1"></a>Session Information</span>
<span id="cb52-424"><a href="#cb52-424" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/summary&gt;</span></span>
<span id="cb52-425"><a href="#cb52-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-428"><a href="#cb52-428" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-429"><a href="#cb52-429" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="st">"attached"</span>)</span>
<span id="cb52-430"><a href="#cb52-430" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-431"><a href="#cb52-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-432"><a href="#cb52-432" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb52-433"><a href="#cb52-433" aria-hidden="true" tabindex="-1"></a>   </span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>