<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-01-16">

<title>Thomas Sandmann’s blog - QuantSeq RNAseq analysis (3): Validating published results (no UMIs)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">QuantSeq RNAseq analysis (3): Validating published results (no UMIs)</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">R</div>
                <div class="quarto-category">nextflow</div>
                <div class="quarto-category">NGS</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 16, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#sample-annotations" id="toc-sample-annotations" class="nav-link" data-scroll-target="#sample-annotations">Sample annotations</a></li>
  <li><a href="#xia-et-als-original-count-data" id="toc-xia-et-als-original-count-data" class="nav-link" data-scroll-target="#xia-et-als-original-count-data">Xia et al’s original count data</a>
  <ul>
  <li><a href="#linear-modeling-with-limmavoom" id="toc-linear-modeling-with-limmavoom" class="nav-link" data-scroll-target="#linear-modeling-with-limmavoom">Linear modeling with limma/voom</a></li>
  </ul></li>
  <li><a href="#nf-corernaseq-results" id="toc-nf-corernaseq-results" class="nav-link" data-scroll-target="#nf-corernaseq-results">nf-core/rnaseq results</a></li>
  <li><a href="#comparing-results-across-preprocessing-workflows" id="toc-comparing-results-across-preprocessing-workflows" class="nav-link" data-scroll-target="#comparing-results-across-preprocessing-workflows">Comparing results across preprocessing workflows</a>
  <ul>
  <li><a href="#normalized-expression" id="toc-normalized-expression" class="nav-link" data-scroll-target="#normalized-expression">Normalized expression</a></li>
  <li><a href="#differential-expression-results" id="toc-differential-expression-results" class="nav-link" data-scroll-target="#differential-expression-results">Differential expression results</a>
  <ul class="collapse">
  <li><a href="#discordant-significance-calls" id="toc-discordant-significance-calls" class="nav-link" data-scroll-target="#discordant-significance-calls">Discordant significance calls</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions">Conclusions</a></li>
  <li><a href="#reproducibility" id="toc-reproducibility" class="nav-link" data-scroll-target="#reproducibility">Reproducibility</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>This is the third of four posts documenting my progress toward processing and analyzing <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq FWD 3’ tag RNAseq</a> data with the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq</a> workflow.</p>
<ol type="1">
<li><a href="../nextflow-core-quantseq-1-settings/index.html">Configuring &amp; executing the nf-core/rnaseq workflow</a></li>
<li><a href="../nextflow-core-quantseq-2-output/index.html">Exploring the workflow outputs</a></li>
<li><strong>Validating the workflow by reproducing results published by Xia et al (no UMIs)</strong></li>
<li><a href="../nextflow-core-quantseq-4-nugent/index.html">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</a></li>
</ol>
<p>Many thanks to <a href="https://github.com/drpatelh">Harshil Patel</a>, <a href="https://github.com/adomingues/">António Miguel de Jesus Domingues</a> and <a href="https://github.com/matthiasZepper/">Matthias Zepper</a> for their generous guidance &amp; input via <a href="nfcore.slack.com">nf-core slack</a>. (Any mistakes are mine.)</p>
</div>
</div>
</div>
<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<ul>
<li>This analysis compares the performance of the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq workflow</a> for QuantSeq FWD 3’ tag RNAseq data without unique molecular identifiers.</li>
<li>The differential expression analysis results are highly concordant with those obtained in the original publication.</li>
<li>With the appropriate settings, the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq workflow</a> is a valid data processing pipeline for this data type.</li>
</ul>
<p>The <a href="../nextflow-core-quantseq-1-settings/index.html">first post in this series</a> walked through the preprocesssing of QuantSeq FWD data published in a preprint by <a href="https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1">Xia et al</a>.</p>
<p>Next, we use Bioconductor/R packages to reproduce the downstream results. We perform the same differential gene expression analysis twice with either</p>
<ol type="1">
<li>the original counts matrix published by the authors <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></li>
<li>the output of the nf-core/rnaseq workflow</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sample-annotations" class="level2">
<h2 class="anchored" data-anchor-id="sample-annotations">Sample annotations</h2>
<p>We start by retrieving the sample annotation table, listing e.g.&nbsp;the <code>sex</code>, and <code>genotype</code> for each mouse, and the <code>batch</code> for each collected sample.</p>
<p>This information is available in the <a href="https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=1&amp;WebEnv=MCID_63c0392862acb114c36e44a3&amp;o=acc_s%3Aa#">SRA Run Explorer</a>. (I saved it as the <a href="sample_metadata.csv">sample_metadata.csv CSV file</a> in case you want to follow along.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>sample_sheet <span class="ot">&lt;-</span> <span class="fu">file.path</span>(work_dir, <span class="st">"sample_metadata.csv"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_sheet, <span class="at">row.names =</span> <span class="st">"Experiment"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno[, <span class="fu">c</span>(<span class="st">"Run"</span>, <span class="st">"Animal.ID"</span>, <span class="st">"Age"</span>, <span class="st">"age_unit"</span>, <span class="st">"Batch"</span>, <span class="st">"sex"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Genotype"</span>, <span class="st">"Sample.Name"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   Run Animal.ID Age age_unit Batch  sex Genotype Sample.Name
SRX9142648 SRR12661924       LA1   8   months  Day1 male       WT  GSM4793335
SRX9142649 SRR12661925       LA1   8   months  Day1 male       WT  GSM4793336
SRX9142650 SRR12661926       LA6   8   months  Day1 male       WT  GSM4793337
SRX9142651 SRR12661927       LA6   8   months  Day1 male       WT  GSM4793338
SRX9142652 SRR12661928       LA9   8   months  Day2 male       WT  GSM4793339
SRX9142653 SRR12661929       LA9   8   months  Day2 male       WT  GSM4793340</code></pre>
</div>
</div>
<p>Because our SRA metadata doesn’t include the GEO sample title, I saved the identifier mappings in the <a href="GEO_sample_ids.csv">GEO_sample_ids.csv CSV file</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>geo_ids <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">file.path</span>(work_dir, <span class="st">"GEO_sample_ids.csv"</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geo_ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  sample_name sample_id
1  GSM4793335 DRN-18429
2  GSM4793336 DRN-18430
3  GSM4793337 DRN-18439
4  GSM4793338 DRN-18440
5  GSM4793339 DRN-18445
6  GSM4793340 DRN-18446</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_anno)<span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">colnames</span>(sample_anno))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_anno) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"."</span>, <span class="st">"_"</span>, <span class="fu">colnames</span>(sample_anno), </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">fixed =</span> <span class="cn">TRUE</span>) </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> sample_anno[, <span class="fu">c</span>(<span class="st">"sample_name"</span>, <span class="st">"animal_id"</span>, <span class="st">"genotype"</span>, <span class="st">"sex"</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"batch"</span>)]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">factor</span>(sample_anno<span class="sc">$</span>genotype, </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                               <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"Het"</span>, <span class="st">"Hom"</span>))</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>sample_title <span class="ot">&lt;-</span> geo_ids[</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">match</span>(sample_anno<span class="sc">$</span>sample_name, geo_ids<span class="sc">$</span>sample_name), <span class="st">"sample_id"</span>]</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>           sample_name animal_id genotype  sex batch sample_title
SRX9142648  GSM4793335       LA1       WT male  Day1    DRN-18429
SRX9142649  GSM4793336       LA1       WT male  Day1    DRN-18430
SRX9142650  GSM4793337       LA6       WT male  Day1    DRN-18439
SRX9142651  GSM4793338       LA6       WT male  Day1    DRN-18440
SRX9142652  GSM4793339       LA9       WT male  Day2    DRN-18445
SRX9142653  GSM4793340       LA9       WT male  Day2    DRN-18446</code></pre>
</div>
</div>
<p>This experiment includes 36 samples of microglia cells obtained from 18 different 8-month old mice. Both male and female animals were included in the study.</p>
<p>The animals carry one of three different genotypes of the gene encoding the APP amyloid beta precursor protein, either</p>
<ul>
<li>the wildtype mouse gene (<code>WT</code>) or</li>
<li>one copy (<code>Het</code>) or</li>
<li>two copies (<code>Hom</code>)</li>
</ul>
<p>of a mutant <em>APP</em> gene carrying mutations associated with familial Alzheimer’s Disease.</p>
<p>Samples from all three genotypes were collected on three days, and we will use this <code>batch</code> information to model the experiment.</p>
<p>Two separate microglia samples were obtained from each animal, and we will include this nested relationship by modeling the <code>animal</code> as random effect in our linear model.</p>
</section>
<section id="xia-et-als-original-count-data" class="level2">
<h2 class="anchored" data-anchor-id="xia-et-als-original-count-data">Xia et al’s original count data</h2>
<p>First, we retrieve the authors’ count matrix from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158152">NCBI GEO</a>, available as a Supplementary Excel file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE158152&amp;"</span>,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>              <span class="st">"format=file&amp;file=GSE158152%5Fdst150%5Fprocessed%2Exlsx"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>temp_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, <span class="at">destfile =</span> temp_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The Excel file has three different worksheets</p>
<ul>
<li><code>sample_annotations</code></li>
<li><code>raw_counts</code></li>
<li><code>normalized_cpm</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(temp_file, <span class="at">sheet =</span> <span class="st">"raw_counts"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(raw_counts), <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "feature_id" "name"       "meta"       "source"     "symbol"    
 [6] "DRN-18429"  "DRN-18430"  "DRN-18439"  "DRN-18440"  "DRN-18445" </code></pre>
</div>
</div>
<p>The <code>raw_counts</code> excel sheet contains information about the detected genes ( <code>feature_ID</code>, <code>name</code>) and the samples are identified by their GEO title (e.g.&nbsp;<code>DRN-18459</code>, <code>DRN-184560</code>). We use the raw counts to populate a new <code>DGEList</code> object and perform Library Size Normalization with the <code>TMM</code> approach.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>count_data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(raw_counts[, <span class="fu">grep</span>(<span class="st">"DRN-"</span>, <span class="fu">colnames</span>(raw_counts))])</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(count_data) <span class="ot">&lt;-</span> raw_counts<span class="sc">$</span>feature_id</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_data) <span class="ot">&lt;-</span> <span class="fu">row.names</span>(sample_anno)[</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">match</span>(<span class="fu">colnames</span>(count_data), sample_anno<span class="sc">$</span>sample_title)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> raw_counts<span class="sc">$</span>feature_id,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_name =</span> raw_counts<span class="sc">$</span>symbol,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> raw_counts<span class="sc">$</span>feature_id</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>col_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  sample_anno[<span class="fu">colnames</span>(count_data),</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>              <span class="fu">c</span>(<span class="st">"sample_title"</span>, <span class="st">"animal_id"</span>, <span class="st">"sex"</span>, <span class="st">"genotype"</span>, <span class="st">"batch"</span>)],</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow =</span> <span class="st">"geo"</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> <span class="fu">as.matrix</span>(count_data), </span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">samples =</span> col_data[<span class="fu">colnames</span>(count_data), ], </span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes =</span> gene_data[<span class="fu">row.names</span>(count_data), ]</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Next, we project the samples into two dimensions by performing multi-dimensional scaling of the top 500 most variable genes. The samples cluster by genotype, with <code>WT</code> and <code>Het</code> segregating from the <code>Hom</code> samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge, <span class="at">labels =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>genotype,</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Multi-dimensional scaling"</span>, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">sub =</span> <span class="st">"Based on the top 500 most variable genes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="528"></p>
</div>
</div>
<p>Let’s identify which genes are significantly differentially expressed between the three genotypes!</p>
<section id="linear-modeling-with-limmavoom" class="level3">
<h3 class="anchored" data-anchor-id="linear-modeling-with-limmavoom">Linear modeling with limma/voom</h3>
<p>First, we use the <code>edgeR::filterByExpr()</code> function to identify genes with sufficiently large counts to be examined for differential expression. (The <code>min.count = 25</code> parameter was determined by examining the mean-variance plot by the <code>voomLmFit()</code> function.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype <span class="sc">+</span> sex <span class="sc">+</span> batch, <span class="at">data =</span> dge<span class="sc">$</span>samples)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"genotype"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(design))</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">design =</span> design, <span class="at">min.count =</span> <span class="dv">25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we fit a linear model to the data using the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">limma/voom</a> approach. The model includes the following fixed effects:</p>
<ul>
<li>The <code>genotype</code> coded as a factor with the <code>WT</code> as the reference level.</li>
<li>The <code>sex</code> and <code>batch</code> covariates, to account for systematic differences in mean gene expression.</li>
</ul>
<p>Because the dataset included two replicate samples from each animal, we model the <code>animal</code> as a random effect (via the <code>block</code> argument of the <code>voomLmFit()</code> function). We then extract the coefficients, log2 fold changes and p-values via limma’s empirical Bayes approach.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We use the <code>limma::treat()</code> function to test the null hypothesis that genes display significant differential expression <em>greater than</em> 1.2-fold. This is more stringent than the conventional null hypothesis of zero change. (Please consult the <code>limma::treat()</code> <a href="https://rdrr.io/bioc/limma/man/ebayes.html">help page</a> for details.)</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">voomLmFit</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  dge[keep, <span class="fu">row.names</span>(design)], </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> design,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">block =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>animal_id, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.weights =</span> <span class="cn">TRUE</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">FALSE</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">treat</span>(fit, <span class="at">robust=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following table displays the number of differentially up- and down-regulated genes after applying a false-discovery (<code>adj.P.Val</code>) threshold of 5%. While we did not detect significant differences between <code>Het</code> and <code>WT</code> animals, the analysis revealed &gt; 450 differentially expressed genes between <code>Hom</code> and <code>WT</code> microglia.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(fit))[, <span class="fu">c</span>(<span class="st">"Het"</span>, <span class="st">"Hom"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Het   Hom
Down       0    73
NotSig 10131  9660
Up         0   398</code></pre>
</div>
</div>
<p>The top 10 genes with the smallest p-values include well known markers of microglia activation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>)[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   gene_name     logFC      P.Value    adj.P.Val
ENSMUSG00000027523      Gnas 1.5904760 3.467080e-24 3.512499e-20
ENSMUSG00000022265       Ank 3.7866023 1.258687e-20 6.375881e-17
ENSMUSG00000021477      Ctsl 1.2829086 2.294556e-20 7.748717e-17
ENSMUSG00000018927      Ccl6 2.7405831 3.514945e-20 8.902477e-17
ENSMUSG00000030579    Tyrobp 1.2880854 9.748473e-19 1.975236e-15
ENSMUSG00000022415    Syngr1 1.5033949 1.367158e-18 2.308446e-15
ENSMUSG00000016256      Ctsz 1.1557870 2.878895e-18 4.166584e-15
ENSMUSG00000023992     Trem2 0.9460719 7.386805e-18 9.354465e-15
ENSMUSG00000056737      Capg 2.4537482 1.927765e-17 2.170021e-14
ENSMUSG00000030342       Cd9 1.0596747 1.328189e-16 1.328093e-13</code></pre>
</div>
</div>
<p>Next we repeat <em>the same analysis</em> with the output of the nf-core/rnaseq workflow.</p>
</section>
</section>
<section id="nf-corernaseq-results" class="level2">
<h2 class="anchored" data-anchor-id="nf-corernaseq-results">nf-core/rnaseq results</h2>
<p>We start with the raw counts contained in the <a href="salmon.merged.gene_counts.rds">salmon.merged.gene_counts.rds</a> file generated by the nf-core/rnaseq workflow.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The nf-core pipeline returned the versioned ENSEMBL gene identifiers (e.g.) <code>ENSMUSG00000000001.4</code>. Because Xia et al only provided the unversioned identifiers (e.g.&nbsp;<code>ENSMUSG00000000001</code>) we trim the numeric suffix.</p>
</div>
</div>
</div>
<p>We TMM-normalize the data, as before. (This step converts the <code>SummarizedExperiment</code> into a <code>DGEList</code> object as well.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>count_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(work_dir, <span class="st">"salmon.merged.gene_counts.rds"</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(count_file)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(se) <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">strsplit</span>(<span class="fu">row.names</span>(se), <span class="at">split =</span> <span class="st">"."</span>, <span class="at">fixed =</span> <span class="cn">TRUE</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="fu">colnames</span>(se) <span class="sc">%in%</span> <span class="fu">row.names</span>(sample_anno)))</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>dge_nfcore <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(se, <span class="at">method =</span> <span class="st">"TMM"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we add the sample metadata and fit the same linear model as before.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dge_nfcore<span class="sc">$</span>genes<span class="sc">$</span>gene_id <span class="ot">&lt;-</span> <span class="fu">row.names</span>(dge_nfcore)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>dge_nfcore<span class="sc">$</span>samples <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  dge_nfcore<span class="sc">$</span>samples,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  sample_anno[<span class="fu">colnames</span>(dge_nfcore),</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>              <span class="fu">c</span>(<span class="st">"sample_title"</span>, <span class="st">"animal_id"</span>, <span class="st">"sex"</span>, <span class="st">"genotype"</span>, <span class="st">"batch"</span>)],</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow =</span> <span class="st">"nfcore"</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="fu">colnames</span>(dge) <span class="sc">%in%</span> <span class="fu">colnames</span>(dge_nfcore)))</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>dge_nfcore <span class="ot">&lt;-</span> dge_nfcore[, <span class="fu">colnames</span>(dge)]</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype <span class="sc">+</span> sex <span class="sc">+</span> batch, <span class="at">data =</span> dge_nfcore<span class="sc">$</span>samples)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"genotype"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(design))</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge_nfcore, <span class="at">design =</span> design, <span class="at">min.count =</span> <span class="dv">25</span>)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>fit_nfcore <span class="ot">&lt;-</span> <span class="fu">voomLmFit</span>(</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  dge_nfcore[keep, <span class="fu">row.names</span>(design)], </span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> design,</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">block =</span> dge_nfcore<span class="sc">$</span>samples<span class="sc">$</span>animal_id, </span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.weights =</span> <span class="cn">TRUE</span>, </span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">FALSE</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>First sample weights (min/max) 0.5573482/2.1680260</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>First intra-block correlation  0.02339845</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Final sample weights (min/max) 0.5367708/2.2366196</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Final intra-block correlation  0.02354808</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>fit_nfcore <span class="ot">&lt;-</span> <span class="fu">treat</span>(fit_nfcore, <span class="at">robust=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>As with the original count data from NCBI GEO, we detect &gt; 450 differentially expressed genes between <code>Hom</code> and <code>WT</code> genotypes (FDR &lt; 5%, null hypothesis: fold change &gt; 1.2).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(fit_nfcore))[, <span class="fu">c</span>(<span class="st">"Het"</span>, <span class="st">"Hom"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Het   Hom
Down       0    76
NotSig 10428  9917
Up         0   435</code></pre>
</div>
</div>
</section>
<section id="comparing-results-across-preprocessing-workflows" class="level2">
<h2 class="anchored" data-anchor-id="comparing-results-across-preprocessing-workflows">Comparing results across preprocessing workflows</h2>
<p>Next, we compare the results obtained with the two datasets. We create the <code>cpms</code> and <code>tt</code> dataframes, holding the combined absolute and differential expression results, respectively.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cpms <span class="ot">&lt;-</span> <span class="fu">local</span>({</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  geo <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(dge<span class="sc">$</span>genes) <span class="sc">%&gt;%</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"SRX"</span>), </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"sample_name"</span>,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"cpm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(dge<span class="sc">$</span>samples, <span class="st">"sample_name"</span>),</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"sample_name"</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"Xia et al"</span>)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  nfcore <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge_nfcore, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(dge_nfcore<span class="sc">$</span>genes) <span class="sc">%&gt;%</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"SRX"</span>), </span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"sample_name"</span>,</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"cpm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(dge_nfcore<span class="sc">$</span>samples, <span class="st">"sample_name"</span>),</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"sample_name"</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"nf-core"</span>)</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(geo, <span class="fu">any_of</span>(<span class="fu">intersect</span>(<span class="fu">colnames</span>(geo), <span class="fu">colnames</span>(nfcore)))),</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(nfcore, <span class="fu">any_of</span>(<span class="fu">intersect</span>(<span class="fu">colnames</span>(geo), <span class="fu">colnames</span>(nfcore))))</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)[</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>    , <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"geo"</span>),</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>   <span class="fu">topTreat</span>(fit_nfcore, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)[</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>     , <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"nfcore"</span>)</span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">adj.P.Val =</span> <span class="fu">signif</span>(adj.P.Val, <span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>), </span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_from =</span> <span class="st">"dataset"</span>, </span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_from =</span> <span class="st">"adj.P.Val"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(nfcore) <span class="sc">%&gt;%</span></span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"gene_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="normalized-expression" class="level3">
<h3 class="anchored" data-anchor-id="normalized-expression">Normalized expression</h3>
<p>First, we examine the correlation between the normalized log-transformed gene expression estimates returned from the two workflows. We focus on those genes that passed the <code>filterByExpr</code> thresholds above, e.g.&nbsp;those genes deemed sufficiently highly expressed to be assessed for differential expression.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>common_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">row.names</span>(fit), <span class="fu">row.names</span>(fit_nfcore))</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>sum_stats <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">%in%</span> common_genes) <span class="sc">%&gt;%</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"sample_name"</span>),</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_from =</span> <span class="st">"cpm"</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_from =</span> <span class="st">"dataset"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(gene_id) <span class="sc">%&gt;%</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">r =</span> <span class="fu">cor</span>(<span class="fu">log1p</span>(<span class="st">`</span><span class="at">Xia et al</span><span class="st">`</span>), <span class="fu">log1p</span>(<span class="st">`</span><span class="at">nf-core</span><span class="st">`</span>)),</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_xia =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">Xia et al</span><span class="st">`</span>),</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_nfcore =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">nf-core</span><span class="st">`</span>))</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sum_stats, <span class="fu">aes</span>(<span class="at">x =</span> r)) <span class="sc">+</span> </span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.02</span>), <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.2</span>)) <span class="sc">+</span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Pearson correlation coefficient (R)"</span>, </span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Number of genes"</span>,</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"Correlation between normalized log2 counts"</span>) <span class="sc">+</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="528"></p>
</div>
</div>
<p>The correlation between normalized log2 expression estimates is very high, with 95% of all genes showing a Pearson correlation coefficient &gt; 0.94.</p>
<p>Most of the 10022 examined genes were detected with &gt; 10 normalized counts per million reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sum_stats, <span class="fu">aes</span>(<span class="at">x =</span> mean_xia <span class="sc">+</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log10_trans</span>(),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">comma_format</span>()) <span class="sc">+</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Mean normalized counts per million"</span>, </span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Number of genes"</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"Average expression"</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Xia et al"</span>) <span class="sc">+</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next, we will examine the results of the <em>differential</em> expression analysis.</p>
</section>
<section id="differential-expression-results" class="level3">
<h3 class="anchored" data-anchor-id="differential-expression-results">Differential expression results</h3>
<p>Analyses based on either preprocessing pipeline yield similar numbers of differentially expressed genes.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">decideTests</span>(fit)[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">decideTests</span>(fit_nfcore)[common_genes, <span class="st">"Hom"</span>]</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(results) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Xia et al"</span>, <span class="st">"nf-core"</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(results) <span class="ot">&lt;-</span> <span class="st">"TestResults"</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   Xia et al nf-core
-1        73      75
0       9554    9528
1        395     419</code></pre>
</div>
</div>
<p>But are these <em>the same genes</em> in both sets of results?</p>
<p>We can visualize the overlap between the sets of significant genes in a Venn diagram (FDR &lt; 5%). The vast majority of differentially expressed genes is detected with both quantitation approaches (for both up- and down-regulated genes.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">vennDiagram</span>(results, <span class="at">include =</span> <span class="fu">c</span>(<span class="st">"up"</span>, <span class="st">"down"</span>),</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">counts.col=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>), <span class="at">mar =</span> <span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="528"></p>
</div>
</div>
<p>For example, the following plots show the normalized expression of a few highly differentially expressed genes (known markers of active microglia).</p>
<div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> <span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="dv">6</span>)[[<span class="st">"gene_id"</span>]]) {</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a> p <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">==</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> genotype, <span class="at">y =</span> cpm)) <span class="sc">+</span> </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">width =</span> <span class="fl">0.05</span>), <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> dge<span class="sc">$</span>genes[gene, <span class="st">"gene_name"</span>],</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"Normalized expression (CPM)"</span>,</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"FDR nf-core: %s</span><span class="sc">\n</span><span class="st">FDR GEO: %s"</span>,</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"nfcore"</span>],</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"geo"</span>]</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>         ) <span class="sc">+</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(p)</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="408"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid" width="408"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-3.png" class="img-fluid" width="408"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-4.png" class="img-fluid" width="408"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-5.png" class="img-fluid" width="408"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-22-6.png" class="img-fluid" width="408"></p>
</div>
</div>
</div>
</div>
<p>Applying a hard FDR threshold can inflate the number of apparent differences, e.g.&nbsp;when a gene is close to the significance threshold (see below).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>p_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The log2 fold estimates for the <code>Hom</code> vs <code>WT</code> comparison are highly correlated across the two analysis workflows (Pearson correlation coefficient R = 0.99 ):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">smoothScatter</span>(</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>],</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"nf-core (log2FC)"</span>,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Xia et al (log2FC)"</span>,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Homozygous APP vs WT (effect size)"</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>, <span class="at">y =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">sprintf</span>(<span class="st">"R = %s"</span>, <span class="fu">signif</span>(p_cor, <span class="dv">2</span>)))</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">0</span>, <span class="at">v =</span> <span class="dv">0</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="432"></p>
</div>
</div>
<p>as are the t-statistics across all examined genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>p_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>])</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">smoothScatter</span>(</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>],</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"nf-core (t-statistic)"</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Xia et al (t-statistic)"</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Homozygous APP vs WT (t-statistic)"</span>)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>, <span class="at">y =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">sprintf</span>(<span class="st">"R = %s"</span>, <span class="fu">signif</span>(p_cor, <span class="dv">2</span>)))</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">0</span>, <span class="at">v =</span> <span class="dv">0</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="432"></p>
</div>
</div>
<section id="discordant-significance-calls" class="level4">
<h4 class="anchored" data-anchor-id="discordant-significance-calls">Discordant significance calls</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># genes detected in GEO, but not significant with nf-core</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">row.names</span>(results)[<span class="fu">which</span>(<span class="fu">abs</span>(results[, <span class="dv">1</span>]) <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> results[, <span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>At FDR &lt; 5% 14 genes were reported as significantly differentially expressed with the original Xia et al count matrix but <em>not</em> with the output of the nf-core/rnaseq workflow.</p>
<p>As side-by-side comparison of the FDR (<code>adj.P.Val</code>) for these genes confirms that the vast majority display significant close to the 5% threshold in the nf-core/rnaseq output as well. (This is in line with the high overall correlation of the t-statistics observed above.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tt[genes, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       gene_name    geo nfcore
ENSMUSG00000078193 RP24-228M19.1 0.0041  0.120
ENSMUSG00000027427        Polr3f 0.0490  0.052
ENSMUSG00000028394         Pole3 0.0480  0.073
ENSMUSG00000029027          Dffb 0.0360  0.082
ENSMUSG00000029649          Pomp 0.0430  0.056
ENSMUSG00000054404         Slfn5 0.0440  0.064
ENSMUSG00000050965         Prkca 0.0310  0.087
ENSMUSG00000020641         Rsad2 0.0440  0.063
ENSMUSG00000021057         Akap5 0.0260  0.074
ENSMUSG00000115230 RP24-123O20.1 0.0400  0.061
ENSMUSG00000042622          Maff 0.0500  0.051
ENSMUSG00000050410         Tcf19 0.0500  0.051
ENSMUSG00000059040         Eno1b 0.0290  0.410
ENSMUSG00000042712        Tceal9 0.0480  0.054</code></pre>
</div>
</div>
<p>Finally, we plot the normalized gene expression estimates for the 14 discordant genes.</p>
<div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> genes) {</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a> p <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">==</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> genotype, <span class="at">y =</span> cpm)) <span class="sc">+</span> </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">width =</span> <span class="fl">0.05</span>), <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> dge<span class="sc">$</span>genes[gene, <span class="st">"gene_name"</span>],</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"Normalized expression (CPM)"</span>,</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"FDR nf-core: %s</span><span class="sc">\n</span><span class="st">FDR GEO: %s"</span>,</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"nfcore"</span>],</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"geo"</span>]</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>         ) <span class="sc">+</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(p)</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-2.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-3.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-4.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-5.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-6.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-7.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-8.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-9.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-10.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-11.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-12.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-13.png" class="img-fluid" width="288"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="index_files/figure-html/unnamed-chunk-28-14.png" class="img-fluid" width="288"></p>
</div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="conclusions" class="level2">
<h2 class="anchored" data-anchor-id="conclusions">Conclusions</h2>
<ul>
<li>Differential expression analyses of raw counts obtained with the <code>nc-core/rnaseq</code> workflow yields results that are highly concordant with those obtained with the raw counts the authors deposited in NCBI GEO.</li>
<li>With appropriate parameters the <code>nf-core/rnaseq</code> workflow can be applied to QuantSeq FWD 3’tag RNA-seq data.</li>
</ul>
</section>
<section id="reproducibility" class="level2">
<h2 class="anchored" data-anchor-id="reproducibility">Reproducibility</h2>
<details>
<summary>
SessionInfo
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.2 (2022-10-31)
 os       macOS Big Sur ... 10.16
 system   x86_64, darwin17.0
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Los_Angeles
 date     2023-01-16
 pandoc   2.19.2 @ /Applications/RStudio.app/Contents/MacOS/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version  date (UTC) lib source
 AnnotationDbi        * 1.60.0   2022-11-01 [1] Bioconductor
 askpass                1.1      2019-01-13 [1] CRAN (R 4.2.0)
 assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.2.0)
 Biobase              * 2.58.0   2022-11-01 [1] Bioconductor
 BiocGenerics         * 0.44.0   2022-11-01 [1] Bioconductor
 Biostrings             2.66.0   2022-11-01 [1] Bioconductor
 bit                    4.0.5    2022-11-15 [1] CRAN (R 4.2.0)
 bit64                  4.0.5    2020-08-30 [1] CRAN (R 4.2.0)
 bitops                 1.0-7    2021-04-24 [1] CRAN (R 4.2.0)
 blob                   1.2.3    2022-04-10 [1] CRAN (R 4.2.0)
 cachem                 1.0.6    2021-08-19 [1] CRAN (R 4.2.0)
 cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.2.0)
 cli                    3.5.0    2022-12-20 [1] CRAN (R 4.2.0)
 colorspace             2.0-3    2022-02-21 [1] CRAN (R 4.2.0)
 crayon                 1.5.2    2022-09-29 [1] CRAN (R 4.2.0)
 credentials            1.3.2    2021-11-29 [1] CRAN (R 4.2.0)
 DBI                    1.1.3    2022-06-18 [1] CRAN (R 4.2.0)
 DelayedArray           0.24.0   2022-11-01 [1] Bioconductor
 digest                 0.6.31   2022-12-11 [1] CRAN (R 4.2.0)
 dplyr                * 1.0.10   2022-09-01 [1] CRAN (R 4.2.0)
 edgeR                * 3.40.1   2022-12-14 [1] Bioconductor
 ellipsis               0.3.2    2021-04-29 [1] CRAN (R 4.2.0)
 evaluate               0.19     2022-12-13 [1] CRAN (R 4.2.0)
 fansi                  1.0.3    2022-03-24 [1] CRAN (R 4.2.0)
 farver                 2.1.1    2022-07-06 [1] CRAN (R 4.2.0)
 fastmap                1.1.0    2021-01-25 [1] CRAN (R 4.2.0)
 generics               0.1.3    2022-07-05 [1] CRAN (R 4.2.0)
 GenomeInfoDb         * 1.34.4   2022-12-01 [1] Bioconductor
 GenomeInfoDbData       1.2.9    2022-12-12 [1] Bioconductor
 GenomicRanges        * 1.50.2   2022-12-16 [1] Bioconductor
 ggplot2              * 3.4.0    2022-11-04 [1] CRAN (R 4.2.0)
 glue                   1.6.2    2022-02-24 [1] CRAN (R 4.2.0)
 gtable                 0.3.1    2022-09-01 [1] CRAN (R 4.2.0)
 here                 * 1.0.1    2020-12-13 [1] CRAN (R 4.2.0)
 htmltools              0.5.4    2022-12-07 [1] CRAN (R 4.2.0)
 htmlwidgets            1.5.4    2021-09-08 [1] CRAN (R 4.2.2)
 httr                   1.4.4    2022-08-17 [1] CRAN (R 4.2.0)
 IRanges              * 2.32.0   2022-11-01 [1] Bioconductor
 jsonlite               1.8.4    2022-12-06 [1] CRAN (R 4.2.0)
 KEGGREST               1.38.0   2022-11-01 [1] Bioconductor
 KernSmooth             2.23-20  2021-05-03 [2] CRAN (R 4.2.2)
 knitr                  1.41     2022-11-18 [1] CRAN (R 4.2.0)
 labeling               0.4.2    2020-10-20 [1] CRAN (R 4.2.0)
 lattice                0.20-45  2021-09-22 [2] CRAN (R 4.2.2)
 lifecycle              1.0.3    2022-10-07 [1] CRAN (R 4.2.0)
 limma                * 3.54.0   2022-11-01 [1] Bioconductor
 locfit                 1.5-9.6  2022-07-11 [1] CRAN (R 4.2.0)
 magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.2.0)
 Matrix                 1.5-3    2022-11-11 [1] CRAN (R 4.2.0)
 MatrixGenerics       * 1.10.0   2022-11-01 [1] Bioconductor
 matrixStats          * 0.63.0   2022-11-18 [1] CRAN (R 4.2.0)
 memoise                2.0.1    2021-11-26 [1] CRAN (R 4.2.0)
 munsell                0.5.0    2018-06-12 [1] CRAN (R 4.2.0)
 openssl                2.0.5    2022-12-06 [1] CRAN (R 4.2.0)
 org.Mm.eg.db         * 3.16.0   2022-12-29 [1] Bioconductor
 pillar                 1.8.1    2022-08-19 [1] CRAN (R 4.2.0)
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.2.0)
 png                    0.1-8    2022-11-29 [1] CRAN (R 4.2.0)
 purrr                  1.0.0    2022-12-20 [1] CRAN (R 4.2.0)
 R6                     2.5.1    2021-08-19 [1] CRAN (R 4.2.0)
 Rcpp                   1.0.9    2022-07-08 [1] CRAN (R 4.2.0)
 RCurl                  1.98-1.9 2022-10-03 [1] CRAN (R 4.2.0)
 readxl               * 1.4.1    2022-08-17 [1] CRAN (R 4.2.0)
 rlang                  1.0.6    2022-09-24 [1] CRAN (R 4.2.0)
 rmarkdown              2.19     2022-12-15 [1] CRAN (R 4.2.0)
 rprojroot              2.0.3    2022-04-02 [1] CRAN (R 4.2.0)
 RSQLite                2.2.19   2022-11-24 [1] CRAN (R 4.2.0)
 rstudioapi             0.14     2022-08-22 [1] CRAN (R 4.2.0)
 S4Vectors            * 0.36.1   2022-12-05 [1] Bioconductor
 scales                 1.2.1    2022-08-20 [1] CRAN (R 4.2.0)
 sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.2.0)
 statmod                1.4.37   2022-08-12 [1] CRAN (R 4.2.0)
 stringi                1.7.8    2022-07-11 [1] CRAN (R 4.2.0)
 stringr                1.5.0    2022-12-02 [1] CRAN (R 4.2.0)
 SummarizedExperiment * 1.28.0   2022-11-01 [1] Bioconductor
 sys                    3.4.1    2022-10-18 [1] CRAN (R 4.2.0)
 tibble               * 3.1.8    2022-07-22 [1] CRAN (R 4.2.0)
 tidyr                * 1.2.1    2022-09-08 [1] CRAN (R 4.2.0)
 tidyselect             1.2.0    2022-10-10 [1] CRAN (R 4.2.0)
 utf8                   1.2.2    2021-07-24 [1] CRAN (R 4.2.0)
 vctrs                  0.5.1    2022-11-16 [1] CRAN (R 4.2.0)
 withr                  2.5.0    2022-03-03 [1] CRAN (R 4.2.0)
 xfun                   0.35     2022-11-16 [1] CRAN (R 4.2.0)
 XVector                0.38.0   2022-11-01 [1] Bioconductor
 yaml                   2.3.6    2022-10-18 [1] CRAN (R 4.2.0)
 zlibbioc               1.44.0   2022-11-01 [1] Bioconductor

 [1] /Users/sandmann/Library/R/x86_64/4.2/library
 [2] /Library/Frameworks/R.framework/Versions/4.2/Resources/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>


<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Full disclosure: I am a co-author of this publication.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb43" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "QuantSeq RNAseq analysis (3): Validating published results (no UMIs)"</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-01-16"</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [R, nextflow, NGS]</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="false"}</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a>This is the third of four posts documenting my progress toward</span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a>processing and analyzing </span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">QuantSeq FWD 3' tag RNAseq</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span> </span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>data </span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>with the </span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/rnaseq</span><span class="co">](https://nf-co.re/rnaseq)</span></span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>workflow. </span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="co">[</span><span class="ot">Configuring &amp; executing the nf-core/rnaseq workflow</span><span class="co">](../nextflow-core-quantseq-1-settings/index.html)</span></span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="co">[</span><span class="ot">Exploring the workflow outputs</span><span class="co">](../nextflow-core-quantseq-2-output/index.html)</span></span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Validating the workflow by reproducing results published by Xia et al (no UMIs)**</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span><span class="co">[</span><span class="ot">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</span><span class="co">](../nextflow-core-quantseq-4-nugent/index.html)</span></span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>Many thanks to <span class="co">[</span><span class="ot">Harshil Patel</span><span class="co">](https://github.com/drpatelh)</span>, </span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">António Miguel de Jesus Domingues</span><span class="co">](https://github.com/adomingues/)</span></span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a>and </span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Matthias Zepper</span><span class="co">](https://github.com/matthiasZepper/)</span></span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>for their generous guidance &amp; input via <span class="co">[</span><span class="ot">nf-core slack</span><span class="co">](nfcore.slack.com)</span>.</span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>(Any mistakes are mine.)</span>
<span id="cb43-43"><a href="#cb43-43" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-44"><a href="#cb43-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-45"><a href="#cb43-45" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb43-46"><a href="#cb43-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-47"><a href="#cb43-47" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>This analysis compares the performance of the </span>
<span id="cb43-48"><a href="#cb43-48" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">nf-core/rnaseq workflow</span><span class="co">](https://nf-co.re/rnaseq)</span></span>
<span id="cb43-49"><a href="#cb43-49" aria-hidden="true" tabindex="-1"></a>  for QuantSeq FWD 3' tag RNAseq data without unique molecular identifiers.</span>
<span id="cb43-50"><a href="#cb43-50" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The differential expression analysis results are highly</span>
<span id="cb43-51"><a href="#cb43-51" aria-hidden="true" tabindex="-1"></a>  concordant with those obtained in the original publication.</span>
<span id="cb43-52"><a href="#cb43-52" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>With the appropriate settings, the </span>
<span id="cb43-53"><a href="#cb43-53" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">nf-core/rnaseq workflow</span><span class="co">](https://nf-co.re/rnaseq)</span> is a valid data processing</span>
<span id="cb43-54"><a href="#cb43-54" aria-hidden="true" tabindex="-1"></a>  pipeline for this data type.</span>
<span id="cb43-55"><a href="#cb43-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-56"><a href="#cb43-56" aria-hidden="true" tabindex="-1"></a>The</span>
<span id="cb43-57"><a href="#cb43-57" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">first post in this series</span><span class="co">](../nextflow-core-quantseq-1-settings/index.html)</span></span>
<span id="cb43-58"><a href="#cb43-58" aria-hidden="true" tabindex="-1"></a>walked through the preprocesssing of QuantSeq FWD data </span>
<span id="cb43-59"><a href="#cb43-59" aria-hidden="true" tabindex="-1"></a>published in a preprint by <span class="co">[</span><span class="ot">Xia et al</span><span class="co">](https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1)</span>.</span>
<span id="cb43-60"><a href="#cb43-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-61"><a href="#cb43-61" aria-hidden="true" tabindex="-1"></a>Next, we use Bioconductor/R packages to reproduce the downstream results.</span>
<span id="cb43-62"><a href="#cb43-62" aria-hidden="true" tabindex="-1"></a>We perform the same differential gene expression analysis twice with either</span>
<span id="cb43-63"><a href="#cb43-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-64"><a href="#cb43-64" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>the original counts matrix published by the authors <span class="ot">[^1]</span></span>
<span id="cb43-65"><a href="#cb43-65" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>the output of the nf-core/rnaseq workflow</span>
<span id="cb43-66"><a href="#cb43-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-67"><a href="#cb43-67" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>Full disclosure: I am a co-author of this publication.</span>
<span id="cb43-68"><a href="#cb43-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-71"><a href="#cb43-71" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-72"><a href="#cb43-72" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb43-73"><a href="#cb43-73" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb43-74"><a href="#cb43-74" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb43-75"><a href="#cb43-75" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb43-76"><a href="#cb43-76" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb43-77"><a href="#cb43-77" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb43-78"><a href="#cb43-78" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb43-79"><a href="#cb43-79" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb43-80"><a href="#cb43-80" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb43-81"><a href="#cb43-81" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb43-82"><a href="#cb43-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-83"><a href="#cb43-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-86"><a href="#cb43-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-87"><a href="#cb43-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb43-88"><a href="#cb43-88" aria-hidden="true" tabindex="-1"></a>work_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"posts"</span>, <span class="st">"nextflow-core-quantseq-3-xia"</span>)</span>
<span id="cb43-89"><a href="#cb43-89" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-90"><a href="#cb43-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-91"><a href="#cb43-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## Sample annotations</span></span>
<span id="cb43-92"><a href="#cb43-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-93"><a href="#cb43-93" aria-hidden="true" tabindex="-1"></a>We start by retrieving the sample annotation table, listing e.g. the <span class="in">`sex`</span>,</span>
<span id="cb43-94"><a href="#cb43-94" aria-hidden="true" tabindex="-1"></a>and <span class="in">`genotype`</span> for each mouse, and the <span class="in">`batch`</span> for each collected sample.</span>
<span id="cb43-95"><a href="#cb43-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-96"><a href="#cb43-96" aria-hidden="true" tabindex="-1"></a>This information is available in the <span class="co">[</span><span class="ot">SRA Run Explorer</span><span class="co">](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=1&amp;WebEnv=MCID_63c0392862acb114c36e44a3&amp;o=acc_s%3Aa#)</span>. </span>
<span id="cb43-97"><a href="#cb43-97" aria-hidden="true" tabindex="-1"></a>(I saved it as the <span class="co">[</span><span class="ot">sample_metadata.csv CSV file</span><span class="co">](sample_metadata.csv)</span> in </span>
<span id="cb43-98"><a href="#cb43-98" aria-hidden="true" tabindex="-1"></a>case you want to follow along.)</span>
<span id="cb43-99"><a href="#cb43-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-102"><a href="#cb43-102" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-103"><a href="#cb43-103" aria-hidden="true" tabindex="-1"></a>sample_sheet <span class="ot">&lt;-</span> <span class="fu">file.path</span>(work_dir, <span class="st">"sample_metadata.csv"</span>)</span>
<span id="cb43-104"><a href="#cb43-104" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_sheet, <span class="at">row.names =</span> <span class="st">"Experiment"</span>)</span>
<span id="cb43-105"><a href="#cb43-105" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno[, <span class="fu">c</span>(<span class="st">"Run"</span>, <span class="st">"Animal.ID"</span>, <span class="st">"Age"</span>, <span class="st">"age_unit"</span>, <span class="st">"Batch"</span>, <span class="st">"sex"</span>,</span>
<span id="cb43-106"><a href="#cb43-106" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Genotype"</span>, <span class="st">"Sample.Name"</span>)])</span>
<span id="cb43-107"><a href="#cb43-107" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-108"><a href="#cb43-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-109"><a href="#cb43-109" aria-hidden="true" tabindex="-1"></a>Because our SRA metadata doesn't include the GEO sample title, I saved the</span>
<span id="cb43-110"><a href="#cb43-110" aria-hidden="true" tabindex="-1"></a>identifier mappings in the </span>
<span id="cb43-111"><a href="#cb43-111" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">GEO_sample_ids.csv CSV file</span><span class="co">](GEO_sample_ids.csv)</span>.</span>
<span id="cb43-112"><a href="#cb43-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-115"><a href="#cb43-115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-116"><a href="#cb43-116" aria-hidden="true" tabindex="-1"></a>geo_ids <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">file.path</span>(work_dir, <span class="st">"GEO_sample_ids.csv"</span>))</span>
<span id="cb43-117"><a href="#cb43-117" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geo_ids)</span>
<span id="cb43-118"><a href="#cb43-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-119"><a href="#cb43-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-122"><a href="#cb43-122" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-123"><a href="#cb43-123" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-124"><a href="#cb43-124" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_anno)<span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">colnames</span>(sample_anno))</span>
<span id="cb43-125"><a href="#cb43-125" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_anno) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"."</span>, <span class="st">"_"</span>, <span class="fu">colnames</span>(sample_anno), </span>
<span id="cb43-126"><a href="#cb43-126" aria-hidden="true" tabindex="-1"></a>                             <span class="at">fixed =</span> <span class="cn">TRUE</span>) </span>
<span id="cb43-127"><a href="#cb43-127" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> sample_anno[, <span class="fu">c</span>(<span class="st">"sample_name"</span>, <span class="st">"animal_id"</span>, <span class="st">"genotype"</span>, <span class="st">"sex"</span>,</span>
<span id="cb43-128"><a href="#cb43-128" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"batch"</span>)]</span>
<span id="cb43-129"><a href="#cb43-129" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">factor</span>(sample_anno<span class="sc">$</span>genotype, </span>
<span id="cb43-130"><a href="#cb43-130" aria-hidden="true" tabindex="-1"></a>                               <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"Het"</span>, <span class="st">"Hom"</span>))</span>
<span id="cb43-131"><a href="#cb43-131" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>sample_title <span class="ot">&lt;-</span> geo_ids[</span>
<span id="cb43-132"><a href="#cb43-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">match</span>(sample_anno<span class="sc">$</span>sample_name, geo_ids<span class="sc">$</span>sample_name), <span class="st">"sample_id"</span>]</span>
<span id="cb43-133"><a href="#cb43-133" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno)</span>
<span id="cb43-134"><a href="#cb43-134" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-135"><a href="#cb43-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-136"><a href="#cb43-136" aria-hidden="true" tabindex="-1"></a>This experiment includes <span class="in">`r nrow(sample_anno)`</span> samples of microglia cells </span>
<span id="cb43-137"><a href="#cb43-137" aria-hidden="true" tabindex="-1"></a>obtained from <span class="in">`r length(unique(sample_anno$animal_id))`</span> different 8-month old </span>
<span id="cb43-138"><a href="#cb43-138" aria-hidden="true" tabindex="-1"></a>mice. Both male and female animals were included in the study.</span>
<span id="cb43-139"><a href="#cb43-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-140"><a href="#cb43-140" aria-hidden="true" tabindex="-1"></a>The animals carry one of three different genotypes of the gene encoding the </span>
<span id="cb43-141"><a href="#cb43-141" aria-hidden="true" tabindex="-1"></a>APP amyloid beta precursor protein, either</span>
<span id="cb43-142"><a href="#cb43-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-143"><a href="#cb43-143" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>the wildtype mouse gene (<span class="in">`WT`</span>) or</span>
<span id="cb43-144"><a href="#cb43-144" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>one copy (<span class="in">`Het`</span>) or</span>
<span id="cb43-145"><a href="#cb43-145" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>two copies (<span class="in">`Hom`</span>) </span>
<span id="cb43-146"><a href="#cb43-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-147"><a href="#cb43-147" aria-hidden="true" tabindex="-1"></a>of a mutant _APP_ gene carrying mutations associated with familial </span>
<span id="cb43-148"><a href="#cb43-148" aria-hidden="true" tabindex="-1"></a>Alzheimer's Disease.</span>
<span id="cb43-149"><a href="#cb43-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-150"><a href="#cb43-150" aria-hidden="true" tabindex="-1"></a>Samples from all three genotypes were collected on three days, and we will</span>
<span id="cb43-151"><a href="#cb43-151" aria-hidden="true" tabindex="-1"></a>use this <span class="in">`batch`</span> information to model the experiment.</span>
<span id="cb43-152"><a href="#cb43-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-153"><a href="#cb43-153" aria-hidden="true" tabindex="-1"></a>Two separate microglia samples were obtained from each animal, and we will</span>
<span id="cb43-154"><a href="#cb43-154" aria-hidden="true" tabindex="-1"></a>include this nested relationship by modeling the <span class="in">`animal`</span> as random effect in </span>
<span id="cb43-155"><a href="#cb43-155" aria-hidden="true" tabindex="-1"></a>our linear model.</span>
<span id="cb43-156"><a href="#cb43-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-157"><a href="#cb43-157" aria-hidden="true" tabindex="-1"></a><span class="fu">## Xia et al's original count data</span></span>
<span id="cb43-158"><a href="#cb43-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-159"><a href="#cb43-159" aria-hidden="true" tabindex="-1"></a>First, we retrieve the authors' count matrix from </span>
<span id="cb43-160"><a href="#cb43-160" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">NCBI GEO</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158152)</span>, </span>
<span id="cb43-161"><a href="#cb43-161" aria-hidden="true" tabindex="-1"></a>available as a Supplementary Excel file.</span>
<span id="cb43-162"><a href="#cb43-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-165"><a href="#cb43-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-166"><a href="#cb43-166" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE158152&amp;"</span>,</span>
<span id="cb43-167"><a href="#cb43-167" aria-hidden="true" tabindex="-1"></a>              <span class="st">"format=file&amp;file=GSE158152%5Fdst150%5Fprocessed%2Exlsx"</span>)</span>
<span id="cb43-168"><a href="#cb43-168" aria-hidden="true" tabindex="-1"></a>temp_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb43-169"><a href="#cb43-169" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, <span class="at">destfile =</span> temp_file)</span>
<span id="cb43-170"><a href="#cb43-170" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-171"><a href="#cb43-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-172"><a href="#cb43-172" aria-hidden="true" tabindex="-1"></a>The Excel file has three different worksheets</span>
<span id="cb43-173"><a href="#cb43-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-174"><a href="#cb43-174" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`sample_annotations`</span></span>
<span id="cb43-175"><a href="#cb43-175" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`raw_counts`</span></span>
<span id="cb43-176"><a href="#cb43-176" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`normalized_cpm`</span></span>
<span id="cb43-177"><a href="#cb43-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-180"><a href="#cb43-180" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-181"><a href="#cb43-181" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(temp_file, <span class="at">sheet =</span> <span class="st">"raw_counts"</span>)</span>
<span id="cb43-182"><a href="#cb43-182" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(raw_counts), <span class="dv">10</span>)</span>
<span id="cb43-183"><a href="#cb43-183" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-184"><a href="#cb43-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-185"><a href="#cb43-185" aria-hidden="true" tabindex="-1"></a>The <span class="in">`raw_counts`</span> excel sheet contains information about the detected genes (</span>
<span id="cb43-186"><a href="#cb43-186" aria-hidden="true" tabindex="-1"></a><span class="in">`feature_ID`</span>, <span class="in">`name`</span>) and the samples are identified by their GEO title </span>
<span id="cb43-187"><a href="#cb43-187" aria-hidden="true" tabindex="-1"></a>(e.g. <span class="in">`DRN-18459`</span>, <span class="in">`DRN-184560`</span>). </span>
<span id="cb43-188"><a href="#cb43-188" aria-hidden="true" tabindex="-1"></a>We use the raw counts to populate a new <span class="in">`DGEList`</span> object and perform</span>
<span id="cb43-189"><a href="#cb43-189" aria-hidden="true" tabindex="-1"></a>Library Size Normalization with the <span class="in">`TMM`</span> approach.</span>
<span id="cb43-190"><a href="#cb43-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-193"><a href="#cb43-193" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-194"><a href="#cb43-194" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-195"><a href="#cb43-195" aria-hidden="true" tabindex="-1"></a>count_data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(raw_counts[, <span class="fu">grep</span>(<span class="st">"DRN-"</span>, <span class="fu">colnames</span>(raw_counts))])</span>
<span id="cb43-196"><a href="#cb43-196" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(count_data) <span class="ot">&lt;-</span> raw_counts<span class="sc">$</span>feature_id</span>
<span id="cb43-197"><a href="#cb43-197" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_data) <span class="ot">&lt;-</span> <span class="fu">row.names</span>(sample_anno)[</span>
<span id="cb43-198"><a href="#cb43-198" aria-hidden="true" tabindex="-1"></a>  <span class="fu">match</span>(<span class="fu">colnames</span>(count_data), sample_anno<span class="sc">$</span>sample_title)</span>
<span id="cb43-199"><a href="#cb43-199" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb43-200"><a href="#cb43-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-201"><a href="#cb43-201" aria-hidden="true" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb43-202"><a href="#cb43-202" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> raw_counts<span class="sc">$</span>feature_id,</span>
<span id="cb43-203"><a href="#cb43-203" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_name =</span> raw_counts<span class="sc">$</span>symbol,</span>
<span id="cb43-204"><a href="#cb43-204" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> raw_counts<span class="sc">$</span>feature_id</span>
<span id="cb43-205"><a href="#cb43-205" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-206"><a href="#cb43-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-207"><a href="#cb43-207" aria-hidden="true" tabindex="-1"></a>col_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb43-208"><a href="#cb43-208" aria-hidden="true" tabindex="-1"></a>  sample_anno[<span class="fu">colnames</span>(count_data),</span>
<span id="cb43-209"><a href="#cb43-209" aria-hidden="true" tabindex="-1"></a>              <span class="fu">c</span>(<span class="st">"sample_title"</span>, <span class="st">"animal_id"</span>, <span class="st">"sex"</span>, <span class="st">"genotype"</span>, <span class="st">"batch"</span>)],</span>
<span id="cb43-210"><a href="#cb43-210" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow =</span> <span class="st">"geo"</span></span>
<span id="cb43-211"><a href="#cb43-211" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-212"><a href="#cb43-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-213"><a href="#cb43-213" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb43-214"><a href="#cb43-214" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> <span class="fu">as.matrix</span>(count_data), </span>
<span id="cb43-215"><a href="#cb43-215" aria-hidden="true" tabindex="-1"></a>  <span class="at">samples =</span> col_data[<span class="fu">colnames</span>(count_data), ], </span>
<span id="cb43-216"><a href="#cb43-216" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes =</span> gene_data[<span class="fu">row.names</span>(count_data), ]</span>
<span id="cb43-217"><a href="#cb43-217" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-218"><a href="#cb43-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-219"><a href="#cb43-219" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span>
<span id="cb43-220"><a href="#cb43-220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-221"><a href="#cb43-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-222"><a href="#cb43-222" aria-hidden="true" tabindex="-1"></a>Next, we project the samples into two dimensions by performing multi-dimensional</span>
<span id="cb43-223"><a href="#cb43-223" aria-hidden="true" tabindex="-1"></a>scaling of the top 500 most variable genes. The samples cluster by genotype, </span>
<span id="cb43-224"><a href="#cb43-224" aria-hidden="true" tabindex="-1"></a>with <span class="in">`WT`</span> and <span class="in">`Het`</span> segregating from the <span class="in">`Hom`</span> samples.</span>
<span id="cb43-225"><a href="#cb43-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-228"><a href="#cb43-228" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-229"><a href="#cb43-229" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 5.5</span></span>
<span id="cb43-230"><a href="#cb43-230" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb43-231"><a href="#cb43-231" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge, <span class="at">labels =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>genotype,</span>
<span id="cb43-232"><a href="#cb43-232" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Multi-dimensional scaling"</span>, </span>
<span id="cb43-233"><a href="#cb43-233" aria-hidden="true" tabindex="-1"></a>        <span class="at">sub =</span> <span class="st">"Based on the top 500 most variable genes"</span>)</span>
<span id="cb43-234"><a href="#cb43-234" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-235"><a href="#cb43-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-236"><a href="#cb43-236" aria-hidden="true" tabindex="-1"></a>Let's identify which genes are significantly differentially expressed between</span>
<span id="cb43-237"><a href="#cb43-237" aria-hidden="true" tabindex="-1"></a>the three genotypes!</span>
<span id="cb43-238"><a href="#cb43-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-239"><a href="#cb43-239" aria-hidden="true" tabindex="-1"></a><span class="fu">### Linear modeling with limma/voom</span></span>
<span id="cb43-240"><a href="#cb43-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-241"><a href="#cb43-241" aria-hidden="true" tabindex="-1"></a>First, we use the <span class="in">`edgeR::filterByExpr()`</span> function to identify genes with</span>
<span id="cb43-242"><a href="#cb43-242" aria-hidden="true" tabindex="-1"></a>sufficiently large counts to be examined for differential expression. (The </span>
<span id="cb43-243"><a href="#cb43-243" aria-hidden="true" tabindex="-1"></a><span class="in">`min.count = 25`</span> parameter was determined by examining the mean-variance plot </span>
<span id="cb43-244"><a href="#cb43-244" aria-hidden="true" tabindex="-1"></a>by the <span class="in">`voomLmFit()`</span> function.)</span>
<span id="cb43-245"><a href="#cb43-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-248"><a href="#cb43-248" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-249"><a href="#cb43-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb43-250"><a href="#cb43-250" aria-hidden="true" tabindex="-1"></a><span class="co">#| results: hide</span></span>
<span id="cb43-251"><a href="#cb43-251" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype <span class="sc">+</span> sex <span class="sc">+</span> batch, <span class="at">data =</span> dge<span class="sc">$</span>samples)</span>
<span id="cb43-252"><a href="#cb43-252" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"genotype"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(design))</span>
<span id="cb43-253"><a href="#cb43-253" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">design =</span> design, <span class="at">min.count =</span> <span class="dv">25</span>)</span>
<span id="cb43-254"><a href="#cb43-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-255"><a href="#cb43-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-256"><a href="#cb43-256" aria-hidden="true" tabindex="-1"></a>Next, we fit a linear model to the data using the </span>
<span id="cb43-257"><a href="#cb43-257" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">limma/voom</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)</span></span>
<span id="cb43-258"><a href="#cb43-258" aria-hidden="true" tabindex="-1"></a>approach. The model includes the following fixed effects:</span>
<span id="cb43-259"><a href="#cb43-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-260"><a href="#cb43-260" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The <span class="in">`genotype`</span> coded as a factor with the <span class="in">`WT`</span> as the reference level.</span>
<span id="cb43-261"><a href="#cb43-261" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The <span class="in">`sex`</span> and <span class="in">`batch`</span> covariates, to account for systematic differences in </span>
<span id="cb43-262"><a href="#cb43-262" aria-hidden="true" tabindex="-1"></a>  mean gene expression.</span>
<span id="cb43-263"><a href="#cb43-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-264"><a href="#cb43-264" aria-hidden="true" tabindex="-1"></a>Because the dataset included two replicate samples from each animal, we model</span>
<span id="cb43-265"><a href="#cb43-265" aria-hidden="true" tabindex="-1"></a>the <span class="in">`animal`</span> as a random effect (via the <span class="in">`block`</span> argument of the <span class="in">`voomLmFit()`</span> </span>
<span id="cb43-266"><a href="#cb43-266" aria-hidden="true" tabindex="-1"></a>function). We then extract the coefficients, log2 fold changes and p-values</span>
<span id="cb43-267"><a href="#cb43-267" aria-hidden="true" tabindex="-1"></a>via limma's empirical Bayes approach.</span>
<span id="cb43-268"><a href="#cb43-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-269"><a href="#cb43-269" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true"}</span>
<span id="cb43-270"><a href="#cb43-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-271"><a href="#cb43-271" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`limma::treat()`</span> function to test the null hypothesis that genes</span>
<span id="cb43-272"><a href="#cb43-272" aria-hidden="true" tabindex="-1"></a>display significant differential expression _greater than_ 1.2-fold. This is</span>
<span id="cb43-273"><a href="#cb43-273" aria-hidden="true" tabindex="-1"></a>more stringent than the conventional null hypothesis of zero change. (Please</span>
<span id="cb43-274"><a href="#cb43-274" aria-hidden="true" tabindex="-1"></a>consult the <span class="in">`limma::treat()`</span> </span>
<span id="cb43-275"><a href="#cb43-275" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">help page</span><span class="co">](https://rdrr.io/bioc/limma/man/ebayes.html)</span> for details.)</span>
<span id="cb43-276"><a href="#cb43-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-277"><a href="#cb43-277" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-278"><a href="#cb43-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-281"><a href="#cb43-281" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-282"><a href="#cb43-282" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb43-283"><a href="#cb43-283" aria-hidden="true" tabindex="-1"></a><span class="co">#| results: hide</span></span>
<span id="cb43-284"><a href="#cb43-284" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">voomLmFit</span>(</span>
<span id="cb43-285"><a href="#cb43-285" aria-hidden="true" tabindex="-1"></a>  dge[keep, <span class="fu">row.names</span>(design)], </span>
<span id="cb43-286"><a href="#cb43-286" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> design,</span>
<span id="cb43-287"><a href="#cb43-287" aria-hidden="true" tabindex="-1"></a>  <span class="at">block =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>animal_id, </span>
<span id="cb43-288"><a href="#cb43-288" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.weights =</span> <span class="cn">TRUE</span>, </span>
<span id="cb43-289"><a href="#cb43-289" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">FALSE</span></span>
<span id="cb43-290"><a href="#cb43-290" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-291"><a href="#cb43-291" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">treat</span>(fit, <span class="at">robust=</span><span class="cn">TRUE</span>)</span>
<span id="cb43-292"><a href="#cb43-292" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-293"><a href="#cb43-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-294"><a href="#cb43-294" aria-hidden="true" tabindex="-1"></a>The following table displays the number of differentially up- and down-regulated</span>
<span id="cb43-295"><a href="#cb43-295" aria-hidden="true" tabindex="-1"></a>genes after applying a false-discovery (<span class="in">`adj.P.Val`</span>) threshold of 5%. While</span>
<span id="cb43-296"><a href="#cb43-296" aria-hidden="true" tabindex="-1"></a>we did not detect significant differences between <span class="in">`Het`</span> and <span class="in">`WT`</span> animals, </span>
<span id="cb43-297"><a href="#cb43-297" aria-hidden="true" tabindex="-1"></a>the analysis revealed &gt; 450 differentially expressed genes between <span class="in">`Hom`</span> and</span>
<span id="cb43-298"><a href="#cb43-298" aria-hidden="true" tabindex="-1"></a><span class="in">`WT`</span> microglia.</span>
<span id="cb43-299"><a href="#cb43-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-302"><a href="#cb43-302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-303"><a href="#cb43-303" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(fit))[, <span class="fu">c</span>(<span class="st">"Het"</span>, <span class="st">"Hom"</span>)]</span>
<span id="cb43-304"><a href="#cb43-304" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-305"><a href="#cb43-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-306"><a href="#cb43-306" aria-hidden="true" tabindex="-1"></a>The top 10 genes with the smallest p-values include well known markers of</span>
<span id="cb43-307"><a href="#cb43-307" aria-hidden="true" tabindex="-1"></a>microglia activation:</span>
<span id="cb43-308"><a href="#cb43-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-311"><a href="#cb43-311" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-312"><a href="#cb43-312" aria-hidden="true" tabindex="-1"></a><span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>)[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)]</span>
<span id="cb43-313"><a href="#cb43-313" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-314"><a href="#cb43-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-315"><a href="#cb43-315" aria-hidden="true" tabindex="-1"></a>Next we repeat *the same analysis* with the output of the nf-core/rnaseq</span>
<span id="cb43-316"><a href="#cb43-316" aria-hidden="true" tabindex="-1"></a>workflow.</span>
<span id="cb43-317"><a href="#cb43-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-318"><a href="#cb43-318" aria-hidden="true" tabindex="-1"></a><span class="fu">## nf-core/rnaseq results</span></span>
<span id="cb43-319"><a href="#cb43-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-320"><a href="#cb43-320" aria-hidden="true" tabindex="-1"></a>We start with the raw counts contained in the </span>
<span id="cb43-321"><a href="#cb43-321" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">salmon.merged.gene_counts.rds</span><span class="co">](salmon.merged.gene_counts.rds)</span> </span>
<span id="cb43-322"><a href="#cb43-322" aria-hidden="true" tabindex="-1"></a>file generated by the nf-core/rnaseq workflow. </span>
<span id="cb43-323"><a href="#cb43-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-324"><a href="#cb43-324" aria-hidden="true" tabindex="-1"></a>::: {.callout-note  collapse="true"}</span>
<span id="cb43-325"><a href="#cb43-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-326"><a href="#cb43-326" aria-hidden="true" tabindex="-1"></a>The nf-core pipeline returned the versioned ENSEMBL gene identifiers (e.g.)</span>
<span id="cb43-327"><a href="#cb43-327" aria-hidden="true" tabindex="-1"></a><span class="in">`ENSMUSG00000000001.4`</span>. Because Xia et al only provided the unversioned</span>
<span id="cb43-328"><a href="#cb43-328" aria-hidden="true" tabindex="-1"></a>identifiers (e.g. <span class="in">`ENSMUSG00000000001`</span>) we trim the numeric suffix.</span>
<span id="cb43-329"><a href="#cb43-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-330"><a href="#cb43-330" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-331"><a href="#cb43-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-332"><a href="#cb43-332" aria-hidden="true" tabindex="-1"></a>We TMM-normalize the data, as before. (This step converts the </span>
<span id="cb43-333"><a href="#cb43-333" aria-hidden="true" tabindex="-1"></a><span class="in">`SummarizedExperiment`</span> into a <span class="in">`DGEList`</span> object as well.)</span>
<span id="cb43-334"><a href="#cb43-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-337"><a href="#cb43-337" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-338"><a href="#cb43-338" aria-hidden="true" tabindex="-1"></a>count_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(work_dir, <span class="st">"salmon.merged.gene_counts.rds"</span>)</span>
<span id="cb43-339"><a href="#cb43-339" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(count_file)</span>
<span id="cb43-340"><a href="#cb43-340" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(se) <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb43-341"><a href="#cb43-341" aria-hidden="true" tabindex="-1"></a>  <span class="fu">strsplit</span>(<span class="fu">row.names</span>(se), <span class="at">split =</span> <span class="st">"."</span>, <span class="at">fixed =</span> <span class="cn">TRUE</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</span>
<span id="cb43-342"><a href="#cb43-342" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="fu">colnames</span>(se) <span class="sc">%in%</span> <span class="fu">row.names</span>(sample_anno)))</span>
<span id="cb43-343"><a href="#cb43-343" aria-hidden="true" tabindex="-1"></a>dge_nfcore <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(se, <span class="at">method =</span> <span class="st">"TMM"</span>)</span>
<span id="cb43-344"><a href="#cb43-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-345"><a href="#cb43-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-346"><a href="#cb43-346" aria-hidden="true" tabindex="-1"></a>Next, we add the sample metadata and fit the same linear model as before.</span>
<span id="cb43-347"><a href="#cb43-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-350"><a href="#cb43-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-351"><a href="#cb43-351" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-352"><a href="#cb43-352" aria-hidden="true" tabindex="-1"></a><span class="co">#| results: hide</span></span>
<span id="cb43-353"><a href="#cb43-353" aria-hidden="true" tabindex="-1"></a>dge_nfcore<span class="sc">$</span>genes<span class="sc">$</span>gene_id <span class="ot">&lt;-</span> <span class="fu">row.names</span>(dge_nfcore)</span>
<span id="cb43-354"><a href="#cb43-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-355"><a href="#cb43-355" aria-hidden="true" tabindex="-1"></a>dge_nfcore<span class="sc">$</span>samples <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb43-356"><a href="#cb43-356" aria-hidden="true" tabindex="-1"></a>  dge_nfcore<span class="sc">$</span>samples,</span>
<span id="cb43-357"><a href="#cb43-357" aria-hidden="true" tabindex="-1"></a>  sample_anno[<span class="fu">colnames</span>(dge_nfcore),</span>
<span id="cb43-358"><a href="#cb43-358" aria-hidden="true" tabindex="-1"></a>              <span class="fu">c</span>(<span class="st">"sample_title"</span>, <span class="st">"animal_id"</span>, <span class="st">"sex"</span>, <span class="st">"genotype"</span>, <span class="st">"batch"</span>)],</span>
<span id="cb43-359"><a href="#cb43-359" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow =</span> <span class="st">"nfcore"</span></span>
<span id="cb43-360"><a href="#cb43-360" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-361"><a href="#cb43-361" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="fu">colnames</span>(dge) <span class="sc">%in%</span> <span class="fu">colnames</span>(dge_nfcore)))</span>
<span id="cb43-362"><a href="#cb43-362" aria-hidden="true" tabindex="-1"></a>dge_nfcore <span class="ot">&lt;-</span> dge_nfcore[, <span class="fu">colnames</span>(dge)]</span>
<span id="cb43-363"><a href="#cb43-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-364"><a href="#cb43-364" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> genotype <span class="sc">+</span> sex <span class="sc">+</span> batch, <span class="at">data =</span> dge_nfcore<span class="sc">$</span>samples)</span>
<span id="cb43-365"><a href="#cb43-365" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"genotype"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(design))</span>
<span id="cb43-366"><a href="#cb43-366" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge_nfcore, <span class="at">design =</span> design, <span class="at">min.count =</span> <span class="dv">25</span>)</span>
<span id="cb43-367"><a href="#cb43-367" aria-hidden="true" tabindex="-1"></a>fit_nfcore <span class="ot">&lt;-</span> <span class="fu">voomLmFit</span>(</span>
<span id="cb43-368"><a href="#cb43-368" aria-hidden="true" tabindex="-1"></a>  dge_nfcore[keep, <span class="fu">row.names</span>(design)], </span>
<span id="cb43-369"><a href="#cb43-369" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> design,</span>
<span id="cb43-370"><a href="#cb43-370" aria-hidden="true" tabindex="-1"></a>  <span class="at">block =</span> dge_nfcore<span class="sc">$</span>samples<span class="sc">$</span>animal_id, </span>
<span id="cb43-371"><a href="#cb43-371" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.weights =</span> <span class="cn">TRUE</span>, </span>
<span id="cb43-372"><a href="#cb43-372" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="cn">FALSE</span></span>
<span id="cb43-373"><a href="#cb43-373" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-374"><a href="#cb43-374" aria-hidden="true" tabindex="-1"></a>fit_nfcore <span class="ot">&lt;-</span> <span class="fu">treat</span>(fit_nfcore, <span class="at">robust=</span><span class="cn">TRUE</span>)</span>
<span id="cb43-375"><a href="#cb43-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-376"><a href="#cb43-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-377"><a href="#cb43-377" aria-hidden="true" tabindex="-1"></a>As with the original count data from NCBI GEO, we detect &gt; 450 differentially</span>
<span id="cb43-378"><a href="#cb43-378" aria-hidden="true" tabindex="-1"></a>expressed genes between <span class="in">`Hom`</span> and <span class="in">`WT`</span> genotypes </span>
<span id="cb43-379"><a href="#cb43-379" aria-hidden="true" tabindex="-1"></a>(FDR &lt; 5%, null hypothesis: fold change &gt; 1.2).</span>
<span id="cb43-380"><a href="#cb43-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-383"><a href="#cb43-383" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-384"><a href="#cb43-384" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(fit_nfcore))[, <span class="fu">c</span>(<span class="st">"Het"</span>, <span class="st">"Hom"</span>)]</span>
<span id="cb43-385"><a href="#cb43-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-386"><a href="#cb43-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-387"><a href="#cb43-387" aria-hidden="true" tabindex="-1"></a><span class="fu">## Comparing results across preprocessing workflows</span></span>
<span id="cb43-388"><a href="#cb43-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-389"><a href="#cb43-389" aria-hidden="true" tabindex="-1"></a>Next, we compare the results obtained with the two datasets. We create the </span>
<span id="cb43-390"><a href="#cb43-390" aria-hidden="true" tabindex="-1"></a><span class="in">`cpms`</span> and <span class="in">`tt`</span> dataframes, holding the combined absolute and differential</span>
<span id="cb43-391"><a href="#cb43-391" aria-hidden="true" tabindex="-1"></a>expression results, respectively.</span>
<span id="cb43-392"><a href="#cb43-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-395"><a href="#cb43-395" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-396"><a href="#cb43-396" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-397"><a href="#cb43-397" aria-hidden="true" tabindex="-1"></a>cpms <span class="ot">&lt;-</span> <span class="fu">local</span>({</span>
<span id="cb43-398"><a href="#cb43-398" aria-hidden="true" tabindex="-1"></a>  geo <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-399"><a href="#cb43-399" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-400"><a href="#cb43-400" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(dge<span class="sc">$</span>genes) <span class="sc">%&gt;%</span></span>
<span id="cb43-401"><a href="#cb43-401" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"SRX"</span>), </span>
<span id="cb43-402"><a href="#cb43-402" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"sample_name"</span>,</span>
<span id="cb43-403"><a href="#cb43-403" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"cpm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-404"><a href="#cb43-404" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb43-405"><a href="#cb43-405" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(dge<span class="sc">$</span>samples, <span class="st">"sample_name"</span>),</span>
<span id="cb43-406"><a href="#cb43-406" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"sample_name"</span></span>
<span id="cb43-407"><a href="#cb43-407" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb43-408"><a href="#cb43-408" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"Xia et al"</span>)</span>
<span id="cb43-409"><a href="#cb43-409" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb43-410"><a href="#cb43-410" aria-hidden="true" tabindex="-1"></a>  nfcore <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge_nfcore, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-411"><a href="#cb43-411" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-412"><a href="#cb43-412" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(dge_nfcore<span class="sc">$</span>genes) <span class="sc">%&gt;%</span></span>
<span id="cb43-413"><a href="#cb43-413" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"SRX"</span>), </span>
<span id="cb43-414"><a href="#cb43-414" aria-hidden="true" tabindex="-1"></a>                 <span class="at">names_to =</span> <span class="st">"sample_name"</span>,</span>
<span id="cb43-415"><a href="#cb43-415" aria-hidden="true" tabindex="-1"></a>                 <span class="at">values_to =</span> <span class="st">"cpm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-416"><a href="#cb43-416" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb43-417"><a href="#cb43-417" aria-hidden="true" tabindex="-1"></a>      tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(dge_nfcore<span class="sc">$</span>samples, <span class="st">"sample_name"</span>),</span>
<span id="cb43-418"><a href="#cb43-418" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"sample_name"</span></span>
<span id="cb43-419"><a href="#cb43-419" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb43-420"><a href="#cb43-420" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"nf-core"</span>)</span>
<span id="cb43-421"><a href="#cb43-421" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb43-422"><a href="#cb43-422" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(</span>
<span id="cb43-423"><a href="#cb43-423" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(geo, <span class="fu">any_of</span>(<span class="fu">intersect</span>(<span class="fu">colnames</span>(geo), <span class="fu">colnames</span>(nfcore)))),</span>
<span id="cb43-424"><a href="#cb43-424" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(nfcore, <span class="fu">any_of</span>(<span class="fu">intersect</span>(<span class="fu">colnames</span>(geo), <span class="fu">colnames</span>(nfcore))))</span>
<span id="cb43-425"><a href="#cb43-425" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb43-426"><a href="#cb43-426" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-427"><a href="#cb43-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-428"><a href="#cb43-428" aria-hidden="true" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb43-429"><a href="#cb43-429" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)[</span>
<span id="cb43-430"><a href="#cb43-430" aria-hidden="true" tabindex="-1"></a>    , <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb43-431"><a href="#cb43-431" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"geo"</span>),</span>
<span id="cb43-432"><a href="#cb43-432" aria-hidden="true" tabindex="-1"></a>   <span class="fu">topTreat</span>(fit_nfcore, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)[</span>
<span id="cb43-433"><a href="#cb43-433" aria-hidden="true" tabindex="-1"></a>     , <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"logFC"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb43-434"><a href="#cb43-434" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"nfcore"</span>)</span>
<span id="cb43-435"><a href="#cb43-435" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb43-436"><a href="#cb43-436" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">adj.P.Val =</span> <span class="fu">signif</span>(adj.P.Val, <span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb43-437"><a href="#cb43-437" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb43-438"><a href="#cb43-438" aria-hidden="true" tabindex="-1"></a>    <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>), </span>
<span id="cb43-439"><a href="#cb43-439" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_from =</span> <span class="st">"dataset"</span>, </span>
<span id="cb43-440"><a href="#cb43-440" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_from =</span> <span class="st">"adj.P.Val"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-441"><a href="#cb43-441" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(nfcore) <span class="sc">%&gt;%</span></span>
<span id="cb43-442"><a href="#cb43-442" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-443"><a href="#cb43-443" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"gene_id"</span>)</span>
<span id="cb43-444"><a href="#cb43-444" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-445"><a href="#cb43-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-446"><a href="#cb43-446" aria-hidden="true" tabindex="-1"></a><span class="fu">### Normalized expression</span></span>
<span id="cb43-447"><a href="#cb43-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-448"><a href="#cb43-448" aria-hidden="true" tabindex="-1"></a>First, we examine the correlation between the normalized log-transformed</span>
<span id="cb43-449"><a href="#cb43-449" aria-hidden="true" tabindex="-1"></a>gene expression estimates returned from the two workflows. We focus on those</span>
<span id="cb43-450"><a href="#cb43-450" aria-hidden="true" tabindex="-1"></a>genes that passed the <span class="in">`filterByExpr`</span> thresholds above, e.g. those genes</span>
<span id="cb43-451"><a href="#cb43-451" aria-hidden="true" tabindex="-1"></a>deemed sufficiently highly expressed to be assessed for differential expression.</span>
<span id="cb43-452"><a href="#cb43-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-455"><a href="#cb43-455" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-456"><a href="#cb43-456" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 5.5</span></span>
<span id="cb43-457"><a href="#cb43-457" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 3.5</span></span>
<span id="cb43-458"><a href="#cb43-458" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-459"><a href="#cb43-459" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb43-460"><a href="#cb43-460" aria-hidden="true" tabindex="-1"></a>common_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">row.names</span>(fit), <span class="fu">row.names</span>(fit_nfcore))</span>
<span id="cb43-461"><a href="#cb43-461" aria-hidden="true" tabindex="-1"></a>sum_stats <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb43-462"><a href="#cb43-462" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">%in%</span> common_genes) <span class="sc">%&gt;%</span></span>
<span id="cb43-463"><a href="#cb43-463" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb43-464"><a href="#cb43-464" aria-hidden="true" tabindex="-1"></a>    <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"sample_name"</span>),</span>
<span id="cb43-465"><a href="#cb43-465" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_from =</span> <span class="st">"cpm"</span>,</span>
<span id="cb43-466"><a href="#cb43-466" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_from =</span> <span class="st">"dataset"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-467"><a href="#cb43-467" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(gene_id) <span class="sc">%&gt;%</span></span>
<span id="cb43-468"><a href="#cb43-468" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb43-469"><a href="#cb43-469" aria-hidden="true" tabindex="-1"></a>    <span class="at">r =</span> <span class="fu">cor</span>(<span class="fu">log1p</span>(<span class="st">`</span><span class="at">Xia et al</span><span class="st">`</span>), <span class="fu">log1p</span>(<span class="st">`</span><span class="at">nf-core</span><span class="st">`</span>)),</span>
<span id="cb43-470"><a href="#cb43-470" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_xia =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">Xia et al</span><span class="st">`</span>),</span>
<span id="cb43-471"><a href="#cb43-471" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_nfcore =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">nf-core</span><span class="st">`</span>))</span>
<span id="cb43-472"><a href="#cb43-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-473"><a href="#cb43-473" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sum_stats, <span class="fu">aes</span>(<span class="at">x =</span> r)) <span class="sc">+</span> </span>
<span id="cb43-474"><a href="#cb43-474" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb43-475"><a href="#cb43-475" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.02</span>), <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.2</span>)) <span class="sc">+</span></span>
<span id="cb43-476"><a href="#cb43-476" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Pearson correlation coefficient (R)"</span>, </span>
<span id="cb43-477"><a href="#cb43-477" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Number of genes"</span>,</span>
<span id="cb43-478"><a href="#cb43-478" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"Correlation between normalized log2 counts"</span>) <span class="sc">+</span></span>
<span id="cb43-479"><a href="#cb43-479" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb43-480"><a href="#cb43-480" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span>
<span id="cb43-481"><a href="#cb43-481" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-482"><a href="#cb43-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-483"><a href="#cb43-483" aria-hidden="true" tabindex="-1"></a>The correlation between normalized log2 expression estimates is very high, with</span>
<span id="cb43-484"><a href="#cb43-484" aria-hidden="true" tabindex="-1"></a>95% of all genes showing a Pearson correlation coefficient &gt; </span>
<span id="cb43-485"><a href="#cb43-485" aria-hidden="true" tabindex="-1"></a><span class="in">`r signif(quantile(sum_stats$r, c(0.05)), 2)`</span>. </span>
<span id="cb43-486"><a href="#cb43-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-487"><a href="#cb43-487" aria-hidden="true" tabindex="-1"></a>Most of the <span class="in">`r length(common_genes)`</span> examined genes were detected with &gt; </span>
<span id="cb43-488"><a href="#cb43-488" aria-hidden="true" tabindex="-1"></a>10 normalized counts per million reads.</span>
<span id="cb43-489"><a href="#cb43-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-492"><a href="#cb43-492" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-493"><a href="#cb43-493" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sum_stats, <span class="fu">aes</span>(<span class="at">x =</span> mean_xia <span class="sc">+</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb43-494"><a href="#cb43-494" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb43-495"><a href="#cb43-495" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log10_trans</span>(),</span>
<span id="cb43-496"><a href="#cb43-496" aria-hidden="true" tabindex="-1"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">comma_format</span>()) <span class="sc">+</span></span>
<span id="cb43-497"><a href="#cb43-497" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Mean normalized counts per million"</span>, </span>
<span id="cb43-498"><a href="#cb43-498" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Number of genes"</span>,</span>
<span id="cb43-499"><a href="#cb43-499" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"Average expression"</span>,</span>
<span id="cb43-500"><a href="#cb43-500" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Xia et al"</span>) <span class="sc">+</span></span>
<span id="cb43-501"><a href="#cb43-501" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb43-502"><a href="#cb43-502" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span>
<span id="cb43-503"><a href="#cb43-503" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-504"><a href="#cb43-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-505"><a href="#cb43-505" aria-hidden="true" tabindex="-1"></a>Next, we will examine the results of the _differential_ expression analysis.</span>
<span id="cb43-506"><a href="#cb43-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-507"><a href="#cb43-507" aria-hidden="true" tabindex="-1"></a><span class="fu">### Differential expression results</span></span>
<span id="cb43-508"><a href="#cb43-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-509"><a href="#cb43-509" aria-hidden="true" tabindex="-1"></a>Analyses based on either preprocessing pipeline yield similar numbers of </span>
<span id="cb43-510"><a href="#cb43-510" aria-hidden="true" tabindex="-1"></a>differentially expressed genes.</span>
<span id="cb43-511"><a href="#cb43-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-514"><a href="#cb43-514" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-515"><a href="#cb43-515" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-516"><a href="#cb43-516" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb43-517"><a href="#cb43-517" aria-hidden="true" tabindex="-1"></a>  <span class="fu">decideTests</span>(fit)[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb43-518"><a href="#cb43-518" aria-hidden="true" tabindex="-1"></a>  <span class="fu">decideTests</span>(fit_nfcore)[common_genes, <span class="st">"Hom"</span>]</span>
<span id="cb43-519"><a href="#cb43-519" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-520"><a href="#cb43-520" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(results) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Xia et al"</span>, <span class="st">"nf-core"</span>)</span>
<span id="cb43-521"><a href="#cb43-521" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(results) <span class="ot">&lt;-</span> <span class="st">"TestResults"</span></span>
<span id="cb43-522"><a href="#cb43-522" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(results)</span>
<span id="cb43-523"><a href="#cb43-523" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-524"><a href="#cb43-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-525"><a href="#cb43-525" aria-hidden="true" tabindex="-1"></a>But are these _the same genes_ in both sets of results?</span>
<span id="cb43-526"><a href="#cb43-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-527"><a href="#cb43-527" aria-hidden="true" tabindex="-1"></a>We can visualize the overlap between the sets of significant genes in a Venn </span>
<span id="cb43-528"><a href="#cb43-528" aria-hidden="true" tabindex="-1"></a>diagram (FDR &lt; 5%). The vast majority of differentially expressed genes is </span>
<span id="cb43-529"><a href="#cb43-529" aria-hidden="true" tabindex="-1"></a>detected with both quantitation approaches (for both up- and down-regulated</span>
<span id="cb43-530"><a href="#cb43-530" aria-hidden="true" tabindex="-1"></a>genes.)</span>
<span id="cb43-531"><a href="#cb43-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-534"><a href="#cb43-534" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-535"><a href="#cb43-535" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 5.5</span></span>
<span id="cb43-536"><a href="#cb43-536" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5.5</span></span>
<span id="cb43-537"><a href="#cb43-537" aria-hidden="true" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">vennDiagram</span>(results, <span class="at">include =</span> <span class="fu">c</span>(<span class="st">"up"</span>, <span class="st">"down"</span>),</span>
<span id="cb43-538"><a href="#cb43-538" aria-hidden="true" tabindex="-1"></a>                   <span class="at">counts.col=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>), <span class="at">mar =</span> <span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</span>
<span id="cb43-539"><a href="#cb43-539" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-540"><a href="#cb43-540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-541"><a href="#cb43-541" aria-hidden="true" tabindex="-1"></a>For example, the following plots show the normalized expression of a few</span>
<span id="cb43-542"><a href="#cb43-542" aria-hidden="true" tabindex="-1"></a>highly differentially expressed genes (known markers of active microglia).</span>
<span id="cb43-543"><a href="#cb43-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-546"><a href="#cb43-546" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-547"><a href="#cb43-547" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4.25</span></span>
<span id="cb43-548"><a href="#cb43-548" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4.25</span></span>
<span id="cb43-549"><a href="#cb43-549" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-550"><a href="#cb43-550" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb43-551"><a href="#cb43-551" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> <span class="fu">topTreat</span>(fit, <span class="at">coef =</span> <span class="st">"Hom"</span>, <span class="at">number =</span> <span class="dv">6</span>)[[<span class="st">"gene_id"</span>]]) {</span>
<span id="cb43-552"><a href="#cb43-552" aria-hidden="true" tabindex="-1"></a> p <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb43-553"><a href="#cb43-553" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">==</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb43-554"><a href="#cb43-554" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> genotype, <span class="at">y =</span> cpm)) <span class="sc">+</span> </span>
<span id="cb43-555"><a href="#cb43-555" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">width =</span> <span class="fl">0.05</span>), <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb43-556"><a href="#cb43-556" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb43-557"><a href="#cb43-557" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> dge<span class="sc">$</span>genes[gene, <span class="st">"gene_name"</span>],</span>
<span id="cb43-558"><a href="#cb43-558" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"Normalized expression (CPM)"</span>,</span>
<span id="cb43-559"><a href="#cb43-559" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb43-560"><a href="#cb43-560" aria-hidden="true" tabindex="-1"></a>         <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"FDR nf-core: %s</span><span class="sc">\n</span><span class="st">FDR GEO: %s"</span>,</span>
<span id="cb43-561"><a href="#cb43-561" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"nfcore"</span>],</span>
<span id="cb43-562"><a href="#cb43-562" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"geo"</span>]</span>
<span id="cb43-563"><a href="#cb43-563" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb43-564"><a href="#cb43-564" aria-hidden="true" tabindex="-1"></a>         ) <span class="sc">+</span></span>
<span id="cb43-565"><a href="#cb43-565" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb43-566"><a href="#cb43-566" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(p)</span>
<span id="cb43-567"><a href="#cb43-567" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-568"><a href="#cb43-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-569"><a href="#cb43-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-570"><a href="#cb43-570" aria-hidden="true" tabindex="-1"></a>Applying a hard FDR threshold can inflate the number of apparent differences,</span>
<span id="cb43-571"><a href="#cb43-571" aria-hidden="true" tabindex="-1"></a>e.g. when a gene is close to the significance threshold (see below).</span>
<span id="cb43-572"><a href="#cb43-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-575"><a href="#cb43-575" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-576"><a href="#cb43-576" aria-hidden="true" tabindex="-1"></a>p_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(</span>
<span id="cb43-577"><a href="#cb43-577" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb43-578"><a href="#cb43-578" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>])</span>
<span id="cb43-579"><a href="#cb43-579" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-580"><a href="#cb43-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-581"><a href="#cb43-581" aria-hidden="true" tabindex="-1"></a>The log2 fold estimates for the <span class="in">`Hom`</span> vs <span class="in">`WT`</span> comparison are highly correlated</span>
<span id="cb43-582"><a href="#cb43-582" aria-hidden="true" tabindex="-1"></a>across the two analysis workflows (Pearson correlation coefficient </span>
<span id="cb43-583"><a href="#cb43-583" aria-hidden="true" tabindex="-1"></a>R = <span class="in">`r signif(p_cor, 2)`</span> ):</span>
<span id="cb43-584"><a href="#cb43-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-587"><a href="#cb43-587" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-588"><a href="#cb43-588" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4.5</span></span>
<span id="cb43-589"><a href="#cb43-589" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4.5</span></span>
<span id="cb43-590"><a href="#cb43-590" aria-hidden="true" tabindex="-1"></a><span class="fu">smoothScatter</span>(</span>
<span id="cb43-591"><a href="#cb43-591" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb43-592"><a href="#cb43-592" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>coefficients[common_genes, <span class="st">"Hom"</span>],</span>
<span id="cb43-593"><a href="#cb43-593" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"nf-core (log2FC)"</span>,</span>
<span id="cb43-594"><a href="#cb43-594" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Xia et al (log2FC)"</span>,</span>
<span id="cb43-595"><a href="#cb43-595" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Homozygous APP vs WT (effect size)"</span></span>
<span id="cb43-596"><a href="#cb43-596" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-597"><a href="#cb43-597" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>, <span class="at">y =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">sprintf</span>(<span class="st">"R = %s"</span>, <span class="fu">signif</span>(p_cor, <span class="dv">2</span>)))</span>
<span id="cb43-598"><a href="#cb43-598" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb43-599"><a href="#cb43-599" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">0</span>, <span class="at">v =</span> <span class="dv">0</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb43-600"><a href="#cb43-600" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-601"><a href="#cb43-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-602"><a href="#cb43-602" aria-hidden="true" tabindex="-1"></a>as are the t-statistics across all examined genes:</span>
<span id="cb43-603"><a href="#cb43-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-606"><a href="#cb43-606" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-607"><a href="#cb43-607" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4.5</span></span>
<span id="cb43-608"><a href="#cb43-608" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4.5</span></span>
<span id="cb43-609"><a href="#cb43-609" aria-hidden="true" tabindex="-1"></a>p_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(</span>
<span id="cb43-610"><a href="#cb43-610" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb43-611"><a href="#cb43-611" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>])</span>
<span id="cb43-612"><a href="#cb43-612" aria-hidden="true" tabindex="-1"></a><span class="fu">smoothScatter</span>(</span>
<span id="cb43-613"><a href="#cb43-613" aria-hidden="true" tabindex="-1"></a>  fit<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>], </span>
<span id="cb43-614"><a href="#cb43-614" aria-hidden="true" tabindex="-1"></a>  fit_nfcore<span class="sc">$</span>t[common_genes, <span class="st">"Hom"</span>],</span>
<span id="cb43-615"><a href="#cb43-615" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"nf-core (t-statistic)"</span>,</span>
<span id="cb43-616"><a href="#cb43-616" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Xia et al (t-statistic)"</span>,</span>
<span id="cb43-617"><a href="#cb43-617" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Homozygous APP vs WT (t-statistic)"</span>)</span>
<span id="cb43-618"><a href="#cb43-618" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>, <span class="at">y =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">sprintf</span>(<span class="st">"R = %s"</span>, <span class="fu">signif</span>(p_cor, <span class="dv">2</span>)))</span>
<span id="cb43-619"><a href="#cb43-619" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb43-620"><a href="#cb43-620" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">0</span>, <span class="at">v =</span> <span class="dv">0</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb43-621"><a href="#cb43-621" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-622"><a href="#cb43-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-623"><a href="#cb43-623" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Discordant significance calls</span></span>
<span id="cb43-624"><a href="#cb43-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-627"><a href="#cb43-627" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-628"><a href="#cb43-628" aria-hidden="true" tabindex="-1"></a><span class="co"># genes detected in GEO, but not significant with nf-core</span></span>
<span id="cb43-629"><a href="#cb43-629" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">row.names</span>(results)[<span class="fu">which</span>(<span class="fu">abs</span>(results[, <span class="dv">1</span>]) <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> results[, <span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>)]</span>
<span id="cb43-630"><a href="#cb43-630" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-631"><a href="#cb43-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-632"><a href="#cb43-632" aria-hidden="true" tabindex="-1"></a>At FDR &lt; 5% <span class="in">`r length(genes)`</span> genes were reported as significantly </span>
<span id="cb43-633"><a href="#cb43-633" aria-hidden="true" tabindex="-1"></a>differentially expressed with the original Xia et al count matrix</span>
<span id="cb43-634"><a href="#cb43-634" aria-hidden="true" tabindex="-1"></a>but *not* with the output of the nf-core/rnaseq workflow.</span>
<span id="cb43-635"><a href="#cb43-635" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-636"><a href="#cb43-636" aria-hidden="true" tabindex="-1"></a>As side-by-side comparison of the FDR (<span class="in">`adj.P.Val`</span>) for these genes confirms</span>
<span id="cb43-637"><a href="#cb43-637" aria-hidden="true" tabindex="-1"></a>that the vast majority display significant close to the 5% threshold in the </span>
<span id="cb43-638"><a href="#cb43-638" aria-hidden="true" tabindex="-1"></a>nf-core/rnaseq output as well. (This is in line with the high overall</span>
<span id="cb43-639"><a href="#cb43-639" aria-hidden="true" tabindex="-1"></a>correlation of the t-statistics observed above.)</span>
<span id="cb43-640"><a href="#cb43-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-643"><a href="#cb43-643" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-644"><a href="#cb43-644" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tt[genes, ])</span>
<span id="cb43-645"><a href="#cb43-645" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-646"><a href="#cb43-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-647"><a href="#cb43-647" aria-hidden="true" tabindex="-1"></a>Finally, we plot the normalized gene expression estimates for the </span>
<span id="cb43-648"><a href="#cb43-648" aria-hidden="true" tabindex="-1"></a><span class="in">`r length(genes)`</span> discordant genes.</span>
<span id="cb43-649"><a href="#cb43-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-652"><a href="#cb43-652" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-653"><a href="#cb43-653" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 3</span></span>
<span id="cb43-654"><a href="#cb43-654" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4.25</span></span>
<span id="cb43-655"><a href="#cb43-655" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb43-656"><a href="#cb43-656" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb43-657"><a href="#cb43-657" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> genes) {</span>
<span id="cb43-658"><a href="#cb43-658" aria-hidden="true" tabindex="-1"></a> p <span class="ot">&lt;-</span> cpms <span class="sc">%&gt;%</span></span>
<span id="cb43-659"><a href="#cb43-659" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">==</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb43-660"><a href="#cb43-660" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> genotype, <span class="at">y =</span> cpm)) <span class="sc">+</span> </span>
<span id="cb43-661"><a href="#cb43-661" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">width =</span> <span class="fl">0.05</span>), <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb43-662"><a href="#cb43-662" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb43-663"><a href="#cb43-663" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> dge<span class="sc">$</span>genes[gene, <span class="st">"gene_name"</span>],</span>
<span id="cb43-664"><a href="#cb43-664" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"Normalized expression (CPM)"</span>,</span>
<span id="cb43-665"><a href="#cb43-665" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb43-666"><a href="#cb43-666" aria-hidden="true" tabindex="-1"></a>         <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"FDR nf-core: %s</span><span class="sc">\n</span><span class="st">FDR GEO: %s"</span>,</span>
<span id="cb43-667"><a href="#cb43-667" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"nfcore"</span>],</span>
<span id="cb43-668"><a href="#cb43-668" aria-hidden="true" tabindex="-1"></a>                       tt[gene, <span class="st">"geo"</span>]</span>
<span id="cb43-669"><a href="#cb43-669" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb43-670"><a href="#cb43-670" aria-hidden="true" tabindex="-1"></a>         ) <span class="sc">+</span></span>
<span id="cb43-671"><a href="#cb43-671" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme_linedraw</span>(<span class="dv">14</span>)</span>
<span id="cb43-672"><a href="#cb43-672" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(p)</span>
<span id="cb43-673"><a href="#cb43-673" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-674"><a href="#cb43-674" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-675"><a href="#cb43-675" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-676"><a href="#cb43-676" aria-hidden="true" tabindex="-1"></a><span class="fu">## Conclusions</span></span>
<span id="cb43-677"><a href="#cb43-677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-678"><a href="#cb43-678" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Differential expression analyses of raw counts obtained with the </span>
<span id="cb43-679"><a href="#cb43-679" aria-hidden="true" tabindex="-1"></a>  <span class="in">`nc-core/rnaseq`</span> workflow yields results that are highly concordant with those</span>
<span id="cb43-680"><a href="#cb43-680" aria-hidden="true" tabindex="-1"></a>  obtained with the raw counts the authors deposited in NCBI GEO.</span>
<span id="cb43-681"><a href="#cb43-681" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>With appropriate parameters the <span class="in">`nf-core/rnaseq`</span> workflow can be applied to</span>
<span id="cb43-682"><a href="#cb43-682" aria-hidden="true" tabindex="-1"></a>  QuantSeq FWD 3'tag RNA-seq data.</span>
<span id="cb43-683"><a href="#cb43-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-684"><a href="#cb43-684" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reproducibility</span></span>
<span id="cb43-685"><a href="#cb43-685" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-686"><a href="#cb43-686" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb43-687"><a href="#cb43-687" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;</span></span>
<span id="cb43-688"><a href="#cb43-688" aria-hidden="true" tabindex="-1"></a>SessionInfo</span>
<span id="cb43-689"><a href="#cb43-689" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/summary&gt;</span></span>
<span id="cb43-690"><a href="#cb43-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-693"><a href="#cb43-693" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-694"><a href="#cb43-694" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb43-695"><a href="#cb43-695" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>()</span>
<span id="cb43-696"><a href="#cb43-696" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-697"><a href="#cb43-697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-698"><a href="#cb43-698" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>