<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-03-12">

<title>Thomas Sandmann’s blog - Lemur: analyzing multi-condition single-cell data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Lemur: analyzing multi-condition single-cell data</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">TIL</div>
                <div class="quarto-category">R</div>
                <div class="quarto-category">scRNAseq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">March 12, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<p>This week, <a href="https://const-ae.name/">Constantin Ahlmann-Eltze</a> and <a href="https://www.embl.org/groups/huber/">Wolfgang Huber</a> published a <a href="https://www.biorxiv.org/content/10.1101/2023.03.06.531268v1">preprint</a> describing <code>LEMUR</code>, a new approach to analyzing single-cell experiments that include samples from multiple conditions, e.g.&nbsp;drug treatments, disease-status, etc.</p>
<p>To date, such analyses often involve two separate steps, e.g.</p>
<ol type="1">
<li>Defining clusters of cells, e.g.&nbsp;cell types or states</li>
<li>Differential expression analysis <em>within</em> each cell type</li>
</ol>
<p>In contrast, <code>LEMUR</code> considers the continuous latent space the individual cells occupy, incorporating the design of the experiment, and then performs differential expression analysis in this embedding space.</p>
<p>An R package implementing <code>LEMUR</code> is available from <a href="https://github.com/const-ae/lemur">github</a> and includes an example dataset <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<section id="ellwanger-et-al-comparing-trem2-wildtype-and-knock-out-mouse-microglia" class="level2">
<h2 class="anchored" data-anchor-id="ellwanger-et-al-comparing-trem2-wildtype-and-knock-out-mouse-microglia">Ellwanger et al: Comparing Trem2 wildtype and knock-out mouse microglia</h2>
<p>Here, I am exploring <code>LEMUR</code> by examining scRNA-seq data published by <a href="https://pubmed.ncbi.nlm.nih.gov/33446504/">Ellwanger et al, 2021</a>, who injected three strains of <a href="https://www.jax.org/strain/008730">5XFAD mice</a>, a murine model of familial Alzheimer’s Disease, either</p>
<ol type="1">
<li>carrying the wild-type (WT) <em>Trem2</em> gene,</li>
<li>carrying the R47H <em>Trem2</em> variant, believed to be a loss-of-function variant,</li>
<li>or completely lacking <em>Trem2</em> expression</li>
</ol>
<p>with either a Trem2 agonist (<code>hT2AB</code>) or a negative control antibody (<code>hIgG1</code>).</p>
<p>48 hours later, the authors isolated CD45-positive <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> cells from the cortex and performed single-cell RNA-seq analysis using the 10X Genomics platform.</p>
<section id="retrieving-the-data" class="level3">
<h3 class="anchored" data-anchor-id="retrieving-the-data">Retrieving the data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Ellwanger et al made both raw and processed data available via the NCBI GEO and SRA repositories under <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156183">GEO accession GSE156183</a>.</p>
<p>They also included complete metadata for each cell, making this a great dataset for re-analysis.</p>
<p>Let’s start by retrieving the</p>
<ol type="1">
<li><a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file&amp;file=GSE156183%5FRAW%2Emtx%2Egz">processed counts (500 Mb)</a> and the</li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file&amp;file=GSE156183%5FCell%5Fmetadata%2Etsv%2Egz">cell metadata (13 Mb)</a></li>
</ol>
<p>files from GEO and store them in a temporary directory:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>temp_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"ellwanger"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(temp_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="dv">360</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>url_root <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc="</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"GSE156183&amp;format=file&amp;file=GSE156183%5F"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"counts.mtx.gz"</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(url_root, <span class="st">"RAW%2Emtx%2Egz"</span>), </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> raw_counts)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>cell_metadata <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"cell_metadata.tsv.gz"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(url_root, <span class="st">"Cell%5Fmetadata%2Etsv%2Egz"</span>), </span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> cell_metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and read the sparse count matrix into our R session:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(raw_counts)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>cell_anno <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(cell_metadata, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">nrow</span>(cell_anno) <span class="sc">==</span> <span class="fu">ncol</span>(m))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Unfortunately, the <code>GSE156183_Feature_metadata.tsv.gz</code> feature (= gene) annotation file the authors deposited with GEO <em>actually</em> contains <em>cell</em> annotations. But luckily, they also deposited counts matrices in TSV format for each sample, which include the ENSEMBL gene identifier for each row.</p>
<p>Here, I download the TAR archive that contains all of the TSV files, and then extract the gene identifiers from one of the files so I can add them to the experiment-wide raw count matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>selected_sample <span class="ot">&lt;-</span> <span class="st">"GSM4726219_RAW-R47H-male-IgG-rep2.tsv.gz"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>tar_archive <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"RAW.tar"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> tar_archive)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>utils<span class="sc">::</span><span class="fu">untar</span>(tar_archive, <span class="at">files =</span> selected_sample, <span class="at">exdir =</span> <span class="fu">tempdir</span>())</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>gene_ids <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="fu">file.path</span>(<span class="fu">tempdir</span>(), selected_sample), </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col_select =</span> <span class="fu">any_of</span>(<span class="st">"feature_id"</span>),</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(feature_id)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(gene_ids) <span class="sc">==</span> <span class="fu">nrow</span>(m))</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(m) <span class="ot">&lt;-</span> gene_ids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="creating-a-singlecellexperiment-object" class="level3">
<h3 class="anchored" data-anchor-id="creating-a-singlecellexperiment-object">Creating a SingleCellExperiment object</h3>
<p>Now I have all the pieces of information required to create a <code>SingleCellExperiment</code>:</p>
<ul>
<li>the raw counts (in the form of a sparse matrix),</li>
<li>the cell annotations (in the form of a data.frame)</li>
<li>the two UMAP dimensions used by the authors (included in the cell metadata).</li>
</ul>
<p>I choose to retain only a subset of the (many) cell-level annotation columns, add gene symbols as row annotations, extract the UMAP coordinates into a separate matrix - and store all of it in the <code>sce</code> object.</p>
<p>Next, I am removing cells without an assigned cell type, and also add a coarser cell type annotation that collapses the different microglia states reported by the authors into a single category. Finally, I filter genes without a valid gene symbol and add an assay slot with the normalized log2 counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>col_data <span class="ot">&lt;-</span> cell_anno <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(cell_id, celltype, sample, sex, genotype, treatment, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                <span class="fu">starts_with</span>(<span class="st">"QC."</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"cell_id"</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(m) <span class="ot">&lt;-</span> <span class="fu">row.names</span>(col_data)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>row_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">symbol =</span> AnnotationDbi<span class="sc">::</span><span class="fu">mapIds</span>(org.Mm.eg.db, <span class="at">keys =</span> gene_ids,</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">column =</span> <span class="st">"SYMBOL"</span>, <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>),</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> gene_ids</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>umap <span class="ot">&lt;-</span> cell_anno <span class="sc">%&gt;%</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">ends_with</span>(<span class="st">"CD45pos"</span>)</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(umap) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(m)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(umap) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"UMAP"</span>, <span class="fu">seq.int</span>(<span class="fu">ncol</span>(umap)))</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> m),</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData =</span> row_data,</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_data,</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">reducedDims =</span> <span class="fu">list</span>(<span class="at">UMAP =</span> umap)</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="sc">!</span><span class="fu">is.na</span>(sce<span class="sc">$</span>celltype)]</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>celltype_coarse <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grepl</span>(<span class="at">x =</span> sce<span class="sc">$</span>celltype, <span class="at">pattern =</span> <span class="st">"Microglia"</span>) <span class="sc">~</span> <span class="st">"Microglia"</span>,</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> sce<span class="sc">$</span>celltype</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>treatment <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>treatment, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IgG"</span>, <span class="st">"hT2AB"</span>))</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>mg_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">sub</span>(<span class="st">"Microglia."</span>, <span class="st">""</span>, sce<span class="sc">$</span>celltype, <span class="at">fixed =</span> <span class="cn">TRUE</span>))</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[<span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>symbol), ]</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">c</span>(<span class="st">"m"</span>, <span class="st">"cell_anno"</span>, <span class="st">"gene_ids"</span>, <span class="st">"row_data"</span>, <span class="st">"col_data"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This <code>SingleCellExperiment</code> object is now ready for downstream analysis.</p>
</section>
</section>
<section id="subsetting-the-experiment-to-samples-of-interest" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-the-experiment-to-samples-of-interest">Subsetting the experiment to samples of interest</h2>
<p>This study contains multiple experimental variables, e.g.&nbsp;each sample is annotated with one of the three <code>genotypes</code>, one of two <code>treatments</code> and the <code>sex</code> for each mouse.</p>
<p>Here, I will focus only on the difference between TREM2 wildtype and TREM2 knock-out animals treated with the <code>IgG</code> control antibody. Only female knock-out animals were included in the study, so I exclude the male animals from the other strain as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="fu">which</span>(sce<span class="sc">$</span>genotype <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"TREM2_CV-5XFAD"</span>, <span class="st">"Trem2_KO-5XFAD"</span>))]</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="fu">which</span>(sce<span class="sc">$</span>sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">&amp;</span> sce<span class="sc">$</span>treatment <span class="sc">==</span> <span class="st">"IgG"</span>)]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(<span class="fu">colData</span>(sce), <span class="fu">table</span>(genotype, treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                treatment
genotype          IgG hT2AB
  TREM2_CV-5XFAD 3781     0
  Trem2_KO-5XFAD 8911     0</code></pre>
</div>
</div>
<p>After subsetting, the experiment now contains 5 samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce) <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(sample, treatment, genotype) <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">remove_rownames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              sample treatment       genotype
1 CV-female-IgG-rep1       IgG TREM2_CV-5XFAD
2 CV-female-IgG-rep2       IgG TREM2_CV-5XFAD
3 KO-female-IgG-rep1       IgG Trem2_KO-5XFAD
4 KO-female-IgG-rep2       IgG Trem2_KO-5XFAD
5 KO-female-IgG-rep3       IgG Trem2_KO-5XFAD</code></pre>
</div>
</div>
<p>At this point, I can reproduce e.g.&nbsp;a version of <a href="https://www.pnas.org/doi/10.1073/pnas.2017742118#fig03">Figure 3E</a> from the original paper, using the UMAP coordinates and cell type labels provided by the authors. (My version of the figure only includes cells from the selected subset of samples, not all cells captured in the study.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Microglia"</span> <span class="ot">=</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"T cells"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Macrophages"</span> <span class="ot">=</span> <span class="st">"firebrick"</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">"MO:T"</span> <span class="ot">=</span> <span class="st">"darkgreen"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Dendritic cells"</span> <span class="ot">=</span> <span class="st">"green"</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Monocytes"</span> <span class="ot">=</span> <span class="st">"orange"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>            <span class="st">"B cells"</span> <span class="ot">=</span> <span class="st">"navy"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Neutrophils"</span> <span class="ot">=</span> <span class="st">"darkblue"</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>            <span class="st">"HCS"</span> <span class="ot">=</span> <span class="st">"grey"</span>, </span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Fibroblasts"</span> <span class="ot">=</span> <span class="st">"yellow"</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"celltype_coarse"</span>) <span class="sc">+</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors, <span class="at">name =</span> <span class="st">"Cell type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Because Ellwanger et al captured all cells with CD45 expression, the dataset includes other immune cell types besides microglia. Let’s remove those to focus only on the latter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, sce<span class="sc">$</span>celltype_coarse <span class="sc">==</span> <span class="st">"Microglia"</span>]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>mg_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>mg_type, </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Resting"</span>, <span class="st">"t1"</span>, <span class="st">"t2"</span>, <span class="st">"t3"</span>, <span class="st">"t4"</span>, <span class="st">"t5"</span>, <span class="st">"t6"</span>, <span class="st">"IFN-R"</span>, <span class="st">"DAM"</span>, </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>             <span class="st">"MHC-II"</span>, <span class="st">"Cyc-M"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Most microglial states were captured in animals from both genotypes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>mg_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cyc-M"</span> <span class="ot">=</span> <span class="st">"navy"</span>,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"DAM"</span> <span class="ot">=</span> <span class="st">"darkgreen"</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"IFN-R"</span> <span class="ot">=</span> <span class="st">"#C12131"</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"MHC-II"</span> <span class="ot">=</span> <span class="st">"green"</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Resting"</span> <span class="ot">=</span> <span class="st">"grey50"</span>,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t1"</span> <span class="ot">=</span> <span class="st">"#FDF5EB"</span>,   </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t2"</span> <span class="ot">=</span>  <span class="st">"#FFE2C0"</span>, </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t3"</span> <span class="ot">=</span> <span class="st">"#FFC08E"</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t4"</span> <span class="ot">=</span> <span class="st">"#FE945C"</span>,</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t5"</span> <span class="ot">=</span>  <span class="st">"#EC5D2F"</span>,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t6"</span> <span class="ot">=</span> <span class="st">"#C12131"</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>) <span class="sc">+</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sce<span class="sc">$</span>genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="differential-expression-analysis-with-lemur" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression-analysis-with-lemur">Differential expression analysis with lemur</h2>
<p>Now I am ready to explore the <code>lemur</code> R package to ask: “which neighborhoods show for differential expression between samples from WT and knock-out animals?”</p>
<p>The following steps closely follow the examples outlined on the <a href="https://github.com/const-ae/lemur">lemor github repository’s README</a> - many thanks for the great documentation, Constantin! (All mistakes and misunderstandings in this post are my own, as always.)</p>
<section id="dependencies-installation" class="level3">
<h3 class="anchored" data-anchor-id="dependencies-installation">Dependencies &amp; installation</h3>
<p>Following the instructions from the <a href="https://github.com/const-ae/lemur">lemu github repository</a> I then installed the latest version of the <a href="https://github.com/const-ae/glmGamPoi">glmGamPoi</a> package, and then the <code>lemur</code> package itself from their github repositories.</p>
<p>To harmonize results across batches (in this case: samples), I will use harmony, so I need to install it from its github repository as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"const-ae/glmGamPoi"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"const-ae/lemur"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"immunogenomics/harmony"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="subsetting-the-experiment" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-the-experiment">Subsetting the experiment</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lemur)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> 1000L</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To speed up my exploration of the <code>LEMUR</code> workflow, I subset the experiment to 1000 random cells from each of the two genotypes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(1L)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>genotypes <span class="ot">&lt;-</span> <span class="fu">unique</span>(sce<span class="sc">$</span>genotype)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>selected_cells <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">sapply</span>(genotypes, \(g) {</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample</span>(<span class="fu">which</span>(sce<span class="sc">$</span>genotype <span class="sc">==</span> g), n_cells)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>}))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As expected, most microglial states described in the paper remain represented in the downsampled dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>celltype[selected_cells], sce<span class="sc">$</span>genotype[selected_cells])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   
                    TREM2_CV-5XFAD Trem2_KO-5XFAD
  Microglia.Cyc-M               17             32
  Microglia.DAM                122            127
  Microglia.IFN-R               46             33
  Microglia.MHC-II              16              0
  Microglia.Resting            101            117
  Microglia.t1                 145            133
  Microglia.t2                 119             71
  Microglia.t3                  84            140
  Microglia.t4                 161             93
  Microglia.t5                 132            209
  Microglia.t6                  57             45</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  sce[, selected_cells], <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>) <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells))) <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sce<span class="sc">$</span>genotype[selected_cells])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="fitting-the-lemur-model" class="level3">
<h3 class="anchored" data-anchor-id="fitting-the-lemur-model">Fitting the LEMUR model</h3>
<p>Next, I fit the latent embedding multivariate regression (LEMUR) model with the <code>lemur()</code> function. Because the dataset is relatively homogeneous, e.g.&nbsp;it contains only microglia, I chose to consider only 25 Principal Components and used 15 dimensions for the LEMUR embedding (e.g.&nbsp;the default number).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> lemur<span class="sc">::</span><span class="fu">lemur</span>(sce[, selected_cells], <span class="at">design =</span> <span class="sc">~</span> genotype,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_ambient =</span> <span class="dv">25</span>, <span class="at">n_embedding =</span> <span class="dv">15</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Because each sample was processed in a separate channel of the 10X Genomics microfluidics device, I am aligning the embeddings of similar cell clusters using <a href="https://www.nature.com/articles/s41592-019-0619-0">harmony</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> lemur<span class="sc">::</span><span class="fu">align_harmony</span>(fit, <span class="at">stretching =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>fit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: lemur_fit 
dim: 23870 2000 
metadata(12): n_ambient n_embedding ... alignment_design
  alignment_design_matrix
assays(1): expr
rownames(23870): ENSMUSG00000051951 ENSMUSG00000025900 ...
  ENSMUSG00000094915 ENSMUSG00000079808
rowData names(1): symbol
colnames(2000): CELL21559 CELL21072 ... CELL45651 CELL52059
colData names(22): celltype sample ... celltype_coarse mg_type
reducedDimNames(2): linearFit embedding
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p>The returned <code>lemur_fit</code> object contains the <code>embedding</code> matrix, the latent space in which the differential expression analysis is performed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(fit<span class="sc">$</span>embedding)  <span class="co"># 15 dimensions, as specified above</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   15 2000</code></pre>
</div>
</div>
<p>Let’s plot the first two dimensions against each other, coloring each cell by the microglial state Ellwanger et al identified through Louvain clustering. (This information has not been used by <code>lemur</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot dim 1 vs dim 2</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  fit, <span class="st">"embedding"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>, <span class="at">shape_by =</span> <span class="st">"genotype"</span>) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Embedding after accounting for genotype"</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Cells group by Ellwanger et al’s subtype labels, and cells from both genotypes are intermixed. We can obtain an alternative visualization by arranging the cells in two dimensions using <a href="https://umap-learn.readthedocs.io/en/latest/">Uniform Manifold Approximation and Projection for Dimension Reduction (UMAP)</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run UMAP on the embedding</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>umap <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">umap</span>(<span class="fu">t</span>(fit<span class="sc">$</span>embedding))</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(umap) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"UMAP 1"</span>, <span class="st">"UMAP 2"</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(fit, <span class="st">"UMAP"</span>) <span class="ot">&lt;-</span> umap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  fit, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>, <span class="at">shape_by =</span> <span class="st">"genotype"</span>) <span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Embedding after accounting for genotype (UMAP)"</span>,</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells))) <span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> <span class="fu">colData</span>(fit)<span class="sc">$</span>genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="testing-for-differential-expression" class="level3">
<h3 class="anchored" data-anchor-id="testing-for-differential-expression">Testing for differential expression</h3>
<p>Next, the <code>test_de</code> function performs a differential expression analysis for locations in the embedding - by default, it will estimate it for each location an original cell was mapped to.</p>
<p>The <code>find_de_neighborhoods</code> function accepts the original counts and will estimate the log2 fold change for each neighborhood, based on aggregating the counts to pseudobulk measures across the cells in each neighborhood.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">test_de</span>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  fit, </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">contrast =</span> <span class="fu">cond</span>(<span class="at">genotype =</span> <span class="st">"Trem2_KO-5XFAD"</span>) <span class="sc">-</span> </span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cond</span>(<span class="at">genotype =</span> <span class="st">"TREM2_CV-5XFAD"</span>))</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>neighborhoods <span class="ot">&lt;-</span> <span class="fu">find_de_neighborhoods</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  fit, </span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> <span class="fu">counts</span>(sce)[, selected_cells],</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">group_by =</span> <span class="fu">vars</span>(sample, genotype),</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">include_complement =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(pval) <span class="sc">%&gt;%</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(fit)), <span class="st">"gene_id"</span>), </span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="at">name =</span> <span class="st">"gene_id"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(symbol, <span class="fu">everything</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>neighborhoods</code> data.frame contains differential expression statistics for each gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(neighborhoods)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 12
  symbol name       region indices n_cells   mean     pval adj_p…¹ f_sta…²   df1
  &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;  &lt;I&lt;lis&gt;   &lt;int&gt;  &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;int&gt;
1 Fxyd5  ENSMUSG00… 1      &lt;int&gt;      1809 -0.251 6.09e-10 1.45e-5   118.      1
2 Il4i1  ENSMUSG00… 1      &lt;int&gt;      1069 -0.181 2.40e- 9 2.87e-5   101.      1
3 Lpl    ENSMUSG00… 1      &lt;int&gt;       830 -0.469 5.82e- 9 4.63e-5    90.6     1
4 Cd74   ENSMUSG00… 1      &lt;int&gt;       611 -1.12  9.25e- 9 5.52e-5    85.7     1
5 Axl    ENSMUSG00… 1      &lt;int&gt;      1584 -0.257 1.78e- 8 8.49e-5    79.2     1
6 H2-Aa  ENSMUSG00… 1      &lt;int&gt;      1172 -0.247 6.81e- 8 2.71e-4    67.1     1
# … with 2 more variables: df2 &lt;dbl&gt;, lfc &lt;dbl&gt;, and abbreviated variable names
#   ¹​adj_pval, ²​f_statistic</code></pre>
</div>
</div>
<p>A volcano plot shows that we recovered a number of genes with differential expression in one or more neighborhoods (after accounting for multiple testing):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># volcano plot</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lfc, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pval))) <span class="sc">+</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color  =</span> adj_pval <span class="sc">&lt;</span> <span class="fl">0.1</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Volcano plot of the neighborhoods"</span>) <span class="sc">+</span> </span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"firebrick"</span>, <span class="st">"FALSE"</span> <span class="ot">=</span> <span class="st">"grey"</span>)) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For example, transcripts of the <a href="https://en.wikipedia.org/wiki/Lipoprotein_lipase">Lipoprotein lipase (Lpl) gene</a> are generally expressed at lower levels in the Trem2 knock-out than in wildtype samples.</p>
<p>But there is also evidence for stronger differences in expression in microglia that have adopted specific states. The largest log2 fold changes are observed in <code>Cyc-M</code>, and <code>DAM</code> microglia, while the <code>Resting</code> microglia are mainly excluded from the neighborhood detected by <code>lemur</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sel_gene <span class="ot">&lt;-</span> <span class="fu">row.names</span>(sce)[<span class="fu">which</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>symbol <span class="sc">==</span> <span class="st">"Lpl"</span>)]</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>neighborhood_coordinates <span class="ot">&lt;-</span> neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> sel_gene) <span class="sc">%&gt;%</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">cell_id =</span> purrr<span class="sc">:::</span><span class="fu">map</span>(indices, \(idx) <span class="fu">colnames</span>(fit)[idx])) <span class="sc">%&gt;%</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">unnest</span>(<span class="fu">c</span>(indices, cell_id)) <span class="sc">%&gt;%</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(<span class="fu">as_tibble</span>(umap, <span class="at">rownames =</span> <span class="st">"cell_id"</span>), <span class="at">by =</span> <span class="st">"cell_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(name, cell_id, <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(umap) <span class="sc">%&gt;%</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">expr =</span> <span class="fu">assay</span>(fit, <span class="st">"DE"</span>)[sel_gene, ]) <span class="sc">%&gt;%</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>() <span class="sc">+</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> expr)) <span class="sc">+</span></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density2d</span>(<span class="at">data =</span> neighborhood_coordinates, <span class="at">breaks =</span> <span class="fl">0.1</span>, </span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>                 <span class="at">contour_var =</span> <span class="st">"ndensity"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>]) <span class="sc">+</span> </span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(umap) <span class="sc">%&gt;%</span></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_cols</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(fit))) <span class="sc">%&gt;%</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> mg_type)) <span class="sc">+</span></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density2d</span>(<span class="at">data =</span> neighborhood_coordinates, <span class="at">breaks =</span> <span class="fl">0.1</span>, </span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a>                 <span class="at">contour_var =</span> <span class="st">"ndensity"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Microglia states"</span>) <span class="sc">+</span> </span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next, I examine each microglial subtype separately, and split cells according to whether they fall into a neighborhood with significant differential <code>Lpl</code> expression (<code>DE</code>) or not (<code>not DE</code>).</p>
<p>Most <code>Cyc-M</code> and <code>DAM</code> microglia are located in neighborhoods with reduced <code>Lpl</code> expression in knock-out samples, e.g.&nbsp;the majority of these cells is in the <code>DE</code> column. The opposite is true for <code>Resting</code> microglia: nearly all of them are outside the significant <code>Lpl</code> differential expression neighborhood.</p>
<p>The transitional subtypes, t1 (most similar to resting microglia) to t6 (most similar to DAM, Cyc-M or IFN-R) fall in between, with a gradual increase along their proposed differentiation sequence.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>within_neighborhood <span class="ot">&lt;-</span> neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> sel_gene) <span class="sc">%&gt;%</span> </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(indices)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(fit)<span class="sc">$</span>neighborhood <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq.int</span>(<span class="fu">ncol</span>(fit)) <span class="sc">%in%</span> within_neighborhood[[<span class="dv">1</span>]], <span class="st">"DE"</span>, <span class="st">"not DE"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(fit)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> neighborhood, <span class="at">fill =</span> mg_type)) <span class="sc">+</span> </span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"count"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> mg_type) <span class="sc">+</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>], </span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Cells per neighborhood"</span>) <span class="sc">+</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally, we can plot the estimated <code>Lpl</code> log2 fold changes for each cells annotated with the various subtype labels. This confirms the observations outlined above, providing an example of how insights from LEMUR can be combined with coarser, clustering-based insights.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">expr =</span> <span class="fu">assay</span>(fit, <span class="st">"DE"</span>)[sel_gene, ],</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(fit)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> mg_type, <span class="at">y =</span> expr, <span class="at">fill =</span> mg_type)) <span class="sc">+</span> </span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span> </span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>],</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Genotype effect (log2 FC)"</span>, </span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
</section>
<section id="conclusions" class="level2">
<h2 class="anchored" data-anchor-id="conclusions">Conclusions</h2>
<ul>
<li>This dataset starts off with a relatively homogeneous cell population, e.g. it is designed to examine subtle differences <em>within</em> a single cell type (microglia.)</li>
<li>Removing <em>Trem2</em> activity is known to shift the composition of the microglial subsets, e.g.&nbsp;depleting <code>DAM</code> microglia and increasing the frequency of <code>Resting</code> in the knock-out versus the wildtype samples. This adds an additional challenge to the task of identifying differential expression.</li>
<li>The <code>lemur()</code> package nevertheless successfully identified genes that track the differential expression of the microglial sub-states reported by Ellwanger et al.&nbsp;(Even with only a subsample of the data.)</li>
<li>I am looking forward to exploring <code>LEMUR</code> in future datasets, e.g.&nbsp;those examining the effects of drug perturbation in single-nuclei RNA-seq datasets sampling a large variety of CNS cell types.</li>
</ul>
<details>
<summary>
Reproducibility
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.2 (2022-10-31)
 os       macOS Big Sur ... 10.16
 system   x86_64, darwin17.0
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Los_Angeles
 date     2023-03-12
 pandoc   2.19.2 @ /Applications/RStudio.app/Contents/MacOS/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version  date (UTC) lib source
 AnnotationDbi        * 1.60.0   2022-11-01 [1] Bioconductor
 askpass                1.1      2019-01-13 [1] CRAN (R 4.2.0)
 assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.2.0)
 beachmat               2.14.0   2022-11-01 [1] Bioconductor
 beeswarm               0.4.0    2021-06-01 [1] CRAN (R 4.2.0)
 Biobase              * 2.58.0   2022-11-01 [1] Bioconductor
 BiocGenerics         * 0.44.0   2022-11-01 [1] Bioconductor
 BiocNeighbors          1.16.0   2022-11-01 [1] Bioconductor
 BiocParallel           1.32.4   2022-12-01 [1] Bioconductor
 BiocSingular           1.14.0   2022-11-01 [1] Bioconductor
 Biostrings             2.66.0   2022-11-01 [1] Bioconductor
 bit                    4.0.5    2022-11-15 [1] CRAN (R 4.2.0)
 bit64                  4.0.5    2020-08-30 [1] CRAN (R 4.2.0)
 bitops                 1.0-7    2021-04-24 [1] CRAN (R 4.2.0)
 blob                   1.2.3    2022-04-10 [1] CRAN (R 4.2.0)
 cachem                 1.0.6    2021-08-19 [1] CRAN (R 4.2.0)
 cli                    3.5.0    2022-12-20 [1] CRAN (R 4.2.0)
 codetools              0.2-18   2020-11-04 [2] CRAN (R 4.2.2)
 colorspace             2.0-3    2022-02-21 [1] CRAN (R 4.2.0)
 cowplot                1.1.1    2020-12-30 [1] CRAN (R 4.2.0)
 crayon                 1.5.2    2022-09-29 [1] CRAN (R 4.2.0)
 credentials            1.3.2    2021-11-29 [1] CRAN (R 4.2.0)
 DBI                    1.1.3    2022-06-18 [1] CRAN (R 4.2.0)
 DelayedArray           0.24.0   2022-11-01 [1] Bioconductor
 DelayedMatrixStats     1.20.0   2022-11-01 [1] Bioconductor
 digest                 0.6.31   2022-12-11 [1] CRAN (R 4.2.0)
 dplyr                * 1.0.10   2022-09-01 [1] CRAN (R 4.2.0)
 ellipsis               0.3.2    2021-04-29 [1] CRAN (R 4.2.0)
 evaluate               0.19     2022-12-13 [1] CRAN (R 4.2.0)
 expm                   0.999-7  2023-01-09 [1] CRAN (R 4.2.0)
 fansi                  1.0.3    2022-03-24 [1] CRAN (R 4.2.0)
 farver                 2.1.1    2022-07-06 [1] CRAN (R 4.2.0)
 fastmap                1.1.0    2021-01-25 [1] CRAN (R 4.2.0)
 FNN                    1.1.3.1  2022-05-23 [1] CRAN (R 4.2.0)
 generics               0.1.3    2022-07-05 [1] CRAN (R 4.2.0)
 GenomeInfoDb         * 1.34.4   2022-12-01 [1] Bioconductor
 GenomeInfoDbData       1.2.9    2022-12-12 [1] Bioconductor
 GenomicRanges        * 1.50.2   2022-12-16 [1] Bioconductor
 ggbeeswarm             0.7.1    2022-12-16 [1] CRAN (R 4.2.0)
 ggplot2              * 3.4.0    2022-11-04 [1] CRAN (R 4.2.0)
 ggrepel                0.9.2    2022-11-06 [1] CRAN (R 4.2.0)
 glmGamPoi              1.11.7   2023-03-11 [1] Github (const-ae/glmGamPoi@78c5cff)
 glue                   1.6.2    2022-02-24 [1] CRAN (R 4.2.0)
 gridExtra              2.3      2017-09-09 [1] CRAN (R 4.2.0)
 gtable                 0.3.1    2022-09-01 [1] CRAN (R 4.2.0)
 harmony                0.1.1    2023-03-11 [1] Github (immunogenomics/harmony@63ebd73)
 here                   1.0.1    2020-12-13 [1] CRAN (R 4.2.0)
 hms                    1.1.2    2022-08-19 [1] CRAN (R 4.2.0)
 htmltools              0.5.4    2022-12-07 [1] CRAN (R 4.2.0)
 htmlwidgets            1.5.4    2021-09-08 [1] CRAN (R 4.2.2)
 httr                   1.4.4    2022-08-17 [1] CRAN (R 4.2.0)
 IRanges              * 2.32.0   2022-11-01 [1] Bioconductor
 irlba                  2.3.5.1  2022-10-03 [1] CRAN (R 4.2.0)
 isoband                0.2.6    2022-10-06 [1] CRAN (R 4.2.0)
 jsonlite               1.8.4    2022-12-06 [1] CRAN (R 4.2.0)
 KEGGREST               1.38.0   2022-11-01 [1] Bioconductor
 knitr                  1.41     2022-11-18 [1] CRAN (R 4.2.0)
 labeling               0.4.2    2020-10-20 [1] CRAN (R 4.2.0)
 lattice                0.20-45  2021-09-22 [2] CRAN (R 4.2.2)
 lemur                * 0.0.9    2023-03-11 [1] Github (const-ae/lemur@efdb6b3)
 lifecycle              1.0.3    2022-10-07 [1] CRAN (R 4.2.0)
 magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.2.0)
 MASS                   7.3-58.1 2022-08-03 [2] CRAN (R 4.2.2)
 Matrix               * 1.5-3    2022-11-11 [1] CRAN (R 4.2.0)
 MatrixGenerics       * 1.10.0   2022-11-01 [1] Bioconductor
 matrixStats          * 0.63.0   2022-11-18 [1] CRAN (R 4.2.0)
 memoise                2.0.1    2021-11-26 [1] CRAN (R 4.2.0)
 munsell                0.5.0    2018-06-12 [1] CRAN (R 4.2.0)
 openssl                2.0.5    2022-12-06 [1] CRAN (R 4.2.0)
 org.Mm.eg.db         * 3.16.0   2022-12-29 [1] Bioconductor
 patchwork            * 1.1.2    2022-08-19 [1] CRAN (R 4.2.0)
 pillar                 1.8.1    2022-08-19 [1] CRAN (R 4.2.0)
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.2.0)
 png                    0.1-8    2022-11-29 [1] CRAN (R 4.2.0)
 purrr                * 1.0.0    2022-12-20 [1] CRAN (R 4.2.0)
 R6                     2.5.1    2021-08-19 [1] CRAN (R 4.2.0)
 Rcpp                   1.0.9    2022-07-08 [1] CRAN (R 4.2.0)
 RCurl                  1.98-1.9 2022-10-03 [1] CRAN (R 4.2.0)
 readr                * 2.1.3    2022-10-01 [1] CRAN (R 4.2.0)
 rlang                  1.0.6    2022-09-24 [1] CRAN (R 4.2.0)
 rmarkdown              2.20     2023-01-19 [1] RSPM (R 4.2.2)
 rprojroot              2.0.3    2022-04-02 [1] CRAN (R 4.2.0)
 RSQLite                2.2.19   2022-11-24 [1] CRAN (R 4.2.0)
 rstudioapi             0.14     2022-08-22 [1] CRAN (R 4.2.0)
 rsvd                   1.0.5    2021-04-16 [1] CRAN (R 4.2.0)
 S4Vectors            * 0.36.1   2022-12-05 [1] Bioconductor
 ScaledMatrix           1.6.0    2022-11-01 [1] Bioconductor
 scales                 1.2.1    2022-08-20 [1] CRAN (R 4.2.0)
 scater               * 1.26.1   2022-11-13 [1] Bioconductor
 scuttle              * 1.8.3    2022-12-14 [1] Bioconductor
 sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.2.0)
 SingleCellExperiment * 1.20.0   2022-11-01 [1] Bioconductor
 sparseMatrixStats      1.10.0   2022-11-01 [1] Bioconductor
 stringi                1.7.8    2022-07-11 [1] CRAN (R 4.2.0)
 stringr                1.5.0    2022-12-02 [1] CRAN (R 4.2.0)
 SummarizedExperiment * 1.28.0   2022-11-01 [1] Bioconductor
 sys                    3.4.1    2022-10-18 [1] CRAN (R 4.2.0)
 tibble                 3.1.8    2022-07-22 [1] CRAN (R 4.2.0)
 tidyr                * 1.2.1    2022-09-08 [1] CRAN (R 4.2.0)
 tidyselect             1.2.0    2022-10-10 [1] CRAN (R 4.2.0)
 tzdb                   0.3.0    2022-03-28 [1] CRAN (R 4.2.0)
 utf8                   1.2.2    2021-07-24 [1] CRAN (R 4.2.0)
 uwot                   0.1.14   2022-08-22 [1] CRAN (R 4.2.0)
 vctrs                  0.5.1    2022-11-16 [1] CRAN (R 4.2.0)
 vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.2.0)
 viridis                0.6.2    2021-10-13 [1] CRAN (R 4.2.0)
 viridisLite            0.4.1    2022-08-22 [1] CRAN (R 4.2.0)
 withr                  2.5.0    2022-03-03 [1] CRAN (R 4.2.0)
 xfun                   0.36     2022-12-21 [1] RSPM (R 4.2.2)
 XVector                0.38.0   2022-11-01 [1] Bioconductor
 yaml                   2.3.6    2022-10-18 [1] CRAN (R 4.2.0)
 zlibbioc               1.44.0   2022-11-01 [1] Bioconductor

 [1] /Users/sandmann/Library/R/x86_64/4.2/library
 [2] /Library/Frameworks/R.framework/Versions/4.2/Resources/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>


<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>scRNA-seq data from glioblastoma slices cultured <em>in vitro</em> with either <a href="https://en.wikipedia.org/wiki/Panobinostat">pamobinostat</a> or a vehicle control, characterized in a terrific paper by <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-021-00894-y">Zhao et al, 2021</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>CD45 is a cell surface antigen that is expressed on most hematopoietic lineage cells, including microglia.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb35" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Lemur: analyzing multi-condition single-cell data"</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-03-12"</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [TIL, R, scRNAseq]</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>sce_cache_does_not_exist <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(</span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>  here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"posts"</span>, <span class="st">"lemur"</span>, <span class="st">"sce.rds"</span>)</span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a>This week, </span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Constantin Ahlmann-Eltze</span><span class="co">](https://const-ae.name/)</span> and </span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Wolfgang Huber</span><span class="co">](https://www.embl.org/groups/huber/)</span></span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>published a</span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">preprint</span><span class="co">](https://www.biorxiv.org/content/10.1101/2023.03.06.531268v1)</span></span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>describing <span class="in">`LEMUR`</span>, a new approach to analyzing single-cell experiments that</span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>include samples from multiple conditions, e.g. drug treatments, disease-status,</span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a>etc.</span>
<span id="cb35-35"><a href="#cb35-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-36"><a href="#cb35-36" aria-hidden="true" tabindex="-1"></a>To date, such analyses often involve two separate steps, e.g.</span>
<span id="cb35-37"><a href="#cb35-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-38"><a href="#cb35-38" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Defining clusters of cells, e.g. cell types or states</span>
<span id="cb35-39"><a href="#cb35-39" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Differential expression analysis _within_ each cell type</span>
<span id="cb35-40"><a href="#cb35-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-41"><a href="#cb35-41" aria-hidden="true" tabindex="-1"></a>In contrast, <span class="in">`LEMUR`</span> considers the continuous latent space the individual cells</span>
<span id="cb35-42"><a href="#cb35-42" aria-hidden="true" tabindex="-1"></a>occupy, incorporating the design of the experiment, and then performs</span>
<span id="cb35-43"><a href="#cb35-43" aria-hidden="true" tabindex="-1"></a>differential expression analysis in this embedding space.</span>
<span id="cb35-44"><a href="#cb35-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-45"><a href="#cb35-45" aria-hidden="true" tabindex="-1"></a>An R package implementing <span class="in">`LEMUR`</span> is available from</span>
<span id="cb35-46"><a href="#cb35-46" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">github</span><span class="co">](https://github.com/const-ae/lemur)</span></span>
<span id="cb35-47"><a href="#cb35-47" aria-hidden="true" tabindex="-1"></a>and includes an example dataset <span class="ot">[^1]</span>.</span>
<span id="cb35-48"><a href="#cb35-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-49"><a href="#cb35-49" aria-hidden="true" tabindex="-1"></a><span class="fu">## Ellwanger et al: Comparing Trem2 wildtype and knock-out mouse microglia</span></span>
<span id="cb35-50"><a href="#cb35-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-51"><a href="#cb35-51" aria-hidden="true" tabindex="-1"></a>Here, I am exploring <span class="in">`LEMUR`</span> by examining scRNA-seq data published by </span>
<span id="cb35-52"><a href="#cb35-52" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Ellwanger et al, 2021</span><span class="co">](https://pubmed.ncbi.nlm.nih.gov/33446504/)</span>,</span>
<span id="cb35-53"><a href="#cb35-53" aria-hidden="true" tabindex="-1"></a>who injected three strains of</span>
<span id="cb35-54"><a href="#cb35-54" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">5XFAD mice</span><span class="co">](https://www.jax.org/strain/008730)</span>,</span>
<span id="cb35-55"><a href="#cb35-55" aria-hidden="true" tabindex="-1"></a>a murine model of familial Alzheimer's Disease, either </span>
<span id="cb35-56"><a href="#cb35-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-57"><a href="#cb35-57" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>carrying the wild-type (WT) _Trem2_ gene,</span>
<span id="cb35-58"><a href="#cb35-58" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>carrying the R47H _Trem2_ variant, believed to be a loss-of-function variant,</span>
<span id="cb35-59"><a href="#cb35-59" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>or completely lacking _Trem2_ expression</span>
<span id="cb35-60"><a href="#cb35-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-61"><a href="#cb35-61" aria-hidden="true" tabindex="-1"></a>with either a Trem2 agonist (<span class="in">`hT2AB`</span>) or a negative control antibody (<span class="in">`hIgG1`</span>). </span>
<span id="cb35-62"><a href="#cb35-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-63"><a href="#cb35-63" aria-hidden="true" tabindex="-1"></a>48 hours later, the authors isolated CD45-positive <span class="ot">[^2]</span> cells from the cortex</span>
<span id="cb35-64"><a href="#cb35-64" aria-hidden="true" tabindex="-1"></a>and performed single-cell RNA-seq analysis using the 10X Genomics platform.</span>
<span id="cb35-65"><a href="#cb35-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-66"><a href="#cb35-66" aria-hidden="true" tabindex="-1"></a><span class="fu">### Retrieving the data</span></span>
<span id="cb35-67"><a href="#cb35-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-70"><a href="#cb35-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-71"><a href="#cb35-71" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-72"><a href="#cb35-72" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb35-73"><a href="#cb35-73" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb35-74"><a href="#cb35-74" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb35-75"><a href="#cb35-75" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb35-76"><a href="#cb35-76" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb35-77"><a href="#cb35-77" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb35-78"><a href="#cb35-78" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb35-79"><a href="#cb35-79" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb35-80"><a href="#cb35-80" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb35-81"><a href="#cb35-81" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-82"><a href="#cb35-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-83"><a href="#cb35-83" aria-hidden="true" tabindex="-1"></a>Ellwanger et al made both raw and processed data available via the NCBI GEO</span>
<span id="cb35-84"><a href="#cb35-84" aria-hidden="true" tabindex="-1"></a>and SRA repositories under </span>
<span id="cb35-85"><a href="#cb35-85" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">GEO accession GSE156183</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156183)</span>.</span>
<span id="cb35-86"><a href="#cb35-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-87"><a href="#cb35-87" aria-hidden="true" tabindex="-1"></a>They also included complete metadata for each cell, making this a great dataset</span>
<span id="cb35-88"><a href="#cb35-88" aria-hidden="true" tabindex="-1"></a>for re-analysis.</span>
<span id="cb35-89"><a href="#cb35-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-90"><a href="#cb35-90" aria-hidden="true" tabindex="-1"></a>Let's start by retrieving the</span>
<span id="cb35-91"><a href="#cb35-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-92"><a href="#cb35-92" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="co">[</span><span class="ot">processed counts (500 Mb)</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file&amp;file=GSE156183%5FRAW%2Emtx%2Egz)</span> </span>
<span id="cb35-93"><a href="#cb35-93" aria-hidden="true" tabindex="-1"></a>  and the</span>
<span id="cb35-94"><a href="#cb35-94" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="co">[</span><span class="ot">cell metadata (13 Mb)</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file&amp;file=GSE156183%5FCell%5Fmetadata%2Etsv%2Egz)</span></span>
<span id="cb35-95"><a href="#cb35-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-96"><a href="#cb35-96" aria-hidden="true" tabindex="-1"></a>files from GEO and store them in a temporary directory:</span>
<span id="cb35-97"><a href="#cb35-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-100"><a href="#cb35-100" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-101"><a href="#cb35-101" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr 'sce_cache_does_not_exist'</span></span>
<span id="cb35-102"><a href="#cb35-102" aria-hidden="true" tabindex="-1"></a>temp_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"ellwanger"</span>)</span>
<span id="cb35-103"><a href="#cb35-103" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(temp_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-104"><a href="#cb35-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-105"><a href="#cb35-105" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="dv">360</span>)</span>
<span id="cb35-106"><a href="#cb35-106" aria-hidden="true" tabindex="-1"></a>url_root <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc="</span>,</span>
<span id="cb35-107"><a href="#cb35-107" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"GSE156183&amp;format=file&amp;file=GSE156183%5F"</span>)</span>
<span id="cb35-108"><a href="#cb35-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-109"><a href="#cb35-109" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"counts.mtx.gz"</span>)</span>
<span id="cb35-110"><a href="#cb35-110" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb35-111"><a href="#cb35-111" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(url_root, <span class="st">"RAW%2Emtx%2Egz"</span>), </span>
<span id="cb35-112"><a href="#cb35-112" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> raw_counts)</span>
<span id="cb35-113"><a href="#cb35-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-114"><a href="#cb35-114" aria-hidden="true" tabindex="-1"></a>cell_metadata <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"cell_metadata.tsv.gz"</span>)</span>
<span id="cb35-115"><a href="#cb35-115" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb35-116"><a href="#cb35-116" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(url_root, <span class="st">"Cell%5Fmetadata%2Etsv%2Egz"</span>), </span>
<span id="cb35-117"><a href="#cb35-117" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> cell_metadata)</span>
<span id="cb35-118"><a href="#cb35-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-119"><a href="#cb35-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-120"><a href="#cb35-120" aria-hidden="true" tabindex="-1"></a>and read the sparse count matrix into our R session:</span>
<span id="cb35-121"><a href="#cb35-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-124"><a href="#cb35-124" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-125"><a href="#cb35-125" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr 'sce_cache_does_not_exist'</span></span>
<span id="cb35-126"><a href="#cb35-126" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(raw_counts)</span>
<span id="cb35-127"><a href="#cb35-127" aria-hidden="true" tabindex="-1"></a>cell_anno <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(cell_metadata, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-128"><a href="#cb35-128" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">nrow</span>(cell_anno) <span class="sc">==</span> <span class="fu">ncol</span>(m))</span>
<span id="cb35-129"><a href="#cb35-129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-130"><a href="#cb35-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-131"><a href="#cb35-131" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="false"}</span>
<span id="cb35-132"><a href="#cb35-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-133"><a href="#cb35-133" aria-hidden="true" tabindex="-1"></a>Unfortunately, the <span class="in">`GSE156183_Feature_metadata.tsv.gz`</span> feature (= gene) </span>
<span id="cb35-134"><a href="#cb35-134" aria-hidden="true" tabindex="-1"></a>annotation file the authors deposited with GEO _actually_ contains _cell_ </span>
<span id="cb35-135"><a href="#cb35-135" aria-hidden="true" tabindex="-1"></a>annotations. But luckily, they also deposited counts matrices in TSV format for</span>
<span id="cb35-136"><a href="#cb35-136" aria-hidden="true" tabindex="-1"></a>each sample, which include the ENSEMBL gene identifier for each row.</span>
<span id="cb35-137"><a href="#cb35-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-138"><a href="#cb35-138" aria-hidden="true" tabindex="-1"></a>Here, I download the TAR archive that contains all of the TSV files, and then</span>
<span id="cb35-139"><a href="#cb35-139" aria-hidden="true" tabindex="-1"></a>extract the gene identifiers from one of the files so I can add them to the</span>
<span id="cb35-140"><a href="#cb35-140" aria-hidden="true" tabindex="-1"></a>experiment-wide raw count matrix.</span>
<span id="cb35-141"><a href="#cb35-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-144"><a href="#cb35-144" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-145"><a href="#cb35-145" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr 'sce_cache_does_not_exist'</span></span>
<span id="cb35-146"><a href="#cb35-146" aria-hidden="true" tabindex="-1"></a>selected_sample <span class="ot">&lt;-</span> <span class="st">"GSM4726219_RAW-R47H-male-IgG-rep2.tsv.gz"</span></span>
<span id="cb35-147"><a href="#cb35-147" aria-hidden="true" tabindex="-1"></a>tar_archive <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temp_dir, <span class="st">"RAW.tar"</span>)</span>
<span id="cb35-148"><a href="#cb35-148" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb35-149"><a href="#cb35-149" aria-hidden="true" tabindex="-1"></a>  <span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156183&amp;format=file"</span>,</span>
<span id="cb35-150"><a href="#cb35-150" aria-hidden="true" tabindex="-1"></a>  <span class="at">destfile =</span> tar_archive)</span>
<span id="cb35-151"><a href="#cb35-151" aria-hidden="true" tabindex="-1"></a>utils<span class="sc">::</span><span class="fu">untar</span>(tar_archive, <span class="at">files =</span> selected_sample, <span class="at">exdir =</span> <span class="fu">tempdir</span>())</span>
<span id="cb35-152"><a href="#cb35-152" aria-hidden="true" tabindex="-1"></a>gene_ids <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="fu">file.path</span>(<span class="fu">tempdir</span>(), selected_sample), </span>
<span id="cb35-153"><a href="#cb35-153" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col_select =</span> <span class="fu">any_of</span>(<span class="st">"feature_id"</span>),</span>
<span id="cb35-154"><a href="#cb35-154" aria-hidden="true" tabindex="-1"></a>                            <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-155"><a href="#cb35-155" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(feature_id)</span>
<span id="cb35-156"><a href="#cb35-156" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(gene_ids) <span class="sc">==</span> <span class="fu">nrow</span>(m))</span>
<span id="cb35-157"><a href="#cb35-157" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(m) <span class="ot">&lt;-</span> gene_ids</span>
<span id="cb35-158"><a href="#cb35-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-159"><a href="#cb35-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-160"><a href="#cb35-160" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb35-161"><a href="#cb35-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-162"><a href="#cb35-162" aria-hidden="true" tabindex="-1"></a><span class="fu">### Creating a SingleCellExperiment object</span></span>
<span id="cb35-163"><a href="#cb35-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-164"><a href="#cb35-164" aria-hidden="true" tabindex="-1"></a>Now I have all the pieces of information required to create a</span>
<span id="cb35-165"><a href="#cb35-165" aria-hidden="true" tabindex="-1"></a><span class="in">`SingleCellExperiment`</span>: </span>
<span id="cb35-166"><a href="#cb35-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-167"><a href="#cb35-167" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>the raw counts (in the form of a sparse matrix),</span>
<span id="cb35-168"><a href="#cb35-168" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>the cell annotations (in the form of a data.frame)</span>
<span id="cb35-169"><a href="#cb35-169" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>the two UMAP dimensions used by the authors (included in the cell metadata).</span>
<span id="cb35-170"><a href="#cb35-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-171"><a href="#cb35-171" aria-hidden="true" tabindex="-1"></a>I choose to retain only a subset of the (many) cell-level annotation columns,</span>
<span id="cb35-172"><a href="#cb35-172" aria-hidden="true" tabindex="-1"></a>add gene symbols as row annotations, extract the UMAP coordinates into a</span>
<span id="cb35-173"><a href="#cb35-173" aria-hidden="true" tabindex="-1"></a>separate matrix - and store all of it in the <span class="in">`sce`</span> object.</span>
<span id="cb35-174"><a href="#cb35-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-175"><a href="#cb35-175" aria-hidden="true" tabindex="-1"></a>Next, I am removing cells without an assigned cell type, and also add a coarser</span>
<span id="cb35-176"><a href="#cb35-176" aria-hidden="true" tabindex="-1"></a>cell type annotation that collapses the different microglia states reported by</span>
<span id="cb35-177"><a href="#cb35-177" aria-hidden="true" tabindex="-1"></a>the authors into a single category. Finally, I filter genes without a valid</span>
<span id="cb35-178"><a href="#cb35-178" aria-hidden="true" tabindex="-1"></a>gene symbol and add an assay slot with the normalized log2 counts.</span>
<span id="cb35-179"><a href="#cb35-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-182"><a href="#cb35-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-183"><a href="#cb35-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr 'sce_cache_does_not_exist'</span></span>
<span id="cb35-184"><a href="#cb35-184" aria-hidden="true" tabindex="-1"></a>col_data <span class="ot">&lt;-</span> cell_anno <span class="sc">%&gt;%</span></span>
<span id="cb35-185"><a href="#cb35-185" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(cell_id, celltype, sample, sex, genotype, treatment, </span>
<span id="cb35-186"><a href="#cb35-186" aria-hidden="true" tabindex="-1"></a>                <span class="fu">starts_with</span>(<span class="st">"QC."</span>)</span>
<span id="cb35-187"><a href="#cb35-187" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb35-188"><a href="#cb35-188" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-189"><a href="#cb35-189" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"cell_id"</span>)</span>
<span id="cb35-190"><a href="#cb35-190" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(m) <span class="ot">&lt;-</span> <span class="fu">row.names</span>(col_data)</span>
<span id="cb35-191"><a href="#cb35-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-192"><a href="#cb35-192" aria-hidden="true" tabindex="-1"></a>row_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb35-193"><a href="#cb35-193" aria-hidden="true" tabindex="-1"></a>  <span class="at">symbol =</span> AnnotationDbi<span class="sc">::</span><span class="fu">mapIds</span>(org.Mm.eg.db, <span class="at">keys =</span> gene_ids,</span>
<span id="cb35-194"><a href="#cb35-194" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">column =</span> <span class="st">"SYMBOL"</span>, <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>),</span>
<span id="cb35-195"><a href="#cb35-195" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> gene_ids</span>
<span id="cb35-196"><a href="#cb35-196" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-197"><a href="#cb35-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-198"><a href="#cb35-198" aria-hidden="true" tabindex="-1"></a>umap <span class="ot">&lt;-</span> cell_anno <span class="sc">%&gt;%</span></span>
<span id="cb35-199"><a href="#cb35-199" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">ends_with</span>(<span class="st">"CD45pos"</span>)</span>
<span id="cb35-200"><a href="#cb35-200" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb35-201"><a href="#cb35-201" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb35-202"><a href="#cb35-202" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(umap) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(m)</span>
<span id="cb35-203"><a href="#cb35-203" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(umap) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"UMAP"</span>, <span class="fu">seq.int</span>(<span class="fu">ncol</span>(umap)))</span>
<span id="cb35-204"><a href="#cb35-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-205"><a href="#cb35-205" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb35-206"><a href="#cb35-206" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> m),</span>
<span id="cb35-207"><a href="#cb35-207" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData =</span> row_data,</span>
<span id="cb35-208"><a href="#cb35-208" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_data,</span>
<span id="cb35-209"><a href="#cb35-209" aria-hidden="true" tabindex="-1"></a>  <span class="at">reducedDims =</span> <span class="fu">list</span>(<span class="at">UMAP =</span> umap)</span>
<span id="cb35-210"><a href="#cb35-210" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-211"><a href="#cb35-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-212"><a href="#cb35-212" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="sc">!</span><span class="fu">is.na</span>(sce<span class="sc">$</span>celltype)]</span>
<span id="cb35-213"><a href="#cb35-213" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>celltype_coarse <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb35-214"><a href="#cb35-214" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grepl</span>(<span class="at">x =</span> sce<span class="sc">$</span>celltype, <span class="at">pattern =</span> <span class="st">"Microglia"</span>) <span class="sc">~</span> <span class="st">"Microglia"</span>,</span>
<span id="cb35-215"><a href="#cb35-215" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> sce<span class="sc">$</span>celltype</span>
<span id="cb35-216"><a href="#cb35-216" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb35-217"><a href="#cb35-217" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>treatment <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>treatment, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IgG"</span>, <span class="st">"hT2AB"</span>))</span>
<span id="cb35-218"><a href="#cb35-218" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>mg_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">sub</span>(<span class="st">"Microglia."</span>, <span class="st">""</span>, sce<span class="sc">$</span>celltype, <span class="at">fixed =</span> <span class="cn">TRUE</span>))</span>
<span id="cb35-219"><a href="#cb35-219" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[<span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>symbol), ]</span>
<span id="cb35-220"><a href="#cb35-220" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span>
<span id="cb35-221"><a href="#cb35-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-222"><a href="#cb35-222" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">c</span>(<span class="st">"m"</span>, <span class="st">"cell_anno"</span>, <span class="st">"gene_ids"</span>, <span class="st">"row_data"</span>, <span class="st">"col_data"</span>))</span>
<span id="cb35-223"><a href="#cb35-223" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-224"><a href="#cb35-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-227"><a href="#cb35-227" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-228"><a href="#cb35-228" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb35-229"><a href="#cb35-229" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (sce_cache_does_not_exist) {</span>
<span id="cb35-230"><a href="#cb35-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(sce, <span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"posts"</span>, <span class="st">"lemur"</span>, <span class="st">"sce.rds"</span>))</span>
<span id="cb35-231"><a href="#cb35-231" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb35-232"><a href="#cb35-232" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"posts"</span>, <span class="st">"lemur"</span>, <span class="st">"sce.rds"</span>))</span>
<span id="cb35-233"><a href="#cb35-233" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb35-234"><a href="#cb35-234" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-235"><a href="#cb35-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-236"><a href="#cb35-236" aria-hidden="true" tabindex="-1"></a>This <span class="in">`SingleCellExperiment`</span> object is now ready for downstream analysis.</span>
<span id="cb35-237"><a href="#cb35-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-238"><a href="#cb35-238" aria-hidden="true" tabindex="-1"></a><span class="fu">## Subsetting the experiment to samples of interest</span></span>
<span id="cb35-239"><a href="#cb35-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-240"><a href="#cb35-240" aria-hidden="true" tabindex="-1"></a>This study contains multiple experimental variables, e.g. each sample is </span>
<span id="cb35-241"><a href="#cb35-241" aria-hidden="true" tabindex="-1"></a>annotated with one of the three <span class="in">`genotypes`</span>, one of two <span class="in">`treatments`</span> and the</span>
<span id="cb35-242"><a href="#cb35-242" aria-hidden="true" tabindex="-1"></a><span class="in">`sex`</span> for each mouse. </span>
<span id="cb35-243"><a href="#cb35-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-244"><a href="#cb35-244" aria-hidden="true" tabindex="-1"></a>Here, I will focus only on the difference between TREM2 wildtype and </span>
<span id="cb35-245"><a href="#cb35-245" aria-hidden="true" tabindex="-1"></a>TREM2 knock-out animals treated with the <span class="in">`IgG`</span> control antibody. Only female</span>
<span id="cb35-246"><a href="#cb35-246" aria-hidden="true" tabindex="-1"></a>knock-out animals were included in the study, so I exclude the male animals</span>
<span id="cb35-247"><a href="#cb35-247" aria-hidden="true" tabindex="-1"></a>from the other strain as well.</span>
<span id="cb35-248"><a href="#cb35-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-251"><a href="#cb35-251" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-252"><a href="#cb35-252" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="fu">which</span>(sce<span class="sc">$</span>genotype <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"TREM2_CV-5XFAD"</span>, <span class="st">"Trem2_KO-5XFAD"</span>))]</span>
<span id="cb35-253"><a href="#cb35-253" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="fu">which</span>(sce<span class="sc">$</span>sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">&amp;</span> sce<span class="sc">$</span>treatment <span class="sc">==</span> <span class="st">"IgG"</span>)]</span>
<span id="cb35-254"><a href="#cb35-254" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(<span class="fu">colData</span>(sce), <span class="fu">table</span>(genotype, treatment))</span>
<span id="cb35-255"><a href="#cb35-255" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-256"><a href="#cb35-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-257"><a href="#cb35-257" aria-hidden="true" tabindex="-1"></a>After subsetting, the experiment now contains <span class="in">`r length(unique(sce$sample))`</span> </span>
<span id="cb35-258"><a href="#cb35-258" aria-hidden="true" tabindex="-1"></a>samples:</span>
<span id="cb35-259"><a href="#cb35-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-262"><a href="#cb35-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-263"><a href="#cb35-263" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce) <span class="sc">%&gt;%</span></span>
<span id="cb35-264"><a href="#cb35-264" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-265"><a href="#cb35-265" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(sample, treatment, genotype) <span class="sc">%&gt;%</span></span>
<span id="cb35-266"><a href="#cb35-266" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-267"><a href="#cb35-267" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">remove_rownames</span>()</span>
<span id="cb35-268"><a href="#cb35-268" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-269"><a href="#cb35-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-270"><a href="#cb35-270" aria-hidden="true" tabindex="-1"></a>At this point, I can reproduce e.g. a version of</span>
<span id="cb35-271"><a href="#cb35-271" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Figure 3E</span><span class="co">](https://www.pnas.org/doi/10.1073/pnas.2017742118#fig03)</span></span>
<span id="cb35-272"><a href="#cb35-272" aria-hidden="true" tabindex="-1"></a>from the original paper, using the UMAP coordinates and cell type labels</span>
<span id="cb35-273"><a href="#cb35-273" aria-hidden="true" tabindex="-1"></a>provided by the authors. (My version of the figure only includes cells</span>
<span id="cb35-274"><a href="#cb35-274" aria-hidden="true" tabindex="-1"></a>from the selected subset of samples, not all cells captured in the study.)</span>
<span id="cb35-275"><a href="#cb35-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-278"><a href="#cb35-278" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-279"><a href="#cb35-279" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-280"><a href="#cb35-280" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-281"><a href="#cb35-281" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-282"><a href="#cb35-282" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Microglia"</span> <span class="ot">=</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb35-283"><a href="#cb35-283" aria-hidden="true" tabindex="-1"></a>            <span class="st">"T cells"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>,</span>
<span id="cb35-284"><a href="#cb35-284" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Macrophages"</span> <span class="ot">=</span> <span class="st">"firebrick"</span>,</span>
<span id="cb35-285"><a href="#cb35-285" aria-hidden="true" tabindex="-1"></a>            <span class="st">"MO:T"</span> <span class="ot">=</span> <span class="st">"darkgreen"</span>,</span>
<span id="cb35-286"><a href="#cb35-286" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Dendritic cells"</span> <span class="ot">=</span> <span class="st">"green"</span>,</span>
<span id="cb35-287"><a href="#cb35-287" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Monocytes"</span> <span class="ot">=</span> <span class="st">"orange"</span>,</span>
<span id="cb35-288"><a href="#cb35-288" aria-hidden="true" tabindex="-1"></a>            <span class="st">"B cells"</span> <span class="ot">=</span> <span class="st">"navy"</span>,</span>
<span id="cb35-289"><a href="#cb35-289" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Neutrophils"</span> <span class="ot">=</span> <span class="st">"darkblue"</span>,</span>
<span id="cb35-290"><a href="#cb35-290" aria-hidden="true" tabindex="-1"></a>            <span class="st">"HCS"</span> <span class="ot">=</span> <span class="st">"grey"</span>, </span>
<span id="cb35-291"><a href="#cb35-291" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Fibroblasts"</span> <span class="ot">=</span> <span class="st">"yellow"</span>)</span>
<span id="cb35-292"><a href="#cb35-292" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"celltype_coarse"</span>) <span class="sc">+</span></span>
<span id="cb35-293"><a href="#cb35-293" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors, <span class="at">name =</span> <span class="st">"Cell type"</span>)</span>
<span id="cb35-294"><a href="#cb35-294" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-295"><a href="#cb35-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-296"><a href="#cb35-296" aria-hidden="true" tabindex="-1"></a>Because Ellwanger et al captured all cells with CD45 expression, the dataset</span>
<span id="cb35-297"><a href="#cb35-297" aria-hidden="true" tabindex="-1"></a>includes other immune cell types besides microglia. Let's remove those to focus</span>
<span id="cb35-298"><a href="#cb35-298" aria-hidden="true" tabindex="-1"></a>only on the latter.</span>
<span id="cb35-299"><a href="#cb35-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-302"><a href="#cb35-302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-303"><a href="#cb35-303" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, sce<span class="sc">$</span>celltype_coarse <span class="sc">==</span> <span class="st">"Microglia"</span>]</span>
<span id="cb35-304"><a href="#cb35-304" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>mg_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb35-305"><a href="#cb35-305" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>mg_type, </span>
<span id="cb35-306"><a href="#cb35-306" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Resting"</span>, <span class="st">"t1"</span>, <span class="st">"t2"</span>, <span class="st">"t3"</span>, <span class="st">"t4"</span>, <span class="st">"t5"</span>, <span class="st">"t6"</span>, <span class="st">"IFN-R"</span>, <span class="st">"DAM"</span>, </span>
<span id="cb35-307"><a href="#cb35-307" aria-hidden="true" tabindex="-1"></a>             <span class="st">"MHC-II"</span>, <span class="st">"Cyc-M"</span>))</span>
<span id="cb35-308"><a href="#cb35-308" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-309"><a href="#cb35-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-310"><a href="#cb35-310" aria-hidden="true" tabindex="-1"></a>Most microglial states were captured in animals from both genotypes:</span>
<span id="cb35-311"><a href="#cb35-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-314"><a href="#cb35-314" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-315"><a href="#cb35-315" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb35-316"><a href="#cb35-316" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-317"><a href="#cb35-317" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-318"><a href="#cb35-318" aria-hidden="true" tabindex="-1"></a>mg_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb35-319"><a href="#cb35-319" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cyc-M"</span> <span class="ot">=</span> <span class="st">"navy"</span>,</span>
<span id="cb35-320"><a href="#cb35-320" aria-hidden="true" tabindex="-1"></a>  <span class="st">"DAM"</span> <span class="ot">=</span> <span class="st">"darkgreen"</span>,</span>
<span id="cb35-321"><a href="#cb35-321" aria-hidden="true" tabindex="-1"></a>  <span class="st">"IFN-R"</span> <span class="ot">=</span> <span class="st">"#C12131"</span>, </span>
<span id="cb35-322"><a href="#cb35-322" aria-hidden="true" tabindex="-1"></a>  <span class="st">"MHC-II"</span> <span class="ot">=</span> <span class="st">"green"</span>,</span>
<span id="cb35-323"><a href="#cb35-323" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Resting"</span> <span class="ot">=</span> <span class="st">"grey50"</span>,</span>
<span id="cb35-324"><a href="#cb35-324" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t1"</span> <span class="ot">=</span> <span class="st">"#FDF5EB"</span>,   </span>
<span id="cb35-325"><a href="#cb35-325" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t2"</span> <span class="ot">=</span>  <span class="st">"#FFE2C0"</span>, </span>
<span id="cb35-326"><a href="#cb35-326" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t3"</span> <span class="ot">=</span> <span class="st">"#FFC08E"</span>,</span>
<span id="cb35-327"><a href="#cb35-327" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t4"</span> <span class="ot">=</span> <span class="st">"#FE945C"</span>,</span>
<span id="cb35-328"><a href="#cb35-328" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t5"</span> <span class="ot">=</span>  <span class="st">"#EC5D2F"</span>,</span>
<span id="cb35-329"><a href="#cb35-329" aria-hidden="true" tabindex="-1"></a>  <span class="st">"t6"</span> <span class="ot">=</span> <span class="st">"#C12131"</span></span>
<span id="cb35-330"><a href="#cb35-330" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-331"><a href="#cb35-331" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>) <span class="sc">+</span></span>
<span id="cb35-332"><a href="#cb35-332" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-333"><a href="#cb35-333" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sce<span class="sc">$</span>genotype)</span>
<span id="cb35-334"><a href="#cb35-334" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-335"><a href="#cb35-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-336"><a href="#cb35-336" aria-hidden="true" tabindex="-1"></a><span class="fu">## Differential expression analysis with lemur</span></span>
<span id="cb35-337"><a href="#cb35-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-338"><a href="#cb35-338" aria-hidden="true" tabindex="-1"></a>Now I am ready to explore the <span class="in">`lemur`</span> R package to ask: "which neighborhoods</span>
<span id="cb35-339"><a href="#cb35-339" aria-hidden="true" tabindex="-1"></a>show for differential expression between samples from WT and knock-out animals?"</span>
<span id="cb35-340"><a href="#cb35-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-341"><a href="#cb35-341" aria-hidden="true" tabindex="-1"></a>The following steps closely follow the examples outlined on the</span>
<span id="cb35-342"><a href="#cb35-342" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">lemor github repository's README</span><span class="co">](https://github.com/const-ae/lemur)</span> - many</span>
<span id="cb35-343"><a href="#cb35-343" aria-hidden="true" tabindex="-1"></a>thanks for the great documentation, Constantin! (All mistakes and </span>
<span id="cb35-344"><a href="#cb35-344" aria-hidden="true" tabindex="-1"></a>misunderstandings in this post are my own, as always.)</span>
<span id="cb35-345"><a href="#cb35-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-346"><a href="#cb35-346" aria-hidden="true" tabindex="-1"></a><span class="fu">### Dependencies &amp; installation</span></span>
<span id="cb35-347"><a href="#cb35-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-348"><a href="#cb35-348" aria-hidden="true" tabindex="-1"></a>Following the instructions from the</span>
<span id="cb35-349"><a href="#cb35-349" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">lemu github repository</span><span class="co">](https://github.com/const-ae/lemur)</span></span>
<span id="cb35-350"><a href="#cb35-350" aria-hidden="true" tabindex="-1"></a>I then installed the latest version of the </span>
<span id="cb35-351"><a href="#cb35-351" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">glmGamPoi</span><span class="co">](https://github.com/const-ae/glmGamPoi)</span> </span>
<span id="cb35-352"><a href="#cb35-352" aria-hidden="true" tabindex="-1"></a>package, and then the <span class="in">`lemur`</span> package itself from their github repositories.</span>
<span id="cb35-353"><a href="#cb35-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-354"><a href="#cb35-354" aria-hidden="true" tabindex="-1"></a>To harmonize results across batches (in this case: samples), I will use</span>
<span id="cb35-355"><a href="#cb35-355" aria-hidden="true" tabindex="-1"></a>harmony, so I need to install it from its github repository as well.</span>
<span id="cb35-356"><a href="#cb35-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-359"><a href="#cb35-359" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-360"><a href="#cb35-360" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb35-361"><a href="#cb35-361" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"const-ae/glmGamPoi"</span>)</span>
<span id="cb35-362"><a href="#cb35-362" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"const-ae/lemur"</span>)</span>
<span id="cb35-363"><a href="#cb35-363" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"immunogenomics/harmony"</span>)</span>
<span id="cb35-364"><a href="#cb35-364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-365"><a href="#cb35-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-366"><a href="#cb35-366" aria-hidden="true" tabindex="-1"></a><span class="fu">### Subsetting the experiment</span></span>
<span id="cb35-367"><a href="#cb35-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-370"><a href="#cb35-370" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-371"><a href="#cb35-371" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-372"><a href="#cb35-372" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lemur)</span>
<span id="cb35-373"><a href="#cb35-373" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> 1000L</span>
<span id="cb35-374"><a href="#cb35-374" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-375"><a href="#cb35-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-376"><a href="#cb35-376" aria-hidden="true" tabindex="-1"></a>To speed up my exploration of the <span class="in">`LEMUR`</span> workflow, I subset the experiment</span>
<span id="cb35-377"><a href="#cb35-377" aria-hidden="true" tabindex="-1"></a>to <span class="in">`r n_cells`</span> random cells from each of the two genotypes.</span>
<span id="cb35-378"><a href="#cb35-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-381"><a href="#cb35-381" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-382"><a href="#cb35-382" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(1L)</span>
<span id="cb35-383"><a href="#cb35-383" aria-hidden="true" tabindex="-1"></a>genotypes <span class="ot">&lt;-</span> <span class="fu">unique</span>(sce<span class="sc">$</span>genotype)</span>
<span id="cb35-384"><a href="#cb35-384" aria-hidden="true" tabindex="-1"></a>selected_cells <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">sapply</span>(genotypes, \(g) {</span>
<span id="cb35-385"><a href="#cb35-385" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample</span>(<span class="fu">which</span>(sce<span class="sc">$</span>genotype <span class="sc">==</span> g), n_cells)</span>
<span id="cb35-386"><a href="#cb35-386" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb35-387"><a href="#cb35-387" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-388"><a href="#cb35-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-389"><a href="#cb35-389" aria-hidden="true" tabindex="-1"></a>As expected, most microglial states described in the paper remain represented in </span>
<span id="cb35-390"><a href="#cb35-390" aria-hidden="true" tabindex="-1"></a>the downsampled dataset:</span>
<span id="cb35-391"><a href="#cb35-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-394"><a href="#cb35-394" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-395"><a href="#cb35-395" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>celltype[selected_cells], sce<span class="sc">$</span>genotype[selected_cells])</span>
<span id="cb35-396"><a href="#cb35-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-397"><a href="#cb35-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-400"><a href="#cb35-400" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-401"><a href="#cb35-401" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-402"><a href="#cb35-402" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-403"><a href="#cb35-403" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-404"><a href="#cb35-404" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb35-405"><a href="#cb35-405" aria-hidden="true" tabindex="-1"></a>  sce[, selected_cells], <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>) <span class="sc">+</span></span>
<span id="cb35-406"><a href="#cb35-406" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells))) <span class="sc">+</span></span>
<span id="cb35-407"><a href="#cb35-407" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-408"><a href="#cb35-408" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sce<span class="sc">$</span>genotype[selected_cells])</span>
<span id="cb35-409"><a href="#cb35-409" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-410"><a href="#cb35-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-411"><a href="#cb35-411" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fitting the LEMUR model</span></span>
<span id="cb35-412"><a href="#cb35-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-413"><a href="#cb35-413" aria-hidden="true" tabindex="-1"></a>Next, I fit the latent embedding multivariate regression (LEMUR) model with the</span>
<span id="cb35-414"><a href="#cb35-414" aria-hidden="true" tabindex="-1"></a><span class="in">`lemur()`</span> function. Because the dataset is relatively homogeneous, e.g. it </span>
<span id="cb35-415"><a href="#cb35-415" aria-hidden="true" tabindex="-1"></a>contains only microglia, I chose to consider only 25 Principal Components and</span>
<span id="cb35-416"><a href="#cb35-416" aria-hidden="true" tabindex="-1"></a>used 15 dimensions for the LEMUR embedding (e.g. the default number).</span>
<span id="cb35-417"><a href="#cb35-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-420"><a href="#cb35-420" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-421"><a href="#cb35-421" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-422"><a href="#cb35-422" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb35-423"><a href="#cb35-423" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> lemur<span class="sc">::</span><span class="fu">lemur</span>(sce[, selected_cells], <span class="at">design =</span> <span class="sc">~</span> genotype,</span>
<span id="cb35-424"><a href="#cb35-424" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_ambient =</span> <span class="dv">25</span>, <span class="at">n_embedding =</span> <span class="dv">15</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-425"><a href="#cb35-425" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-426"><a href="#cb35-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-427"><a href="#cb35-427" aria-hidden="true" tabindex="-1"></a>Because each sample was processed in a separate channel of the 10X Genomics</span>
<span id="cb35-428"><a href="#cb35-428" aria-hidden="true" tabindex="-1"></a>microfluidics device, I am aligning the embeddings of similar cell clusters</span>
<span id="cb35-429"><a href="#cb35-429" aria-hidden="true" tabindex="-1"></a>using</span>
<span id="cb35-430"><a href="#cb35-430" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">harmony</span><span class="co">](https://www.nature.com/articles/s41592-019-0619-0)</span>.</span>
<span id="cb35-431"><a href="#cb35-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-434"><a href="#cb35-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-435"><a href="#cb35-435" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-436"><a href="#cb35-436" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb35-437"><a href="#cb35-437" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> lemur<span class="sc">::</span><span class="fu">align_harmony</span>(fit, <span class="at">stretching =</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-438"><a href="#cb35-438" aria-hidden="true" tabindex="-1"></a>fit</span>
<span id="cb35-439"><a href="#cb35-439" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-440"><a href="#cb35-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-441"><a href="#cb35-441" aria-hidden="true" tabindex="-1"></a>The returned <span class="in">`lemur_fit`</span> object contains the <span class="in">`embedding`</span> matrix,</span>
<span id="cb35-442"><a href="#cb35-442" aria-hidden="true" tabindex="-1"></a>the latent space in which the differential expression analysis is performed.</span>
<span id="cb35-443"><a href="#cb35-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-446"><a href="#cb35-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-447"><a href="#cb35-447" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(fit<span class="sc">$</span>embedding)  <span class="co"># 15 dimensions, as specified above</span></span>
<span id="cb35-448"><a href="#cb35-448" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-449"><a href="#cb35-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-450"><a href="#cb35-450" aria-hidden="true" tabindex="-1"></a>Let's plot the first two dimensions against each other, coloring each cell by</span>
<span id="cb35-451"><a href="#cb35-451" aria-hidden="true" tabindex="-1"></a>the microglial state Ellwanger et al identified through Louvain clustering. </span>
<span id="cb35-452"><a href="#cb35-452" aria-hidden="true" tabindex="-1"></a>(This information has not been used by <span class="in">`lemur`</span>):</span>
<span id="cb35-453"><a href="#cb35-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-456"><a href="#cb35-456" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-457"><a href="#cb35-457" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-458"><a href="#cb35-458" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-459"><a href="#cb35-459" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-460"><a href="#cb35-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-461"><a href="#cb35-461" aria-hidden="true" tabindex="-1"></a><span class="co"># plot dim 1 vs dim 2</span></span>
<span id="cb35-462"><a href="#cb35-462" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb35-463"><a href="#cb35-463" aria-hidden="true" tabindex="-1"></a>  fit, <span class="st">"embedding"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>, <span class="at">shape_by =</span> <span class="st">"genotype"</span>) <span class="sc">+</span></span>
<span id="cb35-464"><a href="#cb35-464" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-465"><a href="#cb35-465" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb35-466"><a href="#cb35-466" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Embedding after accounting for genotype"</span>,</span>
<span id="cb35-467"><a href="#cb35-467" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells)))</span>
<span id="cb35-468"><a href="#cb35-468" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-469"><a href="#cb35-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-470"><a href="#cb35-470" aria-hidden="true" tabindex="-1"></a>Cells group by Ellwanger et al's subtype labels, and cells from both </span>
<span id="cb35-471"><a href="#cb35-471" aria-hidden="true" tabindex="-1"></a>genotypes are intermixed. We can obtain an alternative visualization by</span>
<span id="cb35-472"><a href="#cb35-472" aria-hidden="true" tabindex="-1"></a>arranging the cells in two dimensions using </span>
<span id="cb35-473"><a href="#cb35-473" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Uniform Manifold Approximation and Projection for Dimension Reduction (UMAP)</span><span class="co">](https://umap-learn.readthedocs.io/en/latest/)</span>:</span>
<span id="cb35-474"><a href="#cb35-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-477"><a href="#cb35-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-478"><a href="#cb35-478" aria-hidden="true" tabindex="-1"></a><span class="co"># run UMAP on the embedding</span></span>
<span id="cb35-479"><a href="#cb35-479" aria-hidden="true" tabindex="-1"></a>umap <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">umap</span>(<span class="fu">t</span>(fit<span class="sc">$</span>embedding))</span>
<span id="cb35-480"><a href="#cb35-480" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(umap) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"UMAP 1"</span>, <span class="st">"UMAP 2"</span>)</span>
<span id="cb35-481"><a href="#cb35-481" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(fit, <span class="st">"UMAP"</span>) <span class="ot">&lt;-</span> umap</span>
<span id="cb35-482"><a href="#cb35-482" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-483"><a href="#cb35-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-486"><a href="#cb35-486" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-487"><a href="#cb35-487" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-488"><a href="#cb35-488" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-489"><a href="#cb35-489" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-490"><a href="#cb35-490" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb35-491"><a href="#cb35-491" aria-hidden="true" tabindex="-1"></a>  fit, <span class="st">"UMAP"</span>, <span class="at">colour_by =</span> <span class="st">"mg_type"</span>, <span class="at">shape_by =</span> <span class="st">"genotype"</span>) <span class="sc">+</span></span>
<span id="cb35-492"><a href="#cb35-492" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-493"><a href="#cb35-493" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb35-494"><a href="#cb35-494" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Embedding after accounting for genotype (UMAP)"</span>,</span>
<span id="cb35-495"><a href="#cb35-495" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="fu">sprintf</span>(<span class="st">"Subsampled to %s microglia"</span>, <span class="fu">length</span>(selected_cells))) <span class="sc">+</span></span>
<span id="cb35-496"><a href="#cb35-496" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> <span class="fu">colData</span>(fit)<span class="sc">$</span>genotype)</span>
<span id="cb35-497"><a href="#cb35-497" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-498"><a href="#cb35-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-499"><a href="#cb35-499" aria-hidden="true" tabindex="-1"></a><span class="fu">### Testing for differential expression</span></span>
<span id="cb35-500"><a href="#cb35-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-501"><a href="#cb35-501" aria-hidden="true" tabindex="-1"></a>Next, the <span class="in">`test_de`</span> function performs a differential expression analysis for</span>
<span id="cb35-502"><a href="#cb35-502" aria-hidden="true" tabindex="-1"></a>locations in the embedding - by default, it will estimate it for each location</span>
<span id="cb35-503"><a href="#cb35-503" aria-hidden="true" tabindex="-1"></a>an original cell was mapped to.</span>
<span id="cb35-504"><a href="#cb35-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-505"><a href="#cb35-505" aria-hidden="true" tabindex="-1"></a>The <span class="in">`find_de_neighborhoods`</span> function accepts the original counts and will</span>
<span id="cb35-506"><a href="#cb35-506" aria-hidden="true" tabindex="-1"></a>estimate the log2 fold change for each neighborhood, based on aggregating the</span>
<span id="cb35-507"><a href="#cb35-507" aria-hidden="true" tabindex="-1"></a>counts to pseudobulk measures across the cells in each neighborhood.</span>
<span id="cb35-508"><a href="#cb35-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-511"><a href="#cb35-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-512"><a href="#cb35-512" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb35-513"><a href="#cb35-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb35-514"><a href="#cb35-514" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">test_de</span>(</span>
<span id="cb35-515"><a href="#cb35-515" aria-hidden="true" tabindex="-1"></a>  fit, </span>
<span id="cb35-516"><a href="#cb35-516" aria-hidden="true" tabindex="-1"></a>  <span class="at">contrast =</span> <span class="fu">cond</span>(<span class="at">genotype =</span> <span class="st">"Trem2_KO-5XFAD"</span>) <span class="sc">-</span> </span>
<span id="cb35-517"><a href="#cb35-517" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cond</span>(<span class="at">genotype =</span> <span class="st">"TREM2_CV-5XFAD"</span>))</span>
<span id="cb35-518"><a href="#cb35-518" aria-hidden="true" tabindex="-1"></a>neighborhoods <span class="ot">&lt;-</span> <span class="fu">find_de_neighborhoods</span>(</span>
<span id="cb35-519"><a href="#cb35-519" aria-hidden="true" tabindex="-1"></a>  fit, </span>
<span id="cb35-520"><a href="#cb35-520" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> <span class="fu">counts</span>(sce)[, selected_cells],</span>
<span id="cb35-521"><a href="#cb35-521" aria-hidden="true" tabindex="-1"></a>  <span class="at">group_by =</span> <span class="fu">vars</span>(sample, genotype),</span>
<span id="cb35-522"><a href="#cb35-522" aria-hidden="true" tabindex="-1"></a>  <span class="at">include_complement =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-523"><a href="#cb35-523" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-524"><a href="#cb35-524" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(pval) <span class="sc">%&gt;%</span></span>
<span id="cb35-525"><a href="#cb35-525" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb35-526"><a href="#cb35-526" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(fit)), <span class="st">"gene_id"</span>), </span>
<span id="cb35-527"><a href="#cb35-527" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="at">name =</span> <span class="st">"gene_id"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb35-528"><a href="#cb35-528" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(symbol, <span class="fu">everything</span>())</span>
<span id="cb35-529"><a href="#cb35-529" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-530"><a href="#cb35-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-531"><a href="#cb35-531" aria-hidden="true" tabindex="-1"></a>The <span class="in">`neighborhoods`</span> data.frame contains differential expression statistics</span>
<span id="cb35-532"><a href="#cb35-532" aria-hidden="true" tabindex="-1"></a>for each gene.</span>
<span id="cb35-533"><a href="#cb35-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-536"><a href="#cb35-536" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-537"><a href="#cb35-537" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(neighborhoods)</span>
<span id="cb35-538"><a href="#cb35-538" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-539"><a href="#cb35-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-540"><a href="#cb35-540" aria-hidden="true" tabindex="-1"></a>A volcano plot shows that we recovered a number of genes with differential </span>
<span id="cb35-541"><a href="#cb35-541" aria-hidden="true" tabindex="-1"></a>expression in one or more neighborhoods (after accounting for multiple testing):</span>
<span id="cb35-542"><a href="#cb35-542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-545"><a href="#cb35-545" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-546"><a href="#cb35-546" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-547"><a href="#cb35-547" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-548"><a href="#cb35-548" aria-hidden="true" tabindex="-1"></a><span class="co"># volcano plot</span></span>
<span id="cb35-549"><a href="#cb35-549" aria-hidden="true" tabindex="-1"></a>neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb35-550"><a href="#cb35-550" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> lfc, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pval))) <span class="sc">+</span></span>
<span id="cb35-551"><a href="#cb35-551" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color  =</span> adj_pval <span class="sc">&lt;</span> <span class="fl">0.1</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb35-552"><a href="#cb35-552" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Volcano plot of the neighborhoods"</span>) <span class="sc">+</span> </span>
<span id="cb35-553"><a href="#cb35-553" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"firebrick"</span>, <span class="st">"FALSE"</span> <span class="ot">=</span> <span class="st">"grey"</span>)) <span class="sc">+</span></span>
<span id="cb35-554"><a href="#cb35-554" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb35-555"><a href="#cb35-555" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb35-556"><a href="#cb35-556" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-557"><a href="#cb35-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-558"><a href="#cb35-558" aria-hidden="true" tabindex="-1"></a>For example, transcripts of the </span>
<span id="cb35-559"><a href="#cb35-559" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Lipoprotein lipase (Lpl) gene</span><span class="co">](https://en.wikipedia.org/wiki/Lipoprotein_lipase)</span></span>
<span id="cb35-560"><a href="#cb35-560" aria-hidden="true" tabindex="-1"></a>are generally expressed at lower levels in the Trem2 knock-out than </span>
<span id="cb35-561"><a href="#cb35-561" aria-hidden="true" tabindex="-1"></a>in wildtype samples. </span>
<span id="cb35-562"><a href="#cb35-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-563"><a href="#cb35-563" aria-hidden="true" tabindex="-1"></a>But there is also evidence for stronger differences in expression in microglia</span>
<span id="cb35-564"><a href="#cb35-564" aria-hidden="true" tabindex="-1"></a>that have adopted specific states. The largest log2 fold changes are observed</span>
<span id="cb35-565"><a href="#cb35-565" aria-hidden="true" tabindex="-1"></a>in <span class="in">`Cyc-M`</span>, and <span class="in">`DAM`</span> microglia, while the <span class="in">`Resting`</span> microglia are mainly</span>
<span id="cb35-566"><a href="#cb35-566" aria-hidden="true" tabindex="-1"></a>excluded from the neighborhood detected by <span class="in">`lemur`</span>.</span>
<span id="cb35-567"><a href="#cb35-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-570"><a href="#cb35-570" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-571"><a href="#cb35-571" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb35-572"><a href="#cb35-572" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-573"><a href="#cb35-573" aria-hidden="true" tabindex="-1"></a>sel_gene <span class="ot">&lt;-</span> <span class="fu">row.names</span>(sce)[<span class="fu">which</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>symbol <span class="sc">==</span> <span class="st">"Lpl"</span>)]</span>
<span id="cb35-574"><a href="#cb35-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-575"><a href="#cb35-575" aria-hidden="true" tabindex="-1"></a>neighborhood_coordinates <span class="ot">&lt;-</span> neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb35-576"><a href="#cb35-576" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> sel_gene) <span class="sc">%&gt;%</span></span>
<span id="cb35-577"><a href="#cb35-577" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">cell_id =</span> purrr<span class="sc">:::</span><span class="fu">map</span>(indices, \(idx) <span class="fu">colnames</span>(fit)[idx])) <span class="sc">%&gt;%</span></span>
<span id="cb35-578"><a href="#cb35-578" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">unnest</span>(<span class="fu">c</span>(indices, cell_id)) <span class="sc">%&gt;%</span></span>
<span id="cb35-579"><a href="#cb35-579" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(<span class="fu">as_tibble</span>(umap, <span class="at">rownames =</span> <span class="st">"cell_id"</span>), <span class="at">by =</span> <span class="st">"cell_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-580"><a href="#cb35-580" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(name, cell_id, <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)</span>
<span id="cb35-581"><a href="#cb35-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-582"><a href="#cb35-582" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(umap) <span class="sc">%&gt;%</span></span>
<span id="cb35-583"><a href="#cb35-583" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">expr =</span> <span class="fu">assay</span>(fit, <span class="st">"DE"</span>)[sel_gene, ]) <span class="sc">%&gt;%</span></span>
<span id="cb35-584"><a href="#cb35-584" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb35-585"><a href="#cb35-585" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>() <span class="sc">+</span></span>
<span id="cb35-586"><a href="#cb35-586" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> expr)) <span class="sc">+</span></span>
<span id="cb35-587"><a href="#cb35-587" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density2d</span>(<span class="at">data =</span> neighborhood_coordinates, <span class="at">breaks =</span> <span class="fl">0.1</span>, </span>
<span id="cb35-588"><a href="#cb35-588" aria-hidden="true" tabindex="-1"></a>                 <span class="at">contour_var =</span> <span class="st">"ndensity"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb35-589"><a href="#cb35-589" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>]) <span class="sc">+</span> </span>
<span id="cb35-590"><a href="#cb35-590" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb35-591"><a href="#cb35-591" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb35-592"><a href="#cb35-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-593"><a href="#cb35-593" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(umap) <span class="sc">%&gt;%</span></span>
<span id="cb35-594"><a href="#cb35-594" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_cols</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(fit))) <span class="sc">%&gt;%</span></span>
<span id="cb35-595"><a href="#cb35-595" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">UMAP 1</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">UMAP 2</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb35-596"><a href="#cb35-596" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> mg_type)) <span class="sc">+</span></span>
<span id="cb35-597"><a href="#cb35-597" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-598"><a href="#cb35-598" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density2d</span>(<span class="at">data =</span> neighborhood_coordinates, <span class="at">breaks =</span> <span class="fl">0.1</span>, </span>
<span id="cb35-599"><a href="#cb35-599" aria-hidden="true" tabindex="-1"></a>                 <span class="at">contour_var =</span> <span class="st">"ndensity"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb35-600"><a href="#cb35-600" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Microglia states"</span>) <span class="sc">+</span> </span>
<span id="cb35-601"><a href="#cb35-601" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb35-602"><a href="#cb35-602" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb35-603"><a href="#cb35-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-604"><a href="#cb35-604" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span>
<span id="cb35-605"><a href="#cb35-605" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-606"><a href="#cb35-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-607"><a href="#cb35-607" aria-hidden="true" tabindex="-1"></a>Next, I examine each microglial subtype separately, and split cells according</span>
<span id="cb35-608"><a href="#cb35-608" aria-hidden="true" tabindex="-1"></a>to whether they fall into a neighborhood with significant differential <span class="in">`Lpl`</span> </span>
<span id="cb35-609"><a href="#cb35-609" aria-hidden="true" tabindex="-1"></a>expression (<span class="in">`DE`</span>) or not (<span class="in">`not DE`</span>).</span>
<span id="cb35-610"><a href="#cb35-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-611"><a href="#cb35-611" aria-hidden="true" tabindex="-1"></a>Most <span class="in">`Cyc-M`</span> and <span class="in">`DAM`</span> microglia are located in neighborhoods with </span>
<span id="cb35-612"><a href="#cb35-612" aria-hidden="true" tabindex="-1"></a>reduced <span class="in">`Lpl`</span> expression in knock-out samples, e.g. the majority of these cells</span>
<span id="cb35-613"><a href="#cb35-613" aria-hidden="true" tabindex="-1"></a>is in the <span class="in">`DE`</span> column. The opposite is true for <span class="in">`Resting`</span> microglia: nearly all</span>
<span id="cb35-614"><a href="#cb35-614" aria-hidden="true" tabindex="-1"></a>of them are outside the significant <span class="in">`Lpl`</span> differential expression neighborhood.</span>
<span id="cb35-615"><a href="#cb35-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-616"><a href="#cb35-616" aria-hidden="true" tabindex="-1"></a>The transitional subtypes, t1 (most similar to resting microglia) to t6 </span>
<span id="cb35-617"><a href="#cb35-617" aria-hidden="true" tabindex="-1"></a>(most similar to DAM, Cyc-M or IFN-R) fall in between, with a gradual increase</span>
<span id="cb35-618"><a href="#cb35-618" aria-hidden="true" tabindex="-1"></a>along their proposed differentiation sequence.</span>
<span id="cb35-619"><a href="#cb35-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-622"><a href="#cb35-622" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-623"><a href="#cb35-623" aria-hidden="true" tabindex="-1"></a>within_neighborhood <span class="ot">&lt;-</span> neighborhoods <span class="sc">%&gt;%</span></span>
<span id="cb35-624"><a href="#cb35-624" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> sel_gene) <span class="sc">%&gt;%</span> </span>
<span id="cb35-625"><a href="#cb35-625" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(indices)</span>
<span id="cb35-626"><a href="#cb35-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-627"><a href="#cb35-627" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(fit)<span class="sc">$</span>neighborhood <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb35-628"><a href="#cb35-628" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq.int</span>(<span class="fu">ncol</span>(fit)) <span class="sc">%in%</span> within_neighborhood[[<span class="dv">1</span>]], <span class="st">"DE"</span>, <span class="st">"not DE"</span>)</span>
<span id="cb35-629"><a href="#cb35-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-630"><a href="#cb35-630" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb35-631"><a href="#cb35-631" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(fit)</span>
<span id="cb35-632"><a href="#cb35-632" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb35-633"><a href="#cb35-633" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> neighborhood, <span class="at">fill =</span> mg_type)) <span class="sc">+</span> </span>
<span id="cb35-634"><a href="#cb35-634" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"count"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb35-635"><a href="#cb35-635" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> mg_type) <span class="sc">+</span></span>
<span id="cb35-636"><a href="#cb35-636" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>], </span>
<span id="cb35-637"><a href="#cb35-637" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb35-638"><a href="#cb35-638" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Cells per neighborhood"</span>) <span class="sc">+</span></span>
<span id="cb35-639"><a href="#cb35-639" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-640"><a href="#cb35-640" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb35-641"><a href="#cb35-641" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb35-642"><a href="#cb35-642" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-643"><a href="#cb35-643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-644"><a href="#cb35-644" aria-hidden="true" tabindex="-1"></a>Finally, we can plot the estimated <span class="in">`Lpl`</span> log2 fold changes for each cells</span>
<span id="cb35-645"><a href="#cb35-645" aria-hidden="true" tabindex="-1"></a>annotated with the various subtype labels. This confirms the observations</span>
<span id="cb35-646"><a href="#cb35-646" aria-hidden="true" tabindex="-1"></a>outlined above, providing an example of how insights from LEMUR can be combined</span>
<span id="cb35-647"><a href="#cb35-647" aria-hidden="true" tabindex="-1"></a>with coarser, clustering-based insights.</span>
<span id="cb35-648"><a href="#cb35-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-651"><a href="#cb35-651" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb35-652"><a href="#cb35-652" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb35-653"><a href="#cb35-653" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb35-654"><a href="#cb35-654" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb35-655"><a href="#cb35-655" aria-hidden="true" tabindex="-1"></a>  <span class="at">expr =</span> <span class="fu">assay</span>(fit, <span class="st">"DE"</span>)[sel_gene, ],</span>
<span id="cb35-656"><a href="#cb35-656" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(fit)</span>
<span id="cb35-657"><a href="#cb35-657" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb35-658"><a href="#cb35-658" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> mg_type, <span class="at">y =</span> expr, <span class="at">fill =</span> mg_type)) <span class="sc">+</span> </span>
<span id="cb35-659"><a href="#cb35-659" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb35-660"><a href="#cb35-660" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span> </span>
<span id="cb35-661"><a href="#cb35-661" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> mg_colors, <span class="at">name =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb35-662"><a href="#cb35-662" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">rowData</span>(sce)[sel_gene, <span class="st">"symbol"</span>],</span>
<span id="cb35-663"><a href="#cb35-663" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Genotype effect (log2 FC)"</span>, </span>
<span id="cb35-664"><a href="#cb35-664" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb35-665"><a href="#cb35-665" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>(<span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb35-666"><a href="#cb35-666" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb35-667"><a href="#cb35-667" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-668"><a href="#cb35-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-669"><a href="#cb35-669" aria-hidden="true" tabindex="-1"></a><span class="fu">## Conclusions</span></span>
<span id="cb35-670"><a href="#cb35-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-671"><a href="#cb35-671" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>This dataset starts off with a relatively homogeneous cell population, e.g.</span>
<span id="cb35-672"><a href="#cb35-672" aria-hidden="true" tabindex="-1"></a>  it is designed to examine subtle differences _within_ a single cell type </span>
<span id="cb35-673"><a href="#cb35-673" aria-hidden="true" tabindex="-1"></a>  (microglia.)</span>
<span id="cb35-674"><a href="#cb35-674" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Removing _Trem2_ activity is known to shift the composition of the microglial</span>
<span id="cb35-675"><a href="#cb35-675" aria-hidden="true" tabindex="-1"></a>  subsets, e.g. depleting <span class="in">`DAM`</span> microglia and increasing the frequency of </span>
<span id="cb35-676"><a href="#cb35-676" aria-hidden="true" tabindex="-1"></a>  <span class="in">`Resting`</span> in the knock-out versus the wildtype samples. This adds an</span>
<span id="cb35-677"><a href="#cb35-677" aria-hidden="true" tabindex="-1"></a>  additional challenge to the task of identifying differential expression.</span>
<span id="cb35-678"><a href="#cb35-678" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The <span class="in">`lemur()`</span> package nevertheless successfully identified genes that track</span>
<span id="cb35-679"><a href="#cb35-679" aria-hidden="true" tabindex="-1"></a>  the differential expression of the microglial sub-states reported by Ellwanger </span>
<span id="cb35-680"><a href="#cb35-680" aria-hidden="true" tabindex="-1"></a>  et al. (Even with only a subsample of the data.)</span>
<span id="cb35-681"><a href="#cb35-681" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>I am looking forward to exploring <span class="in">`LEMUR`</span> in future datasets, e.g. those</span>
<span id="cb35-682"><a href="#cb35-682" aria-hidden="true" tabindex="-1"></a>  examining the effects of drug perturbation in single-nuclei RNA-seq datasets</span>
<span id="cb35-683"><a href="#cb35-683" aria-hidden="true" tabindex="-1"></a>  sampling a large variety of CNS cell types.</span>
<span id="cb35-684"><a href="#cb35-684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-685"><a href="#cb35-685" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb35-686"><a href="#cb35-686" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;</span></span>
<span id="cb35-687"><a href="#cb35-687" aria-hidden="true" tabindex="-1"></a>Reproducibility</span>
<span id="cb35-688"><a href="#cb35-688" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/summary&gt;</span></span>
<span id="cb35-689"><a href="#cb35-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-690"><a href="#cb35-690" aria-hidden="true" tabindex="-1"></a><span class="in">```{r sessioninfo}</span></span>
<span id="cb35-691"><a href="#cb35-691" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb35-692"><a href="#cb35-692" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>()</span>
<span id="cb35-693"><a href="#cb35-693" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb35-694"><a href="#cb35-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-695"><a href="#cb35-695" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb35-696"><a href="#cb35-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-697"><a href="#cb35-697" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>scRNA-seq data from glioblastoma slices cultured _in vitro_ with either </span>
<span id="cb35-698"><a href="#cb35-698" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">pamobinostat</span><span class="co">](https://en.wikipedia.org/wiki/Panobinostat)</span></span>
<span id="cb35-699"><a href="#cb35-699" aria-hidden="true" tabindex="-1"></a>or a vehicle control, characterized in a terrific paper by</span>
<span id="cb35-700"><a href="#cb35-700" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Zhao et al, 2021</span><span class="co">](https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-021-00894-y)</span></span>
<span id="cb35-701"><a href="#cb35-701" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span>CD45 is a cell surface antigen that is expressed on most hematopoietic </span>
<span id="cb35-702"><a href="#cb35-702" aria-hidden="true" tabindex="-1"></a>lineage cells, including microglia.</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>