<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-01-16">

<title>Thomas Sandmann’s blog - QuantSeq RNAseq analysis (1): configuring the nf-core/rnaseq workflow</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-XNBDW4FYSC"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-XNBDW4FYSC', { 'anonymize_ip': true});
</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/"><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann"><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">QuantSeq RNAseq analysis (1): configuring the nf-core/rnaseq workflow</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">nextflow</div>
                <div class="quarto-category">NGS</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 16, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#installing-nextflow" id="toc-installing-nextflow" class="nav-link" data-scroll-target="#installing-nextflow">Installing Nextflow</a>
  <ul>
  <li><a href="#nextflow-configs" id="toc-nextflow-configs" class="nav-link" data-scroll-target="#nextflow-configs">Nextflow configs</a></li>
  </ul></li>
  <li><a href="#analysis-of-published-studies" id="toc-analysis-of-published-studies" class="nav-link" data-scroll-target="#analysis-of-published-studies">Analysis of published studies</a></li>
  <li><a href="#quantseq-fwd-without-umis-xia-et-al-dataset" id="toc-quantseq-fwd-without-umis-xia-et-al-dataset" class="nav-link" data-scroll-target="#quantseq-fwd-without-umis-xia-et-al-dataset">Quantseq FWD without UMIs: Xia et al dataset</a>
  <ul>
  <li><a href="#experimental-design" id="toc-experimental-design" class="nav-link" data-scroll-target="#experimental-design">Experimental design</a></li>
  <li><a href="#raw-data-retrieval-with-nf-corefetchngs" id="toc-raw-data-retrieval-with-nf-corefetchngs" class="nav-link" data-scroll-target="#raw-data-retrieval-with-nf-corefetchngs">Raw data retrieval with nf-core/fetchngs</a></li>
  <li><a href="#configuring-the-nf-corernaseq-workflow" id="toc-configuring-the-nf-corernaseq-workflow" class="nav-link" data-scroll-target="#configuring-the-nf-corernaseq-workflow">Configuring the nf-core/rnaseq workflow</a>
  <ul class="collapse">
  <li><a href="#reference-data" id="toc-reference-data" class="nav-link" data-scroll-target="#reference-data">Reference data</a></li>
  <li><a href="#custom-arguments" id="toc-custom-arguments" class="nav-link" data-scroll-target="#custom-arguments">Custom arguments</a></li>
  <li><a href="#extra-star-arguments" id="toc-extra-star-arguments" class="nav-link" data-scroll-target="#extra-star-arguments">Extra STAR arguments</a></li>
  <li><a href="#extra-salmon-arguments" id="toc-extra-salmon-arguments" class="nav-link" data-scroll-target="#extra-salmon-arguments">Extra Salmon arguments</a></li>
  </ul></li>
  <li><a href="#starting-the-workflow" id="toc-starting-the-workflow" class="nav-link" data-scroll-target="#starting-the-workflow">Starting the workflow</a></li>
  <li><a href="#examining-the-output-files" id="toc-examining-the-output-files" class="nav-link" data-scroll-target="#examining-the-output-files">Examining the output files</a></li>
  <li><a href="#reusing-reference-files-and-indices" id="toc-reusing-reference-files-and-indices" class="nav-link" data-scroll-target="#reusing-reference-files-and-indices">Reusing reference files and indices</a></li>
  </ul></li>
  <li><a href="#quantseq-fwd-with-umis-nugent-et-al-dataset" id="toc-quantseq-fwd-with-umis-nugent-et-al-dataset" class="nav-link" data-scroll-target="#quantseq-fwd-with-umis-nugent-et-al-dataset">Quantseq FWD with UMIs: Nugent et al dataset</a>
  <ul>
  <li><a href="#experimental-design-1" id="toc-experimental-design-1" class="nav-link" data-scroll-target="#experimental-design-1">Experimental design</a></li>
  <li><a href="#raw-data-retrieval-with-nf-corefetchngs-1" id="toc-raw-data-retrieval-with-nf-corefetchngs-1" class="nav-link" data-scroll-target="#raw-data-retrieval-with-nf-corefetchngs-1">Raw data retrieval with nf-core/fetchngs</a></li>
  <li><a href="#configuring-the-nf-corernaseq-workflow-1" id="toc-configuring-the-nf-corernaseq-workflow-1" class="nav-link" data-scroll-target="#configuring-the-nf-corernaseq-workflow-1">Configuring the nf-core/rnaseq workflow</a>
  <ul class="collapse">
  <li><a href="#reference-data-1" id="toc-reference-data-1" class="nav-link" data-scroll-target="#reference-data-1">Reference data</a></li>
  <li><a href="#unique-molecular-identifiers" id="toc-unique-molecular-identifiers" class="nav-link" data-scroll-target="#unique-molecular-identifiers">Unique molecular identifiers</a></li>
  </ul></li>
  <li><a href="#starting-the-workflow-1" id="toc-starting-the-workflow-1" class="nav-link" data-scroll-target="#starting-the-workflow-1">Starting the workflow</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>This is the first of four posts documenting my progress toward processing and analyzing <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq FWD 3’ tag RNAseq</a> data with the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq</a> workflow.</p>
<ol type="1">
<li><strong>Configuring the nf-core/rnaseq workflow</strong></li>
<li><a href="../nextflow-core-quantseq-2-output/index.html">Exploring the workflow outputs</a></li>
<li><a href="../nextflow-core-quantseq-3-xia/index.html">Validating the workflow by reproducing results published by Xia et al (no UMIs)</a></li>
<li><a href="../nextflow-core-quantseq-4-nugent/index.html">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</a></li>
</ol>
<p>Many thanks to <a href="https://github.com/drpatelh">Harshil Patel</a>, <a href="https://github.com/adomingues/">António Miguel de Jesus Domingues</a> and <a href="https://github.com/matthiasZepper/">Matthias Zepper</a> for their generous guidance &amp; input via <a href="nfcore.slack.com">nf-core slack</a>. (Any mistakes are mine.)</p>
</div>
</div>
</div>
<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<ul>
<li>This tutorial documents how to configure &amp; execute the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq workflow</a> for the processing of raw QuantSeq 3’ FWD RNA-seq data.</li>
<li>It highlights <a href="#custom-arguments">custom nf-core/rnaseq parameters settings</a> and applies them to two published datasets, one with and one without UMIs.</li>
</ul>
</section>
<section id="installing-nextflow" class="level2">
<h2 class="anchored" data-anchor-id="installing-nextflow">Installing Nextflow</h2>
<p>To run nextflow, we create a new <a href="https://docs.conda.io/en/latest/">conda</a> environment. (See <a href="https://docs.conda.io/en/latest/miniconda.html#latest-miniconda-installer-links">conda installation instructions</a> if you are not yet familiar with the conda package manager.) Then we install <a href="https://www.nextflow.io/">nextflow</a> and its dependencies from the <a href="https://anaconda.org/bioconda/nextflow">bioconda</a> repository.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="ex">create</span> <span class="at">-n</span> nf</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="fu">install</span> <span class="at">-c</span> bioconda nextflow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once the installation is complete, we activate our new conda environment and verify that Nextflow is installed:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="ex">activate</span> nf</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">-v</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> version 22.10.4.5836</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="nextflow-configs" class="level3">
<h3 class="anchored" data-anchor-id="nextflow-configs">Nextflow configs</h3>
<p>Nextflow separates the definition of a workflow from the platform it is executed on. For example, you can run the same workflow on your local computer, a high-performance cluster or using a cloud provider like Amazon Web Services (AWS) or Google Cloud Services (GCS).</p>
<p>Each nf-core workflow defines the required software sources, which can either be installed (e.g.&nbsp;via conda) or be made available through a container engine such as <a href="https://www.docker.com/">docker</a>, <a href="https://en.wikipedia.org/wiki/Singularity_(software)">singularity</a>, etc. For details, please consult the <a href="https://www.nextflow.io/docs/latest/config.html">Nextflow documentation</a>.</p>
<p>The desired configuration is specified in one or more configuration file(s), and passed to the <code>nextflow</code> command via its <code>-profile</code> argument. Here, we execute the workflows on the local system, using docker containers by specifying the built-in <code>-profile docker</code> configuration.</p>
</section>
</section>
<section id="analysis-of-published-studies" class="level2">
<h2 class="anchored" data-anchor-id="analysis-of-published-studies">Analysis of published studies</h2>
<p>To validate the nf-core/rnaseq workflow for QuantSeq 3’ tag RNA-seq data, we reanalyze data from two published studies, one with and one without the use of unique molecular identifiers (UMIs).</p>
<ol type="1">
<li>“Fibrillar Aβ causes profound microglial metabolic perturbations in a novel APP knock-in mouse model” by <a href="https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1">Xia et al, 2021</a></li>
</ol>
<ul>
<li><a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq 3’ mRNA-Seq Library Prep Kit FWD FWD</a></li>
<li>This data does <strong>not</strong> include UMIs.</li>
<li>Raw data are available via
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158152">GEO accession GSE158152</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/sra?term=SRP282921">SRA accession SRP213880</a></li>
</ul></li>
</ul>
<ol start="2" type="1">
<li>“TREM2 regulates microglial lipid metabolism during aging in mice” by <a href="https://www.cell.com/neuron/fulltext/S0896-6273(19)31049-9">Nugent et al, 2020</a>.</li>
</ol>
<ul>
<li><a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq 3’ mRNA-Seq Library Prep Kit FWD FWD</a></li>
<li>This data includes <a href="https://www.lexogen.com/store/umi-second-strand-synthesis-module-for-quantseq-fwd/">UMIs</a>, which and can be used to filter alignments originating from PCR duplicates.</li>
<li>Raw data are available via
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134031">GEO accession GSE134031</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/sra?term=SRP213880">SRA accession SRP213880</a></li>
</ul></li>
</ul>
</section>
<section id="quantseq-fwd-without-umis-xia-et-al-dataset" class="level2">
<h2 class="anchored" data-anchor-id="quantseq-fwd-without-umis-xia-et-al-dataset">Quantseq FWD without UMIs: Xia et al dataset</h2>
<p>Here, we will reanalyze RNA-seq data published in the preprint “Fibrillar Aβ causes profound microglial metabolic perturbations in a novel APP knock-in mouse model” by <a href="https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1">Xia et al, 2021</a>.</p>
<section id="experimental-design" class="level3">
<h3 class="anchored" data-anchor-id="experimental-design">Experimental design</h3>
<p>This study examines FACS-isolated microglia from 18 mice from three genetic backgrounds (N = 6 WT, N =6 heterozygous App-SAA and N = 6 homozygous App-SAA animals). Microglia were isolated from the brain (cortex &amp; hippocampus) of each animal using FACS and gene expression changes were analyzed using 3-tag RNA-seq.</p>
<p>From each mouse, two (technical) replicate pools of microglia were collected and processed into separate libraries with <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">Lexogen’s QuantSeq FWD</a>. This dataset does <strong>not</strong> include unique molecular identifiers (UMIs).</p>
</section>
<section id="raw-data-retrieval-with-nf-corefetchngs" class="level3">
<h3 class="anchored" data-anchor-id="raw-data-retrieval-with-nf-corefetchngs">Raw data retrieval with nf-core/fetchngs</h3>
<p>We start the analysis by downloading the raw data (36 FASTQ files) and the sample metadata with the <a href="https://nf-co.re/fetchngs">nf-core/fetchngs</a> workflow. For datasets stored in SRA, we only need to paste the <code>SRA Study identifier</code> (SRP282921) into a text file for us as the <code>input</code> for the fetchngs workflow. The following command will download all of the FASTQ files associated with this study into the <code>raw_data</code> directory:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> mkdir <span class="ex">SRP282921</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> cd <span class="ex">SRP282921</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">SRP282921</span> <span class="op">&gt;</span> ids.txt</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> nf-core/fetchngs <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">-revision</span> 1.9 <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> raw_data <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> ids.txt <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_pipeline</span> rnaseq <span class="dt">\</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_rnaseq_strandedness</span> forward</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>By specifying the <code>--nf_core_pipeline</code> argument, the workflow creates a sample sheet that can be used as input for the <code>nf-core/rnaseq</code> pipeline. Because the data was generated with the <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq FWD</a> kit, we also know that the reads are from the <code>forward</code> strand and add this information to the sample sheet via the <code>--nf_core_rnaseq_strandedness</code> argument. <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<p>Once the workflow has completed, the <code>raw_data</code> output directory contains the following structure:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> tree <span class="ex">raw_data</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">raw_data/</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> custom</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── user-settings.mkfg</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> fastq</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── md5</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── SRX9142647_SRR12661938.fastq.gz.md5</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── SRX9142648_SRR12661924.fastq.gz.md5</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── 34 additional .fastq.gz.md5 files <span class="er">(</span><span class="ex">not</span> shown<span class="kw">)</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SRX9142647_SRR12661938.fastq.gz</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SRX9142648_SRR12661924.fastq.gz</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── 34 additional .fastq.gz files <span class="er">(</span><span class="ex">not</span> shown<span class="kw">)</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> metadata</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── SRP282921.runinfo_ftp.tsv</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> pipeline_info</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_report_2023-01-12_01-49-02.html</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_timeline_2023-01-12_01-49-02.html</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_trace_2023-01-12_01-49-02.txt</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── pipeline_dag_2023-01-12_01-49-02.html</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── software_versions.yml</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> samplesheet</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> id_mappings.csv</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> multiqc_config.yml</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> nf_params.json</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> run.sh</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> samplesheet.csv</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="ex">6</span> directories, 90 files</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The 36 FASTQ files are in the <code>fastq</code> subfolder, each accompanied by a MD5 checksum file in the <code>md5</code> directory.</li>
<li>In the <code>samplesheet</code> subdirectory, we find the <code>samplesheet.csv</code> file that we can pass to the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq workflow</a>.</li>
</ul>
<p>For more details about the output oft he workflow, please check <a href="../nextflow-core-quantseq-2-output/index.html">the next post in this series</a>] and the <a href="https://nf-co.re/rnaseq/3.10.1/output">nf-core/rnaseq output docs</a>.</p>
</section>
<section id="configuring-the-nf-corernaseq-workflow" class="level3">
<h3 class="anchored" data-anchor-id="configuring-the-nf-corernaseq-workflow">Configuring the nf-core/rnaseq workflow</h3>
<section id="reference-data" class="level4">
<h4 class="anchored" data-anchor-id="reference-data">Reference data</h4>
<p>According to the sample metadata in <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4793335">NCBI GEO</a> the authors mapped the data to the mouse GRCm38 genome version, using Gencode annotations from <a href="https://www.gencodegenes.org/mouse/release_M17.html">release M17</a>. To later compare our results with those obtained in the original publication, we use the same Gencode version (even though it is not the most recent).</p>
<p>The <code>nf-core/rnaseq</code> workflow can automatically generate all necessary indices when provided with</p>
<ol type="1">
<li>A FASTA file with the genomic sequences, e.g.&nbsp;Gencode’s <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/GRCm38.primary_assembly.genome.fa.gz">Primary genome assembly</a> and</li>
<li>A GFP file with matching gene annotations, e.g.&nbsp; Gencode <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/gencode.vM17.primary_assembly.annotation.gtf.gz">Comprehensive gene annotations</a></li>
</ol>
<p>Because downloading and indexing the genome is time consuming, we will save them for later use.</p>
</section>
<section id="custom-arguments" class="level4">
<h4 class="anchored" data-anchor-id="custom-arguments">Custom arguments</h4>
<p>The <code>nf-core/rnaseq</code> pipeline has robust default parameters that are suitable for whole transcriptome RNA-seq data. But this dataset was generated with Lexogen’s QuantSeq FWD 3’ mRNA library preparation kit, which specificially targets the 3’ end of polyadenylated transcripts. As a consequence, reads align primarily to the 3’ UTR and the final exon and coverage of the remainder of the gene body is minimal.</p>
<p>To account for these properties, we supply a number of custom arguments to the <code>nextflow run</code> command, which augment the workflow for this specific data type.</p>
</section>
<section id="extra-star-arguments" class="level4">
<h4 class="anchored" data-anchor-id="extra-star-arguments">Extra STAR arguments</h4>
<p>Lexogen provides an example analysis workflow <a href="https://www.lexogen.com/quantseq-data-analysis/">on their website</a>, which uses the STAR aligner with several non-default parameters. Most of them parameters correspond to the <em>ENCODE standard options</em> listed in the <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR manual</a>:</p>
<ul>
<li><code>--alignIntronMax 1000000</code></li>
<li><code>--alignIntronMin 20</code></li>
<li><code>--alignMatesGapMax 1000000</code></li>
<li><code>--alignSJoverhangMin 8</code></li>
<li><code>--alignSJDBoverhangMin 1</code></li>
<li><code>--outFilterMismatchNmax 999</code></li>
<li><code>--outFilterMultimapNmax 20</code></li>
<li><code>--outFilterType BySJout</code></li>
</ul>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Important
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <a href="https://www.genome.gov/Funded-Programs-Projects/ENCODE-Project-ENCyclopedia-Of-DNA-Elements">ENCODE project</a> originally focussed on data from human and mouse. The parameters shown above are suitable for these genomes or those with comparable characteristics. For analysis of data from other organisms, e.g.&nbsp;those with shorter introns, you might need to modify them.</p>
</div>
</div>
</div>
<p>In addition, Lexogen also specified the</p>
<ul>
<li><code>--outFilterMismatchNoverLmax 0.1</code>, decreasing the tolerance for mismatches.</li>
</ul>
<p>Lexogen’s example workflow uses <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/">bbduk</a> to trim adapters, poly(A) tails and low quality bases from the 3’ end of reads. The nf-core/rnaseq workflow uses <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a> to trim adapters, but does not remove poly(A) tails. Instead we use the STAR ligner to clip poly(A) tails during the alignment stage:</p>
<ul>
<li><code>--clip3pAdapterSeq AAAAAAAA</code></li>
</ul>
<p>We will pass all of the arguments listed above to STAR via nf-core/rnaseq’s <code>--extra_star_align_args</code> argument (as a string, see example below).</p>
</section>
<section id="extra-salmon-arguments" class="level4">
<h4 class="anchored" data-anchor-id="extra-salmon-arguments">Extra Salmon arguments</h4>
<p>The <a href="https://www.nature.com/articles/nmeth.4197">salmon</a> algorithm takes into account transcript length when quantifying gene expression. Because QuantSeq (and other 3’ tag sequencing methods) only capture the 3’ most part of each transcript, not its full length, this feature needs to be deactivated:</p>
<ul>
<li><code>--noLengthCorrection</code></li>
</ul>
<p>The nf-core/rnaseq workflow features a special argument, <code>extra_salmon_quant_args</code> to pass additional arguments to the <code>salmon</code> tool (see syntax below).</p>
</section>
</section>
<section id="starting-the-workflow" class="level3">
<h3 class="anchored" data-anchor-id="starting-the-workflow">Starting the workflow</h3>
<p>We can pass the parameters we discussed above as individual arguments to the <code>nextflow run</code> command (prefixed with <code>--</code>). Alternatively, we can collect them in a JSON file and specify its name via the <code>-params-file</code> argument.</p>
<p>For example, the following JSON string specifies the parameters for analyzing sQuantSeq FWD data <strong>without UMIs</strong>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode json code-with-copy"><code class="sourceCode json"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"save_reference"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"fasta"</span><span class="fu">:</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/GRCm38.primary_assembly.genome.fa.gz"</span><span class="fu">,</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gtf"</span><span class="fu">:</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/gencode.vM17.primary_assembly.annotation.gtf.gz"</span><span class="fu">,</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_star_align_args"</span><span class="fu">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="fu">,</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"with_umi"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_salmon_quant_args"</span><span class="fu">:</span> <span class="st">"--noLengthCorrection"</span><span class="fu">,</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"skip_stringtie"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gencode"</span><span class="fu">:</span> <span class="kw">true</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We store the JSON string in a file called <code>nf_params.json</code>, and then pass it to the <code>nextflow run</code> command <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>      nf-core/rnaseq <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">-r</span> 3.10.1 <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">-params-file</span> nf_params.json <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> raw_data/samplesheet/samplesheet.csv <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> SRP282921</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="examining-the-output-files" class="level3">
<h3 class="anchored" data-anchor-id="examining-the-output-files">Examining the output files</h3>
</section>
<section id="reusing-reference-files-and-indices" class="level3">
<h3 class="anchored" data-anchor-id="reusing-reference-files-and-indices">Reusing reference files and indices</h3>
<p>Retrieving the reference files and indexing them is time consuming. If you are planning to map additional sample against the same genome / transcriptome in the future, you might want to reuse them. Because we set the <code>save_reference</code> argument to <code>true</code> in our workflow, the output directory contains a <code>genome</code> folder, featuring reference files (e.g.&nbsp;genome and transcriptome sequences in FASTQ formats) and indices (e.g.&nbsp;for STAR, salmon and rsem).</p>
<p>Let’s move the genome folder to our current working directory (or any other suitable location). <a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> For future runs, we can include paths to the files and indices in our <code>nf_params.json</code> file, e.g.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode json code-with-copy"><code class="sourceCode json"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"save_reference"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"fasta"</span><span class="fu">:</span> <span class="st">"genome/GRCm38.primary_assembly.genome.fa"</span><span class="fu">,</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gtf"</span><span class="fu">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.gtf"</span><span class="fu">,</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gene_bed"</span><span class="fu">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.bed"</span><span class="fu">,</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"transcript_fasta"</span><span class="fu">:</span> <span class="st">"genome/genome.transcripts.fa"</span><span class="fu">,</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"star_index"</span><span class="fu">:</span> <span class="st">"genome/index/star"</span><span class="fu">,</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"salmon_index"</span><span class="fu">:</span> <span class="st">"genome/index/salmon"</span><span class="fu">,</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"rsem_index"</span><span class="fu">:</span> <span class="st">"genome/rsem"</span><span class="fu">,</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_star_align_args"</span><span class="fu">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="fu">,</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"with_umi"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_salmon_quant_args"</span><span class="fu">:</span> <span class="st">"--noLengthCorrection"</span><span class="fu">,</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"skip_stringtie"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gencode"</span><span class="fu">:</span> <span class="kw">true</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="fu">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Our next run will execute much faster, because these reference files and indices are already available and don’t need to be created from scratch.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you chose a location other than <code>genome</code> to store your reference data, please specify the (absolute) path in the JSON file instead.</p>
</div>
</div>
</section>
</section>
<section id="quantseq-fwd-with-umis-nugent-et-al-dataset" class="level2">
<h2 class="anchored" data-anchor-id="quantseq-fwd-with-umis-nugent-et-al-dataset">Quantseq FWD with UMIs: Nugent et al dataset</h2>
<p>Especially when only low amounts of input material (total RNA) are available, the QuantSeq FWD procotol requires multiple PCR amplification steps. Each step generates clones of the originally captured tags, but does not provide new information about transcript abundance. Inclusion of unique molecular identifiers (UMIs) in the second strand synthesis guarantees that PCR duplicates can be identified and removed in the analysis (based on both the coordinates of their alignment and the UMI sequence).</p>
<p>Nugent et al used the QuantSeq FWD protocol with UMIs, and we will reanalyze this dataset to illustrate how to instruct <a href="https://github.com/CGATOxford/UMI-tools">UMI-tools</a> to extract the UMIs from the raw reads and deduplicated the STAR alignments.</p>
<section id="experimental-design-1" class="level3">
<h3 class="anchored" data-anchor-id="experimental-design-1">Experimental design</h3>
<p>This study examines FACS-isolated astroctyes and microglia from female mice that were either 2- or 16 months of age. The animals are either wildtype (<code>WT</code>) or<br>
homozygous knockouts (<code>KO</code>) for the <a href="https://www.uniprot.org/uniprotkb/Q99NH8/entry">Trem2 gene</a>.</p>
<p>From each mouse, separate microglia and astrocyte samples were collected and processed into separate libraries with <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">Lexogen’s QuantSeq FWD</a> kit <em>and</em> the <a href="https://faqs.lexogen.com/faq/quantseq-umi-module">QuantSeq UMI add-on module</a>.</p>
</section>
<section id="raw-data-retrieval-with-nf-corefetchngs-1" class="level3">
<h3 class="anchored" data-anchor-id="raw-data-retrieval-with-nf-corefetchngs-1">Raw data retrieval with nf-core/fetchngs</h3>
<p>As in the first example, we use the <code>nf-core/fetchngs</code> workflow to retrieve the FASTQ files from SRA.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> mkdir <span class="ex">SRP213880</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> cd <span class="ex">SRP213880</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">SRP213880</span> <span class="op">&gt;</span> ids.txt</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> nf-core/fetchngs <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">-revision</span> 1.9 <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> raw_data <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> ids.txt <span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_pipeline</span> rnaseq <span class="dt">\</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_rnaseq_strandedness</span> forward</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="configuring-the-nf-corernaseq-workflow-1" class="level3">
<h3 class="anchored" data-anchor-id="configuring-the-nf-corernaseq-workflow-1">Configuring the nf-core/rnaseq workflow</h3>
<section id="reference-data-1" class="level4">
<h4 class="anchored" data-anchor-id="reference-data-1">Reference data</h4>
<p>According to the sample metadata in <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3933549">NCBI GEO</a> the authors used the same reference data as Xia et al, <a href="https://www.gencodegenes.org/mouse/release_M17.html">Gencode release M17</a>. We can therefore reuse the reference files and indices we generated in the first analysis (see above).</p>
</section>
<section id="unique-molecular-identifiers" class="level4">
<h4 class="anchored" data-anchor-id="unique-molecular-identifiers">Unique molecular identifiers</h4>
<p>To process QuantSeq data with UMIs, the following parameters need to be provided.</p>
<ul>
<li><code>--with_umi</code></li>
<li><code>--umitools_extract_method "regex"</code>: The type of pattern to detect, either <code>string</code> (default) or <code>regex</code></li>
<li><code>--umitools_bc_pattern "^(?P&lt;umi_1&gt;.{6})(?P&lt;discard_1&gt;.{4}).*"</code>: A regular expression that instructs <code>UMI-tools</code> to extract the first 6 bases (the UMI) from the 5’ end of the read <a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> and to discard the following 4 bases. (Lexogen adds an invariant <code>TATA</code> sequence motif to each UMI.)</li>
<li><code>--umitools_grouping_method "unique"</code>: The method used to group similar UMIs, e.g.&nbsp;to correct sequencing errors. The default method is <code>directional</code>, but it is computationally expensive. Here, we use the simpler <code>unique</code> method, suppressing grouping / error correction entirely.</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>These arguments shown above are suitable for UMIs offered by Lexogen for the QuantSeq protocol. If you use custom UMIs, please consult the <a href="https://nf-co.re/rnaseq/3.10.1/parameters#umi-options">nf-core/rnaseq documentation for full details and additional options</a> for details on additional parameters you can set.</p>
</div>
</div>
</section>
</section>
<section id="starting-the-workflow-1" class="level3">
<h3 class="anchored" data-anchor-id="starting-the-workflow-1">Starting the workflow</h3>
<p>Next, we add the UMI-related parameters to the JSON file we used above, so we can pass them to <code>nextflow run</code> via the <code>-params-file</code> argument. Because Nugent et al used the same reference (Gencode M17), we can reuse the same reference we generated for Xia et al’s dataset. (Here, we assume the precomputed references are in the <code>genome</code> folder to our current working directory. If you chose to store them elsewhere, please provide absolute paths in the JSON file instead.)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode json code-with-copy"><code class="sourceCode json"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"save_reference"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"fasta"</span><span class="fu">:</span> <span class="st">"genome/GRCm38.primary_assembly.genome.fa"</span><span class="fu">,</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gtf"</span><span class="fu">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.gtf"</span><span class="fu">,</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gene_bed"</span><span class="fu">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.bed"</span><span class="fu">,</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"transcript_fasta"</span><span class="fu">:</span> <span class="st">"genome/genome.transcripts.fa"</span><span class="fu">,</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"star_index"</span><span class="fu">:</span> <span class="st">"genome/index/star"</span><span class="fu">,</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"salmon_index"</span><span class="fu">:</span> <span class="st">"genome/index/salmon"</span><span class="fu">,</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"rsem_index"</span><span class="fu">:</span> <span class="st">"genome/rsem"</span><span class="fu">,</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_star_align_args"</span><span class="fu">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="fu">,</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"with_umi"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"umitools_extract_method"</span><span class="fu">:</span> <span class="st">"regex"</span><span class="fu">,</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"umitools_grouping_method"</span><span class="fu">:</span> <span class="st">"unique"</span><span class="fu">,</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"umitools_bc_pattern"</span><span class="fu">:</span> <span class="st">"^(?P&lt;umi_1&gt;.{6})(?P&lt;discard_1&gt;.{4}).*"</span><span class="fu">,</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"extra_salmon_quant_args"</span><span class="fu">:</span> <span class="st">"--noLengthCorrection"</span><span class="fu">,</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"skip_stringtie"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"gencode"</span><span class="fu">:</span> <span class="kw">true</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="fu">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We store the JSON string in a file called <code>nf_params.json</code>, and then pass it to the <code>nextflow run</code> command:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>      nf-core/rnaseq <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">-r</span> 3.10.1 <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">-params-file</span> nf_params.json <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> raw_data/samplesheet/samplesheet.csv <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> SRP213880</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Upon successful completion, the workflow’s output is available in the <code>SRP213880</code> directory.</p>
<p>Next, we will take a closer look at the available results in <a href="../nextflow-core-quantseq-2-output/index.html">the second post in this series</a>].</p>


<!-- -->

</section>
</section>


<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>The <code>strandedness</code> column in the sample sheet CSV file must contain one of <code>forward</code>, <code>reverse</code>, <code>unstranded</code> or <code>auto</code>. The latter will prompt the <code>nf-core/rnaseq</code> workflow to subsample reads and map them to the transcriptome with <a href="https://combine-lab.github.io/salmon/">salmon</a> to determine the likely orientation of the reads. Because this adds additional steps to the workflow execution, it is recommended that you specify the <code>strandedness</code> of your library explicitly whenever possible.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The <code>-resume</code> argument allows you to resume previous executions of the same workflow. Nextflow will reuse any existing outputs and generate only those that are missing.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>Nextflow can interact with different storage backends. For example, you could provide URLs of remote files (e.g.&nbsp;with the <code>ftp://</code> prefix) or in cloud storage (e.g.&nbsp;with the <code>s3://</code> or <code>gs://</code> prefix) and they will automatically be downloaded and <em>staged</em> when you start the run.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>Typically, QuantSeq libraries are single-end (not paired-end) and the UMI is therefore found in the R1 read.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb11" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "QuantSeq RNAseq analysis (1): configuring the nf-core/rnaseq workflow"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-01-16"</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [nextflow, NGS]</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="false"}</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>This is the first of four posts documenting my progress toward</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>processing and analyzing </span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">QuantSeq FWD 3' tag RNAseq</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span> </span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>data </span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>with the </span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/rnaseq</span><span class="co">](https://nf-co.re/rnaseq)</span></span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>workflow. </span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Configuring the nf-core/rnaseq workflow**</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="co">[</span><span class="ot">Exploring the workflow outputs</span><span class="co">](../nextflow-core-quantseq-2-output/index.html)</span></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="co">[</span><span class="ot">Validating the workflow by reproducing results published by Xia et al (no UMIs)</span><span class="co">](../nextflow-core-quantseq-3-xia/index.html)</span></span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span><span class="co">[</span><span class="ot">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</span><span class="co">](../nextflow-core-quantseq-4-nugent/index.html)</span></span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>Many thanks to <span class="co">[</span><span class="ot">Harshil Patel</span><span class="co">](https://github.com/drpatelh)</span>, </span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">António Miguel de Jesus Domingues</span><span class="co">](https://github.com/adomingues/)</span></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>and </span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Matthias Zepper</span><span class="co">](https://github.com/matthiasZepper/)</span></span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>for their generous guidance &amp; input via <span class="co">[</span><span class="ot">nf-core slack</span><span class="co">](nfcore.slack.com)</span>.</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>(Any mistakes are mine.)</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>This tutorial documents how to configure &amp; execute the</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">nf-core/rnaseq workflow</span><span class="co">](https://nf-co.re/rnaseq)</span></span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a>  for the processing of raw QuantSeq 3' FWD RNA-seq data.</span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>It highlights <span class="co">[</span><span class="ot">custom nf-core/rnaseq parameters settings</span><span class="co">](#custom-arguments)</span></span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a>  and applies them to two published datasets, one with and one without UMIs.</span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-53"><a href="#cb11-53" aria-hidden="true" tabindex="-1"></a><span class="fu">## Installing Nextflow</span></span>
<span id="cb11-54"><a href="#cb11-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-55"><a href="#cb11-55" aria-hidden="true" tabindex="-1"></a>To run nextflow, we create a new </span>
<span id="cb11-56"><a href="#cb11-56" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">conda</span><span class="co">](https://docs.conda.io/en/latest/)</span> environment. (See </span>
<span id="cb11-57"><a href="#cb11-57" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">conda installation instructions</span><span class="co">]</span>(https://docs.conda.io/en/latest/miniconda.html#latest-miniconda-installer-links</span>
<span id="cb11-58"><a href="#cb11-58" aria-hidden="true" tabindex="-1"></a>) if you are not yet familiar with the conda package manager.) Then we install </span>
<span id="cb11-59"><a href="#cb11-59" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nextflow</span><span class="co">](https://www.nextflow.io/)</span> and its dependencies from the </span>
<span id="cb11-60"><a href="#cb11-60" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">bioconda</span><span class="co">](https://anaconda.org/bioconda/nextflow)</span> repository.</span>
<span id="cb11-61"><a href="#cb11-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-62"><a href="#cb11-62" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-63"><a href="#cb11-63" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="ex">create</span> <span class="at">-n</span> nf</span>
<span id="cb11-64"><a href="#cb11-64" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="fu">install</span> <span class="at">-c</span> bioconda nextflow</span>
<span id="cb11-65"><a href="#cb11-65" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-66"><a href="#cb11-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-67"><a href="#cb11-67" aria-hidden="true" tabindex="-1"></a>Once the installation is complete, we activate our new conda environment and</span>
<span id="cb11-68"><a href="#cb11-68" aria-hidden="true" tabindex="-1"></a>verify that Nextflow is installed:</span>
<span id="cb11-69"><a href="#cb11-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-70"><a href="#cb11-70" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-71"><a href="#cb11-71" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> conda <span class="ex">activate</span> nf</span>
<span id="cb11-72"><a href="#cb11-72" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">-v</span></span>
<span id="cb11-73"><a href="#cb11-73" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> version 22.10.4.5836</span>
<span id="cb11-74"><a href="#cb11-74" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-75"><a href="#cb11-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-76"><a href="#cb11-76" aria-hidden="true" tabindex="-1"></a><span class="fu">### Nextflow configs</span></span>
<span id="cb11-77"><a href="#cb11-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-78"><a href="#cb11-78" aria-hidden="true" tabindex="-1"></a>Nextflow separates the definition of a workflow from the platform it is executed</span>
<span id="cb11-79"><a href="#cb11-79" aria-hidden="true" tabindex="-1"></a>on. For example, you can run the same workflow on your local computer, a</span>
<span id="cb11-80"><a href="#cb11-80" aria-hidden="true" tabindex="-1"></a>high-performance cluster or using a cloud provider like Amazon Web Services</span>
<span id="cb11-81"><a href="#cb11-81" aria-hidden="true" tabindex="-1"></a>(AWS) or Google Cloud Services (GCS).</span>
<span id="cb11-82"><a href="#cb11-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-83"><a href="#cb11-83" aria-hidden="true" tabindex="-1"></a>Each nf-core workflow defines the required software sources, which can either be installed (e.g. via conda) or be made available through a container engine such as </span>
<span id="cb11-84"><a href="#cb11-84" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">docker</span><span class="co">](https://www.docker.com/)</span>, </span>
<span id="cb11-85"><a href="#cb11-85" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">singularity</span><span class="co">]</span>(https://en.wikipedia.org/wiki/Singularity_(software)), etc. </span>
<span id="cb11-86"><a href="#cb11-86" aria-hidden="true" tabindex="-1"></a>For details, please consult the </span>
<span id="cb11-87"><a href="#cb11-87" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Nextflow documentation</span><span class="co">](https://www.nextflow.io/docs/latest/config.html)</span>.</span>
<span id="cb11-88"><a href="#cb11-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-89"><a href="#cb11-89" aria-hidden="true" tabindex="-1"></a>The desired configuration is specified in one or more configuration file(s), and</span>
<span id="cb11-90"><a href="#cb11-90" aria-hidden="true" tabindex="-1"></a>passed to the <span class="in">`nextflow`</span> command via its <span class="in">`-profile`</span> argument. Here, we execute</span>
<span id="cb11-91"><a href="#cb11-91" aria-hidden="true" tabindex="-1"></a>the workflows on the local system, using docker containers by specifying the</span>
<span id="cb11-92"><a href="#cb11-92" aria-hidden="true" tabindex="-1"></a>built-in <span class="in">`-profile docker`</span> configuration.</span>
<span id="cb11-93"><a href="#cb11-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-94"><a href="#cb11-94" aria-hidden="true" tabindex="-1"></a><span class="fu">## Analysis of published studies</span></span>
<span id="cb11-95"><a href="#cb11-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-96"><a href="#cb11-96" aria-hidden="true" tabindex="-1"></a>To validate the nf-core/rnaseq workflow for QuantSeq 3' tag RNA-seq data, we</span>
<span id="cb11-97"><a href="#cb11-97" aria-hidden="true" tabindex="-1"></a>reanalyze data from two published studies, one with and one without the use of</span>
<span id="cb11-98"><a href="#cb11-98" aria-hidden="true" tabindex="-1"></a>unique molecular identifiers (UMIs).</span>
<span id="cb11-99"><a href="#cb11-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-100"><a href="#cb11-100" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>"Fibrillar Aβ causes profound microglial metabolic perturbations in a novel </span>
<span id="cb11-101"><a href="#cb11-101" aria-hidden="true" tabindex="-1"></a>  APP knock-in mouse model" by </span>
<span id="cb11-102"><a href="#cb11-102" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">Xia et al, 2021</span><span class="co">](https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1)</span></span>
<span id="cb11-103"><a href="#cb11-103" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">QuantSeq 3’ mRNA-Seq Library Prep Kit FWD FWD</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span></span>
<span id="cb11-104"><a href="#cb11-104" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>This data does __not__ include UMIs.</span>
<span id="cb11-105"><a href="#cb11-105" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>Raw data are available via </span>
<span id="cb11-106"><a href="#cb11-106" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span><span class="co">[</span><span class="ot">GEO accession GSE158152</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158152)</span></span>
<span id="cb11-107"><a href="#cb11-107" aria-hidden="true" tabindex="-1"></a><span class="ss">     -  </span><span class="co">[</span><span class="ot">SRA accession SRP213880</span><span class="co">](https://www.ncbi.nlm.nih.gov/sra?term=SRP282921)</span></span>
<span id="cb11-108"><a href="#cb11-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-109"><a href="#cb11-109" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>"TREM2 regulates microglial lipid metabolism during aging in mice" by </span>
<span id="cb11-110"><a href="#cb11-110" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">Nugent et al, 2020</span><span class="co">]</span>(https://www.cell.com/neuron/fulltext/S0896-6273(19)31049-9).</span>
<span id="cb11-111"><a href="#cb11-111" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">QuantSeq 3’ mRNA-Seq Library Prep Kit FWD FWD</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span></span>
<span id="cb11-112"><a href="#cb11-112" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>This data includes <span class="co">[</span><span class="ot">UMIs</span><span class="co">](https://www.lexogen.com/store/umi-second-strand-synthesis-module-for-quantseq-fwd/)</span>,     which and can be used to filter alignments originating from PCR duplicates.</span>
<span id="cb11-113"><a href="#cb11-113" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Raw data are available via </span>
<span id="cb11-114"><a href="#cb11-114" aria-hidden="true" tabindex="-1"></a><span class="ss">      - </span><span class="co">[</span><span class="ot">GEO accession GSE134031</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134031)</span></span>
<span id="cb11-115"><a href="#cb11-115" aria-hidden="true" tabindex="-1"></a><span class="ss">      -  </span><span class="co">[</span><span class="ot">SRA accession SRP213880</span><span class="co">](https://www.ncbi.nlm.nih.gov/sra?term=SRP213880)</span></span>
<span id="cb11-116"><a href="#cb11-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-117"><a href="#cb11-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-118"><a href="#cb11-118" aria-hidden="true" tabindex="-1"></a><span class="fu">## Quantseq FWD without UMIs: Xia et al dataset</span></span>
<span id="cb11-119"><a href="#cb11-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-120"><a href="#cb11-120" aria-hidden="true" tabindex="-1"></a>Here, we will reanalyze RNA-seq data published in the preprint "Fibrillar Aβ</span>
<span id="cb11-121"><a href="#cb11-121" aria-hidden="true" tabindex="-1"></a>causes profound microglial metabolic perturbations in a novel APP knock-in mouse</span>
<span id="cb11-122"><a href="#cb11-122" aria-hidden="true" tabindex="-1"></a>model" by</span>
<span id="cb11-123"><a href="#cb11-123" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Xia et al, 2021</span><span class="co">](https://www.biorxiv.org/content/10.1101/2021.01.19.426731v1)</span>. </span>
<span id="cb11-124"><a href="#cb11-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-125"><a href="#cb11-125" aria-hidden="true" tabindex="-1"></a><span class="fu">### Experimental design</span></span>
<span id="cb11-126"><a href="#cb11-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-127"><a href="#cb11-127" aria-hidden="true" tabindex="-1"></a>This study examines FACS-isolated microglia from 18 mice from three genetic</span>
<span id="cb11-128"><a href="#cb11-128" aria-hidden="true" tabindex="-1"></a>backgrounds (N = 6 WT, N =6 heterozygous App-SAA and N = 6 homozygous App-SAA</span>
<span id="cb11-129"><a href="#cb11-129" aria-hidden="true" tabindex="-1"></a>animals). Microglia were isolated from the brain (cortex &amp; hippocampus) of each</span>
<span id="cb11-130"><a href="#cb11-130" aria-hidden="true" tabindex="-1"></a>animal using FACS and gene expression changes were analyzed using 3-tag RNA-seq.</span>
<span id="cb11-131"><a href="#cb11-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-132"><a href="#cb11-132" aria-hidden="true" tabindex="-1"></a>From each mouse, two (technical) replicate pools of microglia were collected and</span>
<span id="cb11-133"><a href="#cb11-133" aria-hidden="true" tabindex="-1"></a>processed into separate libraries with [Lexogen's QuantSeq</span>
<span id="cb11-134"><a href="#cb11-134" aria-hidden="true" tabindex="-1"></a>FWD](https://www.lexogen.com/quantseq-3mrna-sequencing/). This dataset does</span>
<span id="cb11-135"><a href="#cb11-135" aria-hidden="true" tabindex="-1"></a>__not__ include unique molecular identifiers (UMIs).</span>
<span id="cb11-136"><a href="#cb11-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-137"><a href="#cb11-137" aria-hidden="true" tabindex="-1"></a><span class="fu">### Raw data retrieval with nf-core/fetchngs</span></span>
<span id="cb11-138"><a href="#cb11-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-139"><a href="#cb11-139" aria-hidden="true" tabindex="-1"></a>We start the analysis by downloading the raw data (36 FASTQ files) and the</span>
<span id="cb11-140"><a href="#cb11-140" aria-hidden="true" tabindex="-1"></a>sample metadata with the </span>
<span id="cb11-141"><a href="#cb11-141" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/fetchngs</span><span class="co">](https://nf-co.re/fetchngs)</span> workflow. </span>
<span id="cb11-142"><a href="#cb11-142" aria-hidden="true" tabindex="-1"></a>For datasets stored in SRA, we only need to paste the <span class="in">`SRA Study identifier`</span></span>
<span id="cb11-143"><a href="#cb11-143" aria-hidden="true" tabindex="-1"></a>(SRP282921) into a text file for us as the <span class="in">`input`</span> for the fetchngs workflow. </span>
<span id="cb11-144"><a href="#cb11-144" aria-hidden="true" tabindex="-1"></a>The following command will download all of the FASTQ files associated with this </span>
<span id="cb11-145"><a href="#cb11-145" aria-hidden="true" tabindex="-1"></a>study into the <span class="in">`raw_data`</span> directory:</span>
<span id="cb11-146"><a href="#cb11-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-147"><a href="#cb11-147" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-148"><a href="#cb11-148" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> mkdir <span class="ex">SRP282921</span></span>
<span id="cb11-149"><a href="#cb11-149" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> cd <span class="ex">SRP282921</span></span>
<span id="cb11-150"><a href="#cb11-150" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">SRP282921</span> <span class="op">&gt;</span> ids.txt</span>
<span id="cb11-151"><a href="#cb11-151" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> nf-core/fetchngs <span class="dt">\</span></span>
<span id="cb11-152"><a href="#cb11-152" aria-hidden="true" tabindex="-1"></a>      <span class="at">-revision</span> 1.9 <span class="dt">\</span></span>
<span id="cb11-153"><a href="#cb11-153" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb11-154"><a href="#cb11-154" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> raw_data <span class="dt">\</span></span>
<span id="cb11-155"><a href="#cb11-155" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> ids.txt <span class="dt">\</span></span>
<span id="cb11-156"><a href="#cb11-156" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_pipeline</span> rnaseq <span class="dt">\</span></span>
<span id="cb11-157"><a href="#cb11-157" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_rnaseq_strandedness</span> forward</span>
<span id="cb11-158"><a href="#cb11-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-159"><a href="#cb11-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-160"><a href="#cb11-160" aria-hidden="true" tabindex="-1"></a>By specifying the <span class="in">`--nf_core_pipeline`</span> argument, the workflow creates a sample</span>
<span id="cb11-161"><a href="#cb11-161" aria-hidden="true" tabindex="-1"></a>sheet that can be used as input for the <span class="in">`nf-core/rnaseq`</span> pipeline. Because the</span>
<span id="cb11-162"><a href="#cb11-162" aria-hidden="true" tabindex="-1"></a>data was generated with the </span>
<span id="cb11-163"><a href="#cb11-163" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">QuantSeq FWD</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span> kit, we also</span>
<span id="cb11-164"><a href="#cb11-164" aria-hidden="true" tabindex="-1"></a>know that the reads are from the <span class="in">`forward`</span> strand and add this information to</span>
<span id="cb11-165"><a href="#cb11-165" aria-hidden="true" tabindex="-1"></a>the sample sheet via the <span class="in">`--nf_core_rnaseq_strandedness`</span> argument. <span class="ot">[^1]</span></span>
<span id="cb11-166"><a href="#cb11-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-167"><a href="#cb11-167" aria-hidden="true" tabindex="-1"></a>Once the workflow has completed, the <span class="in">`raw_data`</span> output directory contains the</span>
<span id="cb11-168"><a href="#cb11-168" aria-hidden="true" tabindex="-1"></a>following structure:</span>
<span id="cb11-169"><a href="#cb11-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-170"><a href="#cb11-170" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-171"><a href="#cb11-171" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> tree <span class="ex">raw_data</span></span>
<span id="cb11-172"><a href="#cb11-172" aria-hidden="true" tabindex="-1"></a><span class="ex">raw_data/</span></span>
<span id="cb11-173"><a href="#cb11-173" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> custom</span>
<span id="cb11-174"><a href="#cb11-174" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── user-settings.mkfg</span>
<span id="cb11-175"><a href="#cb11-175" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> fastq</span>
<span id="cb11-176"><a href="#cb11-176" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── md5</span>
<span id="cb11-177"><a href="#cb11-177" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── SRX9142647_SRR12661938.fastq.gz.md5</span>
<span id="cb11-178"><a href="#cb11-178" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── SRX9142648_SRR12661924.fastq.gz.md5</span>
<span id="cb11-179"><a href="#cb11-179" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── 34 additional .fastq.gz.md5 files <span class="er">(</span><span class="ex">not</span> shown<span class="kw">)</span></span>
<span id="cb11-180"><a href="#cb11-180" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SRX9142647_SRR12661938.fastq.gz</span>
<span id="cb11-181"><a href="#cb11-181" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SRX9142648_SRR12661924.fastq.gz</span>
<span id="cb11-182"><a href="#cb11-182" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── 34 additional .fastq.gz files <span class="er">(</span><span class="ex">not</span> shown<span class="kw">)</span></span>
<span id="cb11-183"><a href="#cb11-183" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> metadata</span>
<span id="cb11-184"><a href="#cb11-184" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── SRP282921.runinfo_ftp.tsv</span>
<span id="cb11-185"><a href="#cb11-185" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> pipeline_info</span>
<span id="cb11-186"><a href="#cb11-186" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_report_2023-01-12_01-49-02.html</span>
<span id="cb11-187"><a href="#cb11-187" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_timeline_2023-01-12_01-49-02.html</span>
<span id="cb11-188"><a href="#cb11-188" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── execution_trace_2023-01-12_01-49-02.txt</span>
<span id="cb11-189"><a href="#cb11-189" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── pipeline_dag_2023-01-12_01-49-02.html</span>
<span id="cb11-190"><a href="#cb11-190" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── software_versions.yml</span>
<span id="cb11-191"><a href="#cb11-191" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> samplesheet</span>
<span id="cb11-192"><a href="#cb11-192" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> id_mappings.csv</span>
<span id="cb11-193"><a href="#cb11-193" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> multiqc_config.yml</span>
<span id="cb11-194"><a href="#cb11-194" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> nf_params.json</span>
<span id="cb11-195"><a href="#cb11-195" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> run.sh</span>
<span id="cb11-196"><a href="#cb11-196" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> samplesheet.csv</span>
<span id="cb11-197"><a href="#cb11-197" aria-hidden="true" tabindex="-1"></a><span class="ex">6</span> directories, 90 files</span>
<span id="cb11-198"><a href="#cb11-198" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-199"><a href="#cb11-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-200"><a href="#cb11-200" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The 36 FASTQ files are in the <span class="in">`fastq`</span> subfolder, each accompanied by a MD5</span>
<span id="cb11-201"><a href="#cb11-201" aria-hidden="true" tabindex="-1"></a>  checksum file in the <span class="in">`md5`</span> directory.</span>
<span id="cb11-202"><a href="#cb11-202" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>In the <span class="in">`samplesheet`</span> subdirectory, we find the <span class="in">`samplesheet.csv`</span> file that we</span>
<span id="cb11-203"><a href="#cb11-203" aria-hidden="true" tabindex="-1"></a>  can pass to the <span class="co">[</span><span class="ot">nf-core/rnaseq workflow</span><span class="co">](https://nf-co.re/rnaseq)</span>.</span>
<span id="cb11-204"><a href="#cb11-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-205"><a href="#cb11-205" aria-hidden="true" tabindex="-1"></a>For more details about the output oft he workflow, please check </span>
<span id="cb11-206"><a href="#cb11-206" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">the next post in this series</span><span class="co">](../nextflow-core-quantseq-2-output/index.html)</span>]</span>
<span id="cb11-207"><a href="#cb11-207" aria-hidden="true" tabindex="-1"></a>and the</span>
<span id="cb11-208"><a href="#cb11-208" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/rnaseq output docs</span><span class="co">](https://nf-co.re/rnaseq/3.10.1/output)</span>.</span>
<span id="cb11-209"><a href="#cb11-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-210"><a href="#cb11-210" aria-hidden="true" tabindex="-1"></a><span class="fu">### Configuring the nf-core/rnaseq workflow</span></span>
<span id="cb11-211"><a href="#cb11-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-212"><a href="#cb11-212" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Reference data</span></span>
<span id="cb11-213"><a href="#cb11-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-214"><a href="#cb11-214" aria-hidden="true" tabindex="-1"></a>According to the sample metadata in </span>
<span id="cb11-215"><a href="#cb11-215" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">NCBI GEO</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4793335)</span> </span>
<span id="cb11-216"><a href="#cb11-216" aria-hidden="true" tabindex="-1"></a>the authors mapped the data to the mouse GRCm38 genome version, using Gencode</span>
<span id="cb11-217"><a href="#cb11-217" aria-hidden="true" tabindex="-1"></a>annotations from </span>
<span id="cb11-218"><a href="#cb11-218" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">release M17</span><span class="co">](https://www.gencodegenes.org/mouse/release_M17.html)</span>. </span>
<span id="cb11-219"><a href="#cb11-219" aria-hidden="true" tabindex="-1"></a>To later compare our results with those obtained in the original publication, </span>
<span id="cb11-220"><a href="#cb11-220" aria-hidden="true" tabindex="-1"></a>we use the same Gencode version (even though it is not the most recent).</span>
<span id="cb11-221"><a href="#cb11-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-222"><a href="#cb11-222" aria-hidden="true" tabindex="-1"></a>The <span class="in">`nf-core/rnaseq`</span> workflow can automatically generate all necessary indices</span>
<span id="cb11-223"><a href="#cb11-223" aria-hidden="true" tabindex="-1"></a>when provided with </span>
<span id="cb11-224"><a href="#cb11-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-225"><a href="#cb11-225" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>A FASTA file with the genomic sequences, e.g. Gencode's </span>
<span id="cb11-226"><a href="#cb11-226" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">Primary genome assembly</span><span class="co">](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/GRCm38.primary_assembly.genome.fa.gz)</span> and</span>
<span id="cb11-227"><a href="#cb11-227" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>A GFP file with matching gene annotations, e.g. </span>
<span id="cb11-228"><a href="#cb11-228" aria-hidden="true" tabindex="-1"></a>  Gencode </span>
<span id="cb11-229"><a href="#cb11-229" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">Comprehensive gene annotations</span><span class="co">](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/gencode.vM17.primary_assembly.annotation.gtf.gz)</span></span>
<span id="cb11-230"><a href="#cb11-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-231"><a href="#cb11-231" aria-hidden="true" tabindex="-1"></a>Because downloading and indexing the genome is time consuming, we will save them for later use.</span>
<span id="cb11-232"><a href="#cb11-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-233"><a href="#cb11-233" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Custom arguments</span></span>
<span id="cb11-234"><a href="#cb11-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-235"><a href="#cb11-235" aria-hidden="true" tabindex="-1"></a>The <span class="in">`nf-core/rnaseq`</span> pipeline has robust default parameters that are suitable</span>
<span id="cb11-236"><a href="#cb11-236" aria-hidden="true" tabindex="-1"></a>for whole transcriptome RNA-seq data. But this dataset was generated with</span>
<span id="cb11-237"><a href="#cb11-237" aria-hidden="true" tabindex="-1"></a>Lexogen's QuantSeq FWD 3' mRNA library preparation kit, which specificially</span>
<span id="cb11-238"><a href="#cb11-238" aria-hidden="true" tabindex="-1"></a>targets the 3' end of polyadenylated transcripts. As a consequence, reads align</span>
<span id="cb11-239"><a href="#cb11-239" aria-hidden="true" tabindex="-1"></a>primarily to the 3' UTR and the final exon and coverage of the remainder of the</span>
<span id="cb11-240"><a href="#cb11-240" aria-hidden="true" tabindex="-1"></a>gene body is minimal.</span>
<span id="cb11-241"><a href="#cb11-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-242"><a href="#cb11-242" aria-hidden="true" tabindex="-1"></a>To account for these properties, we supply a number of custom arguments to the</span>
<span id="cb11-243"><a href="#cb11-243" aria-hidden="true" tabindex="-1"></a><span class="in">`nextflow run`</span> command, which augment the workflow for this specific data type.</span>
<span id="cb11-244"><a href="#cb11-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-245"><a href="#cb11-245" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Extra STAR arguments</span></span>
<span id="cb11-246"><a href="#cb11-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-247"><a href="#cb11-247" aria-hidden="true" tabindex="-1"></a>Lexogen provides an example analysis workflow </span>
<span id="cb11-248"><a href="#cb11-248" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">on their website</span><span class="co">](https://www.lexogen.com/quantseq-data-analysis/)</span>, which uses</span>
<span id="cb11-249"><a href="#cb11-249" aria-hidden="true" tabindex="-1"></a>the STAR aligner with several non-default parameters. Most of them parameters </span>
<span id="cb11-250"><a href="#cb11-250" aria-hidden="true" tabindex="-1"></a>correspond to the _ENCODE standard options_ listed in the </span>
<span id="cb11-251"><a href="#cb11-251" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">STAR manual</span><span class="co">](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)</span>:</span>
<span id="cb11-252"><a href="#cb11-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-253"><a href="#cb11-253" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--alignIntronMax 1000000`</span></span>
<span id="cb11-254"><a href="#cb11-254" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--alignIntronMin 20`</span></span>
<span id="cb11-255"><a href="#cb11-255" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--alignMatesGapMax 1000000`</span></span>
<span id="cb11-256"><a href="#cb11-256" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--alignSJoverhangMin 8`</span></span>
<span id="cb11-257"><a href="#cb11-257" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--alignSJDBoverhangMin 1`</span></span>
<span id="cb11-258"><a href="#cb11-258" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--outFilterMismatchNmax 999`</span></span>
<span id="cb11-259"><a href="#cb11-259" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--outFilterMultimapNmax 20`</span></span>
<span id="cb11-260"><a href="#cb11-260" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--outFilterType BySJout`</span></span>
<span id="cb11-261"><a href="#cb11-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-262"><a href="#cb11-262" aria-hidden="true" tabindex="-1"></a>::: {.callout-important collapse="true"}</span>
<span id="cb11-263"><a href="#cb11-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-264"><a href="#cb11-264" aria-hidden="true" tabindex="-1"></a>The </span>
<span id="cb11-265"><a href="#cb11-265" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">ENCODE project</span><span class="co">](https://www.genome.gov/Funded-Programs-Projects/ENCODE-Project-ENCyclopedia-Of-DNA-Elements)</span> </span>
<span id="cb11-266"><a href="#cb11-266" aria-hidden="true" tabindex="-1"></a>originally focussed on data from human and mouse. The parameters shown above are</span>
<span id="cb11-267"><a href="#cb11-267" aria-hidden="true" tabindex="-1"></a>suitable for these genomes or those with comparable characteristics. For</span>
<span id="cb11-268"><a href="#cb11-268" aria-hidden="true" tabindex="-1"></a>analysis of data from other organisms, e.g. those with shorter introns, you</span>
<span id="cb11-269"><a href="#cb11-269" aria-hidden="true" tabindex="-1"></a>might need to modify them.</span>
<span id="cb11-270"><a href="#cb11-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-271"><a href="#cb11-271" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb11-272"><a href="#cb11-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-273"><a href="#cb11-273" aria-hidden="true" tabindex="-1"></a>In addition, Lexogen also specified the </span>
<span id="cb11-274"><a href="#cb11-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-275"><a href="#cb11-275" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--outFilterMismatchNoverLmax 0.1`</span>, decreasing the tolerance for mismatches.</span>
<span id="cb11-276"><a href="#cb11-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-277"><a href="#cb11-277" aria-hidden="true" tabindex="-1"></a>Lexogen's example workflow uses <span class="co">[</span><span class="ot">bbduk</span><span class="co">](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/)</span> </span>
<span id="cb11-278"><a href="#cb11-278" aria-hidden="true" tabindex="-1"></a>to trim adapters, poly(A) tails and low quality bases from the 3' end of reads. </span>
<span id="cb11-279"><a href="#cb11-279" aria-hidden="true" tabindex="-1"></a>The nf-core/rnaseq workflow uses </span>
<span id="cb11-280"><a href="#cb11-280" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Trim Galore</span><span class="co">](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)</span> </span>
<span id="cb11-281"><a href="#cb11-281" aria-hidden="true" tabindex="-1"></a>to trim adapters, but does not remove poly(A) tails. Instead we use the STAR </span>
<span id="cb11-282"><a href="#cb11-282" aria-hidden="true" tabindex="-1"></a>ligner to clip poly(A) tails during the alignment stage:</span>
<span id="cb11-283"><a href="#cb11-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-284"><a href="#cb11-284" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--clip3pAdapterSeq AAAAAAAA`</span></span>
<span id="cb11-285"><a href="#cb11-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-286"><a href="#cb11-286" aria-hidden="true" tabindex="-1"></a>We will pass all of the arguments listed above to STAR via nf-core/rnaseq's</span>
<span id="cb11-287"><a href="#cb11-287" aria-hidden="true" tabindex="-1"></a><span class="in">`--extra_star_align_args`</span> argument (as a string, see example below).</span>
<span id="cb11-288"><a href="#cb11-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-289"><a href="#cb11-289" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Extra Salmon arguments</span></span>
<span id="cb11-290"><a href="#cb11-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-291"><a href="#cb11-291" aria-hidden="true" tabindex="-1"></a>The </span>
<span id="cb11-292"><a href="#cb11-292" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">salmon</span><span class="co">](https://www.nature.com/articles/nmeth.4197)</span></span>
<span id="cb11-293"><a href="#cb11-293" aria-hidden="true" tabindex="-1"></a>algorithm takes into account transcript length when quantifying gene expression.</span>
<span id="cb11-294"><a href="#cb11-294" aria-hidden="true" tabindex="-1"></a>Because QuantSeq (and other 3' tag sequencing methods) only capture the 3' most </span>
<span id="cb11-295"><a href="#cb11-295" aria-hidden="true" tabindex="-1"></a>part of each transcript, not its full length, this feature needs to be </span>
<span id="cb11-296"><a href="#cb11-296" aria-hidden="true" tabindex="-1"></a>deactivated:</span>
<span id="cb11-297"><a href="#cb11-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-298"><a href="#cb11-298" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--noLengthCorrection`</span></span>
<span id="cb11-299"><a href="#cb11-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-300"><a href="#cb11-300" aria-hidden="true" tabindex="-1"></a>The nf-core/rnaseq workflow features a special argument,</span>
<span id="cb11-301"><a href="#cb11-301" aria-hidden="true" tabindex="-1"></a><span class="in">`extra_salmon_quant_args`</span> to pass additional arguments to the <span class="in">`salmon`</span> tool (see</span>
<span id="cb11-302"><a href="#cb11-302" aria-hidden="true" tabindex="-1"></a>syntax below).</span>
<span id="cb11-303"><a href="#cb11-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-304"><a href="#cb11-304" aria-hidden="true" tabindex="-1"></a><span class="fu">### Starting the workflow</span></span>
<span id="cb11-305"><a href="#cb11-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-306"><a href="#cb11-306" aria-hidden="true" tabindex="-1"></a>We can pass the parameters we discussed above as individual arguments to the</span>
<span id="cb11-307"><a href="#cb11-307" aria-hidden="true" tabindex="-1"></a><span class="in">`nextflow run`</span> command (prefixed with <span class="in">`--`</span>). Alternatively, we can collect them</span>
<span id="cb11-308"><a href="#cb11-308" aria-hidden="true" tabindex="-1"></a>in a JSON file and specify its name via the <span class="in">`-params-file`</span> argument.</span>
<span id="cb11-309"><a href="#cb11-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-310"><a href="#cb11-310" aria-hidden="true" tabindex="-1"></a>For example, the following JSON string specifies the parameters for analyzing</span>
<span id="cb11-311"><a href="#cb11-311" aria-hidden="true" tabindex="-1"></a>sQuantSeq FWD data __without UMIs__:</span>
<span id="cb11-312"><a href="#cb11-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-313"><a href="#cb11-313" aria-hidden="true" tabindex="-1"></a><span class="in">```json</span></span>
<span id="cb11-314"><a href="#cb11-314" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb11-315"><a href="#cb11-315" aria-hidden="true" tabindex="-1"></a>    <span class="st">"save_reference"</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb11-316"><a href="#cb11-316" aria-hidden="true" tabindex="-1"></a>    <span class="st">"fasta"</span><span class="op">:</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/GRCm38.primary_assembly.genome.fa.gz"</span><span class="op">,</span></span>
<span id="cb11-317"><a href="#cb11-317" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gtf"</span><span class="op">:</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M17/gencode.vM17.primary_assembly.annotation.gtf.gz"</span><span class="op">,</span></span>
<span id="cb11-318"><a href="#cb11-318" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_star_align_args"</span><span class="op">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="op">,</span></span>
<span id="cb11-319"><a href="#cb11-319" aria-hidden="true" tabindex="-1"></a>    <span class="st">"with_umi"</span><span class="op">:</span> <span class="kw">false</span><span class="op">,</span></span>
<span id="cb11-320"><a href="#cb11-320" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_salmon_quant_args"</span><span class="op">:</span> <span class="st">"--noLengthCorrection"</span><span class="op">,</span></span>
<span id="cb11-321"><a href="#cb11-321" aria-hidden="true" tabindex="-1"></a>    <span class="st">"skip_stringtie"</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb11-322"><a href="#cb11-322" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gencode"</span><span class="op">:</span> <span class="kw">true</span></span>
<span id="cb11-323"><a href="#cb11-323" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-324"><a href="#cb11-324" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-325"><a href="#cb11-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-326"><a href="#cb11-326" aria-hidden="true" tabindex="-1"></a>We store the JSON string in a file called <span class="in">`nf_params.json`</span>, and then pass it to</span>
<span id="cb11-327"><a href="#cb11-327" aria-hidden="true" tabindex="-1"></a>the <span class="in">`nextflow run`</span> command <span class="ot">[^2]</span>:</span>
<span id="cb11-328"><a href="#cb11-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-329"><a href="#cb11-329" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-330"><a href="#cb11-330" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> <span class="dt">\</span></span>
<span id="cb11-331"><a href="#cb11-331" aria-hidden="true" tabindex="-1"></a>      nf-core/rnaseq <span class="dt">\</span></span>
<span id="cb11-332"><a href="#cb11-332" aria-hidden="true" tabindex="-1"></a>      <span class="at">-r</span> 3.10.1 <span class="dt">\</span></span>
<span id="cb11-333"><a href="#cb11-333" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb11-334"><a href="#cb11-334" aria-hidden="true" tabindex="-1"></a>      <span class="at">-params-file</span> nf_params.json <span class="dt">\</span></span>
<span id="cb11-335"><a href="#cb11-335" aria-hidden="true" tabindex="-1"></a>      <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb11-336"><a href="#cb11-336" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> raw_data/samplesheet/samplesheet.csv <span class="dt">\</span></span>
<span id="cb11-337"><a href="#cb11-337" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> SRP282921</span>
<span id="cb11-338"><a href="#cb11-338" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-339"><a href="#cb11-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-340"><a href="#cb11-340" aria-hidden="true" tabindex="-1"></a><span class="fu">### Examining the output files</span></span>
<span id="cb11-341"><a href="#cb11-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-342"><a href="#cb11-342" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reusing reference files and indices</span></span>
<span id="cb11-343"><a href="#cb11-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-344"><a href="#cb11-344" aria-hidden="true" tabindex="-1"></a>Retrieving the reference files and indexing them is time consuming. If you are</span>
<span id="cb11-345"><a href="#cb11-345" aria-hidden="true" tabindex="-1"></a>planning to map additional sample against the same genome / transcriptome in the</span>
<span id="cb11-346"><a href="#cb11-346" aria-hidden="true" tabindex="-1"></a>future, you might want to reuse them. Because we set the <span class="in">`save_reference`</span></span>
<span id="cb11-347"><a href="#cb11-347" aria-hidden="true" tabindex="-1"></a>argument to <span class="in">`true`</span> in our workflow, the output directory contains a <span class="in">`genome`</span></span>
<span id="cb11-348"><a href="#cb11-348" aria-hidden="true" tabindex="-1"></a>folder, featuring reference files (e.g. genome and transcriptome sequences in</span>
<span id="cb11-349"><a href="#cb11-349" aria-hidden="true" tabindex="-1"></a>FASTQ formats) and indices (e.g. for STAR, salmon and rsem).</span>
<span id="cb11-350"><a href="#cb11-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-351"><a href="#cb11-351" aria-hidden="true" tabindex="-1"></a>Let's move the genome folder to our current working directory (or any other</span>
<span id="cb11-352"><a href="#cb11-352" aria-hidden="true" tabindex="-1"></a>suitable location). <span class="ot">[^3]</span> For future runs, we can include paths to the files and</span>
<span id="cb11-353"><a href="#cb11-353" aria-hidden="true" tabindex="-1"></a>indices in our <span class="in">`nf_params.json`</span> file, e.g.</span>
<span id="cb11-354"><a href="#cb11-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-355"><a href="#cb11-355" aria-hidden="true" tabindex="-1"></a><span class="in">```json</span></span>
<span id="cb11-356"><a href="#cb11-356" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb11-357"><a href="#cb11-357" aria-hidden="true" tabindex="-1"></a>    <span class="st">"save_reference"</span><span class="op">:</span> <span class="kw">false</span><span class="op">,</span></span>
<span id="cb11-358"><a href="#cb11-358" aria-hidden="true" tabindex="-1"></a>    <span class="st">"fasta"</span><span class="op">:</span> <span class="st">"genome/GRCm38.primary_assembly.genome.fa"</span><span class="op">,</span></span>
<span id="cb11-359"><a href="#cb11-359" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gtf"</span><span class="op">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.gtf"</span><span class="op">,</span></span>
<span id="cb11-360"><a href="#cb11-360" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gene_bed"</span><span class="op">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.bed"</span><span class="op">,</span></span>
<span id="cb11-361"><a href="#cb11-361" aria-hidden="true" tabindex="-1"></a>    <span class="st">"transcript_fasta"</span><span class="op">:</span> <span class="st">"genome/genome.transcripts.fa"</span><span class="op">,</span></span>
<span id="cb11-362"><a href="#cb11-362" aria-hidden="true" tabindex="-1"></a>    <span class="st">"star_index"</span><span class="op">:</span> <span class="st">"genome/index/star"</span><span class="op">,</span></span>
<span id="cb11-363"><a href="#cb11-363" aria-hidden="true" tabindex="-1"></a>    <span class="st">"salmon_index"</span><span class="op">:</span> <span class="st">"genome/index/salmon"</span><span class="op">,</span></span>
<span id="cb11-364"><a href="#cb11-364" aria-hidden="true" tabindex="-1"></a>    <span class="st">"rsem_index"</span><span class="op">:</span> <span class="st">"genome/rsem"</span><span class="op">,</span></span>
<span id="cb11-365"><a href="#cb11-365" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_star_align_args"</span><span class="op">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="op">,</span></span>
<span id="cb11-366"><a href="#cb11-366" aria-hidden="true" tabindex="-1"></a>    <span class="st">"with_umi"</span><span class="op">:</span> <span class="kw">false</span><span class="op">,</span></span>
<span id="cb11-367"><a href="#cb11-367" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_salmon_quant_args"</span><span class="op">:</span> <span class="st">"--noLengthCorrection"</span><span class="op">,</span></span>
<span id="cb11-368"><a href="#cb11-368" aria-hidden="true" tabindex="-1"></a>    <span class="st">"skip_stringtie"</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb11-369"><a href="#cb11-369" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gencode"</span><span class="op">:</span> <span class="kw">true</span></span>
<span id="cb11-370"><a href="#cb11-370" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-371"><a href="#cb11-371" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-372"><a href="#cb11-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-373"><a href="#cb11-373" aria-hidden="true" tabindex="-1"></a>Our next run will execute much faster, because these reference files and indices</span>
<span id="cb11-374"><a href="#cb11-374" aria-hidden="true" tabindex="-1"></a>are already available and don't need to be created from scratch.</span>
<span id="cb11-375"><a href="#cb11-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-376"><a href="#cb11-376" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb11-377"><a href="#cb11-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-378"><a href="#cb11-378" aria-hidden="true" tabindex="-1"></a>If you chose a location other than <span class="in">`genome`</span> to store your reference data, please</span>
<span id="cb11-379"><a href="#cb11-379" aria-hidden="true" tabindex="-1"></a>specify the (absolute) path in the JSON file instead.</span>
<span id="cb11-380"><a href="#cb11-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-381"><a href="#cb11-381" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb11-382"><a href="#cb11-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-383"><a href="#cb11-383" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>The <span class="in">`strandedness`</span> column in the sample sheet CSV file must contain one of</span>
<span id="cb11-384"><a href="#cb11-384" aria-hidden="true" tabindex="-1"></a><span class="in">`forward`</span>, <span class="in">`reverse`</span>, <span class="in">`unstranded`</span> or <span class="in">`auto`</span>. The latter will prompt the</span>
<span id="cb11-385"><a href="#cb11-385" aria-hidden="true" tabindex="-1"></a><span class="in">`nf-core/rnaseq`</span> workflow to subsample reads and map them to the transcriptome</span>
<span id="cb11-386"><a href="#cb11-386" aria-hidden="true" tabindex="-1"></a>with</span>
<span id="cb11-387"><a href="#cb11-387" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">salmon</span><span class="co">](https://combine-lab.github.io/salmon/)</span> </span>
<span id="cb11-388"><a href="#cb11-388" aria-hidden="true" tabindex="-1"></a>to determine the likely orientation of the reads. Because this adds additional</span>
<span id="cb11-389"><a href="#cb11-389" aria-hidden="true" tabindex="-1"></a>steps to the workflow execution, it is recommended that you specify the</span>
<span id="cb11-390"><a href="#cb11-390" aria-hidden="true" tabindex="-1"></a><span class="in">`strandedness`</span> of your library explicitly whenever possible.</span>
<span id="cb11-391"><a href="#cb11-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-392"><a href="#cb11-392" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span>The <span class="in">`-resume`</span> argument allows you to resume previous executions of the</span>
<span id="cb11-393"><a href="#cb11-393" aria-hidden="true" tabindex="-1"></a>same workflow. Nextflow will reuse any existing outputs and generate only those</span>
<span id="cb11-394"><a href="#cb11-394" aria-hidden="true" tabindex="-1"></a>that are missing.</span>
<span id="cb11-395"><a href="#cb11-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-396"><a href="#cb11-396" aria-hidden="true" tabindex="-1"></a><span class="ot">[^3]: </span>Nextflow can interact with different storage backends. For example, </span>
<span id="cb11-397"><a href="#cb11-397" aria-hidden="true" tabindex="-1"></a>you could provide URLs of remote files (e.g. with the <span class="in">`ftp://`</span> prefix) or in </span>
<span id="cb11-398"><a href="#cb11-398" aria-hidden="true" tabindex="-1"></a>cloud storage (e.g. with the <span class="in">`s3://`</span> or <span class="in">`gs://`</span> prefix) and they will </span>
<span id="cb11-399"><a href="#cb11-399" aria-hidden="true" tabindex="-1"></a>automatically be downloaded and _staged_ when you start the run.</span>
<span id="cb11-400"><a href="#cb11-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-401"><a href="#cb11-401" aria-hidden="true" tabindex="-1"></a><span class="fu">## Quantseq FWD with UMIs: Nugent et al dataset</span></span>
<span id="cb11-402"><a href="#cb11-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-403"><a href="#cb11-403" aria-hidden="true" tabindex="-1"></a>Especially when only low amounts of input material (total RNA) are available,</span>
<span id="cb11-404"><a href="#cb11-404" aria-hidden="true" tabindex="-1"></a>the QuantSeq FWD procotol requires multiple PCR amplification steps. Each step</span>
<span id="cb11-405"><a href="#cb11-405" aria-hidden="true" tabindex="-1"></a>generates clones of the originally captured tags, but does not provide new</span>
<span id="cb11-406"><a href="#cb11-406" aria-hidden="true" tabindex="-1"></a>information about transcript abundance. Inclusion of unique molecular</span>
<span id="cb11-407"><a href="#cb11-407" aria-hidden="true" tabindex="-1"></a>identifiers (UMIs) in the second strand synthesis guarantees that PCR duplicates</span>
<span id="cb11-408"><a href="#cb11-408" aria-hidden="true" tabindex="-1"></a>can be identified and removed in  the analysis (based on both the coordinates of</span>
<span id="cb11-409"><a href="#cb11-409" aria-hidden="true" tabindex="-1"></a>their alignment and the UMI sequence).</span>
<span id="cb11-410"><a href="#cb11-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-411"><a href="#cb11-411" aria-hidden="true" tabindex="-1"></a>Nugent et al used the QuantSeq FWD protocol with UMIs, and we will reanalyze</span>
<span id="cb11-412"><a href="#cb11-412" aria-hidden="true" tabindex="-1"></a>this dataset to illustrate how to instruct</span>
<span id="cb11-413"><a href="#cb11-413" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">UMI-tools</span><span class="co">](https://github.com/CGATOxford/UMI-tools)</span></span>
<span id="cb11-414"><a href="#cb11-414" aria-hidden="true" tabindex="-1"></a>to extract the UMIs from the raw reads and deduplicated the STAR alignments.</span>
<span id="cb11-415"><a href="#cb11-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-416"><a href="#cb11-416" aria-hidden="true" tabindex="-1"></a><span class="fu">### Experimental design</span></span>
<span id="cb11-417"><a href="#cb11-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-418"><a href="#cb11-418" aria-hidden="true" tabindex="-1"></a>This study examines FACS-isolated astroctyes and microglia from female mice that</span>
<span id="cb11-419"><a href="#cb11-419" aria-hidden="true" tabindex="-1"></a>were either 2- or 16 months of age. The animals are either wildtype (<span class="in">`WT`</span>) or  </span>
<span id="cb11-420"><a href="#cb11-420" aria-hidden="true" tabindex="-1"></a>homozygous knockouts (<span class="in">`KO`</span>) for the</span>
<span id="cb11-421"><a href="#cb11-421" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Trem2 gene</span><span class="co">](https://www.uniprot.org/uniprotkb/Q99NH8/entry)</span>.</span>
<span id="cb11-422"><a href="#cb11-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-423"><a href="#cb11-423" aria-hidden="true" tabindex="-1"></a>From each mouse, separate microglia and astrocyte samples were collected and</span>
<span id="cb11-424"><a href="#cb11-424" aria-hidden="true" tabindex="-1"></a>processed into separate libraries with [Lexogen's QuantSeq</span>
<span id="cb11-425"><a href="#cb11-425" aria-hidden="true" tabindex="-1"></a>FWD](https://www.lexogen.com/quantseq-3mrna-sequencing/) kit _and_ the</span>
<span id="cb11-426"><a href="#cb11-426" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">QuantSeq UMI add-on module</span><span class="co">](https://faqs.lexogen.com/faq/quantseq-umi-module)</span>.</span>
<span id="cb11-427"><a href="#cb11-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-428"><a href="#cb11-428" aria-hidden="true" tabindex="-1"></a><span class="fu">### Raw data retrieval with nf-core/fetchngs</span></span>
<span id="cb11-429"><a href="#cb11-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-430"><a href="#cb11-430" aria-hidden="true" tabindex="-1"></a>As in the first example, we use the <span class="in">`nf-core/fetchngs`</span> workflow to retrieve the</span>
<span id="cb11-431"><a href="#cb11-431" aria-hidden="true" tabindex="-1"></a>FASTQ files from SRA.</span>
<span id="cb11-432"><a href="#cb11-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-433"><a href="#cb11-433" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-434"><a href="#cb11-434" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> mkdir <span class="ex">SRP213880</span></span>
<span id="cb11-435"><a href="#cb11-435" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> cd <span class="ex">SRP213880</span></span>
<span id="cb11-436"><a href="#cb11-436" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">SRP213880</span> <span class="op">&gt;</span> ids.txt</span>
<span id="cb11-437"><a href="#cb11-437" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> nf-core/fetchngs <span class="dt">\</span></span>
<span id="cb11-438"><a href="#cb11-438" aria-hidden="true" tabindex="-1"></a>      <span class="at">-revision</span> 1.9 <span class="dt">\</span></span>
<span id="cb11-439"><a href="#cb11-439" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb11-440"><a href="#cb11-440" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> raw_data <span class="dt">\</span></span>
<span id="cb11-441"><a href="#cb11-441" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> ids.txt <span class="dt">\</span></span>
<span id="cb11-442"><a href="#cb11-442" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_pipeline</span> rnaseq <span class="dt">\</span></span>
<span id="cb11-443"><a href="#cb11-443" aria-hidden="true" tabindex="-1"></a>      <span class="at">--nf_core_rnaseq_strandedness</span> forward</span>
<span id="cb11-444"><a href="#cb11-444" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-445"><a href="#cb11-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-446"><a href="#cb11-446" aria-hidden="true" tabindex="-1"></a><span class="fu">### Configuring the nf-core/rnaseq workflow</span></span>
<span id="cb11-447"><a href="#cb11-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-448"><a href="#cb11-448" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Reference data</span></span>
<span id="cb11-449"><a href="#cb11-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-450"><a href="#cb11-450" aria-hidden="true" tabindex="-1"></a>According to the sample metadata in </span>
<span id="cb11-451"><a href="#cb11-451" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">NCBI GEO</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3933549)</span></span>
<span id="cb11-452"><a href="#cb11-452" aria-hidden="true" tabindex="-1"></a>the authors used the same reference data as Xia et al, </span>
<span id="cb11-453"><a href="#cb11-453" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Gencode release M17</span><span class="co">](https://www.gencodegenes.org/mouse/release_M17.html)</span>.</span>
<span id="cb11-454"><a href="#cb11-454" aria-hidden="true" tabindex="-1"></a>We can therefore reuse the reference files and indices we generated in the first</span>
<span id="cb11-455"><a href="#cb11-455" aria-hidden="true" tabindex="-1"></a>analysis (see above).</span>
<span id="cb11-456"><a href="#cb11-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-457"><a href="#cb11-457" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Unique molecular identifiers</span></span>
<span id="cb11-458"><a href="#cb11-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-459"><a href="#cb11-459" aria-hidden="true" tabindex="-1"></a>To process QuantSeq data with UMIs, the following parameters need to be </span>
<span id="cb11-460"><a href="#cb11-460" aria-hidden="true" tabindex="-1"></a>provided.</span>
<span id="cb11-461"><a href="#cb11-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-462"><a href="#cb11-462" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--with_umi`</span></span>
<span id="cb11-463"><a href="#cb11-463" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--umitools_extract_method "regex"`</span>: The type of pattern to detect, either</span>
<span id="cb11-464"><a href="#cb11-464" aria-hidden="true" tabindex="-1"></a>  <span class="in">`string`</span> (default) or <span class="in">`regex`</span></span>
<span id="cb11-465"><a href="#cb11-465" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--umitools_bc_pattern "^(?P&lt;umi_1&gt;.{6})(?P&lt;discard_1&gt;.{4}).*"`</span>: A regular</span>
<span id="cb11-466"><a href="#cb11-466" aria-hidden="true" tabindex="-1"></a>  expression that instructs <span class="in">`UMI-tools`</span> to extract the first 6 bases (the UMI) </span>
<span id="cb11-467"><a href="#cb11-467" aria-hidden="true" tabindex="-1"></a>  from the 5' end of the read <span class="ot">[^4]</span> and to discard the following 4 bases. </span>
<span id="cb11-468"><a href="#cb11-468" aria-hidden="true" tabindex="-1"></a>  (Lexogen adds an invariant <span class="in">`TATA`</span> sequence motif to each UMI.)</span>
<span id="cb11-469"><a href="#cb11-469" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`--umitools_grouping_method "unique"`</span>: The method used to group similar UMIs,</span>
<span id="cb11-470"><a href="#cb11-470" aria-hidden="true" tabindex="-1"></a>  e.g. to correct sequencing errors. The default method is <span class="in">`directional`</span>, but it</span>
<span id="cb11-471"><a href="#cb11-471" aria-hidden="true" tabindex="-1"></a>  is computationally expensive. Here, we use the simpler <span class="in">`unique`</span> method, </span>
<span id="cb11-472"><a href="#cb11-472" aria-hidden="true" tabindex="-1"></a>  suppressing grouping / error correction entirely.</span>
<span id="cb11-473"><a href="#cb11-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-474"><a href="#cb11-474" aria-hidden="true" tabindex="-1"></a><span class="ot">[^4]: </span>Typically, QuantSeq libraries are single-end (not paired-end) and the</span>
<span id="cb11-475"><a href="#cb11-475" aria-hidden="true" tabindex="-1"></a>UMI is therefore found in the R1 read.</span>
<span id="cb11-476"><a href="#cb11-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-477"><a href="#cb11-477" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb11-478"><a href="#cb11-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-479"><a href="#cb11-479" aria-hidden="true" tabindex="-1"></a>These arguments shown above are suitable for UMIs offered by Lexogen for the </span>
<span id="cb11-480"><a href="#cb11-480" aria-hidden="true" tabindex="-1"></a>QuantSeq protocol. If you use custom UMIs, please consult the </span>
<span id="cb11-481"><a href="#cb11-481" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/rnaseq documentation for full details and additional options</span><span class="co">](https://nf-co.re/rnaseq/3.10.1/parameters#umi-options)</span></span>
<span id="cb11-482"><a href="#cb11-482" aria-hidden="true" tabindex="-1"></a>for details on additional parameters you can set.</span>
<span id="cb11-483"><a href="#cb11-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-484"><a href="#cb11-484" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb11-485"><a href="#cb11-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-486"><a href="#cb11-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-487"><a href="#cb11-487" aria-hidden="true" tabindex="-1"></a><span class="fu">### Starting the workflow</span></span>
<span id="cb11-488"><a href="#cb11-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-489"><a href="#cb11-489" aria-hidden="true" tabindex="-1"></a>Next, we add the UMI-related parameters to the JSON file we used above, so we</span>
<span id="cb11-490"><a href="#cb11-490" aria-hidden="true" tabindex="-1"></a>can pass them to <span class="in">`nextflow run`</span> via the <span class="in">`-params-file`</span> argument. Because Nugent</span>
<span id="cb11-491"><a href="#cb11-491" aria-hidden="true" tabindex="-1"></a>et al used the same reference (Gencode M17), we can reuse the same reference we</span>
<span id="cb11-492"><a href="#cb11-492" aria-hidden="true" tabindex="-1"></a>generated for Xia et al's dataset. (Here, we assume the precomputed references</span>
<span id="cb11-493"><a href="#cb11-493" aria-hidden="true" tabindex="-1"></a>are in the <span class="in">`genome`</span> folder to our current working directory. If you chose to</span>
<span id="cb11-494"><a href="#cb11-494" aria-hidden="true" tabindex="-1"></a>store them elsewhere, please provide absolute paths in the JSON file instead.)</span>
<span id="cb11-495"><a href="#cb11-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-496"><a href="#cb11-496" aria-hidden="true" tabindex="-1"></a><span class="in">```json</span></span>
<span id="cb11-497"><a href="#cb11-497" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb11-498"><a href="#cb11-498" aria-hidden="true" tabindex="-1"></a>    <span class="st">"save_reference"</span><span class="op">:</span> <span class="kw">false</span><span class="op">,</span></span>
<span id="cb11-499"><a href="#cb11-499" aria-hidden="true" tabindex="-1"></a>    <span class="st">"fasta"</span><span class="op">:</span> <span class="st">"genome/GRCm38.primary_assembly.genome.fa"</span><span class="op">,</span></span>
<span id="cb11-500"><a href="#cb11-500" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gtf"</span><span class="op">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.gtf"</span><span class="op">,</span></span>
<span id="cb11-501"><a href="#cb11-501" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gene_bed"</span><span class="op">:</span> <span class="st">"genome/gencode.vM17.primary_assembly.annotation.bed"</span><span class="op">,</span></span>
<span id="cb11-502"><a href="#cb11-502" aria-hidden="true" tabindex="-1"></a>    <span class="st">"transcript_fasta"</span><span class="op">:</span> <span class="st">"genome/genome.transcripts.fa"</span><span class="op">,</span></span>
<span id="cb11-503"><a href="#cb11-503" aria-hidden="true" tabindex="-1"></a>    <span class="st">"star_index"</span><span class="op">:</span> <span class="st">"genome/index/star"</span><span class="op">,</span></span>
<span id="cb11-504"><a href="#cb11-504" aria-hidden="true" tabindex="-1"></a>    <span class="st">"salmon_index"</span><span class="op">:</span> <span class="st">"genome/index/salmon"</span><span class="op">,</span></span>
<span id="cb11-505"><a href="#cb11-505" aria-hidden="true" tabindex="-1"></a>    <span class="st">"rsem_index"</span><span class="op">:</span> <span class="st">"genome/rsem"</span><span class="op">,</span></span>
<span id="cb11-506"><a href="#cb11-506" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_star_align_args"</span><span class="op">:</span> <span class="st">"--alignIntronMax 1000000 --alignIntronMin 20 --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --outFilterMismatchNmax 999 --outFilterMultimapNmax 20 --outFilterType BySJout --outFilterMismatchNoverLmax 0.1 --clip3pAdapterSeq AAAAAAAA"</span><span class="op">,</span></span>
<span id="cb11-507"><a href="#cb11-507" aria-hidden="true" tabindex="-1"></a>    <span class="st">"with_umi"</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb11-508"><a href="#cb11-508" aria-hidden="true" tabindex="-1"></a>    <span class="st">"umitools_extract_method"</span><span class="op">:</span> <span class="st">"regex"</span><span class="op">,</span></span>
<span id="cb11-509"><a href="#cb11-509" aria-hidden="true" tabindex="-1"></a>    <span class="st">"umitools_grouping_method"</span><span class="op">:</span> <span class="st">"unique"</span><span class="op">,</span></span>
<span id="cb11-510"><a href="#cb11-510" aria-hidden="true" tabindex="-1"></a>    <span class="st">"umitools_bc_pattern"</span><span class="op">:</span> <span class="st">"^(?P&lt;umi_1&gt;.{6})(?P&lt;discard_1&gt;.{4}).*"</span><span class="op">,</span></span>
<span id="cb11-511"><a href="#cb11-511" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extra_salmon_quant_args"</span><span class="op">:</span> <span class="st">"--noLengthCorrection"</span><span class="op">,</span></span>
<span id="cb11-512"><a href="#cb11-512" aria-hidden="true" tabindex="-1"></a>    <span class="st">"skip_stringtie"</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb11-513"><a href="#cb11-513" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gencode"</span><span class="op">:</span> <span class="kw">true</span></span>
<span id="cb11-514"><a href="#cb11-514" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-515"><a href="#cb11-515" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-516"><a href="#cb11-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-517"><a href="#cb11-517" aria-hidden="true" tabindex="-1"></a>We store the JSON string in a file called <span class="in">`nf_params.json`</span>, and then pass it to</span>
<span id="cb11-518"><a href="#cb11-518" aria-hidden="true" tabindex="-1"></a>the <span class="in">`nextflow run`</span> command:</span>
<span id="cb11-519"><a href="#cb11-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-520"><a href="#cb11-520" aria-hidden="true" tabindex="-1"></a><span class="in">```bash</span></span>
<span id="cb11-521"><a href="#cb11-521" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> nextflow <span class="ex">run</span> <span class="dt">\</span></span>
<span id="cb11-522"><a href="#cb11-522" aria-hidden="true" tabindex="-1"></a>      nf-core/rnaseq <span class="dt">\</span></span>
<span id="cb11-523"><a href="#cb11-523" aria-hidden="true" tabindex="-1"></a>      <span class="at">-r</span> 3.10.1 <span class="dt">\</span></span>
<span id="cb11-524"><a href="#cb11-524" aria-hidden="true" tabindex="-1"></a>      <span class="at">-profile</span> docker <span class="dt">\</span></span>
<span id="cb11-525"><a href="#cb11-525" aria-hidden="true" tabindex="-1"></a>      <span class="at">-params-file</span> nf_params.json <span class="dt">\</span></span>
<span id="cb11-526"><a href="#cb11-526" aria-hidden="true" tabindex="-1"></a>      <span class="at">-resume</span> <span class="dt">\</span></span>
<span id="cb11-527"><a href="#cb11-527" aria-hidden="true" tabindex="-1"></a>      <span class="at">--input</span> raw_data/samplesheet/samplesheet.csv <span class="dt">\</span></span>
<span id="cb11-528"><a href="#cb11-528" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> SRP213880</span>
<span id="cb11-529"><a href="#cb11-529" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb11-530"><a href="#cb11-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-531"><a href="#cb11-531" aria-hidden="true" tabindex="-1"></a>Upon successful completion, the workflow's output is available in the </span>
<span id="cb11-532"><a href="#cb11-532" aria-hidden="true" tabindex="-1"></a><span class="in">`SRP213880`</span> directory. </span>
<span id="cb11-533"><a href="#cb11-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-534"><a href="#cb11-534" aria-hidden="true" tabindex="-1"></a>Next, we will take a closer look at the available results in </span>
<span id="cb11-535"><a href="#cb11-535" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">the second post in this series</span><span class="co">](../nextflow-core-quantseq-2-output/index.html)</span>].</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>