<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-09-05">

<title>Thomas Sandmann’s blog - Adventures with parquet II: Implementing the parquetArray S4 class</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Adventures with parquet II: Implementing the parquetArray S4 class</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">R</div>
                <div class="quarto-category">TIL</div>
                <div class="quarto-category">parquet</div>
                <div class="quarto-category">Bioconductor</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">September 5, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#exporting-gene-expression-data-to-parquet-files" id="toc-exporting-gene-expression-data-to-parquet-files" class="nav-link" data-scroll-target="#exporting-gene-expression-data-to-parquet-files">Exporting gene expression data to parquet files</a></li>
  <li><a href="#creating-the-parquetarrayseed-s4-class" id="toc-creating-the-parquetarrayseed-s4-class" class="nav-link" data-scroll-target="#creating-the-parquetarrayseed-s4-class">Creating the <code>ParquetArraySeed</code> S4 class</a></li>
  <li><a href="#essential-methods" id="toc-essential-methods" class="nav-link" data-scroll-target="#essential-methods">Essential methods</a>
  <ul>
  <li><a href="#dim-and-dimnames-methods" id="toc-dim-and-dimnames-methods" class="nav-link" data-scroll-target="#dim-and-dimnames-methods"><code>dim</code> and <code>dimnames</code> methods</a></li>
  <li><a href="#the-extract_array-method" id="toc-the-extract_array-method" class="nav-link" data-scroll-target="#the-extract_array-method">The <code>extract_array</code> method</a></li>
  </ul></li>
  <li><a href="#creating-a-first-parquetarrayseed-object" id="toc-creating-a-first-parquetarrayseed-object" class="nav-link" data-scroll-target="#creating-a-first-parquetarrayseed-object">Creating a first <code>ParquetArraySeed</code> object</a></li>
  <li><a href="#creating-a-parquet-backed-summarizedexperiment" id="toc-creating-a-parquet-backed-summarizedexperiment" class="nav-link" data-scroll-target="#creating-a-parquet-backed-summarizedexperiment">Creating a parquet-backed SummarizedExperiment</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  <li><a href="#reproducibility" id="toc-reproducibility" class="nav-link" data-scroll-target="#reproducibility">Reproducibility</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<p><a href="../parquet/">Previously</a>, I learned how to export gene expression data from Bioconductor/R objects as parquet files. Today, I am exploring Hervé Pagès’s awesome <a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html">DelayedArray Bioconductor package</a> to represent data stored in parquet files as R matrices - retrieving only those subsets needed for downstream analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"arrow"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"duckdb"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"edgeR"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"glue"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"rnaseqExamples"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"tibble"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"tidyr"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"DelayedArray"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"Mus.musculus"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-gene-expression-data-to-parquet-files" class="level2">
<h2 class="anchored" data-anchor-id="exporting-gene-expression-data-to-parquet-files">Exporting gene expression data to parquet files</h2>
<p>To get some example RNA-seq data, I am loading two <code>SummarizedExperiments</code> (<code>tau</code> and <code>sarm1</code>)with bulk RNA-seq data from my <a href="https://github.com/tomsing1/rnaseq-examples">rnaseqExamples R package</a>.</p>
<p>First, I extract the raw counts into data.frames with a short <code>tidy()</code> helper function. Afterward, I store this data in parquet files in a temporary directory.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Note that I am not including sample annotations in the parquet files, as I found it useful to keep this (relatively small) set of metadata in a separate object, file or (e.g.&nbsp;SQLite) database.</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Coerce a DGEList or a SummarizedExperiment into a tibble</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>tidy <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract raw counts</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(x)<span class="sc">$</span>counts <span class="sc">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(x), <span class="at">names_to =</span> <span class="st">"sample_id"</span>, </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">values_to =</span> <span class="st">"count"</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Store gene expression data for two mouse RNA-seq studies as parquet files"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>out_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"parquet"</span>)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(out_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (dataset <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"sarm1"</span>)) {</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">tidy</span>(<span class="fu">get</span>(dataset))</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>study <span class="ot">&lt;-</span> dataset  <span class="co"># add a columns with the name of the experiment</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  arrow<span class="sc">::</span><span class="fu">write_parquet</span>(</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> df,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">sink =</span> <span class="fu">file.path</span>(out_dir, <span class="fu">paste0</span>(dataset, <span class="st">".parquet"</span>))</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="fu">dir</span>(out_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sarm1.parquet" "tau.parquet"  </code></pre>
</div>
</div>
</section>
<section id="creating-the-parquetarrayseed-s4-class" class="level2">
<h2 class="anchored" data-anchor-id="creating-the-parquetarrayseed-s4-class">Creating the <code>ParquetArraySeed</code> S4 class</h2>
<p>Now, with the two parquet files in place, I am ready to follow <a href="https://bioconductor.org/packages/release/bioc/vignettes/DelayedArray/inst/doc/02-Implementing_a_backend.html">Hervé’s excellent instructions</a> to define my very own <code>ParquetArraySeed</code> S4 class, which can then be passed into the <code>DelayedArray()</code> constructor function (see below).</p>
<p>To keep things simple, my <code>ParquetArraySeed</code> class will only store a single attribute: the path to the directory containing the parquet files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setClass</span>(<span class="st">"ParquetArraySeed"</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">contains =</span> <span class="st">"Array"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">slots =</span> <span class="fu">c</span>(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">filepath =</span> <span class="st">"character"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To simplify the instantiation of new objects, I also define a constructor function, which also ensures that the absolute file path is stored.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom tools file_path_as_absolute</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>ParquetArraySeed <span class="ot">&lt;-</span> <span class="cf">function</span>(filepath) {</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  filepath <span class="ot">&lt;-</span> tools<span class="sc">::</span><span class="fu">file_path_as_absolute</span>(filepath)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">new</span>(<span class="st">"ParquetArraySeed"</span>, <span class="at">filepath =</span> filepath)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="essential-methods" class="level2">
<h2 class="anchored" data-anchor-id="essential-methods">Essential methods</h2>
<p>To power a <code>DelayedArray</code> object, I need to define at least three different S4 methods for my new class:</p>
<ul>
<li><code>dim()</code> - returning an integer vector with the dimensions</li>
<li><code>dimnames()</code> - returning a list of character vectors with the dimension names (if any), e.g.&nbsp;the row and column names of a matrix.</li>
<li><code>extract_array()</code> - returning an <em>ordinary array</em> for a set of indices (see below), e.g.&nbsp;a subset of the dataset to realize in memory.</li>
</ul>
<section id="dim-and-dimnames-methods" class="level3">
<h3 class="anchored" data-anchor-id="dim-and-dimnames-methods"><code>dim</code> and <code>dimnames</code> methods</h3>
<p>To query the set of parquet files, I am using <a href="https://duckdb.org/docs/api/r.html">duckdb</a> to retrieve the unique sample and feature identifiers. In case this information is useful later, I am caching the return values with <a href="https://memoise.r-lib.org/">memoise</a>. That’s handy because I can simply return the lengths of the dimensions names via the <code>dim()</code> method - without accessing the same files again.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>.unique_values <span class="ot">&lt;-</span> <span class="cf">function</span>(x, column, <span class="at">con =</span> <span class="cn">NULL</span>, <span class="at">suffix =</span> <span class="st">".parquet"</span>) {</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(con)) {</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    con <span class="ot">&lt;-</span> duckdb<span class="sc">::</span><span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">on.exit</span>(duckdb<span class="sc">::</span><span class="fu">dbDisconnect</span>(con, <span class="at">shutdown=</span><span class="cn">TRUE</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  data_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(x<span class="sc">@</span>filepath, <span class="fu">paste0</span>(<span class="st">"*"</span>, suffix))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dbGetQuery</span>(</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">con =</span> con,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">glue_sql</span>(</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>      <span class="st">"SELECT DISTINCT {`column`} </span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="st">       ORDER BY {`column`}"</span>, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">.con =</span> con)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  )[, <span class="dv">1</span>]</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>unique_values <span class="ot">&lt;-</span> memoise<span class="sc">::</span><span class="fu">memoise</span>(.unique_values)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"dimnames"</span>, <span class="st">"ParquetArraySeed"</span>, <span class="cf">function</span>(x) {</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  samples <span class="ot">&lt;-</span> <span class="fu">unique_values</span>(x, <span class="at">column =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  features <span class="ot">&lt;-</span> <span class="fu">unique_values</span>(x, <span class="at">column =</span> <span class="st">"feature_id"</span>)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(features, samples)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"dim"</span>, <span class="st">"ParquetArraySeed"</span>, <span class="cf">function</span>(x) {</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lengths</span>(<span class="fu">dimnames</span>(x))</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="the-extract_array-method" class="level3">
<h3 class="anchored" data-anchor-id="the-extract_array-method">The <code>extract_array</code> method</h3>
<p>I also use <code>duckdb()</code> to retrieve the actual data for a subset of features, a subset of samples - or both. In case users only specify one <em>or</em> the other, e.g.&nbsp;by passing an index with <code>NULL</code> values, I explicitely define four different SQL queries returning:</p>
<ol type="1">
<li>The full dataset (when <code>index = list(NULL, NULL)</code>),</li>
<li>All features for a subset of samples,</li>
<li>Selected features for all samples, or</li>
<li>Selected features for selected samples.</li>
</ol>
<p>Finally, I pivot the data into a feature x sample matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>.get_data <span class="ot">&lt;-</span> <span class="cf">function</span>(x, index, <span class="at">con =</span> <span class="cn">NULL</span>, <span class="at">suffix =</span> <span class="st">".parquet"</span>) {</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(con)) {</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    con <span class="ot">&lt;-</span> duckdb<span class="sc">::</span><span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">on.exit</span>(duckdb<span class="sc">::</span><span class="fu">dbDisconnect</span>(con, <span class="at">shutdown=</span><span class="cn">TRUE</span>))</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  data_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(x<span class="sc">@</span>filepath, <span class="fu">paste0</span>(<span class="st">"*"</span>, suffix))</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># match indices to feature and sample identifiers</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  dims <span class="ot">&lt;-</span> <span class="fu">dimnames</span>(x)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no indices =&gt; return the full dataset</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;</span> <span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>   dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir})"</span>, </span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>   )</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>   keep_features <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>feature_id)</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>   keep_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_id)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>   <span class="co"># no sample index =&gt; return all samples</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;&amp;</span> <span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE feature_id IN ({keep_features*})"</span>, </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_id)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># no feature index =&gt; return all features</span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> dims[[<span class="dv">2</span>]][index[[<span class="dv">2</span>]]]</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE sample_id IN ({keep_samples*})"</span>, </span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>feature_id)</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> dims[[<span class="dv">2</span>]][index[[<span class="dv">2</span>]]]</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE feature_id IN ({keep_features*})AND sample_id IN ({keep_samples*})</span></span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a><span class="st">       ORDER BY sample_id, feature_id"</span>, </span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pivot the count data into a regular matrix</span></span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="cn">NA_integer_</span>, </span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">nrow =</span> <span class="fu">length</span>(keep_features),</span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="fu">length</span>(keep_samples),</span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">dimnames =</span> <span class="fu">list</span>(keep_features, keep_samples))</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true" tabindex="-1"></a>  matrix_index <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">match</span>(dataset<span class="sc">$</span>feature_id, <span class="fu">row.names</span>(m)), </span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true" tabindex="-1"></a>    <span class="fu">match</span>(dataset<span class="sc">$</span>sample_id, <span class="fu">colnames</span>(m))</span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true" tabindex="-1"></a>  m[matrix_index] <span class="ot">&lt;-</span> dataset<span class="sc">$</span>count</span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(m)</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true" tabindex="-1"></a>.extract_array_from_ParquetArraySeed <span class="ot">&lt;-</span> <span class="cf">function</span>(x, index) {</span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.get_data</span>(<span class="at">x =</span> x, <span class="at">index =</span> index)</span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"extract_array"</span>, <span class="st">"ParquetArraySeed"</span>, </span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true" tabindex="-1"></a>          .extract_array_from_ParquetArraySeed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="creating-a-first-parquetarrayseed-object" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-first-parquetarrayseed-object">Creating a first <code>ParquetArraySeed</code> object</h2>
<p>With these three methods in place, I can instantiate my first <code>ParquetArraySeed</code> object, which is suitable as input to the <code>DelayedArray</code> constructor from the eponymous R package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>seed <span class="ot">&lt;-</span> <span class="fu">ParquetArraySeed</span>(out_dir)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>da <span class="ot">&lt;-</span> <span class="fu">DelayedArray</span>(seed)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>da</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;53801 x 48&gt; DelayedMatrix object of type "integer":
                   DRN-10415 DRN-10418 DRN-10421 ... DRN-16866 DRN-16867
ENSMUSG00000000001      2754      2764      2907   .      1214      1002
ENSMUSG00000000003         0         0         0   .         0         0
ENSMUSG00000000028        54        61        66   .        39        57
ENSMUSG00000000031         0         0         3   .        16        17
ENSMUSG00000000037         9        50        26   .        58        78
               ...         .         .         .   .         .         .
ENSMUSG00000116519         0         0         0   .         0         0
ENSMUSG00000116520         0         0         0   .         0         0
ENSMUSG00000116521         0         0         0   .         0         0
ENSMUSG00000116525         3         1         2   .        65        44
ENSMUSG00000116528         0         0         0   .         0         0
                   DRN-16868
ENSMUSG00000000001       921
ENSMUSG00000000003         0
ENSMUSG00000000028        38
ENSMUSG00000000031        22
ENSMUSG00000000037        33
               ...         .
ENSMUSG00000116519         0
ENSMUSG00000116520         0
ENSMUSG00000116521         0
ENSMUSG00000116525        45
ENSMUSG00000116528         0</code></pre>
</div>
</div>
<p>When the object is printed, we are presented with the first couple of rows and columns, but most of the data remains on disk (e.g.&nbsp;has not been retrieved from the parquet files).</p>
<p>Subsetting the <code>DelayedArray</code> works as expected, e.g.&nbsp;to retain only samples from the <code>tau</code> study:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset to one dataset - using gene and sample identifiers</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>da[<span class="fu">row.names</span>(tau), <span class="fu">colnames</span>(tau)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;53801 x 32&gt; DelayedMatrix object of type "integer":
                   DRN-10415 DRN-10418 DRN-10421 ... DRN-10502 DRN-10505
ENSMUSG00000000001      2754      2764      2907   .      2260      3273
ENSMUSG00000000003         0         0         0   .         0         0
ENSMUSG00000000028        54        61        66   .        16        82
ENSMUSG00000000031         0         0         3   .         0         0
ENSMUSG00000000037         9        50        26   .        11        11
               ...         .         .         .   .         .         .
ENSMUSG00000116519         0         0         0   .         0         0
ENSMUSG00000116520         0         0         0   .         0         0
ENSMUSG00000116521         0         0         0   .         0         0
ENSMUSG00000116525         3         1         2   .         1         3
ENSMUSG00000116528         0         0         0   .         0         0
                   DRN-10508
ENSMUSG00000000001      2758
ENSMUSG00000000003         0
ENSMUSG00000000028        47
ENSMUSG00000000031         0
ENSMUSG00000000037         6
               ...         .
ENSMUSG00000116519         0
ENSMUSG00000116520         0
ENSMUSG00000116521         0
ENSMUSG00000116525         1
ENSMUSG00000116528         0</code></pre>
</div>
</div>
</section>
<section id="creating-a-parquet-backed-summarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-parquet-backed-summarizedexperiment">Creating a parquet-backed SummarizedExperiment</h2>
<p>For downstream analyses, it is useful to combine sample- and feature-annotations with the gene expression counts in a single <code>SummarizedExperiment</code>.</p>
<p>Let’s extract the sample metadata from the original data objects and add a <code>study</code> column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>pdata <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"sarm1"</span>), \(x) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  col_data <span class="ot">&lt;-</span> <span class="fu">colData</span>(<span class="fu">get</span>(x))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  col_data<span class="sc">$</span>study <span class="ot">&lt;-</span> x</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(col_data)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>pdata <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(rbind, pdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The original datasets only provide ensembl gene identifiers, so let’s retrieve the corresponding gene symbols and entrez identifiers from the <a href="https://bioconductor.org/packages/release/data/annotation/html/Mus.musculus.html">Mus.musculus Bioconductor annotation package</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add Entrez identifiers and gene symbols for both human and mouse genes</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>fdata <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ensembl =</span> <span class="fu">row.names</span>(da),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">entrez =</span> <span class="fu">mapIds</span>(Mus.musculus, </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">row.names</span>(da), </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column =</span> <span class="st">"ENTREZID"</span>, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>, </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals =</span> <span class="st">"first"</span>),</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">symbol =</span> <span class="fu">mapIds</span>(Mus.musculus, </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">row.names</span>(da), </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column =</span> <span class="st">"SYMBOL"</span>, </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>, </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals =</span> <span class="st">"first"</span>),</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="fu">row.names</span>(da)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we are ready to compose meta- and gene expression data into a single object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> da), </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData =</span> fdata[<span class="fu">row.names</span>(da), ],</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> pdata[<span class="fu">colnames</span>(da),,<span class="at">drop=</span><span class="cn">FALSE</span> ]</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 53801 48 
metadata(0):
assays(1): counts
rownames(53801): ENSMUSG00000000001 ENSMUSG00000000003 ...
  ENSMUSG00000116525 ENSMUSG00000116528
rowData names(3): ensembl entrez symbol
colnames(48): DRN-10415 DRN-10418 ... DRN-16867 DRN-16868
colData names(2): group study</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(se<span class="sc">$</span>study)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
sarm1   tau 
   16    32 </code></pre>
</div>
</div>
<p>At this point, we still haven’t retrieved the actual counts from the parquet files, e.g.&nbsp;they information is still on disk. It will automatically be retrieved when downstream functions coerce the <code>DelayedArray</code> into a regular matrix. For example, <a href="https://rdrr.io/bioc/edgeR/man/calcNormFactors.html">edgeR’s calcNormFactors function</a> accepts a SummarizedExperiment and returns a <code>DGEList</code>, ready for exploratory analysis e.g.&nbsp;multi-dimensional scaling.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(se[, se<span class="sc">$</span>study <span class="sc">==</span> <span class="st">"tau"</span>])</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>y<span class="sc">$</span>samples<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"WT_"</span>, y<span class="sc">$</span>samples<span class="sc">$</span>group), <span class="st">"WT"</span>, <span class="st">"Transgenic"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">palette.colors</span>(<span class="at">n =</span> <span class="fu">nlevels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                         <span class="at">palette =</span> <span class="st">"Set1"</span>)[<span class="fu">as.integer</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype)]</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">plotMDS</span>(y, <span class="at">col =</span> colors, <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">levels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">pch =</span> <span class="dv">19</span>, </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">palette.colors</span>(<span class="at">n =</span> <span class="fu">nlevels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), <span class="at">palette =</span> <span class="st">"Set1"</span>),</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">y.intersp=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Because the <code>DelayedArray</code> package is so thoughtfully designed, it was surprisingly straightforward to represent data stored in parquet files as a matrix. This way, I can leverage familiar Bioconductor tools <em>and</em> take advantage of the language-agnostic file format.</p>
<p>A few directions and caveats that I haven’t explored, yet:</p>
<ul>
<li>In this example, I am generating one parquet file for each of the studies, e.g.&nbsp;I am manually partitioning the data by <code>study</code>. When I want to retrieve all data for one of the studies, I could simply parse <em>only</em> the corresponding parquet file, instead of querying &amp; filtering all files in the directory (as a <a href="https://arrow.apache.org/docs/r/articles/dataset.html#opening-datasets">Dataset</a> ).</li>
<li>My parquet files are stored on the local system. Accessing them in a cloud storage location (e.g.&nbsp;on AWS S3) is supported by <code>duckdb</code> as well, but network speed will likely become limiting. I will explore the use of <code>AWS Athena</code>, a cloud-based SQL database, in a future post, as shown e.g.&nbsp;in <a href="https://aws.amazon.com/blogs/industries/fast-queries-of-scrnaseq-datasets-with-amazon-athena/">this AWS blog post</a>.</li>
<li>The parquet files I generated here are relatively small, e.g.&nbsp;~ 5 Mb, reflecting the relatively small size of the example bulk RNA-seq datasets. The <a href="https://arrow.apache.org/docs/r/articles/dataset.html#partitioning-performance-considerations">arrow project recommends</a> avoiding accessing many files smaller than 20MB, but that’s likely not a problem until I need to handle thousands of studies.</li>
<li>Disk-backed arrays are great if computations can be performed on small chunks of data, e.g.&nbsp;to visualize a subset of feature and / or samples. When it is necessary to realize the full dataset in memory, the <code>DelayedArray</code> returns a conventional matrix - e.g.&nbsp;sparse representations don’t seem to be supported right now (as far as I can tell).</li>
<li>Finally, I have only implemented three essential methods, but have not touched on any optimized backend-specific methods. For example, we could issue a SQL query to calculate <code>max</code> or <code>colSums</code> values instead of loading the full data matrix into our R session. (This might be especially interesting if the database engine is running on a remote server.)</li>
</ul>
<p>Lots more to learn - and a big thank you to <a href="https://www.linkedin.com/in/herv%C3%A9-pag%C3%A8s-5038527">Hervé Pagès</a> and the entire Bioconductor team!</p>
</section>
<section id="reproducibility" class="level2">
<h2 class="anchored" data-anchor-id="reproducibility">Reproducibility</h2>
<details>
<summary>
Session Information
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="st">"attached"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       Debian GNU/Linux 12 (bookworm)
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Los_Angeles
 date     2023-09-05
 pandoc   3.1.1 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 ! package                            * version    date (UTC) lib source
 P abind                              * 1.4-5      2016-07-21 [?] CRAN (R 4.3.1)
 P AnnotationDbi                      * 1.62.2     2023-07-02 [?] Bioconductor
 P arrow                              * 13.0.0     2023-08-30 [?] CRAN (R 4.3.1)
 P Biobase                            * 2.60.0     2023-04-25 [?] Bioconductor
 P BiocGenerics                       * 0.46.0     2023-04-25 [?] Bioconductor
 P DBI                                * 1.1.3      2022-06-18 [?] CRAN (R 4.3.1)
 P DelayedArray                       * 0.26.7     2023-07-28 [?] Bioconductor
 P DESeq2                             * 1.40.2     2023-06-23 [?] Bioconductor
 P dplyr                              * 1.1.2      2023-04-20 [?] CRAN (R 4.3.1)
 P duckdb                             * 0.8.1-2    2023-08-25 [?] CRAN (R 4.3.1)
 P edgeR                              * 3.42.4     2023-05-31 [?] Bioconductor
 P GenomeInfoDb                       * 1.36.1     2023-06-21 [?] Bioconductor
 P GenomicFeatures                    * 1.52.2     2023-08-25 [?] Bioconductor
 P GenomicRanges                      * 1.52.0     2023-04-25 [?] Bioconductor
 P glue                               * 1.6.2      2022-02-24 [?] CRAN (R 4.3.1)
 P GO.db                              * 3.17.0     2023-09-04 [?] Bioconductor
 P IRanges                            * 2.34.1     2023-06-22 [?] Bioconductor
 P limma                              * 3.56.2     2023-06-04 [?] Bioconductor
 P Matrix                             * 1.5-4.1    2023-05-18 [?] CRAN (R 4.3.1)
 P MatrixGenerics                     * 1.12.3     2023-07-30 [?] Bioconductor
 P matrixStats                        * 1.0.0      2023-06-02 [?] CRAN (R 4.3.1)
 P Mus.musculus                       * 1.3.1      2023-09-04 [?] Bioconductor
 P org.Mm.eg.db                       * 3.17.0     2023-09-04 [?] Bioconductor
 P OrganismDbi                        * 1.42.0     2023-04-25 [?] Bioconductor
 P rnaseqExamples                     * 0.0.0.9000 2023-09-04 [?] Github (tomsing1/rnaseq-Examples@ac35304)
 P S4Arrays                           * 1.0.5      2023-07-24 [?] Bioconductor
 P S4Vectors                          * 0.38.1     2023-05-02 [?] Bioconductor
 P SummarizedExperiment               * 1.30.2     2023-06-06 [?] Bioconductor
 P tibble                             * 3.2.1      2023-03-20 [?] CRAN (R 4.3.1)
 P tidyr                              * 1.3.0      2023-01-24 [?] CRAN (R 4.3.1)
 P TxDb.Mmusculus.UCSC.mm10.knownGene * 3.10.0     2023-09-04 [?] Bioconductor

 [1] /home/sandmann/repositories/blog/renv/library/R-4.3/x86_64-pc-linux-gnu
 [2] /home/sandmann/.cache/R/renv/sandbox/R-4.3/x86_64-pc-linux-gnu/9a444a72

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>

<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb21" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Adventures with parquet II: Implementing the parquetArray S4 class"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-09-05"</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [R, TIL, parquet, Bioconductor]</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Previously</span><span class="co">](../parquet/)</span>, </span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>I learned how to export gene expression data from Bioconductor/R</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>objects as parquet files. Today, I am exploring Hervé Pagès's awesome</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">DelayedArray Bioconductor package</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/DelayedArray.html)</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>to represent data stored in parquet files as R matrices - retrieving only those</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>subsets needed for downstream analysis.</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"arrow"</span>)</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"duckdb"</span>)</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"edgeR"</span>)</span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"glue"</span>)</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"rnaseqExamples"</span>)</span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"tibble"</span>)</span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"tidyr"</span>)</span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"DelayedArray"</span>)</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(<span class="st">"Mus.musculus"</span>)</span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb21-49"><a href="#cb21-49" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-50"><a href="#cb21-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-51"><a href="#cb21-51" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exporting gene expression data to parquet files</span></span>
<span id="cb21-52"><a href="#cb21-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-53"><a href="#cb21-53" aria-hidden="true" tabindex="-1"></a>To get some example RNA-seq data, I am loading two <span class="in">`SummarizedExperiments`</span></span>
<span id="cb21-54"><a href="#cb21-54" aria-hidden="true" tabindex="-1"></a>(<span class="in">`tau`</span> and <span class="in">`sarm1`</span>)with bulk RNA-seq data from my</span>
<span id="cb21-55"><a href="#cb21-55" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">rnaseqExamples R package</span><span class="co">](https://github.com/tomsing1/rnaseq-examples)</span>. </span>
<span id="cb21-56"><a href="#cb21-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-57"><a href="#cb21-57" aria-hidden="true" tabindex="-1"></a>First, I extract the raw counts into data.frames with a short <span class="in">`tidy()`</span> </span>
<span id="cb21-58"><a href="#cb21-58" aria-hidden="true" tabindex="-1"></a>helper function. Afterward, I store this data in parquet files in a temporary</span>
<span id="cb21-59"><a href="#cb21-59" aria-hidden="true" tabindex="-1"></a>directory.</span>
<span id="cb21-60"><a href="#cb21-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-61"><a href="#cb21-61" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true"}</span>
<span id="cb21-62"><a href="#cb21-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-63"><a href="#cb21-63" aria-hidden="true" tabindex="-1"></a>Note that I am not including sample annotations in the parquet files, as I</span>
<span id="cb21-64"><a href="#cb21-64" aria-hidden="true" tabindex="-1"></a>found it useful to keep this (relatively small) set of metadata in a</span>
<span id="cb21-65"><a href="#cb21-65" aria-hidden="true" tabindex="-1"></a>separate object, file or (e.g. SQLite) database.</span>
<span id="cb21-66"><a href="#cb21-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-67"><a href="#cb21-67" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-68"><a href="#cb21-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-69"><a href="#cb21-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-72"><a href="#cb21-72" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-73"><a href="#cb21-73" aria-hidden="true" tabindex="-1"></a><span class="co"># Coerce a DGEList or a SummarizedExperiment into a tibble</span></span>
<span id="cb21-74"><a href="#cb21-74" aria-hidden="true" tabindex="-1"></a>tidy <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb21-75"><a href="#cb21-75" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract raw counts</span></span>
<span id="cb21-76"><a href="#cb21-76" aria-hidden="true" tabindex="-1"></a>  edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(x)<span class="sc">$</span>counts <span class="sc">%&gt;%</span></span>
<span id="cb21-77"><a href="#cb21-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb21-78"><a href="#cb21-78" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"feature_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-79"><a href="#cb21-79" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">colnames</span>(x), <span class="at">names_to =</span> <span class="st">"sample_id"</span>, </span>
<span id="cb21-80"><a href="#cb21-80" aria-hidden="true" tabindex="-1"></a>                        <span class="at">values_to =</span> <span class="st">"count"</span>)</span>
<span id="cb21-81"><a href="#cb21-81" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-82"><a href="#cb21-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-83"><a href="#cb21-83" aria-hidden="true" tabindex="-1"></a><span class="co"># Store gene expression data for two mouse RNA-seq studies as parquet files"</span></span>
<span id="cb21-84"><a href="#cb21-84" aria-hidden="true" tabindex="-1"></a>out_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"parquet"</span>)</span>
<span id="cb21-85"><a href="#cb21-85" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(out_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-86"><a href="#cb21-86" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (dataset <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"sarm1"</span>)) {</span>
<span id="cb21-87"><a href="#cb21-87" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">tidy</span>(<span class="fu">get</span>(dataset))</span>
<span id="cb21-88"><a href="#cb21-88" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>study <span class="ot">&lt;-</span> dataset  <span class="co"># add a columns with the name of the experiment</span></span>
<span id="cb21-89"><a href="#cb21-89" aria-hidden="true" tabindex="-1"></a>  arrow<span class="sc">::</span><span class="fu">write_parquet</span>(</span>
<span id="cb21-90"><a href="#cb21-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> df,</span>
<span id="cb21-91"><a href="#cb21-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">sink =</span> <span class="fu">file.path</span>(out_dir, <span class="fu">paste0</span>(dataset, <span class="st">".parquet"</span>))</span>
<span id="cb21-92"><a href="#cb21-92" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-93"><a href="#cb21-93" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-94"><a href="#cb21-94" aria-hidden="true" tabindex="-1"></a><span class="fu">dir</span>(out_dir)</span>
<span id="cb21-95"><a href="#cb21-95" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-96"><a href="#cb21-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-97"><a href="#cb21-97" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating the `ParquetArraySeed` S4 class</span></span>
<span id="cb21-98"><a href="#cb21-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-99"><a href="#cb21-99" aria-hidden="true" tabindex="-1"></a>Now, with the two parquet files in place, I am ready to follow </span>
<span id="cb21-100"><a href="#cb21-100" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Hervé's excellent instructions</span><span class="co">](https://bioconductor.org/packages/release/bioc/vignettes/DelayedArray/inst/doc/02-Implementing_a_backend.html)</span></span>
<span id="cb21-101"><a href="#cb21-101" aria-hidden="true" tabindex="-1"></a>to define my very own <span class="in">`ParquetArraySeed`</span> S4 class, which can then be passed</span>
<span id="cb21-102"><a href="#cb21-102" aria-hidden="true" tabindex="-1"></a>into the <span class="in">`DelayedArray()`</span> constructor function (see below).</span>
<span id="cb21-103"><a href="#cb21-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-104"><a href="#cb21-104" aria-hidden="true" tabindex="-1"></a>To keep things simple, my <span class="in">`ParquetArraySeed`</span> class will only store a single</span>
<span id="cb21-105"><a href="#cb21-105" aria-hidden="true" tabindex="-1"></a>attribute: the path to the directory containing the parquet files.</span>
<span id="cb21-106"><a href="#cb21-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-107"><a href="#cb21-107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r seed-class-definition}</span></span>
<span id="cb21-108"><a href="#cb21-108" aria-hidden="true" tabindex="-1"></a><span class="fu">setClass</span>(<span class="st">"ParquetArraySeed"</span>,</span>
<span id="cb21-109"><a href="#cb21-109" aria-hidden="true" tabindex="-1"></a>    <span class="at">contains =</span> <span class="st">"Array"</span>,</span>
<span id="cb21-110"><a href="#cb21-110" aria-hidden="true" tabindex="-1"></a>    <span class="at">slots =</span> <span class="fu">c</span>(</span>
<span id="cb21-111"><a href="#cb21-111" aria-hidden="true" tabindex="-1"></a>        <span class="at">filepath =</span> <span class="st">"character"</span></span>
<span id="cb21-112"><a href="#cb21-112" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-113"><a href="#cb21-113" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-114"><a href="#cb21-114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-115"><a href="#cb21-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-116"><a href="#cb21-116" aria-hidden="true" tabindex="-1"></a>To simplify the instantiation of new objects, I also define a constructor</span>
<span id="cb21-117"><a href="#cb21-117" aria-hidden="true" tabindex="-1"></a>function, which also ensures that the absolute file path is stored.</span>
<span id="cb21-118"><a href="#cb21-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-119"><a href="#cb21-119" aria-hidden="true" tabindex="-1"></a><span class="in">```{r constructor}</span></span>
<span id="cb21-120"><a href="#cb21-120" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom tools file_path_as_absolute</span></span>
<span id="cb21-121"><a href="#cb21-121" aria-hidden="true" tabindex="-1"></a>ParquetArraySeed <span class="ot">&lt;-</span> <span class="cf">function</span>(filepath) {</span>
<span id="cb21-122"><a href="#cb21-122" aria-hidden="true" tabindex="-1"></a>  filepath <span class="ot">&lt;-</span> tools<span class="sc">::</span><span class="fu">file_path_as_absolute</span>(filepath)</span>
<span id="cb21-123"><a href="#cb21-123" aria-hidden="true" tabindex="-1"></a>  <span class="fu">new</span>(<span class="st">"ParquetArraySeed"</span>, <span class="at">filepath =</span> filepath)</span>
<span id="cb21-124"><a href="#cb21-124" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-125"><a href="#cb21-125" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-126"><a href="#cb21-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-127"><a href="#cb21-127" aria-hidden="true" tabindex="-1"></a><span class="fu">## Essential methods</span></span>
<span id="cb21-128"><a href="#cb21-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-129"><a href="#cb21-129" aria-hidden="true" tabindex="-1"></a>To power a <span class="in">`DelayedArray`</span> object, I need to define at least three different</span>
<span id="cb21-130"><a href="#cb21-130" aria-hidden="true" tabindex="-1"></a>S4 methods for my new class:</span>
<span id="cb21-131"><a href="#cb21-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-132"><a href="#cb21-132" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`dim()`</span> - returning an integer vector with the dimensions</span>
<span id="cb21-133"><a href="#cb21-133" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`dimnames()`</span> - returning a list of character vectors with the dimension names</span>
<span id="cb21-134"><a href="#cb21-134" aria-hidden="true" tabindex="-1"></a>  (if any), e.g. the row and column names of a matrix.</span>
<span id="cb21-135"><a href="#cb21-135" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`extract_array()`</span> - returning an _ordinary array_ for a set of indices (see</span>
<span id="cb21-136"><a href="#cb21-136" aria-hidden="true" tabindex="-1"></a>  below), e.g. a subset of the dataset to realize in memory.</span>
<span id="cb21-137"><a href="#cb21-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-138"><a href="#cb21-138" aria-hidden="true" tabindex="-1"></a><span class="fu">### `dim` and `dimnames` methods</span></span>
<span id="cb21-139"><a href="#cb21-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-140"><a href="#cb21-140" aria-hidden="true" tabindex="-1"></a>To query the set of parquet files, I am using </span>
<span id="cb21-141"><a href="#cb21-141" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">duckdb</span><span class="co">](https://duckdb.org/docs/api/r.html)</span></span>
<span id="cb21-142"><a href="#cb21-142" aria-hidden="true" tabindex="-1"></a>to retrieve the unique sample and feature identifiers. In case this information</span>
<span id="cb21-143"><a href="#cb21-143" aria-hidden="true" tabindex="-1"></a>is useful later, I am caching the return values with </span>
<span id="cb21-144"><a href="#cb21-144" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">memoise</span><span class="co">](https://memoise.r-lib.org/)</span>. </span>
<span id="cb21-145"><a href="#cb21-145" aria-hidden="true" tabindex="-1"></a>That's handy because I can simply return the lengths of the dimensions names via</span>
<span id="cb21-146"><a href="#cb21-146" aria-hidden="true" tabindex="-1"></a>the <span class="in">`dim()`</span> method - without accessing the same files again.</span>
<span id="cb21-147"><a href="#cb21-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-148"><a href="#cb21-148" aria-hidden="true" tabindex="-1"></a><span class="in">```{r dim-dimnames}</span></span>
<span id="cb21-149"><a href="#cb21-149" aria-hidden="true" tabindex="-1"></a>.unique_values <span class="ot">&lt;-</span> <span class="cf">function</span>(x, column, <span class="at">con =</span> <span class="cn">NULL</span>, <span class="at">suffix =</span> <span class="st">".parquet"</span>) {</span>
<span id="cb21-150"><a href="#cb21-150" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(con)) {</span>
<span id="cb21-151"><a href="#cb21-151" aria-hidden="true" tabindex="-1"></a>    con <span class="ot">&lt;-</span> duckdb<span class="sc">::</span><span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span>
<span id="cb21-152"><a href="#cb21-152" aria-hidden="true" tabindex="-1"></a>    <span class="fu">on.exit</span>(duckdb<span class="sc">::</span><span class="fu">dbDisconnect</span>(con, <span class="at">shutdown=</span><span class="cn">TRUE</span>))</span>
<span id="cb21-153"><a href="#cb21-153" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-154"><a href="#cb21-154" aria-hidden="true" tabindex="-1"></a>  data_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(x<span class="sc">@</span>filepath, <span class="fu">paste0</span>(<span class="st">"*"</span>, suffix))</span>
<span id="cb21-155"><a href="#cb21-155" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dbGetQuery</span>(</span>
<span id="cb21-156"><a href="#cb21-156" aria-hidden="true" tabindex="-1"></a>    <span class="at">con =</span> con,</span>
<span id="cb21-157"><a href="#cb21-157" aria-hidden="true" tabindex="-1"></a>    <span class="fu">glue_sql</span>(</span>
<span id="cb21-158"><a href="#cb21-158" aria-hidden="true" tabindex="-1"></a>      <span class="st">"SELECT DISTINCT {`column`} </span></span>
<span id="cb21-159"><a href="#cb21-159" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb21-160"><a href="#cb21-160" aria-hidden="true" tabindex="-1"></a><span class="st">       ORDER BY {`column`}"</span>, </span>
<span id="cb21-161"><a href="#cb21-161" aria-hidden="true" tabindex="-1"></a>     <span class="at">.con =</span> con)</span>
<span id="cb21-162"><a href="#cb21-162" aria-hidden="true" tabindex="-1"></a>  )[, <span class="dv">1</span>]</span>
<span id="cb21-163"><a href="#cb21-163" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-164"><a href="#cb21-164" aria-hidden="true" tabindex="-1"></a>unique_values <span class="ot">&lt;-</span> memoise<span class="sc">::</span><span class="fu">memoise</span>(.unique_values)</span>
<span id="cb21-165"><a href="#cb21-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-166"><a href="#cb21-166" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"dimnames"</span>, <span class="st">"ParquetArraySeed"</span>, <span class="cf">function</span>(x) {</span>
<span id="cb21-167"><a href="#cb21-167" aria-hidden="true" tabindex="-1"></a>  samples <span class="ot">&lt;-</span> <span class="fu">unique_values</span>(x, <span class="at">column =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb21-168"><a href="#cb21-168" aria-hidden="true" tabindex="-1"></a>  features <span class="ot">&lt;-</span> <span class="fu">unique_values</span>(x, <span class="at">column =</span> <span class="st">"feature_id"</span>)</span>
<span id="cb21-169"><a href="#cb21-169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(features, samples)</span>
<span id="cb21-170"><a href="#cb21-170" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb21-171"><a href="#cb21-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-172"><a href="#cb21-172" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"dim"</span>, <span class="st">"ParquetArraySeed"</span>, <span class="cf">function</span>(x) {</span>
<span id="cb21-173"><a href="#cb21-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lengths</span>(<span class="fu">dimnames</span>(x))</span>
<span id="cb21-174"><a href="#cb21-174" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb21-175"><a href="#cb21-175" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-176"><a href="#cb21-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-177"><a href="#cb21-177" aria-hidden="true" tabindex="-1"></a><span class="fu">### The `extract_array` method</span></span>
<span id="cb21-178"><a href="#cb21-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-179"><a href="#cb21-179" aria-hidden="true" tabindex="-1"></a>I also use <span class="in">`duckdb()`</span> to retrieve the actual data for a subset of </span>
<span id="cb21-180"><a href="#cb21-180" aria-hidden="true" tabindex="-1"></a>features, a subset of samples - or both. In case users only specify one _or_</span>
<span id="cb21-181"><a href="#cb21-181" aria-hidden="true" tabindex="-1"></a>the other, e.g. by passing an index with <span class="in">`NULL`</span> values, I explicitely define</span>
<span id="cb21-182"><a href="#cb21-182" aria-hidden="true" tabindex="-1"></a>four different SQL queries returning:</span>
<span id="cb21-183"><a href="#cb21-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-184"><a href="#cb21-184" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>The full dataset (when <span class="in">`index = list(NULL, NULL)`</span>),</span>
<span id="cb21-185"><a href="#cb21-185" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>All features for a subset of samples,</span>
<span id="cb21-186"><a href="#cb21-186" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Selected features for all samples, or</span>
<span id="cb21-187"><a href="#cb21-187" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Selected features for selected samples.</span>
<span id="cb21-188"><a href="#cb21-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-189"><a href="#cb21-189" aria-hidden="true" tabindex="-1"></a>Finally, I pivot the data into a feature x sample matrix.</span>
<span id="cb21-190"><a href="#cb21-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-191"><a href="#cb21-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{r extractarray}</span></span>
<span id="cb21-192"><a href="#cb21-192" aria-hidden="true" tabindex="-1"></a>.get_data <span class="ot">&lt;-</span> <span class="cf">function</span>(x, index, <span class="at">con =</span> <span class="cn">NULL</span>, <span class="at">suffix =</span> <span class="st">".parquet"</span>) {</span>
<span id="cb21-193"><a href="#cb21-193" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(con)) {</span>
<span id="cb21-194"><a href="#cb21-194" aria-hidden="true" tabindex="-1"></a>    con <span class="ot">&lt;-</span> duckdb<span class="sc">::</span><span class="fu">dbConnect</span>(duckdb<span class="sc">::</span><span class="fu">duckdb</span>())</span>
<span id="cb21-195"><a href="#cb21-195" aria-hidden="true" tabindex="-1"></a>    <span class="fu">on.exit</span>(duckdb<span class="sc">::</span><span class="fu">dbDisconnect</span>(con, <span class="at">shutdown=</span><span class="cn">TRUE</span>))</span>
<span id="cb21-196"><a href="#cb21-196" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-197"><a href="#cb21-197" aria-hidden="true" tabindex="-1"></a>  data_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(x<span class="sc">@</span>filepath, <span class="fu">paste0</span>(<span class="st">"*"</span>, suffix))</span>
<span id="cb21-198"><a href="#cb21-198" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-199"><a href="#cb21-199" aria-hidden="true" tabindex="-1"></a>  <span class="co"># match indices to feature and sample identifiers</span></span>
<span id="cb21-200"><a href="#cb21-200" aria-hidden="true" tabindex="-1"></a>  dims <span class="ot">&lt;-</span> <span class="fu">dimnames</span>(x)</span>
<span id="cb21-201"><a href="#cb21-201" aria-hidden="true" tabindex="-1"></a>  keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb21-202"><a href="#cb21-202" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-203"><a href="#cb21-203" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no indices =&gt; return the full dataset</span></span>
<span id="cb21-204"><a href="#cb21-204" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;</span> <span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb21-205"><a href="#cb21-205" aria-hidden="true" tabindex="-1"></a>   dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb21-206"><a href="#cb21-206" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb21-207"><a href="#cb21-207" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb21-208"><a href="#cb21-208" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb21-209"><a href="#cb21-209" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir})"</span>, </span>
<span id="cb21-210"><a href="#cb21-210" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb21-211"><a href="#cb21-211" aria-hidden="true" tabindex="-1"></a>   )</span>
<span id="cb21-212"><a href="#cb21-212" aria-hidden="true" tabindex="-1"></a>   keep_features <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>feature_id)</span>
<span id="cb21-213"><a href="#cb21-213" aria-hidden="true" tabindex="-1"></a>   keep_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_id)</span>
<span id="cb21-214"><a href="#cb21-214" aria-hidden="true" tabindex="-1"></a>   <span class="co"># no sample index =&gt; return all samples</span></span>
<span id="cb21-215"><a href="#cb21-215" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;&amp;</span> <span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb21-216"><a href="#cb21-216" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb21-217"><a href="#cb21-217" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb21-218"><a href="#cb21-218" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb21-219"><a href="#cb21-219" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb21-220"><a href="#cb21-220" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb21-221"><a href="#cb21-221" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb21-222"><a href="#cb21-222" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE feature_id IN ({keep_features*})"</span>, </span>
<span id="cb21-223"><a href="#cb21-223" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb21-224"><a href="#cb21-224" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-225"><a href="#cb21-225" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_id)</span>
<span id="cb21-226"><a href="#cb21-226" aria-hidden="true" tabindex="-1"></a>    <span class="co"># no feature index =&gt; return all features</span></span>
<span id="cb21-227"><a href="#cb21-227" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">is.null</span>(index[[<span class="dv">1</span>]]) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(index[[<span class="dv">2</span>]])) {</span>
<span id="cb21-228"><a href="#cb21-228" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> dims[[<span class="dv">2</span>]][index[[<span class="dv">2</span>]]]</span>
<span id="cb21-229"><a href="#cb21-229" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb21-230"><a href="#cb21-230" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb21-231"><a href="#cb21-231" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb21-232"><a href="#cb21-232" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb21-233"><a href="#cb21-233" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb21-234"><a href="#cb21-234" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE sample_id IN ({keep_samples*})"</span>, </span>
<span id="cb21-235"><a href="#cb21-235" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb21-236"><a href="#cb21-236" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-237"><a href="#cb21-237" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>feature_id)</span>
<span id="cb21-238"><a href="#cb21-238" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb21-239"><a href="#cb21-239" aria-hidden="true" tabindex="-1"></a>    keep_features <span class="ot">&lt;-</span> dims[[<span class="dv">1</span>]][index[[<span class="dv">1</span>]]]</span>
<span id="cb21-240"><a href="#cb21-240" aria-hidden="true" tabindex="-1"></a>    keep_samples <span class="ot">&lt;-</span> dims[[<span class="dv">2</span>]][index[[<span class="dv">2</span>]]]</span>
<span id="cb21-241"><a href="#cb21-241" aria-hidden="true" tabindex="-1"></a>    dataset <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(</span>
<span id="cb21-242"><a href="#cb21-242" aria-hidden="true" tabindex="-1"></a>      <span class="at">con =</span> con,</span>
<span id="cb21-243"><a href="#cb21-243" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glue_sql</span>(</span>
<span id="cb21-244"><a href="#cb21-244" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SELECT feature_id, sample_id, count</span></span>
<span id="cb21-245"><a href="#cb21-245" aria-hidden="true" tabindex="-1"></a><span class="st">       FROM read_parquet({data_dir}) </span></span>
<span id="cb21-246"><a href="#cb21-246" aria-hidden="true" tabindex="-1"></a><span class="st">       WHERE feature_id IN ({keep_features*})AND sample_id IN ({keep_samples*})</span></span>
<span id="cb21-247"><a href="#cb21-247" aria-hidden="true" tabindex="-1"></a><span class="st">       ORDER BY sample_id, feature_id"</span>, </span>
<span id="cb21-248"><a href="#cb21-248" aria-hidden="true" tabindex="-1"></a>       <span class="at">.con =</span> con)</span>
<span id="cb21-249"><a href="#cb21-249" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb21-250"><a href="#cb21-250" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-251"><a href="#cb21-251" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pivot the count data into a regular matrix</span></span>
<span id="cb21-252"><a href="#cb21-252" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb21-253"><a href="#cb21-253" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="cn">NA_integer_</span>, </span>
<span id="cb21-254"><a href="#cb21-254" aria-hidden="true" tabindex="-1"></a>    <span class="at">nrow =</span> <span class="fu">length</span>(keep_features),</span>
<span id="cb21-255"><a href="#cb21-255" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="fu">length</span>(keep_samples),</span>
<span id="cb21-256"><a href="#cb21-256" aria-hidden="true" tabindex="-1"></a>    <span class="at">dimnames =</span> <span class="fu">list</span>(keep_features, keep_samples))</span>
<span id="cb21-257"><a href="#cb21-257" aria-hidden="true" tabindex="-1"></a>  matrix_index <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb21-258"><a href="#cb21-258" aria-hidden="true" tabindex="-1"></a>    <span class="fu">match</span>(dataset<span class="sc">$</span>feature_id, <span class="fu">row.names</span>(m)), </span>
<span id="cb21-259"><a href="#cb21-259" aria-hidden="true" tabindex="-1"></a>    <span class="fu">match</span>(dataset<span class="sc">$</span>sample_id, <span class="fu">colnames</span>(m))</span>
<span id="cb21-260"><a href="#cb21-260" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-261"><a href="#cb21-261" aria-hidden="true" tabindex="-1"></a>  m[matrix_index] <span class="ot">&lt;-</span> dataset<span class="sc">$</span>count</span>
<span id="cb21-262"><a href="#cb21-262" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(m)</span>
<span id="cb21-263"><a href="#cb21-263" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-264"><a href="#cb21-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-265"><a href="#cb21-265" aria-hidden="true" tabindex="-1"></a>.extract_array_from_ParquetArraySeed <span class="ot">&lt;-</span> <span class="cf">function</span>(x, index) {</span>
<span id="cb21-266"><a href="#cb21-266" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.get_data</span>(<span class="at">x =</span> x, <span class="at">index =</span> index)</span>
<span id="cb21-267"><a href="#cb21-267" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-268"><a href="#cb21-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-269"><a href="#cb21-269" aria-hidden="true" tabindex="-1"></a><span class="fu">setMethod</span>(<span class="st">"extract_array"</span>, <span class="st">"ParquetArraySeed"</span>, </span>
<span id="cb21-270"><a href="#cb21-270" aria-hidden="true" tabindex="-1"></a>          .extract_array_from_ParquetArraySeed)</span>
<span id="cb21-271"><a href="#cb21-271" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-272"><a href="#cb21-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-273"><a href="#cb21-273" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating a first `ParquetArraySeed` object</span></span>
<span id="cb21-274"><a href="#cb21-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-275"><a href="#cb21-275" aria-hidden="true" tabindex="-1"></a>With these three methods in place, I can instantiate my first <span class="in">`ParquetArraySeed`</span></span>
<span id="cb21-276"><a href="#cb21-276" aria-hidden="true" tabindex="-1"></a>object, which is suitable as input to the <span class="in">`DelayedArray`</span> constructor from</span>
<span id="cb21-277"><a href="#cb21-277" aria-hidden="true" tabindex="-1"></a>the eponymous R package:</span>
<span id="cb21-278"><a href="#cb21-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-281"><a href="#cb21-281" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-282"><a href="#cb21-282" aria-hidden="true" tabindex="-1"></a>seed <span class="ot">&lt;-</span> <span class="fu">ParquetArraySeed</span>(out_dir)</span>
<span id="cb21-283"><a href="#cb21-283" aria-hidden="true" tabindex="-1"></a>da <span class="ot">&lt;-</span> <span class="fu">DelayedArray</span>(seed)</span>
<span id="cb21-284"><a href="#cb21-284" aria-hidden="true" tabindex="-1"></a>da</span>
<span id="cb21-285"><a href="#cb21-285" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-286"><a href="#cb21-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-287"><a href="#cb21-287" aria-hidden="true" tabindex="-1"></a>When the object is printed, we are presented with the first couple of rows</span>
<span id="cb21-288"><a href="#cb21-288" aria-hidden="true" tabindex="-1"></a>and columns, but most of the data remains on disk (e.g. has not been retrieved</span>
<span id="cb21-289"><a href="#cb21-289" aria-hidden="true" tabindex="-1"></a>from the parquet files).</span>
<span id="cb21-290"><a href="#cb21-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-291"><a href="#cb21-291" aria-hidden="true" tabindex="-1"></a>Subsetting the <span class="in">`DelayedArray`</span> works as expected, e.g. to retain only samples</span>
<span id="cb21-292"><a href="#cb21-292" aria-hidden="true" tabindex="-1"></a>from the <span class="in">`tau`</span> study:</span>
<span id="cb21-293"><a href="#cb21-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-296"><a href="#cb21-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-297"><a href="#cb21-297" aria-hidden="true" tabindex="-1"></a><span class="co"># subset to one dataset - using gene and sample identifiers</span></span>
<span id="cb21-298"><a href="#cb21-298" aria-hidden="true" tabindex="-1"></a>da[<span class="fu">row.names</span>(tau), <span class="fu">colnames</span>(tau)]</span>
<span id="cb21-299"><a href="#cb21-299" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-300"><a href="#cb21-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-301"><a href="#cb21-301" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating a parquet-backed SummarizedExperiment</span></span>
<span id="cb21-302"><a href="#cb21-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-303"><a href="#cb21-303" aria-hidden="true" tabindex="-1"></a>For downstream analyses, it is useful to combine sample- and feature-annotations</span>
<span id="cb21-304"><a href="#cb21-304" aria-hidden="true" tabindex="-1"></a>with the gene expression counts in a single <span class="in">`SummarizedExperiment`</span>.</span>
<span id="cb21-305"><a href="#cb21-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-306"><a href="#cb21-306" aria-hidden="true" tabindex="-1"></a>Let's extract the sample metadata from the original data objects and add a</span>
<span id="cb21-307"><a href="#cb21-307" aria-hidden="true" tabindex="-1"></a><span class="in">`study`</span> column.</span>
<span id="cb21-308"><a href="#cb21-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-311"><a href="#cb21-311" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-312"><a href="#cb21-312" aria-hidden="true" tabindex="-1"></a>pdata <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">"tau"</span>, <span class="st">"sarm1"</span>), \(x) {</span>
<span id="cb21-313"><a href="#cb21-313" aria-hidden="true" tabindex="-1"></a>  col_data <span class="ot">&lt;-</span> <span class="fu">colData</span>(<span class="fu">get</span>(x))</span>
<span id="cb21-314"><a href="#cb21-314" aria-hidden="true" tabindex="-1"></a>  col_data<span class="sc">$</span>study <span class="ot">&lt;-</span> x</span>
<span id="cb21-315"><a href="#cb21-315" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(col_data)</span>
<span id="cb21-316"><a href="#cb21-316" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb21-317"><a href="#cb21-317" aria-hidden="true" tabindex="-1"></a>pdata <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(rbind, pdata)</span>
<span id="cb21-318"><a href="#cb21-318" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-319"><a href="#cb21-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-320"><a href="#cb21-320" aria-hidden="true" tabindex="-1"></a>The original datasets only provide ensembl gene identifiers, so let's retrieve</span>
<span id="cb21-321"><a href="#cb21-321" aria-hidden="true" tabindex="-1"></a>the corresponding gene symbols and entrez identifiers from the</span>
<span id="cb21-322"><a href="#cb21-322" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Mus.musculus Bioconductor annotation package</span><span class="co">](https://bioconductor.org/packages/release/data/annotation/html/Mus.musculus.html)</span>.</span>
<span id="cb21-323"><a href="#cb21-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-326"><a href="#cb21-326" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-327"><a href="#cb21-327" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb21-328"><a href="#cb21-328" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb21-329"><a href="#cb21-329" aria-hidden="true" tabindex="-1"></a><span class="co"># add Entrez identifiers and gene symbols for both human and mouse genes</span></span>
<span id="cb21-330"><a href="#cb21-330" aria-hidden="true" tabindex="-1"></a>fdata <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(</span>
<span id="cb21-331"><a href="#cb21-331" aria-hidden="true" tabindex="-1"></a>    <span class="at">ensembl =</span> <span class="fu">row.names</span>(da),</span>
<span id="cb21-332"><a href="#cb21-332" aria-hidden="true" tabindex="-1"></a>    <span class="at">entrez =</span> <span class="fu">mapIds</span>(Mus.musculus, </span>
<span id="cb21-333"><a href="#cb21-333" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">row.names</span>(da), </span>
<span id="cb21-334"><a href="#cb21-334" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column =</span> <span class="st">"ENTREZID"</span>, </span>
<span id="cb21-335"><a href="#cb21-335" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>, </span>
<span id="cb21-336"><a href="#cb21-336" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals =</span> <span class="st">"first"</span>),</span>
<span id="cb21-337"><a href="#cb21-337" aria-hidden="true" tabindex="-1"></a>    <span class="at">symbol =</span> <span class="fu">mapIds</span>(Mus.musculus, </span>
<span id="cb21-338"><a href="#cb21-338" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">row.names</span>(da), </span>
<span id="cb21-339"><a href="#cb21-339" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column =</span> <span class="st">"SYMBOL"</span>, </span>
<span id="cb21-340"><a href="#cb21-340" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>, </span>
<span id="cb21-341"><a href="#cb21-341" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals =</span> <span class="st">"first"</span>),</span>
<span id="cb21-342"><a href="#cb21-342" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="fu">row.names</span>(da)</span>
<span id="cb21-343"><a href="#cb21-343" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-344"><a href="#cb21-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-345"><a href="#cb21-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-346"><a href="#cb21-346" aria-hidden="true" tabindex="-1"></a>Now we are ready to compose meta- and gene expression data into a single</span>
<span id="cb21-347"><a href="#cb21-347" aria-hidden="true" tabindex="-1"></a>object:</span>
<span id="cb21-348"><a href="#cb21-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-351"><a href="#cb21-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-352"><a href="#cb21-352" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb21-353"><a href="#cb21-353" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> da), </span>
<span id="cb21-354"><a href="#cb21-354" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData =</span> fdata[<span class="fu">row.names</span>(da), ],</span>
<span id="cb21-355"><a href="#cb21-355" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> pdata[<span class="fu">colnames</span>(da),,<span class="at">drop=</span><span class="cn">FALSE</span> ]</span>
<span id="cb21-356"><a href="#cb21-356" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-357"><a href="#cb21-357" aria-hidden="true" tabindex="-1"></a>se</span>
<span id="cb21-358"><a href="#cb21-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-359"><a href="#cb21-359" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(se<span class="sc">$</span>study)</span>
<span id="cb21-360"><a href="#cb21-360" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-361"><a href="#cb21-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-362"><a href="#cb21-362" aria-hidden="true" tabindex="-1"></a>At this point, we still haven't retrieved the actual counts from the parquet</span>
<span id="cb21-363"><a href="#cb21-363" aria-hidden="true" tabindex="-1"></a>files, e.g. they information is still on disk. It will automatically be</span>
<span id="cb21-364"><a href="#cb21-364" aria-hidden="true" tabindex="-1"></a>retrieved when downstream functions coerce the <span class="in">`DelayedArray`</span> into a regular</span>
<span id="cb21-365"><a href="#cb21-365" aria-hidden="true" tabindex="-1"></a>matrix. For example, </span>
<span id="cb21-366"><a href="#cb21-366" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">edgeR's calcNormFactors function</span><span class="co">](https://rdrr.io/bioc/edgeR/man/calcNormFactors.html)</span></span>
<span id="cb21-367"><a href="#cb21-367" aria-hidden="true" tabindex="-1"></a>accepts a SummarizedExperiment and returns a <span class="in">`DGEList`</span>, ready for exploratory</span>
<span id="cb21-368"><a href="#cb21-368" aria-hidden="true" tabindex="-1"></a>analysis e.g. multi-dimensional scaling.</span>
<span id="cb21-369"><a href="#cb21-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-372"><a href="#cb21-372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-373"><a href="#cb21-373" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb21-374"><a href="#cb21-374" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4</span></span>
<span id="cb21-375"><a href="#cb21-375" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb21-376"><a href="#cb21-376" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">calcNormFactors</span>(se[, se<span class="sc">$</span>study <span class="sc">==</span> <span class="st">"tau"</span>])</span>
<span id="cb21-377"><a href="#cb21-377" aria-hidden="true" tabindex="-1"></a>y<span class="sc">$</span>samples<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb21-378"><a href="#cb21-378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"WT_"</span>, y<span class="sc">$</span>samples<span class="sc">$</span>group), <span class="st">"WT"</span>, <span class="st">"Transgenic"</span>)</span>
<span id="cb21-379"><a href="#cb21-379" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-380"><a href="#cb21-380" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">palette.colors</span>(<span class="at">n =</span> <span class="fu">nlevels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), </span>
<span id="cb21-381"><a href="#cb21-381" aria-hidden="true" tabindex="-1"></a>                         <span class="at">palette =</span> <span class="st">"Set1"</span>)[<span class="fu">as.integer</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype)]</span>
<span id="cb21-382"><a href="#cb21-382" aria-hidden="true" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">plotMDS</span>(y, <span class="at">col =</span> colors, <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb21-383"><a href="#cb21-383" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">levels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">pch =</span> <span class="dv">19</span>, </span>
<span id="cb21-384"><a href="#cb21-384" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">palette.colors</span>(<span class="at">n =</span> <span class="fu">nlevels</span>(y<span class="sc">$</span>samples<span class="sc">$</span>genotype), <span class="at">palette =</span> <span class="st">"Set1"</span>),</span>
<span id="cb21-385"><a href="#cb21-385" aria-hidden="true" tabindex="-1"></a>       <span class="at">y.intersp=</span><span class="dv">2</span>)</span>
<span id="cb21-386"><a href="#cb21-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-387"><a href="#cb21-387" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-388"><a href="#cb21-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-389"><a href="#cb21-389" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb21-390"><a href="#cb21-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-391"><a href="#cb21-391" aria-hidden="true" tabindex="-1"></a>Because the <span class="in">`DelayedArray`</span> package is so thoughtfully designed, it was</span>
<span id="cb21-392"><a href="#cb21-392" aria-hidden="true" tabindex="-1"></a>surprisingly straightforward to represent data stored in parquet files as a</span>
<span id="cb21-393"><a href="#cb21-393" aria-hidden="true" tabindex="-1"></a>matrix. This way, I can leverage familiar Bioconductor tools _and_ take</span>
<span id="cb21-394"><a href="#cb21-394" aria-hidden="true" tabindex="-1"></a>advantage of the language-agnostic file format.</span>
<span id="cb21-395"><a href="#cb21-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-396"><a href="#cb21-396" aria-hidden="true" tabindex="-1"></a>A few directions and caveats that I haven't explored, yet:</span>
<span id="cb21-397"><a href="#cb21-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-398"><a href="#cb21-398" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>In this example, I am generating one parquet file for each of the studies,</span>
<span id="cb21-399"><a href="#cb21-399" aria-hidden="true" tabindex="-1"></a>  e.g. I am manually partitioning the data by <span class="in">`study`</span>. When I want to</span>
<span id="cb21-400"><a href="#cb21-400" aria-hidden="true" tabindex="-1"></a>  retrieve all data for one of the studies, I could simply parse _only_ the</span>
<span id="cb21-401"><a href="#cb21-401" aria-hidden="true" tabindex="-1"></a>  corresponding parquet file, instead of querying &amp; filtering all files in</span>
<span id="cb21-402"><a href="#cb21-402" aria-hidden="true" tabindex="-1"></a>  the directory (as a </span>
<span id="cb21-403"><a href="#cb21-403" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">Dataset</span><span class="co">](https://arrow.apache.org/docs/r/articles/dataset.html#opening-datasets)</span></span>
<span id="cb21-404"><a href="#cb21-404" aria-hidden="true" tabindex="-1"></a>  ).</span>
<span id="cb21-405"><a href="#cb21-405" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>My parquet files are stored on the local system. Accessing them in a cloud</span>
<span id="cb21-406"><a href="#cb21-406" aria-hidden="true" tabindex="-1"></a>  storage location (e.g. on AWS S3) is supported by <span class="in">`duckdb`</span> as well, but</span>
<span id="cb21-407"><a href="#cb21-407" aria-hidden="true" tabindex="-1"></a>  network speed will likely become limiting. I will explore the use of</span>
<span id="cb21-408"><a href="#cb21-408" aria-hidden="true" tabindex="-1"></a>  <span class="in">`AWS Athena`</span>, a cloud-based SQL database, in a future post, as shown e.g. in </span>
<span id="cb21-409"><a href="#cb21-409" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">this AWS blog post</span><span class="co">](https://aws.amazon.com/blogs/industries/fast-queries-of-scrnaseq-datasets-with-amazon-athena/)</span>.</span>
<span id="cb21-410"><a href="#cb21-410" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The parquet files I generated here are relatively small, e.g. ~ 5 Mb,</span>
<span id="cb21-411"><a href="#cb21-411" aria-hidden="true" tabindex="-1"></a>  reflecting the relatively small size of the example bulk RNA-seq datasets.</span>
<span id="cb21-412"><a href="#cb21-412" aria-hidden="true" tabindex="-1"></a>  The</span>
<span id="cb21-413"><a href="#cb21-413" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">arrow project recommends</span><span class="co">](https://arrow.apache.org/docs/r/articles/dataset.html#partitioning-performance-considerations)</span></span>
<span id="cb21-414"><a href="#cb21-414" aria-hidden="true" tabindex="-1"></a>  avoiding accessing many files smaller than 20MB, but that's likely not a</span>
<span id="cb21-415"><a href="#cb21-415" aria-hidden="true" tabindex="-1"></a>  problem until I need to handle thousands of studies.</span>
<span id="cb21-416"><a href="#cb21-416" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Disk-backed arrays are great if computations can be performed on small chunks</span>
<span id="cb21-417"><a href="#cb21-417" aria-hidden="true" tabindex="-1"></a>  of data, e.g. to visualize a subset of feature and / or samples. When it is </span>
<span id="cb21-418"><a href="#cb21-418" aria-hidden="true" tabindex="-1"></a>  necessary to realize the full dataset in memory, the <span class="in">`DelayedArray`</span> returns</span>
<span id="cb21-419"><a href="#cb21-419" aria-hidden="true" tabindex="-1"></a>  a conventional matrix - e.g. sparse representations don't seem to be supported</span>
<span id="cb21-420"><a href="#cb21-420" aria-hidden="true" tabindex="-1"></a>  right now (as far as I can tell).</span>
<span id="cb21-421"><a href="#cb21-421" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Finally, I have only implemented three essential methods, but have not touched</span>
<span id="cb21-422"><a href="#cb21-422" aria-hidden="true" tabindex="-1"></a>  on any optimized backend-specific methods. For example, we could issue a SQL</span>
<span id="cb21-423"><a href="#cb21-423" aria-hidden="true" tabindex="-1"></a>  query to calculate <span class="in">`max`</span> or <span class="in">`colSums`</span> values instead of loading the full</span>
<span id="cb21-424"><a href="#cb21-424" aria-hidden="true" tabindex="-1"></a>  data matrix into our R session. (This might be especially interesting if</span>
<span id="cb21-425"><a href="#cb21-425" aria-hidden="true" tabindex="-1"></a>  the database engine is running on a remote server.)</span>
<span id="cb21-426"><a href="#cb21-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-427"><a href="#cb21-427" aria-hidden="true" tabindex="-1"></a>Lots more to learn - and a big thank you to</span>
<span id="cb21-428"><a href="#cb21-428" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Hervé Pagès</span><span class="co">](https://www.linkedin.com/in/herv%C3%A9-pag%C3%A8s-5038527)</span></span>
<span id="cb21-429"><a href="#cb21-429" aria-hidden="true" tabindex="-1"></a>and the entire Bioconductor team!</span>
<span id="cb21-430"><a href="#cb21-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-431"><a href="#cb21-431" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reproducibility</span></span>
<span id="cb21-432"><a href="#cb21-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-433"><a href="#cb21-433" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb21-434"><a href="#cb21-434" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;</span></span>
<span id="cb21-435"><a href="#cb21-435" aria-hidden="true" tabindex="-1"></a>Session Information</span>
<span id="cb21-436"><a href="#cb21-436" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/summary&gt;</span></span>
<span id="cb21-437"><a href="#cb21-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-440"><a href="#cb21-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-441"><a href="#cb21-441" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="st">"attached"</span>)</span>
<span id="cb21-442"><a href="#cb21-442" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-443"><a href="#cb21-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-444"><a href="#cb21-444" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb21-445"><a href="#cb21-445" aria-hidden="true" tabindex="-1"></a>   </span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>