<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2023-01-16">

<title>Thomas Sandmann’s blog - QuantSeq RNAseq analysis (2): Exploring nf-core/rnaseq output</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-XNBDW4FYSC"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-XNBDW4FYSC', { 'anonymize_ip': true});
</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/"><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann"><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">QuantSeq RNAseq analysis (2): Exploring nf-core/rnaseq output</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">nextflow</div>
                <div class="quarto-category">NGS</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 16, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#reports" id="toc-reports" class="nav-link" data-scroll-target="#reports">Reports</a>
  <ul>
  <li><a href="#multiqc-report" id="toc-multiqc-report" class="nav-link" data-scroll-target="#multiqc-report">MultiQC report</a></li>
  <li><a href="#pipeline-info" id="toc-pipeline-info" class="nav-link" data-scroll-target="#pipeline-info">Pipeline info</a></li>
  <li><a href="#fastqc-reports" id="toc-fastqc-reports" class="nav-link" data-scroll-target="#fastqc-reports">FastQC reports</a></li>
  <li><a href="#trim-galore-reports" id="toc-trim-galore-reports" class="nav-link" data-scroll-target="#trim-galore-reports">Trim Galore reports</a></li>
  <li><a href="#umitools" id="toc-umitools" class="nav-link" data-scroll-target="#umitools">umitools</a></li>
  </ul></li>
  <li><a href="#workflow-results" id="toc-workflow-results" class="nav-link" data-scroll-target="#workflow-results">Workflow results</a>
  <ul>
  <li><a href="#bigwig" id="toc-bigwig" class="nav-link" data-scroll-target="#bigwig">bigwig</a></li>
  <li><a href="#deseq2-object-qc-metrics" id="toc-deseq2-object-qc-metrics" class="nav-link" data-scroll-target="#deseq2-object-qc-metrics">DESeq2 object &amp; QC metrics</a></li>
  <li><a href="#dupradar" id="toc-dupradar" class="nav-link" data-scroll-target="#dupradar">Dupradar</a></li>
  <li><a href="#featurecounts" id="toc-featurecounts" class="nav-link" data-scroll-target="#featurecounts">Featurecounts</a></li>
  <li><a href="#star-log-files" id="toc-star-log-files" class="nav-link" data-scroll-target="#star-log-files">STAR log files</a></li>
  <li><a href="#qualimap" id="toc-qualimap" class="nav-link" data-scroll-target="#qualimap">Qualimap</a></li>
  <li><a href="#rseqc" id="toc-rseqc" class="nav-link" data-scroll-target="#rseqc">RSEQC</a></li>
  </ul></li>
  <li><a href="#alignments-gene-level-counts" id="toc-alignments-gene-level-counts" class="nav-link" data-scroll-target="#alignments-gene-level-counts">Alignments &amp; gene-level counts</a>
  <ul>
  <li><a href="#salmon-quantitation" id="toc-salmon-quantitation" class="nav-link" data-scroll-target="#salmon-quantitation">Salmon quantitation</a>
  <ul class="collapse">
  <li><a href="#aggregated" id="toc-aggregated" class="nav-link" data-scroll-target="#aggregated">Aggregated</a></li>
  <li><a href="#by-sample" id="toc-by-sample" class="nav-link" data-scroll-target="#by-sample">By sample</a></li>
  </ul></li>
  <li><a href="#star-alignments" id="toc-star-alignments" class="nav-link" data-scroll-target="#star-alignments">STAR alignments</a></li>
  </ul></li>
  <li><a href="#genome-gene-annotations-indices" id="toc-genome-gene-annotations-indices" class="nav-link" data-scroll-target="#genome-gene-annotations-indices">Genome, gene annotations &amp; indices</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>This is the second of four posts documenting my progress toward processing and analyzing <a href="https://www.lexogen.com/quantseq-3mrna-sequencing/">QuantSeq FWD 3’ tag RNAseq</a> data with the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq</a> workflow.</p>
<ol type="1">
<li><a href="../nextflow-core-quantseq-1-settings/index.html">Configuring &amp; executing the nf-core/rnaseq workflow</a></li>
<li><strong>Exploring the workflow outputs</strong></li>
<li><a href="../nextflow-core-quantseq-3-xia/index.html">Validating the workflow by reproducing results published by Xia et al (no UMIs)</a></li>
<li><a href="../nextflow-core-quantseq-4-nugent/index.html">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</a></li>
</ol>
<p>Many thanks to <a href="https://github.com/drpatelh">Harshil Patel</a>, <a href="https://github.com/adomingues/">António Miguel de Jesus Domingues</a> and <a href="https://github.com/matthiasZepper/">Matthias Zepper</a> for their generous guidance &amp; input via <a href="nfcore.slack.com">nf-core slack</a>. (Any mistakes are mine.)</p>
</div>
</div>
</div>
<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<ul>
<li>This post documents the output files &amp; folders of the <a href="https://nf-co.re/rnaseq">nf-core/rnaseq workflow (v 3.10.1)</a>, run with default settings with the <code>star_salmon</code> aligner / quantitation method.</li>
<li>For additional information, e.g.&nbsp;on the content of the <a href="https://multiqc.info/">MultiQC</a> report, please see the <a href="https://nf-co.re/rnaseq/3.10.1/output">official nf-core/rnaseq documentation</a>.</li>
</ul>
</section>
<section id="reports" class="level2">
<h2 class="anchored" data-anchor-id="reports">Reports</h2>
<section id="multiqc-report" class="level3">
<h3 class="anchored" data-anchor-id="multiqc-report">MultiQC report</h3>
<p>The MultiQC HTML report is a one-stop-shop that summarises QC metrics across the workflow. It can be found int he <code>multiqc</code> folder, in a subdirectory named according to the aligner &amp; quantifier combination used (default: <code>star_salmon</code>).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/multiqc.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">multiqc report</figcaption><p></p>
</figure>
</div>
</section>
<section id="pipeline-info" class="level3">
<h3 class="anchored" data-anchor-id="pipeline-info">Pipeline info</h3>
<p>The <code>pipeline_info</code> folder contains html reports and text (CSV, TXT, YML) files with information about the run, including the versions of the software tools used.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/pipeline_info.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">pipeline info</figcaption><p></p>
</figure>
</div>
</section>
<section id="fastqc-reports" class="level3">
<h3 class="anchored" data-anchor-id="fastqc-reports">FastQC reports</h3>
<p>The <code>fastqc</code> folder contains the output of the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a> tool. Most of the reported metrics are included in the MultiQC report as well, but the HTML reports for individual samples are available here if needed.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/fastqc.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">FastQC report</figcaption><p></p>
</figure>
</div>
</section>
<section id="trim-galore-reports" class="level3">
<h3 class="anchored" data-anchor-id="trim-galore-reports">Trim Galore reports</h3>
<p>The <code>trimgalore</code> folder contains</p>
<ul>
<li>trimming reports for each sample</li>
<li>the <code>fastqc</code> sub-folder with quality metrics for the <em>trimmed</em> FASTQ files</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/trimgalore.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Trim Galore</figcaption><p></p>
</figure>
</div>
</section>
<section id="umitools" class="level3">
<h3 class="anchored" data-anchor-id="umitools">umitools</h3>
<p>This folder contains the log files returned by <a href="https://github.com/CGATOxford/UMI-tools">UMI-tools</a></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/umitools.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">UMI-tools</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="workflow-results" class="level2">
<h2 class="anchored" data-anchor-id="workflow-results">Workflow results</h2>
<p>The main output of the workflow is available in the <code>star_salmon</code> subdirectory. (This folder is named after the selected alignment &amp; quantification strategy, e.g.&nbsp;<code>star_salmon</code> is present only if this tool combination was used.)</p>
<p>It contains multiple folders, as well as the (deduplicated) sorted BAM files for each sample.</p>
<section id="bigwig" class="level3">
<h3 class="anchored" data-anchor-id="bigwig">bigwig</h3>
<p>Genome coverage in <a href="https://genome.ucsc.edu/goldenPath/help/bigWig.html">bigWig</a> format for each sample.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bigwig.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">BigWig files</figcaption><p></p>
</figure>
</div>
</section>
<section id="deseq2-object-qc-metrics" class="level3">
<h3 class="anchored" data-anchor-id="deseq2-object-qc-metrics">DESeq2 object &amp; QC metrics</h3>
<p>The workflow aggregates all counts into a DESeq2 R objects, performs QC and exploratory analyses and serializes the object as <code>deseq2.dds.RData</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/deseq2.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">DESeq2 object &amp; QC metrics</figcaption><p></p>
</figure>
</div>
</section>
<section id="dupradar" class="level3">
<h3 class="anchored" data-anchor-id="dupradar">Dupradar</h3>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/dupRadar.html">dupRadar</a> Bioconductor package performs duplication rate quality control.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/dupradar.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">dupRadar</figcaption><p></p>
</figure>
</div>
</section>
<section id="featurecounts" class="level3">
<h3 class="anchored" data-anchor-id="featurecounts">Featurecounts</h3>
<p>Output from <a href="https://academic.oup.com/bioinformatics/article/30/7/923/232889?login=false">featurecounts</a> tool is <em>only</em> used to generate QC metrics. For actual quantitation of the gene-level results, the output of <a href="https://combine-lab.github.io/salmon/">salmon</a> (default) or <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323">rsem</a> are used.</p>
<p>The metrics reported in the <code>featurecounts</code> folder are included in the MultiQC report.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/featurecounts.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">featureCounts</figcaption><p></p>
</figure>
</div>
</section>
<section id="star-log-files" class="level3">
<h3 class="anchored" data-anchor-id="star-log-files">STAR log files</h3>
<p>This folder contains log files output by the <code>STAR</code> aligner.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/log.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">STAR logs</figcaption><p></p>
</figure>
</div>
</section>
<section id="qualimap" class="level3">
<h3 class="anchored" data-anchor-id="qualimap">Qualimap</h3>
<p>The <a href="http://qualimap.conesalab.org/">qualimap</a> package generates QC metrics from BAM files.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/qualimap.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Qualimap</figcaption><p></p>
</figure>
</div>
</section>
<section id="rseqc" class="level3">
<h3 class="anchored" data-anchor-id="rseqc">RSEQC</h3>
<p>The output of the <a href="https://rseqc.sourceforge.net/">rseqc</a> QC control package are in this directory.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/rseqc.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">RSEQC</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="alignments-gene-level-counts" class="level2">
<h2 class="anchored" data-anchor-id="alignments-gene-level-counts">Alignments &amp; gene-level counts</h2>
<p>The <code>star_salmon</code> folder also contains the main results of the workflow: gene-level counts and alignments (BAM files).</p>
<section id="salmon-quantitation" class="level3">
<h3 class="anchored" data-anchor-id="salmon-quantitation">Salmon quantitation</h3>
<section id="aggregated" class="level4">
<h4 class="anchored" data-anchor-id="aggregated">Aggregated</h4>
<p>The workflow outputs salmon quantitation results aggregated across all samples. Different types of counts (e.g.&nbsp;raw, length-scaled, TPMs) are available - the choice for downstream analyses depends on the chosen approach. Please see <a href="https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#Downstream_DGE_in_Bioconductor">tximport</a> for details.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/salmon_aggregated.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Salmon aggrated output</figcaption><p></p>
</figure>
</div>
</section>
<section id="by-sample" class="level4">
<h4 class="anchored" data-anchor-id="by-sample">By sample</h4>
<p>In addition, the salmon outputs for individual samples are avialable in sub-folders, one for each sample.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/salmon_by_sample.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Salmon output for each sample</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="star-alignments" class="level3">
<h3 class="anchored" data-anchor-id="star-alignments">STAR alignments</h3>
<p>The sorted (and, in the case of datasets that include UMIs, deduplicated) BAM files and their indices are available:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/star_bam.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">STAR alignments</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="genome-gene-annotations-indices" class="level2">
<h2 class="anchored" data-anchor-id="genome-gene-annotations-indices">Genome, gene annotations &amp; indices</h2>
<p>If the workflow was exectuted with the <code>"save_reference": true</code> parameter, then all reference files (FASTA, GTF, BED, etc) and the indices generated for STAR, salmon and rsem are returned in the <code>genome</code> folder within the output directory:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/save_reference.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">genome folder</figcaption><p></p>
</figure>
</div>
<p>These files can be reused for future runs, shortening the execuction time of the workflow.</p>
<p>Next, we will compare how the gene-counts returned of the <code>nf-core/rnaseq</code> workflow compare to those posted on GEO by the authors of the two datasets <a href="../nextflow-core-quantseq-1-settings/index.html">we processed in the first post in this series</a> by performing a differential expression analysis in the <a href="../nextflow-core-quantseq-3-xia/index.html">third</a> and <a href="../nextflow-core-quantseq-4-nugent/index.html">fourth</a> posts in this series.</p>


<!-- -->

</section>

<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb1" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "QuantSeq RNAseq analysis (2): Exploring nf-core/rnaseq output"</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2023-01-16"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [nextflow, NGS]</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="false"}</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>This is the second of four posts documenting my progress toward</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>processing and analyzing </span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">QuantSeq FWD 3' tag RNAseq</span><span class="co">](https://www.lexogen.com/quantseq-3mrna-sequencing/)</span> </span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>data </span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>with the </span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">nf-core/rnaseq</span><span class="co">](https://nf-co.re/rnaseq)</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>workflow. </span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="co">[</span><span class="ot">Configuring &amp; executing the nf-core/rnaseq workflow</span><span class="co">](../nextflow-core-quantseq-1-settings/index.html)</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Exploring the workflow outputs**</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="co">[</span><span class="ot">Validating the workflow by reproducing results published by Xia et al (no UMIs)</span><span class="co">](../nextflow-core-quantseq-3-xia/index.html)</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span><span class="co">[</span><span class="ot">Validating the workflow by reproducing results published by Nugent et al (including UMIs)</span><span class="co">](../nextflow-core-quantseq-4-nugent/index.html)</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>Many thanks to <span class="co">[</span><span class="ot">Harshil Patel</span><span class="co">](https://github.com/drpatelh)</span>, </span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">António Miguel de Jesus Domingues</span><span class="co">](https://github.com/adomingues/)</span></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>and </span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Matthias Zepper</span><span class="co">](https://github.com/matthiasZepper/)</span></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>for their generous guidance &amp; input via <span class="co">[</span><span class="ot">nf-core slack</span><span class="co">](nfcore.slack.com)</span>.</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>(Any mistakes are mine.)</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>This post documents the output files &amp; folders of the</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">nf-core/rnaseq workflow (v 3.10.1)</span><span class="co">](https://nf-co.re/rnaseq)</span>,</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>  run with default settings with the <span class="in">`star_salmon`</span> aligner / quantitation method.</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>For additional information, e.g. on the content of the</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">MultiQC</span><span class="co">](https://multiqc.info/)</span></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>  report, please see the</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">official nf-core/rnaseq documentation</span><span class="co">](https://nf-co.re/rnaseq/3.10.1/output)</span>.</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reports</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a><span class="fu">### MultiQC report</span></span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a>The MultiQC HTML report is a one-stop-shop that summarises QC metrics across the workflow. It can be found int he <span class="in">`multiqc`</span> folder, in a subdirectory named according to the aligner &amp; quantifier combination used (default: <span class="in">`star_salmon`</span>).</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a><span class="al">![multiqc report](images/multiqc.png)</span></span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a><span class="fu">### Pipeline info</span></span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a>The <span class="in">`pipeline_info`</span> folder contains html reports and text (CSV, TXT, YML) files with information about the run, including the versions of the software tools used. </span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a><span class="al">![pipeline info](images/pipeline_info.png)</span></span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a><span class="fu">### FastQC reports</span></span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a>The <span class="in">`fastqc`</span> folder contains the output of the</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">fastqc</span><span class="co">](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)</span></span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>tool. Most of the reported metrics are included in the MultiQC report as well, but the HTML reports for individual samples are available here if needed.</span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a><span class="al">![FastQC report](images/fastqc.png)</span></span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a><span class="fu">### Trim Galore reports</span></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a>The <span class="in">`trimgalore`</span> folder contains</span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>trimming reports for each sample</span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>the <span class="in">`fastqc`</span> sub-folder with quality metrics for the _trimmed_ FASTQ files</span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a><span class="al">![Trim Galore](images/trimgalore.png)</span></span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a><span class="fu">### umitools</span></span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>This folder contains the log files returned by</span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">UMI-tools</span><span class="co">](https://github.com/CGATOxford/UMI-tools)</span></span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a><span class="al">![UMI-tools](images/umitools.png)</span></span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a><span class="fu">## Workflow results</span></span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a>The main output of the workflow is available in the <span class="in">`star_salmon`</span> subdirectory. (This folder is named after the selected alignment &amp; quantification strategy, e.g. <span class="in">`star_salmon`</span> is present only if this tool combination was used.)</span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a>It contains multiple folders, as well as the (deduplicated) sorted BAM files for each sample.</span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a><span class="fu">### bigwig</span></span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a>Genome coverage in <span class="co">[</span><span class="ot">bigWig</span><span class="co">](https://genome.ucsc.edu/goldenPath/help/bigWig.html)</span> format for each sample.</span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a><span class="al">![BigWig files](images/bigwig.png)</span></span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a><span class="fu">### DESeq2 object &amp; QC metrics</span></span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a>The workflow aggregates all counts into a DESeq2 R objects, performs QC and exploratory analyses and serializes the object as <span class="in">`deseq2.dds.RData`</span>.</span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a><span class="al">![DESeq2 object &amp; QC metrics](images/deseq2.png)</span></span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a><span class="fu">### Dupradar</span></span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a>The </span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">dupRadar</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/dupRadar.html)</span></span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a>Bioconductor package performs duplication rate quality control.</span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a><span class="al">![dupRadar](images/dupradar.png)</span></span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a><span class="fu">### Featurecounts</span></span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a>Output from </span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">featurecounts</span><span class="co">](https://academic.oup.com/bioinformatics/article/30/7/923/232889?login=false)</span></span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a>tool is _only_ used to generate QC metrics. For actual quantitation of the gene-level results, the output of </span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">salmon</span><span class="co">](https://combine-lab.github.io/salmon/)</span> (default) or</span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">rsem</span><span class="co">](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323)</span></span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a>are used.</span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a>The metrics reported in the <span class="in">`featurecounts`</span> folder are included in the MultiQC report.</span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a><span class="al">![featureCounts](images/featurecounts.png)</span></span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a><span class="fu">### STAR log files</span></span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a>This folder contains log files output by the <span class="in">`STAR`</span> aligner.</span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a><span class="al">![STAR logs](images/log.png)</span></span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a><span class="fu">### Qualimap</span></span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">qualimap</span><span class="co">](http://qualimap.conesalab.org/)</span> package generates QC metrics from BAM files.</span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a><span class="al">![Qualimap](images/qualimap.png)</span></span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a><span class="fu">### RSEQC</span></span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a>The output of the <span class="co">[</span><span class="ot">rseqc</span><span class="co">](https://rseqc.sourceforge.net/)</span> QC control package are in this directory.</span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a><span class="al">![RSEQC](images/rseqc.png)</span></span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a><span class="fu">## Alignments &amp; gene-level counts</span></span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a>The <span class="in">`star_salmon`</span> folder also contains the main results of the workflow: gene-level counts and alignments (BAM files).</span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a><span class="fu">### Salmon quantitation</span></span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Aggregated</span></span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a>The workflow outputs salmon quantitation results aggregated across all samples. Different types of counts (e.g. raw, length-scaled, TPMs) are available - the choice for downstream analyses depends on the chosen approach. Please see <span class="co">[</span><span class="ot">tximport</span><span class="co">](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#Downstream_DGE_in_Bioconductor)</span> for details.</span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a><span class="al">![Salmon aggrated output](images/salmon_aggregated.png)</span></span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a><span class="fu">#### By sample</span></span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a>In addition, the salmon outputs for individual samples are avialable in sub-folders, one for each sample.</span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a><span class="al">![Salmon output for each sample](images/salmon_by_sample.png)</span></span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a><span class="fu">### STAR alignments</span></span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a>The sorted (and, in the case of datasets that include UMIs, deduplicated) BAM files and their indices are available:</span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a><span class="al">![STAR alignments](images/star_bam.png)</span></span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a><span class="fu">## Genome, gene annotations &amp; indices</span></span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a>If the workflow was exectuted with the <span class="in">`"save_reference": true`</span> parameter, then all reference files (FASTA, GTF, BED, etc) and the indices generated for STAR, salmon and rsem are returned in the <span class="in">`genome`</span> folder within the output directory:</span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a><span class="al">![genome folder](images/save_reference.png)</span></span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a>These files can be reused for future runs, shortening the execuction time of the workflow.</span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a>Next, we will compare how the gene-counts returned of the <span class="in">`nf-core/rnaseq`</span></span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a>workflow compare to those posted on GEO by the authors of the two datasets</span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">we processed in the first post in this series</span><span class="co">](../nextflow-core-quantseq-1-settings/index.html)</span> by</span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a>performing a differential expression analysis in the </span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">third</span><span class="co">](../nextflow-core-quantseq-3-xia/index.html)</span> and</span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">fourth</span><span class="co">](../nextflow-core-quantseq-4-nugent/index.html)</span></span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a>posts in this series.</span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>