<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Thomas Sandmann">
<meta name="dcterms.date" content="2024-10-12">

<title>Exploring the tidyHeatmap R package – Thomas Sandmann’s blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Thomas Sandmann’s blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tomsing1"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/thomas-sandmann/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@thomas_sandmann"> <i class="bi bi-mastodon" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Exploring the tidyHeatmap R package</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
            <p class="subtitle lead">Reproducing heatmaps published in Xia et al, Molecular neurodegeneration, 2022</p>
                                <div class="quarto-categories">
                <div class="quarto-category">R</div>
                <div class="quarto-category">TIL</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Thomas Sandmann </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">October 12, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr" id="toc-tldr" class="nav-link active" data-scroll-target="#tldr">tl;dr</a></li>
  <li><a href="#installing-the-tidyheatmap-r-package" id="toc-installing-the-tidyheatmap-r-package" class="nav-link" data-scroll-target="#installing-the-tidyheatmap-r-package">Installing the tidyHeatmap R package</a></li>
  <li><a href="#targeted-lipidomics-data-published-by-xia-et-al" id="toc-targeted-lipidomics-data-published-by-xia-et-al" class="nav-link" data-scroll-target="#targeted-lipidomics-data-published-by-xia-et-al">Targeted lipidomics data published by Xia et al</a></li>
  <li><a href="#available-datasets" id="toc-available-datasets" class="nav-link" data-scroll-target="#available-datasets">Available datasets</a></li>
  <li><a href="#a-first-heatmap" id="toc-a-first-heatmap" class="nav-link" data-scroll-target="#a-first-heatmap">A first heatmap</a></li>
  <li><a href="#adding-annotations" id="toc-adding-annotations" class="nav-link" data-scroll-target="#adding-annotations">Adding annotations</a></li>
  <li><a href="#adding-quantitative-annotations" id="toc-adding-quantitative-annotations" class="nav-link" data-scroll-target="#adding-quantitative-annotations">Adding quantitative annotations</a></li>
  <li><a href="#manually-ordering-rows" id="toc-manually-ordering-rows" class="nav-link" data-scroll-target="#manually-ordering-rows">Manually ordering rows</a></li>
  <li><a href="#manually-scaling-rows" id="toc-manually-scaling-rows" class="nav-link" data-scroll-target="#manually-scaling-rows">Manually scaling rows</a></li>
  <li><a href="#reproducibility" id="toc-reproducibility" class="nav-link" data-scroll-target="#reproducibility">Reproducibility</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">tl;dr</h2>
<p>This week I learned about the <code>tidyHeatmap</code> package, which wraps the functionality of the awesome <a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap R package</a> into functions with a tidyverse interface. No more need for pivoting tidy datasets into matrices and back!</p>
</section>
<section id="installing-the-tidyheatmap-r-package" class="level2">
<h2 class="anchored" data-anchor-id="installing-the-tidyheatmap-r-package">Installing the tidyHeatmap R package</h2>
<p>The <code>tidyHeatmap</code> R package is available from <a href="https://cran.r-project.org/package=tidyHeatmap">CRAN</a> <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<p>The latest release available from CRAN can be installed via:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"tidyHeatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and the latest development version (used in this tutorial) is available from github:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"stemangiola/tidyHeatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the time of writing, the latest version available from CRAN (version 1.8.1) is outdated; please install &amp; use the latest development version from github (currently version 1.10.2) to follow along with this tutorial.</p>
</div>
</div>
<p>In addition, we will use the following R packages, all of which can be installed from CRAN with <code>install.packages()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)  <span class="co"># to create continuous color gradients</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)  <span class="co"># wrangling tidy tables</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)  <span class="co"># to define witdth / height of annotations using grid::unit()</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ragg)  <span class="co"># to generate high quality graphics</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)  <span class="co"># to access predefined color palettes</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)  <span class="co"># to convert data.frames to tibbles</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyHeatmap) <span class="co"># the main act: plotting &amp; annotating heatmaps</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="targeted-lipidomics-data-published-by-xia-et-al" class="level2">
<h2 class="anchored" data-anchor-id="targeted-lipidomics-data-published-by-xia-et-al">Targeted lipidomics data published by Xia et al</h2>
<p>In this tutorial we will used targeted mass spectrometry data published by <a href="https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7">Xia et al, Molecular neurodegeneration, 2022</a> and reproduce Figures 5a and 5b<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_Fig5_HTML.png?as=webp" class="img-fluid figure-img"></p>
<figcaption>Figure 5 from Xia et al, 2022</figcaption>
</figure>
</div>
</section>
<section id="available-datasets" class="level2">
<h2 class="anchored" data-anchor-id="available-datasets">Available datasets</h2>
<ul>
<li>Normalized abundances: <a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_MOESM15_ESM.xlsx">Supplementary table 8</a> is an excel file containing the normalized abundances for 225 analytes measured across 30 samples.
<ul>
<li>The <code>get_abundances()</code> function defined in the callout box below starts by downloading the supplementary excel file, and then wrangles sample annotations, feature annotations and the measurements into a tidy tibble.</li>
<li>Alternatively, you can download a CSV file with the pre-wrangled dataset <a href="https://raw.githubusercontent.com/tomsing1/blog/refs/heads/main/posts/tidyHeatmap/abundances.csv">here</a>.</li>
</ul></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The get_abundances() function
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(checkmate)  <span class="co"># helper function for (optional) helper functions</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)  <span class="co"># reading excel files</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)  <span class="co"># pivoting from wide to long tables (and more!)</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>.parse_sample_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse sample- and feature-annotations, and normalized abundances</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"sample_annotations"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">strsplit</span>(df<span class="sc">$</span>sample_id, <span class="st">"_"</span>), \(x) x[[<span class="dv">2</span>]], <span class="at">FUN.VALUE =</span> <span class="fu">character</span>(<span class="dv">1</span>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  df[, <span class="fu">c</span>(<span class="st">"sample_id"</span>,  <span class="st">"description"</span>, <span class="st">"cell_number"</span>, <span class="st">"genotype"</span>, <span class="st">"condition"</span>,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>         <span class="st">"sex"</span>, <span class="st">"batch"</span>)]</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>.parse_feature_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"feature_annotations"</span>)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[, <span class="fu">grep</span>(<span class="st">"QTRAP|XEVO|m.z"</span>, <span class="fu">colnames</span>(df), <span class="at">invert =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[<span class="sc">!</span><span class="fu">duplicated</span>(df<span class="sc">$</span>component_name), ]</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>feature_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[<span class="sc">!</span>df<span class="sc">$</span>is_internal_standard,]</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>  df[, <span class="fu">setdiff</span>(</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(df), <span class="fu">c</span>(<span class="st">"is_internal_standard"</span>)),</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>    drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom tidyr pivot_longer</span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>.parse_abundances <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"peak_area_ratio_to_is"</span>)</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">component_name =</span> df<span class="sc">$</span>component_name,</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>    df[, <span class="fu">grep</span>(<span class="st">"^NE"</span>, <span class="fu">colnames</span>(df), <span class="at">value =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(df, <span class="at">cols =</span> <span class="fu">setdiff</span>(<span class="fu">colnames</span>(df), <span class="st">"component_name"</span>),</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>                            <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"abundance"</span>)</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">strsplit</span>(df<span class="sc">$</span>sample_id, <span class="st">"_"</span>), \(x) x[[<span class="dv">2</span>]], <span class="at">FUN.VALUE =</span> <span class="fu">character</span>(<span class="dv">1</span>)</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df)</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a><span class="co">#' Retrieve targeted LC/MS data published by Xia et al, 2022 as Suppl. Table 7</span></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom dplyr left_join join_by</span></span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="co">#' @export</span></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a><span class="co">#' @return A tidy tibble with 2,718 rows and 9 columns</span></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a><span class="co">#' @source [Xia et al, Molecular neurodegeneration, 2022](https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7)</span></span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a><span class="co">#' @examples</span></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a><span class="co">#' get_abundances()</span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>get_abundances <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>  url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/"</span>,</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MediaObjects/13024_2022_547_MOESM15_ESM.xlsx"</span>)</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>  <span class="co"># retrieve the first supplementary data file</span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>  tmp_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(url, tmp_file, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse sample- and feature-annotations &amp; normalized abundances</span></span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>  sample_anno <span class="ot">&lt;-</span> <span class="fu">.parse_sample_annotations</span>(tmp_file)</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>  feature_anno <span class="ot">&lt;-</span> <span class="fu">.parse_feature_annotations</span>(tmp_file)</span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>  abundances <span class="ot">&lt;-</span> <span class="fu">.parse_abundances</span>(tmp_file)</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join all tables into a single tidy tibble</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>    feature_anno, abundances, <span class="at">by =</span> dplyr<span class="sc">::</span><span class="fu">join_by</span>(<span class="st">"component_name"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(sample_anno, <span class="at">by =</span> dplyr<span class="sc">::</span><span class="fu">join_by</span>(<span class="st">"sample_id"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>      <span class="at">sample_id =</span> <span class="fu">factor</span>(sample_id, <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="fu">seq.int</span>(<span class="dv">30</span>))),</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>      <span class="at">sex =</span> <span class="fu">factor</span>(sex),</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>      <span class="at">condition =</span> <span class="fu">factor</span>(condition),</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>      <span class="at">ionization =</span> <span class="fu">factor</span>(ionization),</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>      <span class="at">batch =</span> <span class="fu">factor</span>(batch),</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>      <span class="at">genotype =</span> <span class="fu">factor</span>(genotype, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"APP_SAA_Hom"</span>)),</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="fu">factor</span>(<span class="fu">paste</span>(genotype, condition, <span class="at">sep =</span> <span class="st">", "</span>),</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>                     <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT, Methoxy_Neg"</span>, <span class="st">"APP_SAA_Hom, Methoxy_Neg"</span>,</span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>                                <span class="st">"APP_SAA_Hom, Methoxy_Pos"</span>)),</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> dplyr<span class="sc">::</span><span class="fu">recode_factor</span>(</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a>        group, </span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a>        <span class="st">"WT, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"WT, Methoxy-"</span>, </span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a>        <span class="st">"APP_SAA_Hom, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy-"</span>,</span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>        <span class="st">"APP_SAA_Hom, Methoxy_Pos"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy+"</span></span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<ul>
<li>Differential abundance analysis: <a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_MOESM16_ESM.xlsx">Supplementary table 8</a> is an excel file with the results of the differential abundance analysis (performed by fitting a linear model for each analyte with the <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma Bioconductor package</a> ), e.g.&nbsp;it provides log2 fold changes, p-values, and adjusted p-values (FDR) for multiple comparisons of interest.
<ul>
<li>The <code>get_stats()</code> function, shown in the callout box below, retrieves and parses this file for us and returns a tidy tibble.</li>
<li>Alternatively, you can download the pre-wrangled statistics in a CSV file <a href="https://raw.githubusercontent.com/tomsing1/blog/refs/heads/main/posts/tidyHeatmap/statistics.csv">here</a>.</li>
</ul></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The get_stats() function
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>#| message: false</p>
<p>library(readxl) # reading excel files</p>
<p>#’ Retrieve and parse differential abundance statistics from Suppl. Table 8 #’ <span class="citation" data-cites="noRd">@noRd</span> #’ <span class="citation" data-cites="examples">@examples</span> #’ get_stats() get_stats &lt;- function() { url &lt;- paste0( “https://static-content.springer.com/esm/art%3A10.1186%2Fs13024”, “-022-00547-7/MediaObjects/13024_2022_547_MOESM16_ESM.xlsx” ) tmp_file &lt;- tempfile(fileext = “.xlsx”) download.file(url, tmp_file, quiet = TRUE) readxl::read_excel(tmp_file, sheet = “3. APP_pos_vs_WT_neg”) |&gt; dplyr::rename(component_name = “symbol”) }</p>
<pre><code>:::

First, we download and read the two CSV files with the pre-wrangled abundance
and statistics data. For compatibility with the `tidyHeatmap` package,
we convert both tables into tibbles as well:

::: {.cell}

```{.r .cell-code}
df &lt;- read.csv(
  'https://raw.githubusercontent.com/tomsing1/tidy_heatmap/main/abundances.csv',
  stringsAsFactors=TRUE) |&gt;
  tibble::as_tibble()
stats &lt;- read.csv(
  'https://raw.githubusercontent.com/tomsing1/tidy_heatmap/main/statistics.csv',
  stringsAsFactors=TRUE) |&gt;
  tibble::as_tibble()
stopifnot(all(stats$component_name %in% df$component_name))</code></pre>
</div>
</div>
</div>
<p>Xia et al only included analytes with a log2 fold change &gt; 20% and an adjusted p-values &lt; 0.1 in Figure 5. Let’s identify the names of the significant analytes in our <code>stats</code> data.frame by applying these thresholds. Then we order the analytes by their log2 fold change and pull out the <code>component_name</code> column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> stats <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">abs</span>(stats<span class="sc">$</span>logFC) <span class="sc">&gt;</span> <span class="fl">0.2</span>, adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.1</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="fu">desc</span>(logFC)) <span class="sc">|&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(component_name)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(significant)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] GM3(d36:1)    TG(20:4_36:3) Spermine      TG(20:4_32:1) PS(16:0_20:4)
[6] TG(20:4_34:2)
225 Levels: (3-O-sulfo)GalCer(d18:1/16:0) ... Xanthine</code></pre>
</div>
</div>
<p>We retain 117 significant analytes, e.g.&nbsp;<code>GM3(d36:1)</code>, <code>TG(20:4_36:3)</code> and <code>Spermine</code> are most strongly increased in APP-SAA, methoxy+ compared to WT, methoxy- samples.</p>
<p>Next, we subset the abundance table (<code>df</code>) to only retain these significant analytes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">|&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(component_name <span class="sc">%in%</span> significant)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have the data in a tidy tibble with 3510 rows and 13 columns.</p>
<p>To control the order of categorical variables (e.g.&nbsp;<code>genotype</code> or or <code>batch</code>), they are encoded as factors.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. To match the order of the groups in Figure 5, we <code>recode</code> the levels of the <code>group</code> variable, so the <code>WT, Methoxy-</code> group is listed first:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>group <span class="ot">=</span> dplyr<span class="sc">::</span><span class="fu">recode_factor</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    df<span class="sc">$</span>group, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"WT, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"WT, Methoxy-"</span>, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"APP_SAA_Hom, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy-"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"APP_SAA_Hom, Methoxy_Pos"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy+"</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       component_name    ionization  
 (3-O-sulfo)GalCer(d18:1/24:1):  30   negative:1170  
 Acetylcarnitine              :  30   positive:2340  
 Adenine                      :  30                  
 Adenosine diphosphate ribose :  30                  
 alpha-Tocopherol             :  30                  
 Arginine                     :  30                  
 (Other)                      :3330                  
                    assigned_internal_standard        panel        sample_id   
 PC(15:0/18:1(d7))-Lipid-pos     : 330         Lipid     :2610   Min.   : 1.0  
 TG(15:0/18:1(d7)/15:0)-Lipid-pos: 300         Metabolite: 900   1st Qu.: 8.0  
 PE(15:0/18:1(d7))-Lipid-neg     : 240                           Median :15.5  
 PE(15:0/18:1(d7))-Lipid-pos     : 210                           Mean   :15.5  
 Arginine-d4C13-Metabolite-pos   : 150                           3rd Qu.:23.0  
 GlcCer(d18:1(d5)/18:0)-Lipid-pos: 150                           Max.   :30.0  
 (Other)                         :2130                                         
   abundance                  description    cell_number           genotype   
 Min.   :-12.6100   HOM, Methoxy NEG:1170   Min.   :19022   APP_SAA_Hom:2340  
 1st Qu.: -5.2800   HOM, Methoxy POS:1170   1st Qu.:47311   WT         :1170  
 Median : -2.1395   WT, Methoxy NEG :1170   Median :50000                     
 Mean   : -2.1045                           Mean   :45504                     
 3rd Qu.:  0.7407                           3rd Qu.:50000                     
 Max.   : 13.2500                           Max.   :50000                     
 NA's   :38                                                                   
       condition        sex         batch                    group     
 Methoxy_Neg:2340   Female:1287   Day_1: 702   WT, Methoxy-     :1170  
 Methoxy_Pos:1170   Male  :2223   Day_2:1404   APP-SAA, Methoxy-:1170  
                                  Day_3:1404   APP-SAA, Methoxy+:1170  
                                                                       
                                                                       
                                                                       
                                                                       </code></pre>
</div>
</div>
<p>We are ready to plot our first heatmap!</p>
</section>
<section id="a-first-heatmap" class="level2">
<h2 class="anchored" data-anchor-id="a-first-heatmap">A first heatmap</h2>
<p>The <code>tidyHeatmap</code> R package provides wrapper functions that make the functionality of the <a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap Bioconductor package</a> available within a tidy workflow. <a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
<p>Creating a first heatmap is as easy as passing our tidy tibble into the <code>tidyHeatmap::heatmap()</code> function, and specifying which columns contain row identifiers, column identifiers and values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(<span class="at">.data =</span> df, <span class="at">.row =</span> component_name, <span class="at">.column =</span> sample_id, </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">.value =</span> abundance, <span class="at">scale =</span> <span class="st">"row"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>The normalized abundances are hard to interpret (because analytes were normalized to different internal standards and don’t correspond to absolute concentrations). Luckily, we are mainly interested in more interested in the relative changes across samples (and groups of samples). By setting <code>scale = "row"</code>, the values for each analyte are converted to z-scores - and the color scale indicates variation around the mean of each row.</p>
</div>
</div>
</div>
<p>We can pass additional arguments through to the <code>ComplexHeatmap::Heatmap()</code> function, e.g.&nbsp;specify another method to cluster rows via the <code>clustering_method_rows</code> argument, by suppressing column clustering entirely via <code>cluster_columns=FALSE</code>, or by defining the color palette to use <a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>),</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># seq(2, -2, length.out = 11), </span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"navy"</span>, <span class="st">"white"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># RColorBrewer::brewer.pal(11, "RdBu")</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(<span class="at">.data =</span> df,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">.row =</span> component_name, </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">.column =</span> sample_id, </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">.value =</span> abundance, </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">palette_value =</span> colors</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<p>We can also introduce visual splits between the columns, e.g.&nbsp;separating the three genotypes from each other, and assign a different color to each one:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_columns =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Splitting features
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The dendrogram (on the left) indicates that there might be two major groups of features in this dataset, e.g.&nbsp;there are two main branches.</p>
<section id="cutting-the-hierarchical-clustering-result" class="level3">
<h3 class="anchored" data-anchor-id="cutting-the-hierarchical-clustering-result">Cutting the hierarchical clustering result</h3>
<p>We can reinforce this result by cutting the dendrogram after two branches have arisen.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_columns =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">split_rows</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="k-means-clustering" class="level3">
<h3 class="anchored" data-anchor-id="k-means-clustering">K-means clustering</h3>
<p>Alternatively, we can can perform <a href="https://en.wikipedia.org/wiki/K-means_clustering">k-means clustering</a> on the rows (or columns) by specifying the number of clusters (k) to isolate via the <code>row_km</code> argument. Our current dataset contains missing values (<code>NA</code>), which is incompatible with the k-means clustering algorithm. To use this method we first need to decide how to handle missing values, e.g.&nbsp;drop analytes or impute reasonable values.</p>
</section>
</div>
</div>
</div>
</section>
<section id="adding-annotations" class="level2">
<h2 class="anchored" data-anchor-id="adding-annotations">Adding annotations</h2>
<p>The <code>annotation_tile()</code> function<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> allows us to add additional annotations to our heatmap, e.g.&nbsp;to identify</p>
<ul>
<li>columns from the <code>batches</code> or correspoding to animals with male or female <code>sex</code> in an annotation bar on top of the heatmap, or</li>
<li>rows with values from different <code>panels</code> (lipds, metabolits) in an annotation strip on the left of the heatmap.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_columns =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotation_tile</span>(</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>        batch, </span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">palette =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Dark2"</span>)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotation_tile</span>(sex, </span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>             <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"pink"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotation_tile</span>(panel, <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">"points"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Row or column annotations?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <code>annotation_tile()</code> function does not require us to specify whether we are adding row or column annotations. Instead, it infers it automatically:</p>
<ul>
<li>Does each value used to define a row of the heatmap (e.g.&nbsp;<code>component_name</code>) map to a single value in the selected column of the data.frame?
<ul>
<li>If yes, create a row annotation strip.</li>
</ul></li>
<li>Does each value used to define a column of the heatmap (e.g.&nbsp;<code>sample_id</code>) map to a single value in the selected column of the data.frame?
<ul>
<li>If yes, create a column annotation.</li>
</ul></li>
<li>If neither of the above are true, an error is raised.</li>
</ul>
</div>
</div>
</div>
</section>
<section id="adding-quantitative-annotations" class="level2">
<h2 class="anchored" data-anchor-id="adding-quantitative-annotations">Adding quantitative annotations</h2>
<p>Some annotations are quantitative, and may be better communicated in a plot. Let’s display the number of cells that were analyzed in each sample on top of our heatmap with the <code>add_point()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_columns =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>    batch, </span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Dark2"</span>)</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>    sex, </span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"pink"</span>)</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(panel, <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">"points"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_point</span>(cell_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Overlaying symbols
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To highlight specific cells in our heatmap, the <code>layer_point()</code>, <code>layer_arrow_up()</code>, <code>layer_arrow_down()</code>, <code>layer_diamond()</code>, <code>layer_square()</code>, <code>layer_star()</code>, <code>layer_text()</code> or <code>layer_asterisk()</code> functions can be used, together with a filter condition <a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>.</p>
<p>Can you spot the asterisks for <code>GM3(d36:1)</code> for samples 1, 3, and 6 in the heatmap below?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_rows =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">clustering_method_columns =</span> <span class="st">"ward.D"</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>    batch, </span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Dark2"</span>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(sex, </span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>                  <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"pink"</span>))  <span class="sc">|&gt;</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(panel, <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">"points"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">layer_asterisk</span>(</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    component_name <span class="sc">==</span> <span class="st">"GM3(d36:1)"</span> <span class="sc">&amp;</span> sample_id <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"1"</span>, <span class="st">"3"</span>, <span class="st">"6"</span>)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="manually-ordering-rows" class="level2">
<h2 class="anchored" data-anchor-id="manually-ordering-rows">Manually ordering rows</h2>
<p>For Figure 5, Xia et al manually ordered the rows, with those analytes showing the largest log2 fold increases at the top, and those with the largest decreases (e.g.&nbsp;negative log2 fold changes) at the bottom.</p>
<p>To reproduce this, we switch off the clustering of the rows, and change the order of the <code>component_name</code> factor to the <code>significant</code> character vector - which is ordered from largest to smallest fold change (see above) - with the <code>dplyr::mutate()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">component_name =</span> <span class="fu">factor</span>(component_name, <span class="at">levels =</span> significant)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group, panel) <span class="sc">|&gt;</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name, </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">palette.colors</span>(<span class="dv">2</span>, <span class="st">"Set1"</span>),</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    batch, </span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Dark2"</span>),</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"points"</span>)</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>    sex, </span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"pink"</span>),</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"points"</span>)</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<p>In this step, we also changed the <code>group_by</code> command: we grouped by both the <code>group</code> and the <code>panel</code> annotations, splitting columns and rows of our heatmap into separate panels, respectively.</p>
</section>
<section id="manually-scaling-rows" class="level2">
<h2 class="anchored" data-anchor-id="manually-scaling-rows">Manually scaling rows</h2>
<p>So far, we have scaled each row by converting the measurements for each analyte to <a href="https://en.wikipedia.org/wiki/Standard_score">z-scores</a>, and the colors reflect differences from the overall mean across all samples.</p>
<p>In many experiments, including this one, there is an explicit experimental group that provides a natural baseline. The effects of the <code>APP-SAA</code> genotype is usually interpret relative to the <code>WT</code> control group. We therefore center the abundances by subtracting the median value observed in this (e.g.&nbsp;the <code>WT, methoxy-</code>) group. Now the colors indicate log2 fold changes relative to the median in the control group.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Here, we use functions from the <code>dplyr</code> package to</p>
<ol type="1">
<li><code>group_by</code>: group the tibble by <code>component_name</code>, so we can perform calculations separately for each analyte,</li>
<li><code>mutate</code>: calculate the median for the abundances observed in the <code>WT, Methoxy-</code> group (for this analyte)</li>
<li><code>mutate</code>: subtract the control median from each value (for this analyte).</li>
<li><code>ungroup</code>: ungroup the tibble, so we can continue with the full dataset.</li>
</ol>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">component_name =</span> <span class="fu">factor</span>(component_name, <span class="at">levels =</span> significant)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(component_name) <span class="sc">|&gt;</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ctrl_median =</span> <span class="fu">median</span>(abundance[group <span class="sc">==</span> <span class="st">"WT, Methoxy-"</span>], <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">abundance =</span> abundance <span class="sc">-</span> ctrl_median</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">|&gt;</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group, panel) <span class="sc">|&gt;</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(<span class="at">.row =</span> component_name,</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">.column =</span> sample_id, </span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>          <span class="at">.value =</span> abundance, </span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>          <span class="at">column_title =</span> <span class="st">"Xia et al: targeted LC/MS data"</span>,</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_title =</span> <span class="st">"Analyte"</span>,</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>          <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>          <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>          <span class="at">row_names_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">6</span>),</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette_value =</span> colors,</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>           <span class="at">palette_grouping =</span> <span class="fu">list</span>(</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>            <span class="fu">palette.colors</span>(<span class="dv">2</span>, <span class="st">"Set1"</span>),</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="st">"darkgrey"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick"</span>)</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>    batch, </span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Dark2"</span>),</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"points"</span>)</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_tile</span>(</span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>    sex, </span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"pink"</span>),</span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"points"</span>)</span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="720"></p>
</figure>
</div>
</div>
</div>
<p>We also added the <code>row_names_gp = grid::gpar(fontsize = 6)</code> argument to the <code>heatmap()</code> call, increasing the font size of the row labels. (They are still a bit on the small side, so you might want to consider labelling only a subset.)</p>
</section>
<section id="reproducibility" class="level2">
<h2 class="anchored" data-anchor-id="reproducibility">Reproducibility</h2>
<details>
<summary>
Session information
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 12 (bookworm)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/Los_Angeles
tzcode source: system (glibc)

attached base packages:
[1] grid      stats     graphics  grDevices datasets  utils     methods  
[8] base     

other attached packages:
[1] tidyr_1.3.1        readxl_1.4.3       checkmate_2.3.2    tidyHeatmap_1.10.2
[5] tibble_3.2.1       RColorBrewer_1.1-3 ragg_1.3.3         dplyr_1.1.4       
[9] circlize_0.4.16   

loaded via a namespace (and not attached):
 [1] gtable_0.3.5          shape_1.4.6.1         rjson_0.2.23         
 [4] xfun_0.37             ggplot2_3.5.1         htmlwidgets_1.6.1    
 [7] GlobalOptions_0.1.2   bspm_0.5.5            vctrs_0.6.5          
[10] tools_4.3.2           generics_0.1.3        stats4_4.3.2         
[13] parallel_4.3.2        fansi_1.0.6           cluster_2.1.6        
[16] pkgconfig_2.0.3       S4Vectors_0.40.2      lifecycle_1.0.3      
[19] compiler_4.3.2        farver_2.1.2          textshaping_0.4.0    
[22] munsell_0.5.1         codetools_0.2-19      ComplexHeatmap_2.18.0
[25] clue_0.3-65           htmltools_0.5.4       yaml_2.3.7           
[28] pillar_1.9.0          crayon_1.5.3          iterators_1.0.14     
[31] viridis_0.6.5         foreach_1.5.2         tidyselect_1.2.1     
[34] digest_0.6.31         purrr_1.0.2           fastmap_1.1.1        
[37] colorspace_2.1-1      cli_3.6.3             magrittr_2.0.3       
[40] patchwork_1.3.0       utf8_1.2.4            withr_3.0.1          
[43] backports_1.5.0       scales_1.3.0          rmarkdown_2.20       
[46] matrixStats_1.4.1     gridExtra_2.3         cellranger_1.1.0     
[49] png_0.1-8             GetoptLong_1.0.5      evaluate_0.20        
[52] knitr_1.42            IRanges_2.36.0        doParallel_1.0.17    
[55] viridisLite_0.4.2     rlang_1.1.4           dendextend_1.18.0    
[58] glue_1.6.2            BiocManager_1.30.25   renv_1.0.11          
[61] BiocGenerics_0.48.1   rstudioapi_0.14       jsonlite_1.8.4       
[64] R6_2.5.1              systemfonts_1.1.0    </code></pre>
</div>
</div>
</details>


<!-- -->

</section>


<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png"></a><br>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Because <code>tidyHeatmap</code> relies heavily on other packages from the <a href="https://www.tidyverse.org/">tidyverse</a>, e.g.&nbsp;the <a href="https://dplyr.tidyverse.org//">dplyr</a> and <a href="https://tidyr.tidyverse.org//">tidyr</a> R packages, it has a relatively deep dependency tree. But that’s par for the course if you want to adopt the <a href="https://tidyverse.tidyverse.org/articles/manifesto.html">tidy tools API</a> by using Posit’s toolchain.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>Well approximately: <a href="https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7/figures/5">In the publication</a> the authors estimated and subtracted effects due to <code>batch</code>, <code>sex</code> and two <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1994707/">surrogate variables</a> with a linear model, and then plotted the <a href="https://en.wikipedia.org/wiki/Errors_and_residuals">residuals</a> for each data point. This adjustment was motivated by the experimental design and the statistical analysis, which also incorporated these covariates to determine whether differences between the experimental groups were statistically significant. Here, we will plot the normalized abundances for each analyte directly, without any adjustment. Nevertheless, the heatmap will look quite similar (albeit not quite as <em>smooth</em> as the original).<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>For example, the two possible values of <code>genotype</code> are: <code>WT</code> and <code>APP_SAA_Hom</code>, and the three levels of the <code>group</code> variable are <code>WT, Methoxy-</code>, <code>APP-SAA, Methoxy-</code> and <code>APP-SAA, Methoxy+</code>.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>The documentation of the <code>ComplexHeatmap</code> package is amazing: the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">Complete Reference</a> contains to 15 chapters.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>See <a href="https://www.rdocumentation.org/packages/ComplexHeatmap/versions/1.10.2/topics/Heatmap">Heatmap’s help page</a> for a full list of arguments.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>Prior to <code>tidyHeatmap</code> version 1.9.0, the <code>add_tile()</code> function was used instead. It is now deprecated in favor of <code>annotation_tile()</code>. (At the time of writing, the latest version available from CRAN is version 1.8.1; please install &amp; use the latest development version from github to follow along this tutorial.)<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p>At the time of writing, the latest version of the <code>tidyHeatmap</code> package available from CRAN is version 1.8.1. The development version used here, available <a href="https://github.com/stemangiola/tidyHeatmap">from github</a>, contains additional functionality, including e.g.&nbsp;the <code>layer_star()</code> and - most importantly - <code>layer_asterisk()</code>.<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/tomsing1\.github\.io\/blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="tomsing1/blog" data-repo-id="R_kgDOIZ9LJw" data-category="Announcements" data-category-id="DIC_kwDOIZ9LJ84CTvnC" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark"><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb23" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Exploring the tidyHeatmap R package"</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Reproducing heatmaps published in Xia et al, Molecular neurodegeneration, 2022"</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Thomas Sandmann"</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2024-10-12"</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [R, TIL]</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown:</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools:</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co">      source: true</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co">      toggle: false</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="co">      caption: none</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a><span class="fu">## tl;dr</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>This week I learned about the <span class="in">`tidyHeatmap`</span> package, which wraps the</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>functionality of the awesome </span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">ComplexHeatmap R package</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html)</span></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>into functions with a tidyverse interface. No more need for pivoting tidy </span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>datasets into matrices and back!</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">use</span>(<span class="at">lockfile =</span> <span class="fu">file.path</span>(<span class="fu">path.expand</span>(<span class="st">"~"</span>), <span class="st">"repositories"</span>, <span class="st">"blog"</span>,</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"posts"</span>, <span class="st">"tidyHeatmap"</span>, <span class="st">"renv.lock"</span>))</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a><span class="fu">## Installing the tidyHeatmap R package</span></span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a>The <span class="in">`tidyHeatmap`</span> R package is available from</span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">CRAN</span><span class="co">](https://cran.r-project.org/package=tidyHeatmap)</span> <span class="ot">[^1]</span>.</span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>Because <span class="in">`tidyHeatmap`</span> relies heavily on other packages from the</span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">tidyverse</span><span class="co">](https://www.tidyverse.org/)</span>,</span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a>e.g. the </span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">dplyr</span><span class="co">](https://dplyr.tidyverse.org//)</span></span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a>and </span>
<span id="cb23-49"><a href="#cb23-49" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">tidyr</span><span class="co">](https://tidyr.tidyverse.org//)</span></span>
<span id="cb23-50"><a href="#cb23-50" aria-hidden="true" tabindex="-1"></a>R packages, it has a relatively deep dependency tree. But that's </span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a>par for the course if you want to adopt the </span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">tidy tools API</span><span class="co">](https://tidyverse.tidyverse.org/articles/manifesto.html)</span> </span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a>by using Posit's toolchain.</span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-55"><a href="#cb23-55" aria-hidden="true" tabindex="-1"></a>The latest release available from CRAN can be installed via: </span>
<span id="cb23-56"><a href="#cb23-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-59"><a href="#cb23-59" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-60"><a href="#cb23-60" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb23-61"><a href="#cb23-61" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"tidyHeatmap"</span>)</span>
<span id="cb23-62"><a href="#cb23-62" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-63"><a href="#cb23-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-64"><a href="#cb23-64" aria-hidden="true" tabindex="-1"></a>and the latest development version (used in this tutorial) is available</span>
<span id="cb23-65"><a href="#cb23-65" aria-hidden="true" tabindex="-1"></a>from github:</span>
<span id="cb23-66"><a href="#cb23-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-69"><a href="#cb23-69" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-70"><a href="#cb23-70" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb23-71"><a href="#cb23-71" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"stemangiola/tidyHeatmap"</span>)</span>
<span id="cb23-72"><a href="#cb23-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-73"><a href="#cb23-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-74"><a href="#cb23-74" aria-hidden="true" tabindex="-1"></a>::: { .callout-note }</span>
<span id="cb23-75"><a href="#cb23-75" aria-hidden="true" tabindex="-1"></a> At the time of writing, the latest version available from CRAN (version </span>
<span id="cb23-76"><a href="#cb23-76" aria-hidden="true" tabindex="-1"></a>1.8.1) is outdated; please install &amp; use the latest development version from</span>
<span id="cb23-77"><a href="#cb23-77" aria-hidden="true" tabindex="-1"></a>github (currently version 1.10.2) to follow along with this tutorial.</span>
<span id="cb23-78"><a href="#cb23-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-79"><a href="#cb23-79" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb23-80"><a href="#cb23-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-81"><a href="#cb23-81" aria-hidden="true" tabindex="-1"></a>In addition, we will use the following R packages, all of which can be installed</span>
<span id="cb23-82"><a href="#cb23-82" aria-hidden="true" tabindex="-1"></a>from CRAN with <span class="in">`install.packages()`</span>:</span>
<span id="cb23-83"><a href="#cb23-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-86"><a href="#cb23-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-87"><a href="#cb23-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb23-88"><a href="#cb23-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-89"><a href="#cb23-89" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)  <span class="co"># to create continuous color gradients</span></span>
<span id="cb23-90"><a href="#cb23-90" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)  <span class="co"># wrangling tidy tables</span></span>
<span id="cb23-91"><a href="#cb23-91" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)  <span class="co"># to define witdth / height of annotations using grid::unit()</span></span>
<span id="cb23-92"><a href="#cb23-92" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ragg)  <span class="co"># to generate high quality graphics</span></span>
<span id="cb23-93"><a href="#cb23-93" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)  <span class="co"># to access predefined color palettes</span></span>
<span id="cb23-94"><a href="#cb23-94" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)  <span class="co"># to convert data.frames to tibbles</span></span>
<span id="cb23-95"><a href="#cb23-95" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyHeatmap) <span class="co"># the main act: plotting &amp; annotating heatmaps</span></span>
<span id="cb23-96"><a href="#cb23-96" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-97"><a href="#cb23-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-98"><a href="#cb23-98" aria-hidden="true" tabindex="-1"></a><span class="fu">## Targeted lipidomics data published by Xia et al</span></span>
<span id="cb23-99"><a href="#cb23-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-100"><a href="#cb23-100" aria-hidden="true" tabindex="-1"></a>In this tutorial we will used targeted mass spectrometry data </span>
<span id="cb23-101"><a href="#cb23-101" aria-hidden="true" tabindex="-1"></a>published by </span>
<span id="cb23-102"><a href="#cb23-102" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Xia et al, Molecular neurodegeneration, 2022</span><span class="co">](https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7)</span></span>
<span id="cb23-103"><a href="#cb23-103" aria-hidden="true" tabindex="-1"></a>and reproduce </span>
<span id="cb23-104"><a href="#cb23-104" aria-hidden="true" tabindex="-1"></a>Figures 5a and 5b<span class="ot">[^2]</span>:</span>
<span id="cb23-105"><a href="#cb23-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-106"><a href="#cb23-106" aria-hidden="true" tabindex="-1"></a><span class="al">![Figure 5 from Xia et al, 2022](https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_Fig5_HTML.png?as=webp)</span></span>
<span id="cb23-107"><a href="#cb23-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-108"><a href="#cb23-108" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span>Well approximately: </span>
<span id="cb23-109"><a href="#cb23-109" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">In the publication</span><span class="co">](https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7/figures/5)</span> the authors estimated and</span>
<span id="cb23-110"><a href="#cb23-110" aria-hidden="true" tabindex="-1"></a>subtracted effects due to <span class="in">`batch`</span>, <span class="in">`sex`</span> and two</span>
<span id="cb23-111"><a href="#cb23-111" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">surrogate variables</span><span class="co">](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1994707/)</span></span>
<span id="cb23-112"><a href="#cb23-112" aria-hidden="true" tabindex="-1"></a>with a linear model, and then plotted the </span>
<span id="cb23-113"><a href="#cb23-113" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">residuals</span><span class="co">](https://en.wikipedia.org/wiki/Errors_and_residuals)</span></span>
<span id="cb23-114"><a href="#cb23-114" aria-hidden="true" tabindex="-1"></a>for each data point. This adjustment was motivated by the</span>
<span id="cb23-115"><a href="#cb23-115" aria-hidden="true" tabindex="-1"></a>experimental design and the statistical analysis, which also incorporated</span>
<span id="cb23-116"><a href="#cb23-116" aria-hidden="true" tabindex="-1"></a>these covariates to determine whether differences between the experimental</span>
<span id="cb23-117"><a href="#cb23-117" aria-hidden="true" tabindex="-1"></a>groups were statistically significant. Here, we will plot the normalized</span>
<span id="cb23-118"><a href="#cb23-118" aria-hidden="true" tabindex="-1"></a>abundances for each analyte directly, without any adjustment. Nevertheless,</span>
<span id="cb23-119"><a href="#cb23-119" aria-hidden="true" tabindex="-1"></a>the heatmap will look quite similar (albeit not quite as _smooth_ as the </span>
<span id="cb23-120"><a href="#cb23-120" aria-hidden="true" tabindex="-1"></a>original).</span>
<span id="cb23-121"><a href="#cb23-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-122"><a href="#cb23-122" aria-hidden="true" tabindex="-1"></a><span class="fu">## Available datasets</span></span>
<span id="cb23-123"><a href="#cb23-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-124"><a href="#cb23-124" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Normalized abundances: <span class="co">[</span><span class="ot">Supplementary table 8</span><span class="co">](https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_MOESM15_ESM.xlsx)</span></span>
<span id="cb23-125"><a href="#cb23-125" aria-hidden="true" tabindex="-1"></a>is an excel file containing the normalized abundances for</span>
<span id="cb23-126"><a href="#cb23-126" aria-hidden="true" tabindex="-1"></a>225 analytes measured across 30 samples.</span>
<span id="cb23-127"><a href="#cb23-127" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>The <span class="in">`get_abundances()`</span> function defined in the callout box below starts by</span>
<span id="cb23-128"><a href="#cb23-128" aria-hidden="true" tabindex="-1"></a>  downloading the supplementary excel file, and then wrangles sample</span>
<span id="cb23-129"><a href="#cb23-129" aria-hidden="true" tabindex="-1"></a>  annotations, feature annotations and the measurements into a tidy tibble.</span>
<span id="cb23-130"><a href="#cb23-130" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Alternatively, you can download a CSV file with the pre-wrangled dataset</span>
<span id="cb23-131"><a href="#cb23-131" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">here</span><span class="co">](https://raw.githubusercontent.com/tomsing1/blog/refs/heads/main/posts/tidyHeatmap/abundances.csv)</span>.</span>
<span id="cb23-132"><a href="#cb23-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-133"><a href="#cb23-133" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip  collapse="true"}</span>
<span id="cb23-134"><a href="#cb23-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-135"><a href="#cb23-135" aria-hidden="true" tabindex="-1"></a><span class="fu">#### The get_abundances() function</span></span>
<span id="cb23-136"><a href="#cb23-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-139"><a href="#cb23-139" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-140"><a href="#cb23-140" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb23-141"><a href="#cb23-141" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(checkmate)  <span class="co"># helper function for (optional) helper functions</span></span>
<span id="cb23-142"><a href="#cb23-142" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)  <span class="co"># reading excel files</span></span>
<span id="cb23-143"><a href="#cb23-143" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)  <span class="co"># pivoting from wide to long tables (and more!)</span></span>
<span id="cb23-144"><a href="#cb23-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-145"><a href="#cb23-145" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb23-146"><a href="#cb23-146" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb23-147"><a href="#cb23-147" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb23-148"><a href="#cb23-148" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb23-149"><a href="#cb23-149" aria-hidden="true" tabindex="-1"></a>.parse_sample_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb23-150"><a href="#cb23-150" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb23-151"><a href="#cb23-151" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse sample- and feature-annotations, and normalized abundances</span></span>
<span id="cb23-152"><a href="#cb23-152" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"sample_annotations"</span>)</span>
<span id="cb23-153"><a href="#cb23-153" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb23-154"><a href="#cb23-154" aria-hidden="true" tabindex="-1"></a>    <span class="fu">strsplit</span>(df<span class="sc">$</span>sample_id, <span class="st">"_"</span>), \(x) x[[<span class="dv">2</span>]], <span class="at">FUN.VALUE =</span> <span class="fu">character</span>(<span class="dv">1</span>)</span>
<span id="cb23-155"><a href="#cb23-155" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb23-156"><a href="#cb23-156" aria-hidden="true" tabindex="-1"></a>  df[, <span class="fu">c</span>(<span class="st">"sample_id"</span>,  <span class="st">"description"</span>, <span class="st">"cell_number"</span>, <span class="st">"genotype"</span>, <span class="st">"condition"</span>,</span>
<span id="cb23-157"><a href="#cb23-157" aria-hidden="true" tabindex="-1"></a>         <span class="st">"sex"</span>, <span class="st">"batch"</span>)]</span>
<span id="cb23-158"><a href="#cb23-158" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-159"><a href="#cb23-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-160"><a href="#cb23-160" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb23-161"><a href="#cb23-161" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb23-162"><a href="#cb23-162" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb23-163"><a href="#cb23-163" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb23-164"><a href="#cb23-164" aria-hidden="true" tabindex="-1"></a>.parse_feature_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb23-165"><a href="#cb23-165" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb23-166"><a href="#cb23-166" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"feature_annotations"</span>)</span>
<span id="cb23-167"><a href="#cb23-167" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[, <span class="fu">grep</span>(<span class="st">"QTRAP|XEVO|m.z"</span>, <span class="fu">colnames</span>(df), <span class="at">invert =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb23-168"><a href="#cb23-168" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[<span class="sc">!</span><span class="fu">duplicated</span>(df<span class="sc">$</span>component_name), ]</span>
<span id="cb23-169"><a href="#cb23-169" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>feature_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb23-170"><a href="#cb23-170" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> df[<span class="sc">!</span>df<span class="sc">$</span>is_internal_standard,]</span>
<span id="cb23-171"><a href="#cb23-171" aria-hidden="true" tabindex="-1"></a>  df[, <span class="fu">setdiff</span>(</span>
<span id="cb23-172"><a href="#cb23-172" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(df), <span class="fu">c</span>(<span class="st">"is_internal_standard"</span>)),</span>
<span id="cb23-173"><a href="#cb23-173" aria-hidden="true" tabindex="-1"></a>    drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb23-174"><a href="#cb23-174" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-175"><a href="#cb23-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-176"><a href="#cb23-176" aria-hidden="true" tabindex="-1"></a><span class="co">#' @noRd</span></span>
<span id="cb23-177"><a href="#cb23-177" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param path Scalar character, the path to the supplementary excel file</span></span>
<span id="cb23-178"><a href="#cb23-178" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom checkmate assert_file</span></span>
<span id="cb23-179"><a href="#cb23-179" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom tidyr pivot_longer</span></span>
<span id="cb23-180"><a href="#cb23-180" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom readxl read_excel</span></span>
<span id="cb23-181"><a href="#cb23-181" aria-hidden="true" tabindex="-1"></a>.parse_abundances <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb23-182"><a href="#cb23-182" aria-hidden="true" tabindex="-1"></a>  checkmate<span class="sc">::</span><span class="fu">assert_file</span>(path)</span>
<span id="cb23-183"><a href="#cb23-183" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(path, <span class="at">sheet =</span> <span class="st">"peak_area_ratio_to_is"</span>)</span>
<span id="cb23-184"><a href="#cb23-184" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb23-185"><a href="#cb23-185" aria-hidden="true" tabindex="-1"></a>    <span class="at">component_name =</span> df<span class="sc">$</span>component_name,</span>
<span id="cb23-186"><a href="#cb23-186" aria-hidden="true" tabindex="-1"></a>    df[, <span class="fu">grep</span>(<span class="st">"^NE"</span>, <span class="fu">colnames</span>(df), <span class="at">value =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb23-187"><a href="#cb23-187" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb23-188"><a href="#cb23-188" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(df, <span class="at">cols =</span> <span class="fu">setdiff</span>(<span class="fu">colnames</span>(df), <span class="st">"component_name"</span>),</span>
<span id="cb23-189"><a href="#cb23-189" aria-hidden="true" tabindex="-1"></a>                            <span class="at">names_to =</span> <span class="st">"sample_id"</span>, <span class="at">values_to =</span> <span class="st">"abundance"</span>)</span>
<span id="cb23-190"><a href="#cb23-190" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb23-191"><a href="#cb23-191" aria-hidden="true" tabindex="-1"></a>    <span class="fu">strsplit</span>(df<span class="sc">$</span>sample_id, <span class="st">"_"</span>), \(x) x[[<span class="dv">2</span>]], <span class="at">FUN.VALUE =</span> <span class="fu">character</span>(<span class="dv">1</span>)</span>
<span id="cb23-192"><a href="#cb23-192" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb23-193"><a href="#cb23-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df)</span>
<span id="cb23-194"><a href="#cb23-194" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-195"><a href="#cb23-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-196"><a href="#cb23-196" aria-hidden="true" tabindex="-1"></a><span class="co">#' Retrieve targeted LC/MS data published by Xia et al, 2022 as Suppl. Table 7</span></span>
<span id="cb23-197"><a href="#cb23-197" aria-hidden="true" tabindex="-1"></a><span class="co">#' @importFrom dplyr left_join join_by</span></span>
<span id="cb23-198"><a href="#cb23-198" aria-hidden="true" tabindex="-1"></a><span class="co">#' @export</span></span>
<span id="cb23-199"><a href="#cb23-199" aria-hidden="true" tabindex="-1"></a><span class="co">#' @return A tidy tibble with 2,718 rows and 9 columns</span></span>
<span id="cb23-200"><a href="#cb23-200" aria-hidden="true" tabindex="-1"></a><span class="co">#' @source [Xia et al, Molecular neurodegeneration, 2022](https://molecularneurodegeneration.biomedcentral.com/articles/10.1186/s13024-022-00547-7)</span></span>
<span id="cb23-201"><a href="#cb23-201" aria-hidden="true" tabindex="-1"></a><span class="co">#' @examples</span></span>
<span id="cb23-202"><a href="#cb23-202" aria-hidden="true" tabindex="-1"></a><span class="co">#' get_abundances()</span></span>
<span id="cb23-203"><a href="#cb23-203" aria-hidden="true" tabindex="-1"></a>get_abundances <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb23-204"><a href="#cb23-204" aria-hidden="true" tabindex="-1"></a>  url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb23-205"><a href="#cb23-205" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/"</span>,</span>
<span id="cb23-206"><a href="#cb23-206" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MediaObjects/13024_2022_547_MOESM15_ESM.xlsx"</span>)</span>
<span id="cb23-207"><a href="#cb23-207" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-208"><a href="#cb23-208" aria-hidden="true" tabindex="-1"></a>  <span class="co"># retrieve the first supplementary data file</span></span>
<span id="cb23-209"><a href="#cb23-209" aria-hidden="true" tabindex="-1"></a>  tmp_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb23-210"><a href="#cb23-210" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(url, tmp_file, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-211"><a href="#cb23-211" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse sample- and feature-annotations &amp; normalized abundances</span></span>
<span id="cb23-212"><a href="#cb23-212" aria-hidden="true" tabindex="-1"></a>  sample_anno <span class="ot">&lt;-</span> <span class="fu">.parse_sample_annotations</span>(tmp_file)</span>
<span id="cb23-213"><a href="#cb23-213" aria-hidden="true" tabindex="-1"></a>  feature_anno <span class="ot">&lt;-</span> <span class="fu">.parse_feature_annotations</span>(tmp_file)</span>
<span id="cb23-214"><a href="#cb23-214" aria-hidden="true" tabindex="-1"></a>  abundances <span class="ot">&lt;-</span> <span class="fu">.parse_abundances</span>(tmp_file)</span>
<span id="cb23-215"><a href="#cb23-215" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join all tables into a single tidy tibble</span></span>
<span id="cb23-216"><a href="#cb23-216" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb23-217"><a href="#cb23-217" aria-hidden="true" tabindex="-1"></a>    feature_anno, abundances, <span class="at">by =</span> dplyr<span class="sc">::</span><span class="fu">join_by</span>(<span class="st">"component_name"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb23-218"><a href="#cb23-218" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(sample_anno, <span class="at">by =</span> dplyr<span class="sc">::</span><span class="fu">join_by</span>(<span class="st">"sample_id"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb23-219"><a href="#cb23-219" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb23-220"><a href="#cb23-220" aria-hidden="true" tabindex="-1"></a>      <span class="at">sample_id =</span> <span class="fu">factor</span>(sample_id, <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="fu">seq.int</span>(<span class="dv">30</span>))),</span>
<span id="cb23-221"><a href="#cb23-221" aria-hidden="true" tabindex="-1"></a>      <span class="at">sex =</span> <span class="fu">factor</span>(sex),</span>
<span id="cb23-222"><a href="#cb23-222" aria-hidden="true" tabindex="-1"></a>      <span class="at">condition =</span> <span class="fu">factor</span>(condition),</span>
<span id="cb23-223"><a href="#cb23-223" aria-hidden="true" tabindex="-1"></a>      <span class="at">ionization =</span> <span class="fu">factor</span>(ionization),</span>
<span id="cb23-224"><a href="#cb23-224" aria-hidden="true" tabindex="-1"></a>      <span class="at">batch =</span> <span class="fu">factor</span>(batch),</span>
<span id="cb23-225"><a href="#cb23-225" aria-hidden="true" tabindex="-1"></a>      <span class="at">genotype =</span> <span class="fu">factor</span>(genotype, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"APP_SAA_Hom"</span>)),</span>
<span id="cb23-226"><a href="#cb23-226" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="fu">factor</span>(<span class="fu">paste</span>(genotype, condition, <span class="at">sep =</span> <span class="st">", "</span>),</span>
<span id="cb23-227"><a href="#cb23-227" aria-hidden="true" tabindex="-1"></a>                     <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT, Methoxy_Neg"</span>, <span class="st">"APP_SAA_Hom, Methoxy_Neg"</span>,</span>
<span id="cb23-228"><a href="#cb23-228" aria-hidden="true" tabindex="-1"></a>                                <span class="st">"APP_SAA_Hom, Methoxy_Pos"</span>)),</span>
<span id="cb23-229"><a href="#cb23-229" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> dplyr<span class="sc">::</span><span class="fu">recode_factor</span>(</span>
<span id="cb23-230"><a href="#cb23-230" aria-hidden="true" tabindex="-1"></a>        group, </span>
<span id="cb23-231"><a href="#cb23-231" aria-hidden="true" tabindex="-1"></a>        <span class="st">"WT, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"WT, Methoxy-"</span>, </span>
<span id="cb23-232"><a href="#cb23-232" aria-hidden="true" tabindex="-1"></a>        <span class="st">"APP_SAA_Hom, Methoxy_Neg"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy-"</span>,</span>
<span id="cb23-233"><a href="#cb23-233" aria-hidden="true" tabindex="-1"></a>        <span class="st">"APP_SAA_Hom, Methoxy_Pos"</span> <span class="ot">=</span> <span class="st">"APP-SAA, Methoxy+"</span></span>
<span id="cb23-234"><a href="#cb23-234" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb23-235"><a href="#cb23-235" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb23-236"><a href="#cb23-236" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-237"><a href="#cb23-237" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-238"><a href="#cb23-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-239"><a href="#cb23-239" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb23-240"><a href="#cb23-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-241"><a href="#cb23-241" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Differential abundance analysis: <span class="co">[</span><span class="ot">Supplementary table 8</span><span class="co">](https://static-content.springer.com/esm/art%3A10.1186%2Fs13024-022-00547-7/MediaObjects/13024_2022_547_MOESM16_ESM.xlsx)</span></span>
<span id="cb23-242"><a href="#cb23-242" aria-hidden="true" tabindex="-1"></a>is an excel file with the results of the differential abundance analysis</span>
<span id="cb23-243"><a href="#cb23-243" aria-hidden="true" tabindex="-1"></a>(performed by fitting a linear model for each analyte with the </span>
<span id="cb23-244"><a href="#cb23-244" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">limma Bioconductor package</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/limma.html)</span></span>
<span id="cb23-245"><a href="#cb23-245" aria-hidden="true" tabindex="-1"></a>), e.g. it provides log2 fold changes, p-values, and adjusted p-values (FDR)</span>
<span id="cb23-246"><a href="#cb23-246" aria-hidden="true" tabindex="-1"></a>for multiple comparisons of interest. </span>
<span id="cb23-247"><a href="#cb23-247" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>The <span class="in">`get_stats()`</span> function, shown in the</span>
<span id="cb23-248"><a href="#cb23-248" aria-hidden="true" tabindex="-1"></a>  callout box below, retrieves and parses this file for us and returns a </span>
<span id="cb23-249"><a href="#cb23-249" aria-hidden="true" tabindex="-1"></a>  tidy tibble.</span>
<span id="cb23-250"><a href="#cb23-250" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Alternatively, you can download the pre-wrangled statistics in a CSV file</span>
<span id="cb23-251"><a href="#cb23-251" aria-hidden="true" tabindex="-1"></a>  <span class="co">[</span><span class="ot">here</span><span class="co">](https://raw.githubusercontent.com/tomsing1/blog/refs/heads/main/posts/tidyHeatmap/statistics.csv)</span>.</span>
<span id="cb23-252"><a href="#cb23-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-253"><a href="#cb23-253" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip  collapse="true"}</span>
<span id="cb23-254"><a href="#cb23-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-255"><a href="#cb23-255" aria-hidden="true" tabindex="-1"></a><span class="fu">#### The get_stats() function</span></span>
<span id="cb23-256"><a href="#cb23-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-257"><a href="#cb23-257" aria-hidden="true" tabindex="-1"></a><span class="in">````{r}</span></span>
<span id="cb23-258"><a href="#cb23-258" aria-hidden="true" tabindex="-1"></a><span class="in">#| message: false</span></span>
<span id="cb23-259"><a href="#cb23-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-260"><a href="#cb23-260" aria-hidden="true" tabindex="-1"></a><span class="in">library(readxl)  # reading excel files</span></span>
<span id="cb23-261"><a href="#cb23-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-262"><a href="#cb23-262" aria-hidden="true" tabindex="-1"></a><span class="in">#' Retrieve and parse differential abundance statistics from Suppl. Table 8</span></span>
<span id="cb23-263"><a href="#cb23-263" aria-hidden="true" tabindex="-1"></a><span class="in">#' @noRd</span></span>
<span id="cb23-264"><a href="#cb23-264" aria-hidden="true" tabindex="-1"></a><span class="in">#' @examples</span></span>
<span id="cb23-265"><a href="#cb23-265" aria-hidden="true" tabindex="-1"></a><span class="in">#' get_stats()</span></span>
<span id="cb23-266"><a href="#cb23-266" aria-hidden="true" tabindex="-1"></a><span class="in">get_stats &lt;- function() {</span></span>
<span id="cb23-267"><a href="#cb23-267" aria-hidden="true" tabindex="-1"></a><span class="in">  url &lt;- paste0(</span></span>
<span id="cb23-268"><a href="#cb23-268" aria-hidden="true" tabindex="-1"></a><span class="in">    "https://static-content.springer.com/esm/art%3A10.1186%2Fs13024",</span></span>
<span id="cb23-269"><a href="#cb23-269" aria-hidden="true" tabindex="-1"></a><span class="in">    "-022-00547-7/MediaObjects/13024_2022_547_MOESM16_ESM.xlsx"</span></span>
<span id="cb23-270"><a href="#cb23-270" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb23-271"><a href="#cb23-271" aria-hidden="true" tabindex="-1"></a><span class="in">  tmp_file &lt;- tempfile(fileext = ".xlsx")</span></span>
<span id="cb23-272"><a href="#cb23-272" aria-hidden="true" tabindex="-1"></a><span class="in">  download.file(url, tmp_file, quiet = TRUE)</span></span>
<span id="cb23-273"><a href="#cb23-273" aria-hidden="true" tabindex="-1"></a><span class="in">  readxl::read_excel(tmp_file, sheet = "3. APP_pos_vs_WT_neg") |&gt;</span></span>
<span id="cb23-274"><a href="#cb23-274" aria-hidden="true" tabindex="-1"></a><span class="in">    dplyr::rename(component_name = "symbol")</span></span>
<span id="cb23-275"><a href="#cb23-275" aria-hidden="true" tabindex="-1"></a><span class="in">}</span></span>
<span id="cb23-276"><a href="#cb23-276" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-277"><a href="#cb23-277" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-278"><a href="#cb23-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-279"><a href="#cb23-279" aria-hidden="true" tabindex="-1"></a><span class="in">First, we download and read the two CSV files with the pre-wrangled abundance</span></span>
<span id="cb23-280"><a href="#cb23-280" aria-hidden="true" tabindex="-1"></a><span class="in">and statistics data. For compatibility with the `tidyHeatmap` package,</span></span>
<span id="cb23-281"><a href="#cb23-281" aria-hidden="true" tabindex="-1"></a><span class="in">we convert both tables into tibbles as well:</span></span>
<span id="cb23-282"><a href="#cb23-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-285"><a href="#cb23-285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-286"><a href="#cb23-286" aria-hidden="true" tabindex="-1"></a><span class="in">df &lt;- read.csv(</span></span>
<span id="cb23-287"><a href="#cb23-287" aria-hidden="true" tabindex="-1"></a><span class="in">  'https://raw.githubusercontent.com/tomsing1/tidy_heatmap/main/abundances.csv',</span></span>
<span id="cb23-288"><a href="#cb23-288" aria-hidden="true" tabindex="-1"></a><span class="in">  stringsAsFactors=TRUE) |&gt;</span></span>
<span id="cb23-289"><a href="#cb23-289" aria-hidden="true" tabindex="-1"></a><span class="in">  tibble::as_tibble()</span></span>
<span id="cb23-290"><a href="#cb23-290" aria-hidden="true" tabindex="-1"></a><span class="in">stats &lt;- read.csv(</span></span>
<span id="cb23-291"><a href="#cb23-291" aria-hidden="true" tabindex="-1"></a><span class="in">  'https://raw.githubusercontent.com/tomsing1/tidy_heatmap/main/statistics.csv',</span></span>
<span id="cb23-292"><a href="#cb23-292" aria-hidden="true" tabindex="-1"></a><span class="in">  stringsAsFactors=TRUE) |&gt;</span></span>
<span id="cb23-293"><a href="#cb23-293" aria-hidden="true" tabindex="-1"></a><span class="in">  tibble::as_tibble()</span></span>
<span id="cb23-294"><a href="#cb23-294" aria-hidden="true" tabindex="-1"></a><span class="in">stopifnot(all(stats$component_name %in% df$component_name))</span></span>
<span id="cb23-295"><a href="#cb23-295" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-296"><a href="#cb23-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-297"><a href="#cb23-297" aria-hidden="true" tabindex="-1"></a><span class="in">Xia et al only included analytes with a log2 fold change &gt; 20% and an adjusted</span></span>
<span id="cb23-298"><a href="#cb23-298" aria-hidden="true" tabindex="-1"></a><span class="in">p-values &lt; 0.1 in Figure 5. Let's identify the names of the significant analytes</span></span>
<span id="cb23-299"><a href="#cb23-299" aria-hidden="true" tabindex="-1"></a><span class="in">in our `stats` data.frame by applying these thresholds. Then we order the</span></span>
<span id="cb23-300"><a href="#cb23-300" aria-hidden="true" tabindex="-1"></a><span class="in">analytes by their log2 fold change and pull out the `component_name` column.</span></span>
<span id="cb23-301"><a href="#cb23-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-304"><a href="#cb23-304" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-305"><a href="#cb23-305" aria-hidden="true" tabindex="-1"></a><span class="in">significant &lt;- stats |&gt;</span></span>
<span id="cb23-306"><a href="#cb23-306" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::filter(abs(stats$logFC) &gt; 0.2, adj.P.Val &lt; 0.1) |&gt;</span></span>
<span id="cb23-307"><a href="#cb23-307" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::arrange(desc(logFC)) |&gt;</span></span>
<span id="cb23-308"><a href="#cb23-308" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::pull(component_name)</span></span>
<span id="cb23-309"><a href="#cb23-309" aria-hidden="true" tabindex="-1"></a><span class="in">head(significant)</span></span>
<span id="cb23-310"><a href="#cb23-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-311"><a href="#cb23-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-312"><a href="#cb23-312" aria-hidden="true" tabindex="-1"></a><span class="in">We retain `r length(significant)` significant analytes, e.g. `GM3(d36:1)`,</span></span>
<span id="cb23-313"><a href="#cb23-313" aria-hidden="true" tabindex="-1"></a><span class="in">`TG(20:4_36:3)` and `Spermine` are most strongly increased in</span></span>
<span id="cb23-314"><a href="#cb23-314" aria-hidden="true" tabindex="-1"></a><span class="in">APP-SAA, methoxy+ compared to WT, methoxy- samples.</span></span>
<span id="cb23-315"><a href="#cb23-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-316"><a href="#cb23-316" aria-hidden="true" tabindex="-1"></a><span class="in">Next, we subset the abundance table (`df`) to only retain these significant</span></span>
<span id="cb23-317"><a href="#cb23-317" aria-hidden="true" tabindex="-1"></a><span class="in">analytes:</span></span>
<span id="cb23-318"><a href="#cb23-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-321"><a href="#cb23-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-322"><a href="#cb23-322" aria-hidden="true" tabindex="-1"></a><span class="in">df &lt;- df |&gt;</span></span>
<span id="cb23-323"><a href="#cb23-323" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::filter(component_name %in% significant)</span></span>
<span id="cb23-324"><a href="#cb23-324" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-325"><a href="#cb23-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-326"><a href="#cb23-326" aria-hidden="true" tabindex="-1"></a><span class="in">Now we have the data in a tidy tibble with `r nrow(df)` rows and `r ncol(df)`</span></span>
<span id="cb23-327"><a href="#cb23-327" aria-hidden="true" tabindex="-1"></a><span class="in">columns.</span></span>
<span id="cb23-328"><a href="#cb23-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-329"><a href="#cb23-329" aria-hidden="true" tabindex="-1"></a><span class="in">To control the order of categorical variables (e.g. `genotype` or or `batch`),</span></span>
<span id="cb23-330"><a href="#cb23-330" aria-hidden="true" tabindex="-1"></a><span class="in">they are encoded as factors.[^3]. To match the order of the groups in Figure 5,</span></span>
<span id="cb23-331"><a href="#cb23-331" aria-hidden="true" tabindex="-1"></a><span class="in">we `recode` the levels of the `group` variable, so the `WT, Methoxy-` group</span></span>
<span id="cb23-332"><a href="#cb23-332" aria-hidden="true" tabindex="-1"></a><span class="in">is listed first:</span></span>
<span id="cb23-333"><a href="#cb23-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-336"><a href="#cb23-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-337"><a href="#cb23-337" aria-hidden="true" tabindex="-1"></a><span class="in">df$group = dplyr::recode_factor(</span></span>
<span id="cb23-338"><a href="#cb23-338" aria-hidden="true" tabindex="-1"></a><span class="in">    df$group, </span></span>
<span id="cb23-339"><a href="#cb23-339" aria-hidden="true" tabindex="-1"></a><span class="in">    "WT, Methoxy_Neg" = "WT, Methoxy-", </span></span>
<span id="cb23-340"><a href="#cb23-340" aria-hidden="true" tabindex="-1"></a><span class="in">    "APP_SAA_Hom, Methoxy_Neg" = "APP-SAA, Methoxy-",</span></span>
<span id="cb23-341"><a href="#cb23-341" aria-hidden="true" tabindex="-1"></a><span class="in">    "APP_SAA_Hom, Methoxy_Pos" = "APP-SAA, Methoxy+"</span></span>
<span id="cb23-342"><a href="#cb23-342" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb23-343"><a href="#cb23-343" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-344"><a href="#cb23-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-345"><a href="#cb23-345" aria-hidden="true" tabindex="-1"></a><span class="in">[^3]: For example, the two possible values of `genotype` are: `WT` and</span></span>
<span id="cb23-346"><a href="#cb23-346" aria-hidden="true" tabindex="-1"></a><span class="in">`APP_SAA_Hom`, and the three levels of the `group` variable are</span></span>
<span id="cb23-347"><a href="#cb23-347" aria-hidden="true" tabindex="-1"></a><span class="in">`WT, Methoxy-`, `APP-SAA, Methoxy-` and `APP-SAA, Methoxy+`.</span></span>
<span id="cb23-348"><a href="#cb23-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-351"><a href="#cb23-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-352"><a href="#cb23-352" aria-hidden="true" tabindex="-1"></a><span class="in">summary(df)</span></span>
<span id="cb23-353"><a href="#cb23-353" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-354"><a href="#cb23-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-355"><a href="#cb23-355" aria-hidden="true" tabindex="-1"></a><span class="in">We are ready to plot our first heatmap!</span></span>
<span id="cb23-356"><a href="#cb23-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-357"><a href="#cb23-357" aria-hidden="true" tabindex="-1"></a><span class="in">## A first heatmap</span></span>
<span id="cb23-358"><a href="#cb23-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-359"><a href="#cb23-359" aria-hidden="true" tabindex="-1"></a><span class="in">The `tidyHeatmap` R package provides wrapper functions that make the</span></span>
<span id="cb23-360"><a href="#cb23-360" aria-hidden="true" tabindex="-1"></a><span class="in">functionality of the</span></span>
<span id="cb23-361"><a href="#cb23-361" aria-hidden="true" tabindex="-1"></a><span class="in">[ComplexHeatmap Bioconductor package](https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html)</span></span>
<span id="cb23-362"><a href="#cb23-362" aria-hidden="true" tabindex="-1"></a><span class="in">available within a tidy workflow. [^4]</span></span>
<span id="cb23-363"><a href="#cb23-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-364"><a href="#cb23-364" aria-hidden="true" tabindex="-1"></a><span class="in">[^4]: The documentation of the `ComplexHeatmap` package is amazing: the</span></span>
<span id="cb23-365"><a href="#cb23-365" aria-hidden="true" tabindex="-1"></a><span class="in">[Complete Reference](https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html) contains to 15 chapters.</span></span>
<span id="cb23-366"><a href="#cb23-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-367"><a href="#cb23-367" aria-hidden="true" tabindex="-1"></a><span class="in">Creating a first heatmap is as easy as passing our tidy tibble into the </span></span>
<span id="cb23-368"><a href="#cb23-368" aria-hidden="true" tabindex="-1"></a><span class="in">`tidyHeatmap::heatmap()` function, and specifying which columns contain</span></span>
<span id="cb23-369"><a href="#cb23-369" aria-hidden="true" tabindex="-1"></a><span class="in">row identifiers, column identifiers and values:</span></span>
<span id="cb23-370"><a href="#cb23-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-373"><a href="#cb23-373" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-374"><a href="#cb23-374" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-375"><a href="#cb23-375" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-376"><a href="#cb23-376" aria-hidden="true" tabindex="-1"></a><span class="in">heatmap(.data = df, .row = component_name, .column = sample_id, </span></span>
<span id="cb23-377"><a href="#cb23-377" aria-hidden="true" tabindex="-1"></a><span class="in">        .value = abundance, scale = "row")</span></span>
<span id="cb23-378"><a href="#cb23-378" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-379"><a href="#cb23-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-380"><a href="#cb23-380" aria-hidden="true" tabindex="-1"></a><span class="in">::: {.callout-tip  collapse="false"}</span></span>
<span id="cb23-381"><a href="#cb23-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-382"><a href="#cb23-382" aria-hidden="true" tabindex="-1"></a><span class="in">The normalized abundances are hard to interpret (because analytes were</span></span>
<span id="cb23-383"><a href="#cb23-383" aria-hidden="true" tabindex="-1"></a><span class="in">normalized to different internal standards and don't correspond to absolute</span></span>
<span id="cb23-384"><a href="#cb23-384" aria-hidden="true" tabindex="-1"></a><span class="in">concentrations). Luckily, we are mainly interested in more interested in the</span></span>
<span id="cb23-385"><a href="#cb23-385" aria-hidden="true" tabindex="-1"></a><span class="in">relative changes across samples (and groups of samples). By setting </span></span>
<span id="cb23-386"><a href="#cb23-386" aria-hidden="true" tabindex="-1"></a><span class="in">`scale = "row"`, the values for each analyte are converted to z-scores - and</span></span>
<span id="cb23-387"><a href="#cb23-387" aria-hidden="true" tabindex="-1"></a><span class="in">the color scale indicates variation around the mean of each row.</span></span>
<span id="cb23-388"><a href="#cb23-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-389"><a href="#cb23-389" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-390"><a href="#cb23-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-391"><a href="#cb23-391" aria-hidden="true" tabindex="-1"></a><span class="in">We can pass additional arguments through to the `ComplexHeatmap::Heatmap()`</span></span>
<span id="cb23-392"><a href="#cb23-392" aria-hidden="true" tabindex="-1"></a><span class="in">function, e.g. specify another method to cluster rows via the </span></span>
<span id="cb23-393"><a href="#cb23-393" aria-hidden="true" tabindex="-1"></a><span class="in">`clustering_method_rows` argument, by suppressing column clustering entirely</span></span>
<span id="cb23-394"><a href="#cb23-394" aria-hidden="true" tabindex="-1"></a><span class="in">via `cluster_columns=FALSE`, or by defining the color palette to use [^5].</span></span>
<span id="cb23-395"><a href="#cb23-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-396"><a href="#cb23-396" aria-hidden="true" tabindex="-1"></a><span class="in">[^5]: See </span></span>
<span id="cb23-397"><a href="#cb23-397" aria-hidden="true" tabindex="-1"></a><span class="in">[Heatmap's help page](https://www.rdocumentation.org/packages/ComplexHeatmap/versions/1.10.2/topics/Heatmap) </span></span>
<span id="cb23-398"><a href="#cb23-398" aria-hidden="true" tabindex="-1"></a><span class="in">for a full list of arguments.</span></span>
<span id="cb23-399"><a href="#cb23-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-402"><a href="#cb23-402" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-403"><a href="#cb23-403" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-404"><a href="#cb23-404" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-405"><a href="#cb23-405" aria-hidden="true" tabindex="-1"></a><span class="in">colors &lt;- circlize::colorRamp2(</span></span>
<span id="cb23-406"><a href="#cb23-406" aria-hidden="true" tabindex="-1"></a><span class="in">  breaks = c(-2, 0, 2),</span></span>
<span id="cb23-407"><a href="#cb23-407" aria-hidden="true" tabindex="-1"></a><span class="in">  # seq(2, -2, length.out = 11), </span></span>
<span id="cb23-408"><a href="#cb23-408" aria-hidden="true" tabindex="-1"></a><span class="in">  colors = c("navy", "white", "firebrick")</span></span>
<span id="cb23-409"><a href="#cb23-409" aria-hidden="true" tabindex="-1"></a><span class="in">  # RColorBrewer::brewer.pal(11, "RdBu")</span></span>
<span id="cb23-410"><a href="#cb23-410" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb23-411"><a href="#cb23-411" aria-hidden="true" tabindex="-1"></a><span class="in">heatmap(.data = df,</span></span>
<span id="cb23-412"><a href="#cb23-412" aria-hidden="true" tabindex="-1"></a><span class="in">        .row = component_name, </span></span>
<span id="cb23-413"><a href="#cb23-413" aria-hidden="true" tabindex="-1"></a><span class="in">        .column = sample_id, </span></span>
<span id="cb23-414"><a href="#cb23-414" aria-hidden="true" tabindex="-1"></a><span class="in">        .value = abundance, </span></span>
<span id="cb23-415"><a href="#cb23-415" aria-hidden="true" tabindex="-1"></a><span class="in">        column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-416"><a href="#cb23-416" aria-hidden="true" tabindex="-1"></a><span class="in">        row_title = "Analyte",</span></span>
<span id="cb23-417"><a href="#cb23-417" aria-hidden="true" tabindex="-1"></a><span class="in">        scale = "row",</span></span>
<span id="cb23-418"><a href="#cb23-418" aria-hidden="true" tabindex="-1"></a><span class="in">        cluster_columns = FALSE,</span></span>
<span id="cb23-419"><a href="#cb23-419" aria-hidden="true" tabindex="-1"></a><span class="in">        clustering_method_rows = "ward.D",</span></span>
<span id="cb23-420"><a href="#cb23-420" aria-hidden="true" tabindex="-1"></a><span class="in">        palette_value = colors</span></span>
<span id="cb23-421"><a href="#cb23-421" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb23-422"><a href="#cb23-422" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-423"><a href="#cb23-423" aria-hidden="true" tabindex="-1"></a><span class="in">We can also introduce visual splits between the columns, e.g. separating</span></span>
<span id="cb23-424"><a href="#cb23-424" aria-hidden="true" tabindex="-1"></a><span class="in">the three genotypes from each other, and assign a different color to each one:</span></span>
<span id="cb23-425"><a href="#cb23-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-428"><a href="#cb23-428" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-429"><a href="#cb23-429" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-430"><a href="#cb23-430" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-431"><a href="#cb23-431" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-432"><a href="#cb23-432" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group) |&gt;</span></span>
<span id="cb23-433"><a href="#cb23-433" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-434"><a href="#cb23-434" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-435"><a href="#cb23-435" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-436"><a href="#cb23-436" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-437"><a href="#cb23-437" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-438"><a href="#cb23-438" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-439"><a href="#cb23-439" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-440"><a href="#cb23-440" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_rows = "ward.D",</span></span>
<span id="cb23-441"><a href="#cb23-441" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_columns = "ward.D",</span></span>
<span id="cb23-442"><a href="#cb23-442" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-443"><a href="#cb23-443" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-444"><a href="#cb23-444" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-445"><a href="#cb23-445" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-446"><a href="#cb23-446" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb23-447"><a href="#cb23-447" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-448"><a href="#cb23-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-449"><a href="#cb23-449" aria-hidden="true" tabindex="-1"></a><span class="in">::: {.callout-tip  collapse="true"}</span></span>
<span id="cb23-450"><a href="#cb23-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-451"><a href="#cb23-451" aria-hidden="true" tabindex="-1"></a><span class="in">## Splitting features</span></span>
<span id="cb23-452"><a href="#cb23-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-453"><a href="#cb23-453" aria-hidden="true" tabindex="-1"></a><span class="in">The dendrogram (on the left) indicates that there might be two major groups</span></span>
<span id="cb23-454"><a href="#cb23-454" aria-hidden="true" tabindex="-1"></a><span class="in">of features in this dataset, e.g. there are two main branches. </span></span>
<span id="cb23-455"><a href="#cb23-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-456"><a href="#cb23-456" aria-hidden="true" tabindex="-1"></a><span class="in">### Cutting the hierarchical clustering result</span></span>
<span id="cb23-457"><a href="#cb23-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-458"><a href="#cb23-458" aria-hidden="true" tabindex="-1"></a><span class="in">We can reinforce this result by cutting the dendrogram after two branches have</span></span>
<span id="cb23-459"><a href="#cb23-459" aria-hidden="true" tabindex="-1"></a><span class="in">arisen.</span></span>
<span id="cb23-460"><a href="#cb23-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-463"><a href="#cb23-463" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-464"><a href="#cb23-464" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-465"><a href="#cb23-465" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-466"><a href="#cb23-466" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-467"><a href="#cb23-467" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group) |&gt;</span></span>
<span id="cb23-468"><a href="#cb23-468" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-469"><a href="#cb23-469" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-470"><a href="#cb23-470" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-471"><a href="#cb23-471" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-472"><a href="#cb23-472" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-473"><a href="#cb23-473" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-474"><a href="#cb23-474" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-475"><a href="#cb23-475" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_rows = "ward.D",</span></span>
<span id="cb23-476"><a href="#cb23-476" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_columns = "ward.D",</span></span>
<span id="cb23-477"><a href="#cb23-477" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-478"><a href="#cb23-478" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-479"><a href="#cb23-479" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-480"><a href="#cb23-480" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-481"><a href="#cb23-481" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-482"><a href="#cb23-482" aria-hidden="true" tabindex="-1"></a><span class="in">    split_rows(2)</span></span>
<span id="cb23-483"><a href="#cb23-483" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-484"><a href="#cb23-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-485"><a href="#cb23-485" aria-hidden="true" tabindex="-1"></a><span class="in">### K-means clustering</span></span>
<span id="cb23-486"><a href="#cb23-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-487"><a href="#cb23-487" aria-hidden="true" tabindex="-1"></a><span class="in">Alternatively, we can can perform</span></span>
<span id="cb23-488"><a href="#cb23-488" aria-hidden="true" tabindex="-1"></a><span class="in">[k-means clustering](https://en.wikipedia.org/wiki/K-means_clustering)</span></span>
<span id="cb23-489"><a href="#cb23-489" aria-hidden="true" tabindex="-1"></a><span class="in">on the rows (or columns) by specifying the number of clusters (k) to isolate</span></span>
<span id="cb23-490"><a href="#cb23-490" aria-hidden="true" tabindex="-1"></a><span class="in">via the `row_km` argument. Our current dataset contains missing values </span></span>
<span id="cb23-491"><a href="#cb23-491" aria-hidden="true" tabindex="-1"></a><span class="in">(`NA`), which is incompatible with the k-means clustering algorithm. To use this</span></span>
<span id="cb23-492"><a href="#cb23-492" aria-hidden="true" tabindex="-1"></a><span class="in">method we first need to decide how to handle missing values, e.g. drop analytes</span></span>
<span id="cb23-493"><a href="#cb23-493" aria-hidden="true" tabindex="-1"></a><span class="in">or impute reasonable values.</span></span>
<span id="cb23-494"><a href="#cb23-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-495"><a href="#cb23-495" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-496"><a href="#cb23-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-497"><a href="#cb23-497" aria-hidden="true" tabindex="-1"></a><span class="in">## Adding annotations</span></span>
<span id="cb23-498"><a href="#cb23-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-499"><a href="#cb23-499" aria-hidden="true" tabindex="-1"></a><span class="in">The `annotation_tile()` function[^6] allows us to add additional annotations to</span></span>
<span id="cb23-500"><a href="#cb23-500" aria-hidden="true" tabindex="-1"></a><span class="in">our heatmap, e.g. to identify </span></span>
<span id="cb23-501"><a href="#cb23-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-502"><a href="#cb23-502" aria-hidden="true" tabindex="-1"></a><span class="in">- columns from the `batches` or correspoding to animals with male or female </span></span>
<span id="cb23-503"><a href="#cb23-503" aria-hidden="true" tabindex="-1"></a><span class="in">  `sex` in an annotation bar on top of the heatmap, or</span></span>
<span id="cb23-504"><a href="#cb23-504" aria-hidden="true" tabindex="-1"></a><span class="in">- rows with values from different `panels` (lipds, metabolits) in an annotation</span></span>
<span id="cb23-505"><a href="#cb23-505" aria-hidden="true" tabindex="-1"></a><span class="in">  strip on the left of the heatmap.</span></span>
<span id="cb23-506"><a href="#cb23-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-507"><a href="#cb23-507" aria-hidden="true" tabindex="-1"></a><span class="in">[^6]: Prior to `tidyHeatmap` version 1.9.0, the `add_tile()` function was used</span></span>
<span id="cb23-508"><a href="#cb23-508" aria-hidden="true" tabindex="-1"></a><span class="in">instead. It is now deprecated in favor of `annotation_tile()`. (At the time of</span></span>
<span id="cb23-509"><a href="#cb23-509" aria-hidden="true" tabindex="-1"></a><span class="in">writing, the latest version available from CRAN is version 1.8.1; please</span></span>
<span id="cb23-510"><a href="#cb23-510" aria-hidden="true" tabindex="-1"></a><span class="in">install &amp; use the latest development version from github to follow along this</span></span>
<span id="cb23-511"><a href="#cb23-511" aria-hidden="true" tabindex="-1"></a><span class="in">tutorial.)</span></span>
<span id="cb23-512"><a href="#cb23-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-513"><a href="#cb23-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-516"><a href="#cb23-516" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-517"><a href="#cb23-517" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-518"><a href="#cb23-518" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-519"><a href="#cb23-519" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-520"><a href="#cb23-520" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group) |&gt;</span></span>
<span id="cb23-521"><a href="#cb23-521" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-522"><a href="#cb23-522" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-523"><a href="#cb23-523" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-524"><a href="#cb23-524" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-525"><a href="#cb23-525" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-526"><a href="#cb23-526" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-527"><a href="#cb23-527" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-528"><a href="#cb23-528" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_rows = "ward.D",</span></span>
<span id="cb23-529"><a href="#cb23-529" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_columns = "ward.D",</span></span>
<span id="cb23-530"><a href="#cb23-530" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-531"><a href="#cb23-531" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-532"><a href="#cb23-532" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-533"><a href="#cb23-533" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-534"><a href="#cb23-534" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-535"><a href="#cb23-535" aria-hidden="true" tabindex="-1"></a><span class="in">    annotation_tile(</span></span>
<span id="cb23-536"><a href="#cb23-536" aria-hidden="true" tabindex="-1"></a><span class="in">        batch, </span></span>
<span id="cb23-537"><a href="#cb23-537" aria-hidden="true" tabindex="-1"></a><span class="in">        palette = RColorBrewer::brewer.pal(3, "Dark2")</span></span>
<span id="cb23-538"><a href="#cb23-538" aria-hidden="true" tabindex="-1"></a><span class="in">    ) |&gt;</span></span>
<span id="cb23-539"><a href="#cb23-539" aria-hidden="true" tabindex="-1"></a><span class="in">    annotation_tile(sex, </span></span>
<span id="cb23-540"><a href="#cb23-540" aria-hidden="true" tabindex="-1"></a><span class="in">             palette = c("skyblue", "pink")) |&gt;</span></span>
<span id="cb23-541"><a href="#cb23-541" aria-hidden="true" tabindex="-1"></a><span class="in">    annotation_tile(panel, size = grid::unit(5, "points"))</span></span>
<span id="cb23-542"><a href="#cb23-542" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-543"><a href="#cb23-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-544"><a href="#cb23-544" aria-hidden="true" tabindex="-1"></a><span class="in">::: {.callout-tip  collapse="true"}</span></span>
<span id="cb23-545"><a href="#cb23-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-546"><a href="#cb23-546" aria-hidden="true" tabindex="-1"></a><span class="in">### Row or column annotations?</span></span>
<span id="cb23-547"><a href="#cb23-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-548"><a href="#cb23-548" aria-hidden="true" tabindex="-1"></a><span class="in">The `annotation_tile()` function does not require us to specify whether we</span></span>
<span id="cb23-549"><a href="#cb23-549" aria-hidden="true" tabindex="-1"></a><span class="in">are adding row or column annotations. Instead, it infers it automatically:</span></span>
<span id="cb23-550"><a href="#cb23-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-551"><a href="#cb23-551" aria-hidden="true" tabindex="-1"></a><span class="in">- Does each value used to define a row of the heatmap (e.g. `component_name`)</span></span>
<span id="cb23-552"><a href="#cb23-552" aria-hidden="true" tabindex="-1"></a><span class="in">  map to a single value in the selected column of the data.frame?</span></span>
<span id="cb23-553"><a href="#cb23-553" aria-hidden="true" tabindex="-1"></a><span class="in">  - If yes, create a row annotation strip.</span></span>
<span id="cb23-554"><a href="#cb23-554" aria-hidden="true" tabindex="-1"></a><span class="in">- Does each value used to define a column of the heatmap (e.g. `sample_id`)</span></span>
<span id="cb23-555"><a href="#cb23-555" aria-hidden="true" tabindex="-1"></a><span class="in">  map to a single value in the selected column of the data.frame?</span></span>
<span id="cb23-556"><a href="#cb23-556" aria-hidden="true" tabindex="-1"></a><span class="in">  - If yes, create a column annotation.</span></span>
<span id="cb23-557"><a href="#cb23-557" aria-hidden="true" tabindex="-1"></a><span class="in">- If neither of the above are true, an error is raised.</span></span>
<span id="cb23-558"><a href="#cb23-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-559"><a href="#cb23-559" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-560"><a href="#cb23-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-561"><a href="#cb23-561" aria-hidden="true" tabindex="-1"></a><span class="in">## Adding quantitative annotations</span></span>
<span id="cb23-562"><a href="#cb23-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-563"><a href="#cb23-563" aria-hidden="true" tabindex="-1"></a><span class="in">Some annotations are quantitative, and may be better communicated in a plot.</span></span>
<span id="cb23-564"><a href="#cb23-564" aria-hidden="true" tabindex="-1"></a><span class="in">Let's display the number of cells that were analyzed in each sample on top of</span></span>
<span id="cb23-565"><a href="#cb23-565" aria-hidden="true" tabindex="-1"></a><span class="in">our heatmap with the `add_point()` function:</span></span>
<span id="cb23-566"><a href="#cb23-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-569"><a href="#cb23-569" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-570"><a href="#cb23-570" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-571"><a href="#cb23-571" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-572"><a href="#cb23-572" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-573"><a href="#cb23-573" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group) |&gt;</span></span>
<span id="cb23-574"><a href="#cb23-574" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-575"><a href="#cb23-575" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-576"><a href="#cb23-576" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-577"><a href="#cb23-577" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-578"><a href="#cb23-578" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-579"><a href="#cb23-579" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-580"><a href="#cb23-580" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-581"><a href="#cb23-581" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_rows = "ward.D",</span></span>
<span id="cb23-582"><a href="#cb23-582" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_columns = "ward.D",</span></span>
<span id="cb23-583"><a href="#cb23-583" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-584"><a href="#cb23-584" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-585"><a href="#cb23-585" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-586"><a href="#cb23-586" aria-hidden="true" tabindex="-1"></a><span class="in">          )</span></span>
<span id="cb23-587"><a href="#cb23-587" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-588"><a href="#cb23-588" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-589"><a href="#cb23-589" aria-hidden="true" tabindex="-1"></a><span class="in">    batch, </span></span>
<span id="cb23-590"><a href="#cb23-590" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = RColorBrewer::brewer.pal(3, "Dark2")</span></span>
<span id="cb23-591"><a href="#cb23-591" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-592"><a href="#cb23-592" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-593"><a href="#cb23-593" aria-hidden="true" tabindex="-1"></a><span class="in">    sex, </span></span>
<span id="cb23-594"><a href="#cb23-594" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = c("skyblue", "pink")</span></span>
<span id="cb23-595"><a href="#cb23-595" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-596"><a href="#cb23-596" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(panel, size = grid::unit(5, "points")) |&gt;</span></span>
<span id="cb23-597"><a href="#cb23-597" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_point(cell_number)</span></span>
<span id="cb23-598"><a href="#cb23-598" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-599"><a href="#cb23-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-600"><a href="#cb23-600" aria-hidden="true" tabindex="-1"></a><span class="in">::: {.callout-tip  collapse="true"}</span></span>
<span id="cb23-601"><a href="#cb23-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-602"><a href="#cb23-602" aria-hidden="true" tabindex="-1"></a><span class="in">## Overlaying symbols</span></span>
<span id="cb23-603"><a href="#cb23-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-604"><a href="#cb23-604" aria-hidden="true" tabindex="-1"></a><span class="in">To highlight specific cells in our heatmap, the `layer_point()`, </span></span>
<span id="cb23-605"><a href="#cb23-605" aria-hidden="true" tabindex="-1"></a><span class="in">`layer_arrow_up()`, `layer_arrow_down()`, `layer_diamond()`, `layer_square()`,</span></span>
<span id="cb23-606"><a href="#cb23-606" aria-hidden="true" tabindex="-1"></a><span class="in">`layer_star()`, `layer_text()` or `layer_asterisk()`</span></span>
<span id="cb23-607"><a href="#cb23-607" aria-hidden="true" tabindex="-1"></a><span class="in">functions can be used, together with a filter condition [^7].</span></span>
<span id="cb23-608"><a href="#cb23-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-609"><a href="#cb23-609" aria-hidden="true" tabindex="-1"></a><span class="in">[^7]: At the time of writing, the latest version of the `tidyHeatmap` package</span></span>
<span id="cb23-610"><a href="#cb23-610" aria-hidden="true" tabindex="-1"></a><span class="in">available from CRAN is version 1.8.1. The development version used here, </span></span>
<span id="cb23-611"><a href="#cb23-611" aria-hidden="true" tabindex="-1"></a><span class="in">available [from github](https://github.com/stemangiola/tidyHeatmap),</span></span>
<span id="cb23-612"><a href="#cb23-612" aria-hidden="true" tabindex="-1"></a><span class="in">contains additional functionality, including e.g. the `layer_star()` and - most</span></span>
<span id="cb23-613"><a href="#cb23-613" aria-hidden="true" tabindex="-1"></a><span class="in">importantly - `layer_asterisk()`.</span></span>
<span id="cb23-614"><a href="#cb23-614" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-615"><a href="#cb23-615" aria-hidden="true" tabindex="-1"></a><span class="in">Can you spot the asterisks for `GM3(d36:1)` for samples 1, 3, and 6 in the </span></span>
<span id="cb23-616"><a href="#cb23-616" aria-hidden="true" tabindex="-1"></a><span class="in">heatmap below?</span></span>
<span id="cb23-617"><a href="#cb23-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-620"><a href="#cb23-620" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-621"><a href="#cb23-621" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-622"><a href="#cb23-622" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-623"><a href="#cb23-623" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-624"><a href="#cb23-624" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group) |&gt;</span></span>
<span id="cb23-625"><a href="#cb23-625" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-626"><a href="#cb23-626" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-627"><a href="#cb23-627" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-628"><a href="#cb23-628" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-629"><a href="#cb23-629" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-630"><a href="#cb23-630" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-631"><a href="#cb23-631" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-632"><a href="#cb23-632" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_rows = "ward.D",</span></span>
<span id="cb23-633"><a href="#cb23-633" aria-hidden="true" tabindex="-1"></a><span class="in">          clustering_method_columns = "ward.D",</span></span>
<span id="cb23-634"><a href="#cb23-634" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-635"><a href="#cb23-635" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-636"><a href="#cb23-636" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-637"><a href="#cb23-637" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-638"><a href="#cb23-638" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-639"><a href="#cb23-639" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-640"><a href="#cb23-640" aria-hidden="true" tabindex="-1"></a><span class="in">    batch, </span></span>
<span id="cb23-641"><a href="#cb23-641" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = RColorBrewer::brewer.pal(3, "Dark2")</span></span>
<span id="cb23-642"><a href="#cb23-642" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-643"><a href="#cb23-643" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(sex, </span></span>
<span id="cb23-644"><a href="#cb23-644" aria-hidden="true" tabindex="-1"></a><span class="in">                  palette = c("skyblue", "pink"))  |&gt;</span></span>
<span id="cb23-645"><a href="#cb23-645" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(panel, size = grid::unit(5, "points")) |&gt;</span></span>
<span id="cb23-646"><a href="#cb23-646" aria-hidden="true" tabindex="-1"></a><span class="in">  layer_asterisk(</span></span>
<span id="cb23-647"><a href="#cb23-647" aria-hidden="true" tabindex="-1"></a><span class="in">    component_name == "GM3(d36:1)" &amp; sample_id %in% c("1", "3", "6")</span></span>
<span id="cb23-648"><a href="#cb23-648" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb23-649"><a href="#cb23-649" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-650"><a href="#cb23-650" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-651"><a href="#cb23-651" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-652"><a href="#cb23-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-653"><a href="#cb23-653" aria-hidden="true" tabindex="-1"></a><span class="in">## Manually ordering rows</span></span>
<span id="cb23-654"><a href="#cb23-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-655"><a href="#cb23-655" aria-hidden="true" tabindex="-1"></a><span class="in">For Figure 5, Xia et al manually ordered the rows, with those analytes showing</span></span>
<span id="cb23-656"><a href="#cb23-656" aria-hidden="true" tabindex="-1"></a><span class="in">the largest log2 fold increases at the top, and those with the largest decreases</span></span>
<span id="cb23-657"><a href="#cb23-657" aria-hidden="true" tabindex="-1"></a><span class="in">(e.g. negative log2 fold changes) at the bottom. </span></span>
<span id="cb23-658"><a href="#cb23-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-659"><a href="#cb23-659" aria-hidden="true" tabindex="-1"></a><span class="in">To reproduce this, we switch off the clustering of the rows, and change the</span></span>
<span id="cb23-660"><a href="#cb23-660" aria-hidden="true" tabindex="-1"></a><span class="in">order of the `component_name` factor to the `significant` character vector - </span></span>
<span id="cb23-661"><a href="#cb23-661" aria-hidden="true" tabindex="-1"></a><span class="in">which is ordered from largest to smallest fold change (see above) - with the</span></span>
<span id="cb23-662"><a href="#cb23-662" aria-hidden="true" tabindex="-1"></a><span class="in">`dplyr::mutate()` function.</span></span>
<span id="cb23-663"><a href="#cb23-663" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-666"><a href="#cb23-666" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-667"><a href="#cb23-667" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-668"><a href="#cb23-668" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-669"><a href="#cb23-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-670"><a href="#cb23-670" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-671"><a href="#cb23-671" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::mutate(</span></span>
<span id="cb23-672"><a href="#cb23-672" aria-hidden="true" tabindex="-1"></a><span class="in">    component_name = factor(component_name, levels = significant)</span></span>
<span id="cb23-673"><a href="#cb23-673" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-674"><a href="#cb23-674" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group, panel) |&gt;</span></span>
<span id="cb23-675"><a href="#cb23-675" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name, </span></span>
<span id="cb23-676"><a href="#cb23-676" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-677"><a href="#cb23-677" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-678"><a href="#cb23-678" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-679"><a href="#cb23-679" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-680"><a href="#cb23-680" aria-hidden="true" tabindex="-1"></a><span class="in">          scale = "row",</span></span>
<span id="cb23-681"><a href="#cb23-681" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-682"><a href="#cb23-682" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_rows = FALSE,</span></span>
<span id="cb23-683"><a href="#cb23-683" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-684"><a href="#cb23-684" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_grouping = list(</span></span>
<span id="cb23-685"><a href="#cb23-685" aria-hidden="true" tabindex="-1"></a><span class="in">            palette.colors(2, "Set1"),</span></span>
<span id="cb23-686"><a href="#cb23-686" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-687"><a href="#cb23-687" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-688"><a href="#cb23-688" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-689"><a href="#cb23-689" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-690"><a href="#cb23-690" aria-hidden="true" tabindex="-1"></a><span class="in">    batch, </span></span>
<span id="cb23-691"><a href="#cb23-691" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = RColorBrewer::brewer.pal(3, "Dark2"),</span></span>
<span id="cb23-692"><a href="#cb23-692" aria-hidden="true" tabindex="-1"></a><span class="in">    size = grid::unit(10, "points")</span></span>
<span id="cb23-693"><a href="#cb23-693" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-694"><a href="#cb23-694" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-695"><a href="#cb23-695" aria-hidden="true" tabindex="-1"></a><span class="in">    sex, </span></span>
<span id="cb23-696"><a href="#cb23-696" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = c("skyblue", "pink"),</span></span>
<span id="cb23-697"><a href="#cb23-697" aria-hidden="true" tabindex="-1"></a><span class="in">    size = grid::unit(10, "points")</span></span>
<span id="cb23-698"><a href="#cb23-698" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb23-699"><a href="#cb23-699" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-700"><a href="#cb23-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-701"><a href="#cb23-701" aria-hidden="true" tabindex="-1"></a><span class="in">In this step, we also changed the `group_by` command: we grouped by both</span></span>
<span id="cb23-702"><a href="#cb23-702" aria-hidden="true" tabindex="-1"></a><span class="in">the `group` and the `panel` annotations, splitting columns and rows of our</span></span>
<span id="cb23-703"><a href="#cb23-703" aria-hidden="true" tabindex="-1"></a><span class="in">heatmap into separate panels, respectively.</span></span>
<span id="cb23-704"><a href="#cb23-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-705"><a href="#cb23-705" aria-hidden="true" tabindex="-1"></a><span class="in">## Manually scaling rows</span></span>
<span id="cb23-706"><a href="#cb23-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-707"><a href="#cb23-707" aria-hidden="true" tabindex="-1"></a><span class="in">So far, we have scaled each row by converting the measurements for each</span></span>
<span id="cb23-708"><a href="#cb23-708" aria-hidden="true" tabindex="-1"></a><span class="in">analyte to </span></span>
<span id="cb23-709"><a href="#cb23-709" aria-hidden="true" tabindex="-1"></a><span class="in">[z-scores](https://en.wikipedia.org/wiki/Standard_score),</span></span>
<span id="cb23-710"><a href="#cb23-710" aria-hidden="true" tabindex="-1"></a><span class="in">and the colors reflect differences from the overall mean across all samples.</span></span>
<span id="cb23-711"><a href="#cb23-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-712"><a href="#cb23-712" aria-hidden="true" tabindex="-1"></a><span class="in">In many experiments, including this one, there is an explicit experimental </span></span>
<span id="cb23-713"><a href="#cb23-713" aria-hidden="true" tabindex="-1"></a><span class="in">group that provides a natural baseline. The effects of the `APP-SAA` genotype</span></span>
<span id="cb23-714"><a href="#cb23-714" aria-hidden="true" tabindex="-1"></a><span class="in">is usually interpret relative to the `WT` control group. We therefore center</span></span>
<span id="cb23-715"><a href="#cb23-715" aria-hidden="true" tabindex="-1"></a><span class="in">the abundances by subtracting the median value observed in this </span></span>
<span id="cb23-716"><a href="#cb23-716" aria-hidden="true" tabindex="-1"></a><span class="in">(e.g. the `WT, methoxy-`) group. Now the colors indicate log2 fold changes</span></span>
<span id="cb23-717"><a href="#cb23-717" aria-hidden="true" tabindex="-1"></a><span class="in">relative to the median in the control group.</span></span>
<span id="cb23-718"><a href="#cb23-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-719"><a href="#cb23-719" aria-hidden="true" tabindex="-1"></a><span class="in">::: {.callout-tip  collapse="false"}</span></span>
<span id="cb23-720"><a href="#cb23-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-721"><a href="#cb23-721" aria-hidden="true" tabindex="-1"></a><span class="in">Here, we use functions from the `dplyr` package to </span></span>
<span id="cb23-722"><a href="#cb23-722" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-723"><a href="#cb23-723" aria-hidden="true" tabindex="-1"></a><span class="in">1. `group_by`: group the tibble by `component_name`, so we can perform</span></span>
<span id="cb23-724"><a href="#cb23-724" aria-hidden="true" tabindex="-1"></a><span class="in">  calculations separately for each analyte,</span></span>
<span id="cb23-725"><a href="#cb23-725" aria-hidden="true" tabindex="-1"></a><span class="in">2. `mutate`: calculate the median for the abundances observed in the</span></span>
<span id="cb23-726"><a href="#cb23-726" aria-hidden="true" tabindex="-1"></a><span class="in">   `WT, Methoxy-` group (for this analyte)</span></span>
<span id="cb23-727"><a href="#cb23-727" aria-hidden="true" tabindex="-1"></a><span class="in">3. `mutate`: subtract the control median from each value (for this analyte).</span></span>
<span id="cb23-728"><a href="#cb23-728" aria-hidden="true" tabindex="-1"></a><span class="in">4. `ungroup`: ungroup the tibble, so we can continue with the full dataset.</span></span>
<span id="cb23-729"><a href="#cb23-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-730"><a href="#cb23-730" aria-hidden="true" tabindex="-1"></a><span class="in">:::</span></span>
<span id="cb23-731"><a href="#cb23-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-734"><a href="#cb23-734" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-735"><a href="#cb23-735" aria-hidden="true" tabindex="-1"></a><span class="in">#| message: false</span></span>
<span id="cb23-736"><a href="#cb23-736" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-width: 7.5</span></span>
<span id="cb23-737"><a href="#cb23-737" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-height: 10</span></span>
<span id="cb23-738"><a href="#cb23-738" aria-hidden="true" tabindex="-1"></a><span class="in">df |&gt;</span></span>
<span id="cb23-739"><a href="#cb23-739" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::mutate(</span></span>
<span id="cb23-740"><a href="#cb23-740" aria-hidden="true" tabindex="-1"></a><span class="in">    component_name = factor(component_name, levels = significant)</span></span>
<span id="cb23-741"><a href="#cb23-741" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-742"><a href="#cb23-742" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::group_by(component_name) |&gt;</span></span>
<span id="cb23-743"><a href="#cb23-743" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::mutate(</span></span>
<span id="cb23-744"><a href="#cb23-744" aria-hidden="true" tabindex="-1"></a><span class="in">    ctrl_median = median(abundance[group == "WT, Methoxy-"], na.rm = TRUE),</span></span>
<span id="cb23-745"><a href="#cb23-745" aria-hidden="true" tabindex="-1"></a><span class="in">    abundance = abundance - ctrl_median</span></span>
<span id="cb23-746"><a href="#cb23-746" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-747"><a href="#cb23-747" aria-hidden="true" tabindex="-1"></a><span class="in">  dplyr::ungroup() |&gt;</span></span>
<span id="cb23-748"><a href="#cb23-748" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(group, panel) |&gt;</span></span>
<span id="cb23-749"><a href="#cb23-749" aria-hidden="true" tabindex="-1"></a><span class="in">  heatmap(.row = component_name,</span></span>
<span id="cb23-750"><a href="#cb23-750" aria-hidden="true" tabindex="-1"></a><span class="in">          .column = sample_id, </span></span>
<span id="cb23-751"><a href="#cb23-751" aria-hidden="true" tabindex="-1"></a><span class="in">          .value = abundance, </span></span>
<span id="cb23-752"><a href="#cb23-752" aria-hidden="true" tabindex="-1"></a><span class="in">          column_title = "Xia et al: targeted LC/MS data",</span></span>
<span id="cb23-753"><a href="#cb23-753" aria-hidden="true" tabindex="-1"></a><span class="in">          row_title = "Analyte",</span></span>
<span id="cb23-754"><a href="#cb23-754" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_columns = FALSE,</span></span>
<span id="cb23-755"><a href="#cb23-755" aria-hidden="true" tabindex="-1"></a><span class="in">          cluster_rows = FALSE,</span></span>
<span id="cb23-756"><a href="#cb23-756" aria-hidden="true" tabindex="-1"></a><span class="in">          show_column_names = FALSE,</span></span>
<span id="cb23-757"><a href="#cb23-757" aria-hidden="true" tabindex="-1"></a><span class="in">          row_names_gp = grid::gpar(fontsize = 6),</span></span>
<span id="cb23-758"><a href="#cb23-758" aria-hidden="true" tabindex="-1"></a><span class="in">          palette_value = colors,</span></span>
<span id="cb23-759"><a href="#cb23-759" aria-hidden="true" tabindex="-1"></a><span class="in">           palette_grouping = list(</span></span>
<span id="cb23-760"><a href="#cb23-760" aria-hidden="true" tabindex="-1"></a><span class="in">            palette.colors(2, "Set1"),</span></span>
<span id="cb23-761"><a href="#cb23-761" aria-hidden="true" tabindex="-1"></a><span class="in">            c("darkgrey", "orange", "firebrick")</span></span>
<span id="cb23-762"><a href="#cb23-762" aria-hidden="true" tabindex="-1"></a><span class="in">        )</span></span>
<span id="cb23-763"><a href="#cb23-763" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-764"><a href="#cb23-764" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-765"><a href="#cb23-765" aria-hidden="true" tabindex="-1"></a><span class="in">    batch, </span></span>
<span id="cb23-766"><a href="#cb23-766" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = RColorBrewer::brewer.pal(3, "Dark2"),</span></span>
<span id="cb23-767"><a href="#cb23-767" aria-hidden="true" tabindex="-1"></a><span class="in">    size = grid::unit(10, "points")</span></span>
<span id="cb23-768"><a href="#cb23-768" aria-hidden="true" tabindex="-1"></a><span class="in">  ) |&gt;</span></span>
<span id="cb23-769"><a href="#cb23-769" aria-hidden="true" tabindex="-1"></a><span class="in">  annotation_tile(</span></span>
<span id="cb23-770"><a href="#cb23-770" aria-hidden="true" tabindex="-1"></a><span class="in">    sex, </span></span>
<span id="cb23-771"><a href="#cb23-771" aria-hidden="true" tabindex="-1"></a><span class="in">    palette = c("skyblue", "pink"),</span></span>
<span id="cb23-772"><a href="#cb23-772" aria-hidden="true" tabindex="-1"></a><span class="in">    size = grid::unit(10, "points")</span></span>
<span id="cb23-773"><a href="#cb23-773" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb23-774"><a href="#cb23-774" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-775"><a href="#cb23-775" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-776"><a href="#cb23-776" aria-hidden="true" tabindex="-1"></a><span class="in">We also added the `row_names_gp = grid::gpar(fontsize = 6)` argument to the</span></span>
<span id="cb23-777"><a href="#cb23-777" aria-hidden="true" tabindex="-1"></a><span class="in">`heatmap()` call, increasing the font size of the row labels. (They are still</span></span>
<span id="cb23-778"><a href="#cb23-778" aria-hidden="true" tabindex="-1"></a><span class="in">a bit on the small side, so you might want to consider labelling only a subset.)</span></span>
<span id="cb23-779"><a href="#cb23-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-780"><a href="#cb23-780" aria-hidden="true" tabindex="-1"></a><span class="in">## Reproducibility</span></span>
<span id="cb23-781"><a href="#cb23-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-782"><a href="#cb23-782" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;details&gt;</span></span>
<span id="cb23-783"><a href="#cb23-783" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;summary&gt;</span></span>
<span id="cb23-784"><a href="#cb23-784" aria-hidden="true" tabindex="-1"></a><span class="in">Session information</span></span>
<span id="cb23-785"><a href="#cb23-785" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;/summary&gt;</span></span>
<span id="cb23-786"><a href="#cb23-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-789"><a href="#cb23-789" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-790"><a href="#cb23-790" aria-hidden="true" tabindex="-1"></a><span class="in">sessionInfo()</span></span>
<span id="cb23-791"><a href="#cb23-791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-792"><a href="#cb23-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-793"><a href="#cb23-793" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;/details&gt;</span></span>
<span id="cb23-794"><a href="#cb23-794" aria-hidden="true" tabindex="-1"></a><span class="in">   </span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>